--- ../src-base/minecraft/net/minecraft/entity/projectile/EntityFireball.java
+++ ../src-work/minecraft/net/minecraft/entity/projectile/EntityFireball.java
@@ -14,30 +14,34 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.entity.Projectile;
+import org.bukkit.projectiles.ProjectileSource;
 
 public abstract class EntityFireball extends Entity
 {
-    public EntityLivingBase field_70235_a;
-    private int field_70236_j;
-    private int field_70234_an;
-    public double field_70232_b;
-    public double field_70233_c;
-    public double field_70230_d;
-
-    public EntityFireball(World p_i1759_1_)
+    public EntityLivingBase shootingEntity;
+    private int ticksAlive;
+    private int ticksInAir;
+    public double accelerationX;
+    public double accelerationY;
+    public double accelerationZ;
+    public float bukkitYield = 1; // CraftBukkit
+    public boolean isIncendiary = true; // CraftBukkit
+    public EntityFireball(World worldIn)
     {
-        super(p_i1759_1_);
-        this.func_70105_a(1.0F, 1.0F);
+        super(worldIn);
+        this.setSize(1.0F, 1.0F);
     }
 
-    protected void func_70088_a()
+    protected void entityInit()
     {
     }
 
     @SideOnly(Side.CLIENT)
-    public boolean func_70112_a(double p_70112_1_)
+    public boolean isInRangeToRenderDist(double distance)
     {
-        double d0 = this.func_174813_aQ().func_72320_b() * 4.0D;
+        double d0 = this.getEntityBoundingBox().getAverageEdgeLength() * 4.0D;
 
         if (Double.isNaN(d0))
         {
@@ -45,185 +49,211 @@
         }
 
         d0 = d0 * 64.0D;
-        return p_70112_1_ < d0 * d0;
+        return distance < d0 * d0;
     }
 
-    public EntityFireball(World p_i1760_1_, double p_i1760_2_, double p_i1760_4_, double p_i1760_6_, double p_i1760_8_, double p_i1760_10_, double p_i1760_12_)
+    public EntityFireball(World worldIn, double x, double y, double z, double accelX, double accelY, double accelZ)
     {
-        super(p_i1760_1_);
-        this.func_70105_a(1.0F, 1.0F);
-        this.func_70012_b(p_i1760_2_, p_i1760_4_, p_i1760_6_, this.field_70177_z, this.field_70125_A);
-        this.func_70107_b(p_i1760_2_, p_i1760_4_, p_i1760_6_);
-        double d0 = (double)MathHelper.func_76133_a(p_i1760_8_ * p_i1760_8_ + p_i1760_10_ * p_i1760_10_ + p_i1760_12_ * p_i1760_12_);
-        this.field_70232_b = p_i1760_8_ / d0 * 0.1D;
-        this.field_70233_c = p_i1760_10_ / d0 * 0.1D;
-        this.field_70230_d = p_i1760_12_ / d0 * 0.1D;
+        super(worldIn);
+        this.setSize(1.0F, 1.0F);
+        this.setLocationAndAngles(x, y, z, this.rotationYaw, this.rotationPitch);
+        this.setPosition(x, y, z);
+        double d0 = (double)MathHelper.sqrt(accelX * accelX + accelY * accelY + accelZ * accelZ);
+        this.accelerationX = accelX / d0 * 0.1D;
+        this.accelerationY = accelY / d0 * 0.1D;
+        this.accelerationZ = accelZ / d0 * 0.1D;
+        this.setDirection(accelX, accelY, accelZ);
     }
 
-    public EntityFireball(World p_i1761_1_, EntityLivingBase p_i1761_2_, double p_i1761_3_, double p_i1761_5_, double p_i1761_7_)
+    public EntityFireball(World worldIn, EntityLivingBase shooter, double accelX, double accelY, double accelZ)
     {
-        super(p_i1761_1_);
-        this.field_70235_a = p_i1761_2_;
-        this.func_70105_a(1.0F, 1.0F);
-        this.func_70012_b(p_i1761_2_.field_70165_t, p_i1761_2_.field_70163_u, p_i1761_2_.field_70161_v, p_i1761_2_.field_70177_z, p_i1761_2_.field_70125_A);
-        this.func_70107_b(this.field_70165_t, this.field_70163_u, this.field_70161_v);
-        this.field_70159_w = 0.0D;
-        this.field_70181_x = 0.0D;
-        this.field_70179_y = 0.0D;
-        p_i1761_3_ = p_i1761_3_ + this.field_70146_Z.nextGaussian() * 0.4D;
-        p_i1761_5_ = p_i1761_5_ + this.field_70146_Z.nextGaussian() * 0.4D;
-        p_i1761_7_ = p_i1761_7_ + this.field_70146_Z.nextGaussian() * 0.4D;
-        double d0 = (double)MathHelper.func_76133_a(p_i1761_3_ * p_i1761_3_ + p_i1761_5_ * p_i1761_5_ + p_i1761_7_ * p_i1761_7_);
-        this.field_70232_b = p_i1761_3_ / d0 * 0.1D;
-        this.field_70233_c = p_i1761_5_ / d0 * 0.1D;
-        this.field_70230_d = p_i1761_7_ / d0 * 0.1D;
+        super(worldIn);
+        this.shootingEntity = shooter;
+        this.setSize(1.0F, 1.0F);
+        this.setLocationAndAngles(shooter.posX, shooter.posY, shooter.posZ, shooter.rotationYaw, shooter.rotationPitch);
+        this.setPosition(this.posX, this.posY, this.posZ);
+        this.motionX = 0.0D;
+        this.motionY = 0.0D;
+        this.motionZ = 0.0D;
+        accelX = accelX + this.rand.nextGaussian() * 0.4D;
+        accelY = accelY + this.rand.nextGaussian() * 0.4D;
+        accelZ = accelZ + this.rand.nextGaussian() * 0.4D;
+        double d0 = (double)MathHelper.sqrt(accelX * accelX + accelY * accelY + accelZ * accelZ);
+        this.accelerationX = accelX / d0 * 0.1D;
+        this.accelerationY = accelY / d0 * 0.1D;
+        this.accelerationZ = accelZ / d0 * 0.1D;
+        this.projectileSource = (ProjectileSource) shooter.getBukkitEntity(); // CraftBukkit
+
     }
+    public void setDirection(double d0, double d1, double d2) {
+        // CraftBukkit end
+        d0 += this.rand.nextGaussian() * 0.4D;
+        d1 += this.rand.nextGaussian() * 0.4D;
+        d2 += this.rand.nextGaussian() * 0.4D;
+        double d3 = (double) MathHelper.sqrt(d0 * d0 + d1 * d1 + d2 * d2);
 
-    public void func_70071_h_()
+        this.accelerationX = d0 / d3 * 0.1D;
+        this.accelerationY = d1 / d3 * 0.1D;
+        this.accelerationZ = d2 / d3 * 0.1D;
+    }
+
+    public void onUpdate()
     {
-        if (this.field_70170_p.field_72995_K || (this.field_70235_a == null || !this.field_70235_a.field_70128_L) && this.field_70170_p.func_175667_e(new BlockPos(this)))
+        if (this.world.isRemote || (this.shootingEntity == null || !this.shootingEntity.isDead) && this.world.isBlockLoaded(new BlockPos(this)))
         {
-            super.func_70071_h_();
+            super.onUpdate();
 
-            if (this.func_184564_k())
+            if (this.isFireballFiery())
             {
-                this.func_70015_d(1);
+                this.setFire(1);
             }
 
-            ++this.field_70234_an;
-            RayTraceResult raytraceresult = ProjectileHelper.func_188802_a(this, true, this.field_70234_an >= 25, this.field_70235_a);
+            ++this.ticksInAir;
+            RayTraceResult raytraceresult = ProjectileHelper.forwardsRaycast(this, true, this.ticksInAir >= 25, this.shootingEntity);
 
             if (raytraceresult != null)
             {
-                this.func_70227_a(raytraceresult);
+                this.onImpact(raytraceresult);
+                // CraftBukkit start - Fire ProjectileHitEvent
+                if (this.isDead) {
+                    CraftEventFactory.callProjectileHitEvent(this, raytraceresult);
+                }
+                // CraftBukkit end
             }
 
-            this.field_70165_t += this.field_70159_w;
-            this.field_70163_u += this.field_70181_x;
-            this.field_70161_v += this.field_70179_y;
-            ProjectileHelper.func_188803_a(this, 0.2F);
-            float f = this.func_82341_c();
+            this.posX += this.motionX;
+            this.posY += this.motionY;
+            this.posZ += this.motionZ;
+            ProjectileHelper.rotateTowardsMovement(this, 0.2F);
+            float f = this.getMotionFactor();
 
-            if (this.func_70090_H())
+            if (this.isInWater())
             {
                 for (int i = 0; i < 4; ++i)
                 {
                     float f1 = 0.25F;
-                    this.field_70170_p.func_175688_a(EnumParticleTypes.WATER_BUBBLE, this.field_70165_t - this.field_70159_w * 0.25D, this.field_70163_u - this.field_70181_x * 0.25D, this.field_70161_v - this.field_70179_y * 0.25D, this.field_70159_w, this.field_70181_x, this.field_70179_y, new int[0]);
+                    this.world.spawnParticle(EnumParticleTypes.WATER_BUBBLE, this.posX - this.motionX * 0.25D, this.posY - this.motionY * 0.25D, this.posZ - this.motionZ * 0.25D, this.motionX, this.motionY, this.motionZ, new int[0]);
                 }
 
                 f = 0.8F;
             }
 
-            this.field_70159_w += this.field_70232_b;
-            this.field_70181_x += this.field_70233_c;
-            this.field_70179_y += this.field_70230_d;
-            this.field_70159_w *= (double)f;
-            this.field_70181_x *= (double)f;
-            this.field_70179_y *= (double)f;
-            this.field_70170_p.func_175688_a(this.func_184563_j(), this.field_70165_t, this.field_70163_u + 0.5D, this.field_70161_v, 0.0D, 0.0D, 0.0D, new int[0]);
-            this.func_70107_b(this.field_70165_t, this.field_70163_u, this.field_70161_v);
+            this.motionX += this.accelerationX;
+            this.motionY += this.accelerationY;
+            this.motionZ += this.accelerationZ;
+            this.motionX *= (double)f;
+            this.motionY *= (double)f;
+            this.motionZ *= (double)f;
+            this.world.spawnParticle(this.getParticleType(), this.posX, this.posY + 0.5D, this.posZ, 0.0D, 0.0D, 0.0D, new int[0]);
+            this.setPosition(this.posX, this.posY, this.posZ);
         }
         else
         {
-            this.func_70106_y();
+            this.setDead();
         }
     }
 
-    protected boolean func_184564_k()
+    protected boolean isFireballFiery()
     {
         return true;
     }
 
-    protected EnumParticleTypes func_184563_j()
+    protected EnumParticleTypes getParticleType()
     {
         return EnumParticleTypes.SMOKE_NORMAL;
     }
 
-    protected float func_82341_c()
+    protected float getMotionFactor()
     {
         return 0.95F;
     }
 
-    protected abstract void func_70227_a(RayTraceResult p_70227_1_);
+    protected abstract void onImpact(RayTraceResult result);
 
-    public static void func_189743_a(DataFixer p_189743_0_, String p_189743_1_)
+    public static void registerFixesFireball(DataFixer fixer, String name)
     {
     }
 
-    public void func_70014_b(NBTTagCompound p_70014_1_)
+    public void writeEntityToNBT(NBTTagCompound compound)
     {
-        p_70014_1_.func_74782_a("direction", this.func_70087_a(new double[] {this.field_70159_w, this.field_70181_x, this.field_70179_y}));
-        p_70014_1_.func_74782_a("power", this.func_70087_a(new double[] {this.field_70232_b, this.field_70233_c, this.field_70230_d}));
-        p_70014_1_.func_74768_a("life", this.field_70236_j);
+        compound.setTag("direction", this.newDoubleNBTList(new double[] {this.motionX, this.motionY, this.motionZ}));
+        compound.setTag("power", this.newDoubleNBTList(new double[] {this.accelerationX, this.accelerationY, this.accelerationZ}));
+        compound.setInteger("life", this.ticksAlive);
     }
 
-    public void func_70037_a(NBTTagCompound p_70037_1_)
+    public void readEntityFromNBT(NBTTagCompound compound)
     {
-        if (p_70037_1_.func_150297_b("power", 9))
+        if (compound.hasKey("power", 9))
         {
-            NBTTagList nbttaglist = p_70037_1_.func_150295_c("power", 6);
+            NBTTagList nbttaglist = compound.getTagList("power", 6);
 
-            if (nbttaglist.func_74745_c() == 3)
+            if (nbttaglist.tagCount() == 3)
             {
-                this.field_70232_b = nbttaglist.func_150309_d(0);
-                this.field_70233_c = nbttaglist.func_150309_d(1);
-                this.field_70230_d = nbttaglist.func_150309_d(2);
+                this.accelerationX = nbttaglist.getDoubleAt(0);
+                this.accelerationY = nbttaglist.getDoubleAt(1);
+                this.accelerationZ = nbttaglist.getDoubleAt(2);
             }
         }
 
-        this.field_70236_j = p_70037_1_.func_74762_e("life");
+        this.ticksAlive = compound.getInteger("life");
 
-        if (p_70037_1_.func_150297_b("direction", 9) && p_70037_1_.func_150295_c("direction", 6).func_74745_c() == 3)
+        if (compound.hasKey("direction", 9) && compound.getTagList("direction", 6).tagCount() == 3)
         {
-            NBTTagList nbttaglist1 = p_70037_1_.func_150295_c("direction", 6);
-            this.field_70159_w = nbttaglist1.func_150309_d(0);
-            this.field_70181_x = nbttaglist1.func_150309_d(1);
-            this.field_70179_y = nbttaglist1.func_150309_d(2);
+            NBTTagList nbttaglist1 = compound.getTagList("direction", 6);
+            this.motionX = nbttaglist1.getDoubleAt(0);
+            this.motionY = nbttaglist1.getDoubleAt(1);
+            this.motionZ = nbttaglist1.getDoubleAt(2);
         }
         else
         {
-            this.func_70106_y();
+            this.setDead();
         }
     }
 
-    public boolean func_70067_L()
+    public boolean canBeCollidedWith()
     {
         return true;
     }
 
-    public float func_70111_Y()
+    public float getCollisionBorderSize()
     {
         return 1.0F;
     }
 
-    public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
+    public boolean attackEntityFrom(DamageSource source, float amount)
     {
-        if (this.func_180431_b(p_70097_1_))
+        if (this.isEntityInvulnerable(source))
         {
             return false;
         }
         else
         {
-            this.func_70018_K();
+            this.setBeenAttacked();
 
-            if (p_70097_1_.func_76346_g() != null)
+            if (source.getEntity() != null)
             {
-                Vec3d vec3d = p_70097_1_.func_76346_g().func_70040_Z();
+                // CraftBukkit start
+                if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, amount)) {
+                    return false;
+                }
+                // CraftBukkit end
+                Vec3d vec3d = source.getEntity().getLookVec();
 
                 if (vec3d != null)
                 {
-                    this.field_70159_w = vec3d.field_72450_a;
-                    this.field_70181_x = vec3d.field_72448_b;
-                    this.field_70179_y = vec3d.field_72449_c;
-                    this.field_70232_b = this.field_70159_w * 0.1D;
-                    this.field_70233_c = this.field_70181_x * 0.1D;
-                    this.field_70230_d = this.field_70179_y * 0.1D;
+                    this.motionX = vec3d.xCoord;
+                    this.motionY = vec3d.yCoord;
+                    this.motionZ = vec3d.zCoord;
+                    this.accelerationX = this.motionX * 0.1D;
+                    this.accelerationY = this.motionY * 0.1D;
+                    this.accelerationZ = this.motionZ * 0.1D;
                 }
 
-                if (p_70097_1_.func_76346_g() instanceof EntityLivingBase)
+                if (source.getEntity() instanceof EntityLivingBase)
                 {
-                    this.field_70235_a = (EntityLivingBase)p_70097_1_.func_76346_g();
+                    this.shootingEntity = (EntityLivingBase)source.getEntity();
+                    this.projectileSource = (org.bukkit.projectiles.ProjectileSource) this.shootingEntity.getBukkitEntity();
+
                 }
 
                 return true;
@@ -235,13 +265,13 @@
         }
     }
 
-    public float func_70013_c(float p_70013_1_)
+    public float getBrightness(float partialTicks)
     {
         return 1.0F;
     }
 
     @SideOnly(Side.CLIENT)
-    public int func_70070_b(float p_70070_1_)
+    public int getBrightnessForRender(float partialTicks)
     {
         return 15728880;
     }
