--- ../src-base/minecraft/net/minecraft/entity/projectile/EntityShulkerBullet.java
+++ ../src-work/minecraft/net/minecraft/entity/projectile/EntityShulkerBullet.java
@@ -28,158 +28,179 @@
 
 public class EntityShulkerBullet extends Entity
 {
-    private EntityLivingBase field_184570_a;
-    private Entity field_184571_b;
+    private EntityLivingBase owner;
+    private Entity target;
     @Nullable
-    private EnumFacing field_184573_c;
-    private int field_184575_d;
-    private double field_184577_e;
-    private double field_184578_f;
-    private double field_184579_g;
+    private EnumFacing direction;
+    private int steps;
+    private double targetDeltaX;
+    private double targetDeltaY;
+    private double targetDeltaZ;
     @Nullable
-    private UUID field_184580_h;
-    private BlockPos field_184572_as;
+    private UUID ownerUniqueId;
+    private BlockPos ownerBlockPos;
     @Nullable
-    private UUID field_184574_at;
-    private BlockPos field_184576_au;
+    private UUID targetUniqueId;
+    private BlockPos targetBlockPos;
 
-    public EntityShulkerBullet(World p_i46770_1_)
+    public EntityShulkerBullet(World worldIn)
     {
-        super(p_i46770_1_);
-        this.func_70105_a(0.3125F, 0.3125F);
-        this.field_70145_X = true;
+        super(worldIn);
+        this.setSize(0.3125F, 0.3125F);
+        this.noClip = true;
     }
 
-    public SoundCategory func_184176_by()
+    public SoundCategory getSoundCategory()
     {
         return SoundCategory.HOSTILE;
     }
 
     @SideOnly(Side.CLIENT)
-    public EntityShulkerBullet(World p_i46771_1_, double p_i46771_2_, double p_i46771_4_, double p_i46771_6_, double p_i46771_8_, double p_i46771_10_, double p_i46771_12_)
+    public EntityShulkerBullet(World worldIn, double x, double y, double z, double motionXIn, double motionYIn, double motionZIn)
     {
-        this(p_i46771_1_);
-        this.func_70012_b(p_i46771_2_, p_i46771_4_, p_i46771_6_, this.field_70177_z, this.field_70125_A);
-        this.field_70159_w = p_i46771_8_;
-        this.field_70181_x = p_i46771_10_;
-        this.field_70179_y = p_i46771_12_;
+        this(worldIn);
+        this.setLocationAndAngles(x, y, z, this.rotationYaw, this.rotationPitch);
+        this.motionX = motionXIn;
+        this.motionY = motionYIn;
+        this.motionZ = motionZIn;
     }
 
-    public EntityShulkerBullet(World p_i46772_1_, EntityLivingBase p_i46772_2_, Entity p_i46772_3_, EnumFacing.Axis p_i46772_4_)
+    public EntityShulkerBullet(World worldIn, EntityLivingBase ownerIn, Entity targetIn, EnumFacing.Axis p_i46772_4_)
     {
-        this(p_i46772_1_);
-        this.field_184570_a = p_i46772_2_;
-        BlockPos blockpos = new BlockPos(p_i46772_2_);
-        double d0 = (double)blockpos.func_177958_n() + 0.5D;
-        double d1 = (double)blockpos.func_177956_o() + 0.5D;
-        double d2 = (double)blockpos.func_177952_p() + 0.5D;
-        this.func_70012_b(d0, d1, d2, this.field_70177_z, this.field_70125_A);
-        this.field_184571_b = p_i46772_3_;
-        this.field_184573_c = EnumFacing.UP;
-        this.func_184569_a(p_i46772_4_);
+        this(worldIn);
+        this.owner = ownerIn;
+        BlockPos blockpos = new BlockPos(ownerIn);
+        double d0 = (double)blockpos.getX() + 0.5D;
+        double d1 = (double)blockpos.getY() + 0.5D;
+        double d2 = (double)blockpos.getZ() + 0.5D;
+        this.setLocationAndAngles(d0, d1, d2, this.rotationYaw, this.rotationPitch);
+        this.target = targetIn;
+        this.direction = EnumFacing.UP;
+        this.selectNextMoveDirection(p_i46772_4_);
+        projectileSource = (org.bukkit.entity.LivingEntity) ownerIn.getBukkitEntity(); // CraftBukkit
+
     }
+    // CraftBukkit start
+    public EntityLivingBase getShooter() {
+        return this.owner;
+    }
 
-    protected void func_70014_b(NBTTagCompound p_70014_1_)
+    public void setShooter(EntityLivingBase e) {
+        this.owner = e;
+    }
+
+    public Entity getTarget() {
+        return this.target;
+    }
+
+    public void setTarget(Entity e) {
+        this.target = e;
+        this.direction = EnumFacing.UP;
+        this.selectNextMoveDirection(EnumFacing.Axis.X);
+    }
+    // CraftBukkit end
+
+    protected void writeEntityToNBT(NBTTagCompound compound)
     {
-        if (this.field_184570_a != null)
+        if (this.owner != null)
         {
-            BlockPos blockpos = new BlockPos(this.field_184570_a);
-            NBTTagCompound nbttagcompound = NBTUtil.func_186862_a(this.field_184570_a.func_110124_au());
-            nbttagcompound.func_74768_a("X", blockpos.func_177958_n());
-            nbttagcompound.func_74768_a("Y", blockpos.func_177956_o());
-            nbttagcompound.func_74768_a("Z", blockpos.func_177952_p());
-            p_70014_1_.func_74782_a("Owner", nbttagcompound);
+            BlockPos blockpos = new BlockPos(this.owner);
+            NBTTagCompound nbttagcompound = NBTUtil.createUUIDTag(this.owner.getUniqueID());
+            nbttagcompound.setInteger("X", blockpos.getX());
+            nbttagcompound.setInteger("Y", blockpos.getY());
+            nbttagcompound.setInteger("Z", blockpos.getZ());
+            compound.setTag("Owner", nbttagcompound);
         }
 
-        if (this.field_184571_b != null)
+        if (this.target != null)
         {
-            BlockPos blockpos1 = new BlockPos(this.field_184571_b);
-            NBTTagCompound nbttagcompound1 = NBTUtil.func_186862_a(this.field_184571_b.func_110124_au());
-            nbttagcompound1.func_74768_a("X", blockpos1.func_177958_n());
-            nbttagcompound1.func_74768_a("Y", blockpos1.func_177956_o());
-            nbttagcompound1.func_74768_a("Z", blockpos1.func_177952_p());
-            p_70014_1_.func_74782_a("Target", nbttagcompound1);
+            BlockPos blockpos1 = new BlockPos(this.target);
+            NBTTagCompound nbttagcompound1 = NBTUtil.createUUIDTag(this.target.getUniqueID());
+            nbttagcompound1.setInteger("X", blockpos1.getX());
+            nbttagcompound1.setInteger("Y", blockpos1.getY());
+            nbttagcompound1.setInteger("Z", blockpos1.getZ());
+            compound.setTag("Target", nbttagcompound1);
         }
 
-        if (this.field_184573_c != null)
+        if (this.direction != null)
         {
-            p_70014_1_.func_74768_a("Dir", this.field_184573_c.func_176745_a());
+            compound.setInteger("Dir", this.direction.getIndex());
         }
 
-        p_70014_1_.func_74768_a("Steps", this.field_184575_d);
-        p_70014_1_.func_74780_a("TXD", this.field_184577_e);
-        p_70014_1_.func_74780_a("TYD", this.field_184578_f);
-        p_70014_1_.func_74780_a("TZD", this.field_184579_g);
+        compound.setInteger("Steps", this.steps);
+        compound.setDouble("TXD", this.targetDeltaX);
+        compound.setDouble("TYD", this.targetDeltaY);
+        compound.setDouble("TZD", this.targetDeltaZ);
     }
 
-    protected void func_70037_a(NBTTagCompound p_70037_1_)
+    protected void readEntityFromNBT(NBTTagCompound compound)
     {
-        this.field_184575_d = p_70037_1_.func_74762_e("Steps");
-        this.field_184577_e = p_70037_1_.func_74769_h("TXD");
-        this.field_184578_f = p_70037_1_.func_74769_h("TYD");
-        this.field_184579_g = p_70037_1_.func_74769_h("TZD");
+        this.steps = compound.getInteger("Steps");
+        this.targetDeltaX = compound.getDouble("TXD");
+        this.targetDeltaY = compound.getDouble("TYD");
+        this.targetDeltaZ = compound.getDouble("TZD");
 
-        if (p_70037_1_.func_150297_b("Dir", 99))
+        if (compound.hasKey("Dir", 99))
         {
-            this.field_184573_c = EnumFacing.func_82600_a(p_70037_1_.func_74762_e("Dir"));
+            this.direction = EnumFacing.getFront(compound.getInteger("Dir"));
         }
 
-        if (p_70037_1_.func_150297_b("Owner", 10))
+        if (compound.hasKey("Owner", 10))
         {
-            NBTTagCompound nbttagcompound = p_70037_1_.func_74775_l("Owner");
-            this.field_184580_h = NBTUtil.func_186860_b(nbttagcompound);
-            this.field_184572_as = new BlockPos(nbttagcompound.func_74762_e("X"), nbttagcompound.func_74762_e("Y"), nbttagcompound.func_74762_e("Z"));
+            NBTTagCompound nbttagcompound = compound.getCompoundTag("Owner");
+            this.ownerUniqueId = NBTUtil.getUUIDFromTag(nbttagcompound);
+            this.ownerBlockPos = new BlockPos(nbttagcompound.getInteger("X"), nbttagcompound.getInteger("Y"), nbttagcompound.getInteger("Z"));
         }
 
-        if (p_70037_1_.func_150297_b("Target", 10))
+        if (compound.hasKey("Target", 10))
         {
-            NBTTagCompound nbttagcompound1 = p_70037_1_.func_74775_l("Target");
-            this.field_184574_at = NBTUtil.func_186860_b(nbttagcompound1);
-            this.field_184576_au = new BlockPos(nbttagcompound1.func_74762_e("X"), nbttagcompound1.func_74762_e("Y"), nbttagcompound1.func_74762_e("Z"));
+            NBTTagCompound nbttagcompound1 = compound.getCompoundTag("Target");
+            this.targetUniqueId = NBTUtil.getUUIDFromTag(nbttagcompound1);
+            this.targetBlockPos = new BlockPos(nbttagcompound1.getInteger("X"), nbttagcompound1.getInteger("Y"), nbttagcompound1.getInteger("Z"));
         }
     }
 
-    protected void func_70088_a()
+    protected void entityInit()
     {
     }
 
-    private void func_184568_a(@Nullable EnumFacing p_184568_1_)
+    private void setDirection(@Nullable EnumFacing directionIn)
     {
-        this.field_184573_c = p_184568_1_;
+        this.direction = directionIn;
     }
 
-    private void func_184569_a(@Nullable EnumFacing.Axis p_184569_1_)
+    private void selectNextMoveDirection(@Nullable EnumFacing.Axis p_184569_1_)
     {
         double d0 = 0.5D;
         BlockPos blockpos;
 
-        if (this.field_184571_b == null)
+        if (this.target == null)
         {
-            blockpos = (new BlockPos(this)).func_177977_b();
+            blockpos = (new BlockPos(this)).down();
         }
         else
         {
-            d0 = (double)this.field_184571_b.field_70131_O * 0.5D;
-            blockpos = new BlockPos(this.field_184571_b.field_70165_t, this.field_184571_b.field_70163_u + d0, this.field_184571_b.field_70161_v);
+            d0 = (double)this.target.height * 0.5D;
+            blockpos = new BlockPos(this.target.posX, this.target.posY + d0, this.target.posZ);
         }
 
-        double d1 = (double)blockpos.func_177958_n() + 0.5D;
-        double d2 = (double)blockpos.func_177956_o() + d0;
-        double d3 = (double)blockpos.func_177952_p() + 0.5D;
+        double d1 = (double)blockpos.getX() + 0.5D;
+        double d2 = (double)blockpos.getY() + d0;
+        double d3 = (double)blockpos.getZ() + 0.5D;
         EnumFacing enumfacing = null;
 
-        if (blockpos.func_177957_d(this.field_70165_t, this.field_70163_u, this.field_70161_v) >= 4.0D)
+        if (blockpos.distanceSqToCenter(this.posX, this.posY, this.posZ) >= 4.0D)
         {
             BlockPos blockpos1 = new BlockPos(this);
             List<EnumFacing> list = Lists.<EnumFacing>newArrayList();
 
             if (p_184569_1_ != EnumFacing.Axis.X)
             {
-                if (blockpos1.func_177958_n() < blockpos.func_177958_n() && this.field_70170_p.func_175623_d(blockpos1.func_177974_f()))
+                if (blockpos1.getX() < blockpos.getX() && this.world.isAirBlock(blockpos1.east()))
                 {
                     list.add(EnumFacing.EAST);
                 }
-                else if (blockpos1.func_177958_n() > blockpos.func_177958_n() && this.field_70170_p.func_175623_d(blockpos1.func_177976_e()))
+                else if (blockpos1.getX() > blockpos.getX() && this.world.isAirBlock(blockpos1.west()))
                 {
                     list.add(EnumFacing.WEST);
                 }
@@ -187,11 +208,11 @@
 
             if (p_184569_1_ != EnumFacing.Axis.Y)
             {
-                if (blockpos1.func_177956_o() < blockpos.func_177956_o() && this.field_70170_p.func_175623_d(blockpos1.func_177984_a()))
+                if (blockpos1.getY() < blockpos.getY() && this.world.isAirBlock(blockpos1.up()))
                 {
                     list.add(EnumFacing.UP);
                 }
-                else if (blockpos1.func_177956_o() > blockpos.func_177956_o() && this.field_70170_p.func_175623_d(blockpos1.func_177977_b()))
+                else if (blockpos1.getY() > blockpos.getY() && this.world.isAirBlock(blockpos1.down()))
                 {
                     list.add(EnumFacing.DOWN);
                 }
@@ -199,158 +220,158 @@
 
             if (p_184569_1_ != EnumFacing.Axis.Z)
             {
-                if (blockpos1.func_177952_p() < blockpos.func_177952_p() && this.field_70170_p.func_175623_d(blockpos1.func_177968_d()))
+                if (blockpos1.getZ() < blockpos.getZ() && this.world.isAirBlock(blockpos1.south()))
                 {
                     list.add(EnumFacing.SOUTH);
                 }
-                else if (blockpos1.func_177952_p() > blockpos.func_177952_p() && this.field_70170_p.func_175623_d(blockpos1.func_177978_c()))
+                else if (blockpos1.getZ() > blockpos.getZ() && this.world.isAirBlock(blockpos1.north()))
                 {
                     list.add(EnumFacing.NORTH);
                 }
             }
 
-            enumfacing = EnumFacing.func_176741_a(this.field_70146_Z);
+            enumfacing = EnumFacing.random(this.rand);
 
             if (list.isEmpty())
             {
-                for (int i = 5; !this.field_70170_p.func_175623_d(blockpos1.func_177972_a(enumfacing)) && i > 0; --i)
+                for (int i = 5; !this.world.isAirBlock(blockpos1.offset(enumfacing)) && i > 0; --i)
                 {
-                    enumfacing = EnumFacing.func_176741_a(this.field_70146_Z);
+                    enumfacing = EnumFacing.random(this.rand);
                 }
             }
             else
             {
-                enumfacing = (EnumFacing)list.get(this.field_70146_Z.nextInt(list.size()));
+                enumfacing = (EnumFacing)list.get(this.rand.nextInt(list.size()));
             }
 
-            d1 = this.field_70165_t + (double)enumfacing.func_82601_c();
-            d2 = this.field_70163_u + (double)enumfacing.func_96559_d();
-            d3 = this.field_70161_v + (double)enumfacing.func_82599_e();
+            d1 = this.posX + (double)enumfacing.getFrontOffsetX();
+            d2 = this.posY + (double)enumfacing.getFrontOffsetY();
+            d3 = this.posZ + (double)enumfacing.getFrontOffsetZ();
         }
 
-        this.func_184568_a(enumfacing);
-        double d6 = d1 - this.field_70165_t;
-        double d7 = d2 - this.field_70163_u;
-        double d4 = d3 - this.field_70161_v;
-        double d5 = (double)MathHelper.func_76133_a(d6 * d6 + d7 * d7 + d4 * d4);
+        this.setDirection(enumfacing);
+        double d6 = d1 - this.posX;
+        double d7 = d2 - this.posY;
+        double d4 = d3 - this.posZ;
+        double d5 = (double)MathHelper.sqrt(d6 * d6 + d7 * d7 + d4 * d4);
 
         if (d5 == 0.0D)
         {
-            this.field_184577_e = 0.0D;
-            this.field_184578_f = 0.0D;
-            this.field_184579_g = 0.0D;
+            this.targetDeltaX = 0.0D;
+            this.targetDeltaY = 0.0D;
+            this.targetDeltaZ = 0.0D;
         }
         else
         {
-            this.field_184577_e = d6 / d5 * 0.15D;
-            this.field_184578_f = d7 / d5 * 0.15D;
-            this.field_184579_g = d4 / d5 * 0.15D;
+            this.targetDeltaX = d6 / d5 * 0.15D;
+            this.targetDeltaY = d7 / d5 * 0.15D;
+            this.targetDeltaZ = d4 / d5 * 0.15D;
         }
 
-        this.field_70160_al = true;
-        this.field_184575_d = 10 + this.field_70146_Z.nextInt(5) * 10;
+        this.isAirBorne = true;
+        this.steps = 10 + this.rand.nextInt(5) * 10;
     }
 
-    public void func_70071_h_()
+    public void onUpdate()
     {
-        if (!this.field_70170_p.field_72995_K && this.field_70170_p.func_175659_aa() == EnumDifficulty.PEACEFUL)
+        if (!this.world.isRemote && this.world.getDifficulty() == EnumDifficulty.PEACEFUL)
         {
-            this.func_70106_y();
+            this.setDead();
         }
         else
         {
-            super.func_70071_h_();
+            super.onUpdate();
 
-            if (!this.field_70170_p.field_72995_K)
+            if (!this.world.isRemote)
             {
-                if (this.field_184571_b == null && this.field_184574_at != null)
+                if (this.target == null && this.targetUniqueId != null)
                 {
-                    for (EntityLivingBase entitylivingbase : this.field_70170_p.func_72872_a(EntityLivingBase.class, new AxisAlignedBB(this.field_184576_au.func_177982_a(-2, -2, -2), this.field_184576_au.func_177982_a(2, 2, 2))))
+                    for (EntityLivingBase entitylivingbase : this.world.getEntitiesWithinAABB(EntityLivingBase.class, new AxisAlignedBB(this.targetBlockPos.add(-2, -2, -2), this.targetBlockPos.add(2, 2, 2))))
                     {
-                        if (entitylivingbase.func_110124_au().equals(this.field_184574_at))
+                        if (entitylivingbase.getUniqueID().equals(this.targetUniqueId))
                         {
-                            this.field_184571_b = entitylivingbase;
+                            this.target = entitylivingbase;
                             break;
                         }
                     }
 
-                    this.field_184574_at = null;
+                    this.targetUniqueId = null;
                 }
 
-                if (this.field_184570_a == null && this.field_184580_h != null)
+                if (this.owner == null && this.ownerUniqueId != null)
                 {
-                    for (EntityLivingBase entitylivingbase1 : this.field_70170_p.func_72872_a(EntityLivingBase.class, new AxisAlignedBB(this.field_184572_as.func_177982_a(-2, -2, -2), this.field_184572_as.func_177982_a(2, 2, 2))))
+                    for (EntityLivingBase entitylivingbase1 : this.world.getEntitiesWithinAABB(EntityLivingBase.class, new AxisAlignedBB(this.ownerBlockPos.add(-2, -2, -2), this.ownerBlockPos.add(2, 2, 2))))
                     {
-                        if (entitylivingbase1.func_110124_au().equals(this.field_184580_h))
+                        if (entitylivingbase1.getUniqueID().equals(this.ownerUniqueId))
                         {
-                            this.field_184570_a = entitylivingbase1;
+                            this.owner = entitylivingbase1;
                             break;
                         }
                     }
 
-                    this.field_184580_h = null;
+                    this.ownerUniqueId = null;
                 }
 
-                if (this.field_184571_b == null || !this.field_184571_b.func_70089_S() || this.field_184571_b instanceof EntityPlayer && ((EntityPlayer)this.field_184571_b).func_175149_v())
+                if (this.target == null || !this.target.isEntityAlive() || this.target instanceof EntityPlayer && ((EntityPlayer)this.target).isSpectator())
                 {
-                    if (!this.func_189652_ae())
+                    if (!this.hasNoGravity())
                     {
-                        this.field_70181_x -= 0.04D;
+                        this.motionY -= 0.04D;
                     }
                 }
                 else
                 {
-                    this.field_184577_e = MathHelper.func_151237_a(this.field_184577_e * 1.025D, -1.0D, 1.0D);
-                    this.field_184578_f = MathHelper.func_151237_a(this.field_184578_f * 1.025D, -1.0D, 1.0D);
-                    this.field_184579_g = MathHelper.func_151237_a(this.field_184579_g * 1.025D, -1.0D, 1.0D);
-                    this.field_70159_w += (this.field_184577_e - this.field_70159_w) * 0.2D;
-                    this.field_70181_x += (this.field_184578_f - this.field_70181_x) * 0.2D;
-                    this.field_70179_y += (this.field_184579_g - this.field_70179_y) * 0.2D;
+                    this.targetDeltaX = MathHelper.clamp(this.targetDeltaX * 1.025D, -1.0D, 1.0D);
+                    this.targetDeltaY = MathHelper.clamp(this.targetDeltaY * 1.025D, -1.0D, 1.0D);
+                    this.targetDeltaZ = MathHelper.clamp(this.targetDeltaZ * 1.025D, -1.0D, 1.0D);
+                    this.motionX += (this.targetDeltaX - this.motionX) * 0.2D;
+                    this.motionY += (this.targetDeltaY - this.motionY) * 0.2D;
+                    this.motionZ += (this.targetDeltaZ - this.motionZ) * 0.2D;
                 }
 
-                RayTraceResult raytraceresult = ProjectileHelper.func_188802_a(this, true, false, this.field_184570_a);
+                RayTraceResult raytraceresult = ProjectileHelper.forwardsRaycast(this, true, false, this.owner);
 
                 if (raytraceresult != null)
                 {
-                    this.func_184567_a(raytraceresult);
+                    this.bulletHit(raytraceresult);
                 }
             }
 
-            this.func_70107_b(this.field_70165_t + this.field_70159_w, this.field_70163_u + this.field_70181_x, this.field_70161_v + this.field_70179_y);
-            ProjectileHelper.func_188803_a(this, 0.5F);
+            this.setPosition(this.posX + this.motionX, this.posY + this.motionY, this.posZ + this.motionZ);
+            ProjectileHelper.rotateTowardsMovement(this, 0.5F);
 
-            if (this.field_70170_p.field_72995_K)
+            if (this.world.isRemote)
             {
-                this.field_70170_p.func_175688_a(EnumParticleTypes.END_ROD, this.field_70165_t - this.field_70159_w, this.field_70163_u - this.field_70181_x + 0.15D, this.field_70161_v - this.field_70179_y, 0.0D, 0.0D, 0.0D, new int[0]);
+                this.world.spawnParticle(EnumParticleTypes.END_ROD, this.posX - this.motionX, this.posY - this.motionY + 0.15D, this.posZ - this.motionZ, 0.0D, 0.0D, 0.0D, new int[0]);
             }
-            else if (this.field_184571_b != null && !this.field_184571_b.field_70128_L)
+            else if (this.target != null && !this.target.isDead)
             {
-                if (this.field_184575_d > 0)
+                if (this.steps > 0)
                 {
-                    --this.field_184575_d;
+                    --this.steps;
 
-                    if (this.field_184575_d == 0)
+                    if (this.steps == 0)
                     {
-                        this.func_184569_a(this.field_184573_c == null ? null : this.field_184573_c.func_176740_k());
+                        this.selectNextMoveDirection(this.direction == null ? null : this.direction.getAxis());
                     }
                 }
 
-                if (this.field_184573_c != null)
+                if (this.direction != null)
                 {
                     BlockPos blockpos = new BlockPos(this);
-                    EnumFacing.Axis enumfacing$axis = this.field_184573_c.func_176740_k();
+                    EnumFacing.Axis enumfacing$axis = this.direction.getAxis();
 
-                    if (this.field_70170_p.func_175677_d(blockpos.func_177972_a(this.field_184573_c), false))
+                    if (this.world.isBlockNormalCube(blockpos.offset(this.direction), false))
                     {
-                        this.func_184569_a(enumfacing$axis);
+                        this.selectNextMoveDirection(enumfacing$axis);
                     }
                     else
                     {
-                        BlockPos blockpos1 = new BlockPos(this.field_184571_b);
+                        BlockPos blockpos1 = new BlockPos(this.target);
 
-                        if (enumfacing$axis == EnumFacing.Axis.X && blockpos.func_177958_n() == blockpos1.func_177958_n() || enumfacing$axis == EnumFacing.Axis.Z && blockpos.func_177952_p() == blockpos1.func_177952_p() || enumfacing$axis == EnumFacing.Axis.Y && blockpos.func_177956_o() == blockpos1.func_177956_o())
+                        if (enumfacing$axis == EnumFacing.Axis.X && blockpos.getX() == blockpos1.getX() || enumfacing$axis == EnumFacing.Axis.Z && blockpos.getZ() == blockpos1.getZ() || enumfacing$axis == EnumFacing.Axis.Y && blockpos.getY() == blockpos1.getY())
                         {
-                            this.func_184569_a(enumfacing$axis);
+                            this.selectNextMoveDirection(enumfacing$axis);
                         }
                     }
                 }
@@ -358,65 +379,65 @@
         }
     }
 
-    public boolean func_70027_ad()
+    public boolean isBurning()
     {
         return false;
     }
 
     @SideOnly(Side.CLIENT)
-    public boolean func_70112_a(double p_70112_1_)
+    public boolean isInRangeToRenderDist(double distance)
     {
-        return p_70112_1_ < 16384.0D;
+        return distance < 16384.0D;
     }
 
-    public float func_70013_c(float p_70013_1_)
+    public float getBrightness(float partialTicks)
     {
         return 1.0F;
     }
 
     @SideOnly(Side.CLIENT)
-    public int func_70070_b(float p_70070_1_)
+    public int getBrightnessForRender(float partialTicks)
     {
         return 15728880;
     }
 
-    protected void func_184567_a(RayTraceResult p_184567_1_)
+    protected void bulletHit(RayTraceResult result)
     {
-        if (p_184567_1_.field_72308_g == null)
+        if (result.entityHit == null)
         {
-            ((WorldServer)this.field_70170_p).func_175739_a(EnumParticleTypes.EXPLOSION_LARGE, this.field_70165_t, this.field_70163_u, this.field_70161_v, 2, 0.2D, 0.2D, 0.2D, 0.0D, new int[0]);
-            this.func_184185_a(SoundEvents.field_187775_eP, 1.0F, 1.0F);
+            ((WorldServer)this.world).spawnParticle(EnumParticleTypes.EXPLOSION_LARGE, this.posX, this.posY, this.posZ, 2, 0.2D, 0.2D, 0.2D, 0.0D, new int[0]);
+            this.playSound(SoundEvents.ENTITY_SHULKER_BULLET_HIT, 1.0F, 1.0F);
         }
         else
         {
-            boolean flag = p_184567_1_.field_72308_g.func_70097_a(DamageSource.func_188403_a(this, this.field_184570_a).func_76349_b(), 4.0F);
+            boolean flag = result.entityHit.attackEntityFrom(DamageSource.causeIndirectDamage(this, this.owner).setProjectile(), 4.0F);
 
             if (flag)
             {
-                this.func_174815_a(this.field_184570_a, p_184567_1_.field_72308_g);
+                this.applyEnchantments(this.owner, result.entityHit);
 
-                if (p_184567_1_.field_72308_g instanceof EntityLivingBase)
+                if (result.entityHit instanceof EntityLivingBase)
                 {
-                    ((EntityLivingBase)p_184567_1_.field_72308_g).func_70690_d(new PotionEffect(MobEffects.field_188424_y, 200));
+                    ((EntityLivingBase)result.entityHit).addPotionEffect(new PotionEffect(MobEffects.LEVITATION, 200));
                 }
             }
         }
 
-        this.func_70106_y();
+        this.setDead();
     }
 
-    public boolean func_70067_L()
+    public boolean canBeCollidedWith()
     {
         return true;
     }
 
-    public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
+    public boolean attackEntityFrom(DamageSource source, float amount)
     {
-        if (!this.field_70170_p.field_72995_K)
+        if (!this.world.isRemote)
         {
-            this.func_184185_a(SoundEvents.field_187777_eQ, 1.0F, 1.0F);
-            ((WorldServer)this.field_70170_p).func_175739_a(EnumParticleTypes.CRIT, this.field_70165_t, this.field_70163_u, this.field_70161_v, 15, 0.2D, 0.2D, 0.2D, 0.0D, new int[0]);
-            this.func_70106_y();
+            this.playSound(SoundEvents.ENTITY_SHULKER_BULLET_HURT, 1.0F, 1.0F);
+            ((WorldServer)this.world).spawnParticle(EnumParticleTypes.CRIT, this.posX, this.posY, this.posZ, 15, 0.2D, 0.2D, 0.2D, 0.0D, new int[0]);
+            this.setDead();
         }
 
         return true;
