--- ../src-base/minecraft/net/minecraft/entity/projectile/EntityArrow.java
+++ ../src-work/minecraft/net/minecraft/entity/projectile/EntityArrow.java
@@ -12,6 +12,7 @@
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.IProjectile;
 import net.minecraft.entity.MoverType;
+import net.minecraft.entity.item.EntityItem;
 import net.minecraft.entity.monster.EntityEnderman;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.player.EntityPlayerMP;
@@ -36,6 +37,13 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.entity.CraftItem;
+import org.bukkit.entity.Arrow;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.entity.Player;
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
+import org.bukkit.event.entity.EntityCombustEvent;
+import org.bukkit.event.player.PlayerPickupArrowEvent;
 
 public abstract class EntityArrow extends Entity implements IProjectile
 {
@@ -84,7 +92,7 @@
     {
         this(p_i46777_1_, p_i46777_2_.field_70165_t, p_i46777_2_.field_70163_u + (double)p_i46777_2_.func_70047_e() - 0.10000000149011612D, p_i46777_2_.field_70161_v);
         this.field_70250_c = p_i46777_2_;
-
+        this.projectileSource = (LivingEntity) p_i46777_2_.getBukkitEntity(); // CraftBukkit
         if (p_i46777_2_ instanceof EntityPlayer)
         {
             this.field_70251_a = EntityArrow.PickupStatus.ALLOWED;
@@ -364,7 +372,13 @@
 
             if (this.func_70027_ad() && !(entity instanceof EntityEnderman))
             {
-                entity.func_70015_d(5);
+                // CraftBukkit start
+                EntityCombustByEntityEvent combustEvent = new EntityCombustByEntityEvent(this.getBukkitEntity(), entity.getBukkitEntity(), 5);
+                org.bukkit.Bukkit.getPluginManager().callEvent(combustEvent);
+                if (!combustEvent.isCancelled()) {
+                    entity.func_70015_d(combustEvent.getDuration());
+                }
+                // CraftBukkit end
             }
 
             if (entity.func_70097_a(damagesource, (float)i))
@@ -571,10 +585,22 @@
     {
         if (!this.field_70170_p.field_72995_K && this.field_70254_i && this.field_70249_b <= 0)
         {
-            boolean flag = this.field_70251_a == EntityArrow.PickupStatus.ALLOWED || this.field_70251_a == EntityArrow.PickupStatus.CREATIVE_ONLY && p_70100_1_.field_71075_bZ.field_75098_d;
+            // CraftBukkit start
+            ItemStack itemstack = this.func_184550_j(); // PAIL: rename
+            EntityItem item = new EntityItem(this.field_70170_p, this.field_70165_t, this.field_70163_u, this.field_70161_v, itemstack);
+            if (this.field_70251_a == PickupStatus.ALLOWED && p_70100_1_.field_71071_by.canHold(itemstack) > 0) {
+                PlayerPickupArrowEvent event = new PlayerPickupArrowEvent((Player) p_70100_1_.getBukkitEntity(), new CraftItem(this.field_70170_p.getServer(), this, item), (Arrow) this.getBukkitEntity());
+                // event.setCancelled(!entityhuman.canPickUpLoot); TODO
+                this.field_70170_p.getServer().getPluginManager().callEvent(event);
 
-            if (this.field_70251_a == EntityArrow.PickupStatus.ALLOWED && !p_70100_1_.field_71071_by.func_70441_a(this.func_184550_j()))
-            {
+                if (event.isCancelled()) {
+                    return;
+                }
+            }
+            boolean flag = this.field_70251_a == PickupStatus.ALLOWED || this.field_70251_a == PickupStatus.CREATIVE_ONLY && p_70100_1_.field_71075_bZ.field_75098_d;
+
+            if (this.field_70251_a == PickupStatus.ALLOWED && !p_70100_1_.field_71071_by.func_70441_a(item.func_92059_d())) {
+                // CraftBukkit end
                 flag = false;
             }
 
@@ -656,8 +682,15 @@
 
         if (EnchantmentHelper.func_185284_a(Enchantments.field_185311_w, p_190547_1_) > 0)
         {
-            this.func_70015_d(100);
-        }
+            // CraftBukkit start - call EntityCombustEvent
+            EntityCombustEvent event = new EntityCombustEvent(this.getBukkitEntity(), 100);
+            this.field_70170_p.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled()) {
+                this.func_70015_d(event.getDuration());
+            }
+            // CraftBukkit end
+            }
     }
 
     public static enum PickupStatus
