--- ../src-base/minecraft/net/minecraft/entity/player/EntityPlayerMP.java
+++ ../src-work/minecraft/net/minecraft/entity/player/EntityPlayerMP.java
@@ -4,6 +4,8 @@
 import com.google.common.collect.Sets;
 import com.mojang.authlib.GameProfile;
 import io.netty.buffer.Unpooled;
+
+import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Set;
@@ -16,21 +18,17 @@
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.crash.CrashReport;
 import net.minecraft.crash.CrashReportCategory;
+import net.minecraft.enchantment.EnchantmentHelper;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityList;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.IMerchant;
+import net.minecraft.entity.item.EntityItem;
 import net.minecraft.entity.passive.AbstractHorse;
 import net.minecraft.entity.projectile.EntityArrow;
 import net.minecraft.init.Items;
 import net.minecraft.init.SoundEvents;
-import net.minecraft.inventory.Container;
-import net.minecraft.inventory.ContainerChest;
-import net.minecraft.inventory.ContainerHorseInventory;
-import net.minecraft.inventory.ContainerMerchant;
-import net.minecraft.inventory.IContainerListener;
-import net.minecraft.inventory.IInventory;
-import net.minecraft.inventory.SlotCrafting;
+import net.minecraft.inventory.*;
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemMapBase;
 import net.minecraft.item.ItemStack;
@@ -100,19 +98,28 @@
 import net.minecraft.util.text.TextComponentTranslation;
 import net.minecraft.util.text.TextFormatting;
 import net.minecraft.village.MerchantRecipeList;
-import net.minecraft.world.GameType;
-import net.minecraft.world.IInteractionObject;
-import net.minecraft.world.ILockableContainer;
-import net.minecraft.world.WorldServer;
+import net.minecraft.world.*;
 import net.minecraft.world.biome.Biome;
 import net.minecraft.world.storage.loot.ILootContainer;
+import net.minecraftforge.event.entity.player.PlayerContainerEvent;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.WeatherType;
+import org.bukkit.craftbukkit.CraftWorld;
+import org.bukkit.craftbukkit.entity.CraftItem;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftInventory;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.inventory.InventoryOpenEvent;
+import org.bukkit.event.player.PlayerEditBookEvent;
+import org.bukkit.inventory.Inventory;
 
 public class EntityPlayerMP extends EntityPlayer implements IContainerListener
 {
     private static final Logger field_147102_bM = LogManager.getLogger();
-    private String field_71148_cg = "en_US";
+    public String field_71148_cg = "en_US";
     public NetHandlerPlayServer field_71135_a;
     public final MinecraftServer field_71133_b;
     public final PlayerInteractionManager field_71134_c;
@@ -140,15 +147,26 @@
     public boolean field_71137_h;
     public int field_71138_i;
     public boolean field_71136_j;
-
+    // CraftBukkit start
+    public String displayName;
+    public ITextComponent listName;
+    public org.bukkit.Location compassTarget;
+    public int newExp = 0;
+    public int newLevel = 0;
+    public int newTotalExp = 0;
+    public boolean keepLevel = false;
+    public double maxHealthCache;
+    public boolean joining = true;
+    public boolean sentListPacket = false;
+    // CraftBukkit end
     public EntityPlayerMP(MinecraftServer p_i45285_1_, WorldServer p_i45285_2_, GameProfile p_i45285_3_, PlayerInteractionManager p_i45285_4_)
     {
         super(p_i45285_2_, p_i45285_3_);
         p_i45285_4_.field_73090_b = this;
         this.field_71134_c = p_i45285_4_;
-        BlockPos blockpos = p_i45285_2_.func_175694_M();
+        BlockPos blockpos = p_i45285_2_.field_73011_w.getRandomizedSpawnPoint();
 
-        if (p_i45285_2_.field_73011_w.func_191066_m() && p_i45285_2_.func_72912_H().func_76077_q() != GameType.ADVENTURE)
+        if (false && p_i45285_2_.field_73011_w.func_191066_m() && p_i45285_2_.func_72912_H().func_76077_q() != GameType.ADVENTURE)
         {
             int i = Math.max(0, p_i45285_1_.func_184108_a(p_i45285_2_));
             int j = MathHelper.func_76128_c(p_i45285_2_.func_175723_af().func_177729_b((double)blockpos.func_177958_n(), (double)blockpos.func_177952_p()));
@@ -175,6 +193,11 @@
         {
             this.func_70107_b(this.field_70165_t, this.field_70163_u + 1.0D, this.field_70161_v);
         }
+        // CraftBukkit start
+        this.displayName = this.func_70005_c_();
+        // this.canPickUpLoot = true; TODO
+        this.maxHealthCache = this.func_110138_aP();
+        // CraftBukkit end
     }
 
     public void func_70037_a(NBTTagCompound p_70037_1_)
@@ -193,7 +216,30 @@
             }
         }
     }
-
+    // CraftBukkit start - World fallback code, either respawn location or global spawn
+    public void func_70029_a(World world) {
+        super.func_70029_a(world);
+        if (world == null) {
+            this.field_70128_L = false;
+            BlockPos position = null;
+            if (this.spawnWorld != null && !this.spawnWorld.equals("")) {
+                CraftWorld cworld = (CraftWorld) Bukkit.getServer().getWorld(this.spawnWorld);
+                if (cworld != null && this.func_180470_cg() != null) {
+                    world = cworld.getHandle();
+                    position = net.minecraft.entity.player.EntityPlayer.func_180467_a(cworld.getHandle(), this.func_180470_cg(), false);
+                }
+            }
+            if (world == null || position == null) {
+                world = ((CraftWorld) Bukkit.getServer().getWorlds().get(0)).getHandle();
+                position = world.func_175694_M();
+            }
+            this.field_70170_p = world;
+            this.func_70107_b(position.func_177958_n() + 0.5, position.func_177956_o(), position.func_177952_p() + 0.5);
+        }
+        this.field_71093_bK = ((WorldServer) this.field_70170_p).dimension;
+        this.field_71134_c.func_73080_a((WorldServer) world);
+    }
+    // CraftBukkit end
     public static void func_191522_a(DataFixer p_191522_0_)
     {
         p_191522_0_.func_188258_a(FixTypes.PLAYER, new IDataWalker()
@@ -269,6 +315,7 @@
 
     public void func_70071_h_()
     {
+
         this.field_71134_c.func_73075_a();
         --this.field_147101_bU;
 
@@ -279,7 +326,7 @@
 
         this.field_71070_bA.func_75142_b();
 
-        if (!this.field_70170_p.field_72995_K && !this.field_71070_bA.func_75145_c(this))
+        if (!this.field_70170_p.field_72995_K && this.field_71070_bA != null && !this.field_71070_bA.func_75145_c(this))
         {
             this.func_71053_j();
             this.field_71070_bA = this.field_71069_bz;
@@ -324,6 +371,12 @@
 
     public void func_71127_g()
     {
+        // CraftBukkit start
+        if (this.joining) {
+            this.joining = false;
+        }
+        // CraftBukkit end
+
         try
         {
             super.func_70071_h_();
@@ -368,7 +421,11 @@
                 this.field_184853_bW = this.func_70086_ai();
                 this.func_184849_a(IScoreCriteria.field_186699_i, MathHelper.func_76123_f((float)this.field_184853_bW));
             }
-
+            // CraftBukkit start - Force max health updates
+            //if (this.maxHealthCache != this.getMaxHealth()) {
+            //    this.getBukkitEntity().updateScaledHealth();
+            //}
+            // CraftBukkit end
             if (this.func_70658_aO() != this.field_184854_bX)
             {
                 this.field_184854_bX = this.func_70658_aO();
@@ -397,6 +454,16 @@
             {
                 this.func_147098_j();
             }
+            // CraftBukkit start - initialize oldLevel and fire PlayerLevelChangeEvent
+            if (this.oldLevel == -1) {
+                this.oldLevel = this.field_71068_ca;
+            }
+
+            if (this.oldLevel != this.field_71068_ca) {
+                CraftEventFactory.callPlayerLevelChangeEvent(this.field_70170_p.getServer().getPlayer((net.minecraft.entity.player.EntityPlayerMP) this), this.oldLevel, this.field_71068_ca);
+                this.oldLevel = this.field_71068_ca;
+            }
+            // CraftBukkit end
         }
         catch (Throwable throwable)
         {
@@ -462,9 +529,72 @@
 
     public void func_70645_a(DamageSource p_70645_1_)
     {
+        if (net.minecraftforge.common.ForgeHooks.onLivingDeath(this, p_70645_1_)) return;
         boolean flag = this.field_70170_p.func_82736_K().func_82766_b("showDeathMessages");
         this.field_71135_a.func_147359_a(new SPacketCombatEvent(this.func_110142_aN(), SPacketCombatEvent.Event.ENTITY_DIED, flag));
+        //craftbukkit  start DeathEvent
+        if(field_70128_L)
+        {
+            return;
+        }
+        List<org.bukkit.inventory.ItemStack> loot = new java.util.ArrayList<org.bukkit.inventory.ItemStack>(this.field_71071_by.func_70302_i_());
+        List<net.minecraft.item.ItemStack> bukkit_processed_loot = new ArrayList<ItemStack>();
+        boolean keepInventory = this.field_70170_p.func_82736_K().func_82766_b("keepInventory");
 
+        if (!keepInventory) {
+            for (ItemStack item : this.field_71071_by.getContents()) {
+                if (!item.func_190926_b() && !EnchantmentHelper.func_190939_c(item)) { // PAIL: shouldNotDrop (Vanishing enchant)
+                    loot.add(CraftItemStack.asCraftMirror(item));
+                }
+            }
+        }
+        ITextComponent chatmessage = this.func_110142_aN().func_151521_b();
+        String deathmessage = chatmessage.func_150260_c();
+        org.bukkit.event.entity.PlayerDeathEvent event = CraftEventFactory.callPlayerDeathEvent(this, loot, deathmessage, keepInventory);
+        for(org.bukkit.inventory.ItemStack itemStack : loot)
+        {
+            bukkit_processed_loot.add(CraftItemStack.asNMSCopy(itemStack));
+        }
+        String deathMessage = event.getDeathMessage();
+
+        if (deathMessage != null && deathMessage.length() > 0 && flag) { // TODO: allow plugins to override?
+            if (deathMessage.equals(deathmessage)) {
+                Team scoreboardteambase = this.func_96124_cp();
+
+                if (scoreboardteambase != null && scoreboardteambase.func_178771_j() != Team.EnumVisible.ALWAYS) {
+                    if (scoreboardteambase.func_178771_j() == Team.EnumVisible.HIDE_FOR_OTHER_TEAMS) {
+                        this.field_71133_b.func_184103_al().func_177453_a((net.minecraft.entity.player.EntityPlayer) this, chatmessage);
+                    } else if (scoreboardteambase.func_178771_j() == Team.EnumVisible.HIDE_FOR_OWN_TEAM) {
+                        this.field_71133_b.func_184103_al().func_177452_b((net.minecraft.entity.player.EntityPlayer) this, chatmessage);
+                    }
+                } else {
+                    this.field_71133_b.func_184103_al().func_148539_a(chatmessage);
+                }
+            } else {
+                this.field_71133_b.func_184103_al().sendMessage(org.bukkit.craftbukkit.util.CraftChatMessage.fromString(deathMessage));
+            }
+        }
+        //CraftBukkit end
+        if (!event.getKeepInventory() && !this.func_175149_v())
+        {
+            captureDrops = true;
+            capturedDrops.clear();
+            this.func_190776_cN();
+            this.field_71071_by.func_70436_m();
+            captureDrops = false;
+            //MCPCRevive start : minor modification
+            net.minecraftforge.event.entity.player.PlayerDropsEvent forgeevent = new net.minecraftforge.event.entity.player.PlayerDropsEvent(this, p_70645_1_, capturedDrops, field_70718_bc > 0);
+            if (!net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(forgeevent))
+            {
+                for (net.minecraft.item.ItemStack item : bukkit_processed_loot)
+                {
+                    this.field_70170_p.func_72838_d(new EntityItem(func_71121_q(),field_70165_t,field_70163_u,field_70161_v,item));
+                }
+            }
+            //spawn Item after forge event
+        }
+
+
         if (flag)
         {
             Team team = this.func_96124_cp();
@@ -486,11 +616,6 @@
             }
         }
 
-        if (!this.field_70170_p.func_82736_K().func_82766_b("keepInventory") && !this.func_175149_v())
-        {
-            this.func_190776_cN();
-            this.field_71071_by.func_70436_m();
-        }
 
         for (ScoreObjective scoreobjective : this.field_70170_p.func_96441_U().func_96520_a(IScoreCriteria.field_96642_c))
         {
@@ -573,6 +698,12 @@
     @Nullable
     public Entity func_184204_a(int p_184204_1_)
     {
+        if (!net.minecraftforge.common.ForgeHooks.onTravelToDimension(this, p_184204_1_)) return this;
+        //MCPCRevive start
+        if(this.field_71093_bK != p_184204_1_) {
+            this.field_71133_b.func_184103_al().func_187242_a(this, p_184204_1_);
+        }
+        //MCPCRevive end
         this.field_184851_cj = true;
 
         if (this.field_71093_bK == 1 && p_184204_1_ == 1)
@@ -607,8 +738,7 @@
             {
                 this.func_71029_a(AchievementList.field_187997_y);
             }
-
-            this.field_71133_b.func_184103_al().func_187242_a(this, p_184204_1_);
+            //this.mcServer.getPlayerList().changePlayerDimension(this, dimensionIn); MCPCRevive moved up
             this.field_71135_a.func_147359_a(new SPacketEffect(1032, BlockPos.field_177992_a, 0, false));
             this.field_71144_ck = -1;
             this.field_71149_ch = -1.0F;
@@ -730,7 +860,7 @@
         BlockPos blockpos = new BlockPos(i, j, k);
         IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos);
 
-        if (iblockstate.func_185904_a() == Material.field_151579_a)
+        if (iblockstate.func_177230_c().isAir(iblockstate, this.field_70170_p, blockpos))
         {
             BlockPos blockpos1 = blockpos.func_177977_b();
             IBlockState iblockstate1 = this.field_70170_p.func_180495_p(blockpos1);
@@ -765,11 +895,19 @@
         }
         else
         {
+            //MCPCRevive start
+            Container ct = CraftEventFactory.callInventoryOpenEvent(this,p_180468_1_.func_174876_a(field_71071_by,this));
+            if(ct == null)
+            {
+                return;
+            }
+            //MCPCRevive end
             this.func_71117_bO();
             this.field_71135_a.func_147359_a(new SPacketOpenWindow(this.field_71139_cq, p_180468_1_.func_174875_k(), p_180468_1_.func_145748_c_()));
             this.field_71070_bA = p_180468_1_.func_174876_a(this.field_71071_by, this);
             this.field_71070_bA.field_75152_c = this.field_71139_cq;
             this.field_71070_bA.func_75132_a(this);
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.PlayerContainerEvent.Open(this, this.field_71070_bA));
         }
     }
 
@@ -785,7 +923,13 @@
             {
                 this.func_71053_j();
             }
-
+            //MCPCRevive start
+            Container ct = CraftEventFactory.callInventoryOpenEvent(this,new ContainerChest(this.field_71071_by,p_71007_1_,this));
+            if(ct == null)
+            {
+                return;
+            }
+            //MCPCRevive end
             if (p_71007_1_ instanceof ILockableContainer)
             {
                 ILockableContainer ilockablecontainer = (ILockableContainer)p_71007_1_;
@@ -813,6 +957,7 @@
 
             this.field_71070_bA.field_75152_c = this.field_71139_cq;
             this.field_71070_bA.func_75132_a(this);
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.PlayerContainerEvent.Open(this, this.field_71070_bA));
         }
     }
 
@@ -822,6 +967,7 @@
         this.field_71070_bA = new ContainerMerchant(this.field_71071_by, p_180472_1_, this.field_70170_p);
         this.field_71070_bA.field_75152_c = this.field_71139_cq;
         this.field_71070_bA.func_75132_a(this);
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.PlayerContainerEvent.Open(this, this.field_71070_bA));
         IInventory iinventory = ((ContainerMerchant)this.field_71070_bA).func_75174_d();
         ITextComponent itextcomponent = p_180472_1_.func_145748_c_();
         this.field_71135_a.func_147359_a(new SPacketOpenWindow(this.field_71139_cq, "minecraft:villager", itextcomponent, iinventory.func_70302_i_()));
@@ -838,11 +984,18 @@
 
     public void func_184826_a(AbstractHorse p_184826_1_, IInventory p_184826_2_)
     {
+
         if (this.field_71070_bA != this.field_71069_bz)
         {
             this.func_71053_j();
         }
-
+        //MCPCRevive start
+        Container ct = CraftEventFactory.callInventoryOpenEvent(this,new ContainerChest(this.field_71071_by,p_184826_2_,this));
+        if(ct == null)
+        {
+            return;
+        }
+        //MCPCRevive end
         this.func_71117_bO();
         this.field_71135_a.func_147359_a(new SPacketOpenWindow(this.field_71139_cq, "EntityHorse", p_184826_2_.func_145748_c_(), p_184826_2_.func_70302_i_(), p_184826_1_.func_145782_y()));
         this.field_71070_bA = new ContainerHorseInventory(this.field_71071_by, p_184826_2_, p_184826_1_, this);
@@ -853,7 +1006,6 @@
     public void func_184814_a(ItemStack p_184814_1_, EnumHand p_184814_2_)
     {
         Item item = p_184814_1_.func_77973_b();
-
         if (item == Items.field_151164_bB)
         {
             PacketBuffer packetbuffer = new PacketBuffer(Unpooled.buffer());
@@ -920,6 +1072,7 @@
     public void func_71128_l()
     {
         this.field_71070_bA.func_75134_a(this);
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.PlayerContainerEvent.Close(this, this.field_71070_bA));
         this.field_71070_bA = this.field_71069_bz;
     }
 
@@ -951,6 +1104,7 @@
     {
         if (p_71064_1_ != null)
         {
+            if (p_71064_1_.func_75967_d() && net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.player.AchievementEvent(this, (net.minecraft.stats.Achievement) p_71064_1_))) return;
             this.field_147103_bO.func_150871_b(this, p_71064_1_, p_71064_2_);
 
             for (ScoreObjective scoreobjective : this.func_96123_co().func_96520_a(p_71064_1_.func_150952_k()))
@@ -1100,6 +1254,11 @@
     {
         this.field_71135_a.func_147359_a(new SPacketChat(p_145747_1_));
     }
+    public void sendMessage(ITextComponent[] components)
+    {
+        for(ITextComponent component : components)
+        this.field_71135_a.func_147359_a(new SPacketChat(component));
+    }
 
     public boolean func_70003_b(int p_70003_1_, String p_70003_2_)
     {
@@ -1273,4 +1432,87 @@
         this.func_70052_a(7, true);
         this.func_70052_a(7, false);
     }
+    // CraftBukkit start - Add per-player time and weather.
+    public long timeOffset = 0;
+    public boolean relativeTime = true;
+
+    public long getPlayerTime() {
+        if (this.relativeTime) {
+            // Adds timeOffset to the current server time.
+            return this.field_70170_p.func_72820_D() + this.timeOffset;
+        } else {
+            // Adds timeOffset to the beginning of this day.
+            return this.field_70170_p.func_72820_D() - (this.field_70170_p.func_72820_D() % 24000) + this.timeOffset;
+        }
+    }
+
+    public WeatherType weather = null;
+
+    public WeatherType getPlayerWeather() {
+        return this.weather;
+    }
+
+    public void setPlayerWeather(WeatherType type, boolean plugin) {
+        if (!plugin && this.weather != null) {
+            return;
+        }
+
+        if (plugin) {
+            this.weather = type;
+        }
+
+        if (type == WeatherType.DOWNFALL) {
+            this.field_71135_a.func_147359_a(new SPacketChangeGameState(2, 0));
+        } else {
+            this.field_71135_a.func_147359_a(new SPacketChangeGameState(1, 0));
+        }
+    }
+
+    private float pluginRainPosition;
+    private float pluginRainPositionPrevious;
+
+    public void updateWeather(float oldRain, float newRain, float oldThunder, float newThunder) {
+        if (this.weather == null) {
+            // Vanilla
+            if (oldRain != newRain) {
+                this.field_71135_a.func_147359_a(new SPacketChangeGameState(7, newRain));
+            }
+        } else {
+            // Plugin
+            if (pluginRainPositionPrevious != pluginRainPosition) {
+                this.field_71135_a.func_147359_a(new SPacketChangeGameState(7, pluginRainPosition));
+            }
+        }
+
+        if (oldThunder != newThunder) {
+            if (weather == WeatherType.DOWNFALL || weather == null) {
+                this.field_71135_a.func_147359_a(new SPacketChangeGameState(8, newThunder));
+            } else {
+                this.field_71135_a.func_147359_a(new SPacketChangeGameState(8, 0));
+            }
+        }
+    }
+
+    public void tickWeather() {
+        if (this.weather == null) return;
+
+        pluginRainPositionPrevious = pluginRainPosition;
+        if (weather == WeatherType.DOWNFALL) {
+            pluginRainPosition += 0.01;
+        } else {
+            pluginRainPosition -= 0.01;
+        }
+
+        pluginRainPosition = MathHelper.func_76131_a(pluginRainPosition, 0.0F, 1.0F);
+    }
+
+    public void resetPlayerWeather() {
+        this.weather = null;
+        this.setPlayerWeather(this.field_70170_p.func_72912_H().func_76059_o() ? WeatherType.DOWNFALL : WeatherType.CLEAR, false);
+    }
+    public int nextContainerCounter() { // CraftBukkit - void -> int
+        this.field_71139_cq = this.field_71139_cq % 100 + 1;
+        return field_71139_cq; // CraftBukkit
+    }
+    //CraftBukkit end
 }
