--- ../src-base/minecraft/net/minecraft/item/crafting/CraftingManager.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/CraftingManager.java
@@ -20,17 +20,20 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class CraftingManager
 {
     private static final CraftingManager field_77598_a = new CraftingManager();
     public final List<IRecipe> field_77597_b = Lists.<IRecipe>newArrayList();
-
+    // CraftBukkit start
+    public net.minecraft.item.crafting.IRecipe lastRecipe;
+    public org.bukkit.inventory.InventoryView lastCraftView;
+    // CraftBukkit end
     public static CraftingManager func_77594_a()
     {
         return field_77598_a;
     }
-
     public CraftingManager()
     {
         (new RecipesTools()).func_77586_a(this);
@@ -314,7 +317,12 @@
         {
             if (irecipe.func_77569_a(p_82787_1_, p_82787_2_))
             {
-                return irecipe.func_77572_b(p_82787_1_);
+
+                // CraftBukkit start - INVENTORY_PRE_CRAFT event
+                p_82787_1_.currentRecipe = irecipe;
+                ItemStack result = irecipe.func_77572_b(p_82787_1_);
+                return CraftEventFactory.callPreCraftEvent(p_82787_1_, result, lastCraftView, false);
+                // CraftBukkit end
             }
         }
 
