--- ../src-base/minecraft/net/minecraft/item/crafting/RecipeBookCloning.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipeBookCloning.java
@@ -7,22 +7,29 @@
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
 
-public class RecipeBookCloning implements IRecipe
+public class RecipeBookCloning extends ShapelessRecipes implements IRecipe
 {
-    public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
+    // CraftBukkit - added extends
+
+    public RecipeBookCloning() {
+        super(new ItemStack(Items.WRITTEN_BOOK, 0, -1), java.util.Arrays.asList(new ItemStack(Items.WRITABLE_BOOK, 0, 0)));
+    }
+    // CraftBukkit end
+
+    public boolean matches(InventoryCrafting inv, World worldIn)
     {
         int i = 0;
-        ItemStack itemstack = ItemStack.field_190927_a;
+        ItemStack itemstack = ItemStack.EMPTY;
 
-        for (int j = 0; j < p_77569_1_.func_70302_i_(); ++j)
+        for (int j = 0; j < inv.getSizeInventory(); ++j)
         {
-            ItemStack itemstack1 = p_77569_1_.func_70301_a(j);
+            ItemStack itemstack1 = inv.getStackInSlot(j);
 
-            if (!itemstack1.func_190926_b())
+            if (!itemstack1.isEmpty())
             {
-                if (itemstack1.func_77973_b() == Items.field_151164_bB)
+                if (itemstack1.getItem() == Items.WRITTEN_BOOK)
                 {
-                    if (!itemstack.func_190926_b())
+                    if (!itemstack.isEmpty())
                     {
                         return false;
                     }
@@ -31,7 +38,7 @@
                 }
                 else
                 {
-                    if (itemstack1.func_77973_b() != Items.field_151099_bA)
+                    if (itemstack1.getItem() != Items.WRITABLE_BOOK)
                     {
                         return false;
                     }
@@ -41,34 +48,34 @@
             }
         }
 
-        return !itemstack.func_190926_b() && itemstack.func_77942_o() && i > 0;
+        return !itemstack.isEmpty() && itemstack.hasTagCompound() && i > 0;
     }
 
-    public ItemStack func_77572_b(InventoryCrafting p_77572_1_)
+    public ItemStack getCraftingResult(InventoryCrafting inv)
     {
         int i = 0;
-        ItemStack itemstack = ItemStack.field_190927_a;
+        ItemStack itemstack = ItemStack.EMPTY;
 
-        for (int j = 0; j < p_77572_1_.func_70302_i_(); ++j)
+        for (int j = 0; j < inv.getSizeInventory(); ++j)
         {
-            ItemStack itemstack1 = p_77572_1_.func_70301_a(j);
+            ItemStack itemstack1 = inv.getStackInSlot(j);
 
-            if (!itemstack1.func_190926_b())
+            if (!itemstack1.isEmpty())
             {
-                if (itemstack1.func_77973_b() == Items.field_151164_bB)
+                if (itemstack1.getItem() == Items.WRITTEN_BOOK)
                 {
-                    if (!itemstack.func_190926_b())
+                    if (!itemstack.isEmpty())
                     {
-                        return ItemStack.field_190927_a;
+                        return ItemStack.EMPTY;
                     }
 
                     itemstack = itemstack1;
                 }
                 else
                 {
-                    if (itemstack1.func_77973_b() != Items.field_151099_bA)
+                    if (itemstack1.getItem() != Items.WRITABLE_BOOK)
                     {
-                        return ItemStack.field_190927_a;
+                        return ItemStack.EMPTY;
                     }
 
                     ++i;
@@ -76,47 +83,47 @@
             }
         }
 
-        if (!itemstack.func_190926_b() && itemstack.func_77942_o() && i >= 1 && ItemWrittenBook.func_179230_h(itemstack) < 2)
+        if (!itemstack.isEmpty() && itemstack.hasTagCompound() && i >= 1 && ItemWrittenBook.getGeneration(itemstack) < 2)
         {
-            ItemStack itemstack2 = new ItemStack(Items.field_151164_bB, i);
-            itemstack2.func_77982_d(itemstack.func_77978_p().func_74737_b());
-            itemstack2.func_77978_p().func_74768_a("generation", ItemWrittenBook.func_179230_h(itemstack) + 1);
+            ItemStack itemstack2 = new ItemStack(Items.WRITTEN_BOOK, i);
+            itemstack2.setTagCompound(itemstack.getTagCompound().copy());
+            itemstack2.getTagCompound().setInteger("generation", ItemWrittenBook.getGeneration(itemstack) + 1);
 
-            if (itemstack.func_82837_s())
+            if (itemstack.hasDisplayName())
             {
-                itemstack2.func_151001_c(itemstack.func_82833_r());
+                itemstack2.setStackDisplayName(itemstack.getDisplayName());
             }
 
             return itemstack2;
         }
         else
         {
-            return ItemStack.field_190927_a;
+            return ItemStack.EMPTY;
         }
     }
 
-    public int func_77570_a()
+    public int getRecipeSize()
     {
         return 9;
     }
 
-    public ItemStack func_77571_b()
+    public ItemStack getRecipeOutput()
     {
-        return ItemStack.field_190927_a;
+        return ItemStack.EMPTY;
     }
 
-    public NonNullList<ItemStack> func_179532_b(InventoryCrafting p_179532_1_)
+    public NonNullList<ItemStack> getRemainingItems(InventoryCrafting inv)
     {
-        NonNullList<ItemStack> nonnulllist = NonNullList.<ItemStack>func_191197_a(p_179532_1_.func_70302_i_(), ItemStack.field_190927_a);
+        NonNullList<ItemStack> nonnulllist = NonNullList.<ItemStack>withSize(inv.getSizeInventory(), ItemStack.EMPTY);
 
         for (int i = 0; i < nonnulllist.size(); ++i)
         {
-            ItemStack itemstack = p_179532_1_.func_70301_a(i);
+            ItemStack itemstack = inv.getStackInSlot(i);
 
-            if (itemstack.func_77973_b() instanceof ItemWrittenBook)
+            if (itemstack.getItem() instanceof ItemWrittenBook)
             {
-                ItemStack itemstack1 = itemstack.func_77946_l();
-                itemstack1.func_190920_e(1);
+                ItemStack itemstack1 = itemstack.copy();
+                itemstack1.setCount(1);
                 nonnulllist.set(i, itemstack1);
                 break;
             }
