--- ../src-base/minecraft/net/minecraft/item/crafting/RecipeFireworks.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipeFireworks.java
@@ -11,10 +11,15 @@
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
 
-public class RecipeFireworks implements IRecipe
+public class RecipeFireworks extends ShapelessRecipes implements IRecipe
 {
     private ItemStack field_92102_a = ItemStack.field_190927_a;
-
+    // CraftBukkit start - Delegate to new parent class with bogus info
+    public RecipeFireworks() {
+        super(new ItemStack(Items.field_151152_bP, 0, 0), java.util.Arrays.asList(new ItemStack(Items.field_151016_H, 0, 5)));
+        this.field_92102_a = ItemStack.field_190927_a;
+    }
+    // CraftBukkit end
     public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
     {
         this.field_92102_a = ItemStack.field_190927_a;
@@ -251,10 +256,7 @@
         {
             ItemStack itemstack = p_179532_1_.func_70301_a(i);
 
-            if (itemstack.func_77973_b().func_77634_r())
-            {
-                nonnulllist.set(i, new ItemStack(itemstack.func_77973_b().func_77668_q()));
-            }
+            nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
         }
 
         return nonnulllist;
