--- ../src-base/minecraft/net/minecraft/item/crafting/RecipeTippedArrow.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipeTippedArrow.java
@@ -7,34 +7,44 @@
 import net.minecraft.potion.PotionUtils;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-
-public class RecipeTippedArrow implements IRecipe
+import net.minecraft.item.crafting.ShapedRecipes;
+public class RecipeTippedArrow extends ShapedRecipes implements IRecipe
 {
-    public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
+    // CraftBukkit start
+    public RecipeTippedArrow() {
+        super(3, 3, new ItemStack[]{
+                new ItemStack(Items.ARROW, 0), new ItemStack(Items.ARROW, 0), new ItemStack(Items.ARROW, 0),
+                new ItemStack(Items.ARROW, 0), new ItemStack(Items.LINGERING_POTION, 0), new ItemStack(Items.ARROW, 0),
+                new ItemStack(Items.ARROW, 0), new ItemStack(Items.ARROW, 0), new ItemStack(Items.ARROW, 0)
+        }, new ItemStack(Items.TIPPED_ARROW, 8));
+    }
+    // CraftBukkit end
+
+    public boolean matches(InventoryCrafting inv, World worldIn)
     {
-        if (p_77569_1_.func_174922_i() == 3 && p_77569_1_.func_174923_h() == 3)
+        if (inv.getWidth() == 3 && inv.getHeight() == 3)
         {
-            for (int i = 0; i < p_77569_1_.func_174922_i(); ++i)
+            for (int i = 0; i < inv.getWidth(); ++i)
             {
-                for (int j = 0; j < p_77569_1_.func_174923_h(); ++j)
+                for (int j = 0; j < inv.getHeight(); ++j)
                 {
-                    ItemStack itemstack = p_77569_1_.func_70463_b(i, j);
+                    ItemStack itemstack = inv.getStackInRowAndColumn(i, j);
 
-                    if (itemstack.func_190926_b())
+                    if (itemstack.isEmpty())
                     {
                         return false;
                     }
 
-                    Item item = itemstack.func_77973_b();
+                    Item item = itemstack.getItem();
 
                     if (i == 1 && j == 1)
                     {
-                        if (item != Items.field_185156_bI)
+                        if (item != Items.LINGERING_POTION)
                         {
                             return false;
                         }
                     }
-                    else if (item != Items.field_151032_g)
+                    else if (item != Items.ARROW)
                     {
                         return false;
                     }
@@ -49,35 +59,35 @@
         }
     }
 
-    public ItemStack func_77572_b(InventoryCrafting p_77572_1_)
+    public ItemStack getCraftingResult(InventoryCrafting inv)
     {
-        ItemStack itemstack = p_77572_1_.func_70463_b(1, 1);
+        ItemStack itemstack = inv.getStackInRowAndColumn(1, 1);
 
-        if (itemstack.func_77973_b() != Items.field_185156_bI)
+        if (itemstack.getItem() != Items.LINGERING_POTION)
         {
-            return ItemStack.field_190927_a;
+            return ItemStack.EMPTY;
         }
         else
         {
-            ItemStack itemstack1 = new ItemStack(Items.field_185167_i, 8);
-            PotionUtils.func_185188_a(itemstack1, PotionUtils.func_185191_c(itemstack));
-            PotionUtils.func_185184_a(itemstack1, PotionUtils.func_185190_b(itemstack));
+            ItemStack itemstack1 = new ItemStack(Items.TIPPED_ARROW, 8);
+            PotionUtils.addPotionToItemStack(itemstack1, PotionUtils.getPotionFromItem(itemstack));
+            PotionUtils.appendEffects(itemstack1, PotionUtils.getFullEffectsFromItem(itemstack));
             return itemstack1;
         }
     }
 
-    public int func_77570_a()
+    public int getRecipeSize()
     {
         return 9;
     }
 
-    public ItemStack func_77571_b()
+    public ItemStack getRecipeOutput()
     {
-        return ItemStack.field_190927_a;
+        return ItemStack.EMPTY;
     }
 
-    public NonNullList<ItemStack> func_179532_b(InventoryCrafting p_179532_1_)
+    public NonNullList<ItemStack> getRemainingItems(InventoryCrafting inv)
     {
-        return NonNullList.<ItemStack>func_191197_a(p_179532_1_.func_70302_i_(), ItemStack.field_190927_a);
+        return NonNullList.<ItemStack>withSize(inv.getSizeInventory(), ItemStack.EMPTY);
     }
 }
