--- ../src-base/minecraft/net/minecraft/item/crafting/RecipeBookCloning.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipeBookCloning.java
@@ -7,8 +7,15 @@
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
 
-public class RecipeBookCloning implements IRecipe
+public class RecipeBookCloning extends ShapelessRecipes implements IRecipe
 {
+    // CraftBukkit - added extends
+
+    public RecipeBookCloning() {
+        super(new ItemStack(Items.field_151164_bB, 0, -1), java.util.Arrays.asList(new ItemStack(Items.field_151099_bA, 0, 0)));
+    }
+    // CraftBukkit end
+
     public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
     {
         int i = 0;
