--- ../src-base/minecraft/net/minecraft/item/crafting/RecipeFireworks.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipeFireworks.java
@@ -11,13 +11,18 @@
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
 
-public class RecipeFireworks implements IRecipe
+public class RecipeFireworks extends ShapelessRecipes implements IRecipe
 {
-    private ItemStack field_92102_a = ItemStack.field_190927_a;
-
-    public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
+    private ItemStack resultItem = ItemStack.EMPTY;
+    // CraftBukkit start - Delegate to new parent class with bogus info
+    public RecipeFireworks() {
+        super(new ItemStack(Items.FIREWORKS, 0, 0), java.util.Arrays.asList(new ItemStack(Items.GUNPOWDER, 0, 5)));
+        this.resultItem = ItemStack.EMPTY;
+    }
+    // CraftBukkit end
+    public boolean matches(InventoryCrafting inv, World worldIn)
     {
-        this.field_92102_a = ItemStack.field_190927_a;
+        this.resultItem = ItemStack.EMPTY;
         int i = 0;
         int j = 0;
         int k = 0;
@@ -25,51 +30,51 @@
         int i1 = 0;
         int j1 = 0;
 
-        for (int k1 = 0; k1 < p_77569_1_.func_70302_i_(); ++k1)
+        for (int k1 = 0; k1 < inv.getSizeInventory(); ++k1)
         {
-            ItemStack itemstack = p_77569_1_.func_70301_a(k1);
+            ItemStack itemstack = inv.getStackInSlot(k1);
 
-            if (!itemstack.func_190926_b())
+            if (!itemstack.isEmpty())
             {
-                if (itemstack.func_77973_b() == Items.field_151016_H)
+                if (itemstack.getItem() == Items.GUNPOWDER)
                 {
                     ++j;
                 }
-                else if (itemstack.func_77973_b() == Items.field_151154_bQ)
+                else if (itemstack.getItem() == Items.FIREWORK_CHARGE)
                 {
                     ++l;
                 }
-                else if (itemstack.func_77973_b() == Items.field_151100_aR)
+                else if (itemstack.getItem() == Items.DYE)
                 {
                     ++k;
                 }
-                else if (itemstack.func_77973_b() == Items.field_151121_aF)
+                else if (itemstack.getItem() == Items.PAPER)
                 {
                     ++i;
                 }
-                else if (itemstack.func_77973_b() == Items.field_151114_aO)
+                else if (itemstack.getItem() == Items.GLOWSTONE_DUST)
                 {
                     ++i1;
                 }
-                else if (itemstack.func_77973_b() == Items.field_151045_i)
+                else if (itemstack.getItem() == Items.DIAMOND)
                 {
                     ++i1;
                 }
-                else if (itemstack.func_77973_b() == Items.field_151059_bz)
+                else if (itemstack.getItem() == Items.FIRE_CHARGE)
                 {
                     ++j1;
                 }
-                else if (itemstack.func_77973_b() == Items.field_151008_G)
+                else if (itemstack.getItem() == Items.FEATHER)
                 {
                     ++j1;
                 }
-                else if (itemstack.func_77973_b() == Items.field_151074_bl)
+                else if (itemstack.getItem() == Items.GOLD_NUGGET)
                 {
                     ++j1;
                 }
                 else
                 {
-                    if (itemstack.func_77973_b() != Items.field_151144_bL)
+                    if (itemstack.getItem() != Items.SKULL)
                     {
                         return false;
                     }
@@ -85,71 +90,71 @@
         {
             if (j >= 1 && i == 1 && i1 == 0)
             {
-                this.field_92102_a = new ItemStack(Items.field_151152_bP, 3);
+                this.resultItem = new ItemStack(Items.FIREWORKS, 3);
                 NBTTagCompound nbttagcompound1 = new NBTTagCompound();
 
                 if (l > 0)
                 {
                     NBTTagList nbttaglist = new NBTTagList();
 
-                    for (int k2 = 0; k2 < p_77569_1_.func_70302_i_(); ++k2)
+                    for (int k2 = 0; k2 < inv.getSizeInventory(); ++k2)
                     {
-                        ItemStack itemstack3 = p_77569_1_.func_70301_a(k2);
+                        ItemStack itemstack3 = inv.getStackInSlot(k2);
 
-                        if (itemstack3.func_77973_b() == Items.field_151154_bQ && itemstack3.func_77942_o() && itemstack3.func_77978_p().func_150297_b("Explosion", 10))
+                        if (itemstack3.getItem() == Items.FIREWORK_CHARGE && itemstack3.hasTagCompound() && itemstack3.getTagCompound().hasKey("Explosion", 10))
                         {
-                            nbttaglist.func_74742_a(itemstack3.func_77978_p().func_74775_l("Explosion"));
+                            nbttaglist.appendTag(itemstack3.getTagCompound().getCompoundTag("Explosion"));
                         }
                     }
 
-                    nbttagcompound1.func_74782_a("Explosions", nbttaglist);
+                    nbttagcompound1.setTag("Explosions", nbttaglist);
                 }
 
-                nbttagcompound1.func_74774_a("Flight", (byte)j);
+                nbttagcompound1.setByte("Flight", (byte)j);
                 NBTTagCompound nbttagcompound3 = new NBTTagCompound();
-                nbttagcompound3.func_74782_a("Fireworks", nbttagcompound1);
-                this.field_92102_a.func_77982_d(nbttagcompound3);
+                nbttagcompound3.setTag("Fireworks", nbttagcompound1);
+                this.resultItem.setTagCompound(nbttagcompound3);
                 return true;
             }
             else if (j == 1 && i == 0 && l == 0 && k > 0 && j1 <= 1)
             {
-                this.field_92102_a = new ItemStack(Items.field_151154_bQ);
+                this.resultItem = new ItemStack(Items.FIREWORK_CHARGE);
                 NBTTagCompound nbttagcompound = new NBTTagCompound();
                 NBTTagCompound nbttagcompound2 = new NBTTagCompound();
                 byte b0 = 0;
                 List<Integer> list = Lists.<Integer>newArrayList();
 
-                for (int l1 = 0; l1 < p_77569_1_.func_70302_i_(); ++l1)
+                for (int l1 = 0; l1 < inv.getSizeInventory(); ++l1)
                 {
-                    ItemStack itemstack2 = p_77569_1_.func_70301_a(l1);
+                    ItemStack itemstack2 = inv.getStackInSlot(l1);
 
-                    if (!itemstack2.func_190926_b())
+                    if (!itemstack2.isEmpty())
                     {
-                        if (itemstack2.func_77973_b() == Items.field_151100_aR)
+                        if (itemstack2.getItem() == Items.DYE)
                         {
-                            list.add(Integer.valueOf(ItemDye.field_150922_c[itemstack2.func_77960_j() & 15]));
+                            list.add(Integer.valueOf(ItemDye.DYE_COLORS[itemstack2.getMetadata() & 15]));
                         }
-                        else if (itemstack2.func_77973_b() == Items.field_151114_aO)
+                        else if (itemstack2.getItem() == Items.GLOWSTONE_DUST)
                         {
-                            nbttagcompound2.func_74757_a("Flicker", true);
+                            nbttagcompound2.setBoolean("Flicker", true);
                         }
-                        else if (itemstack2.func_77973_b() == Items.field_151045_i)
+                        else if (itemstack2.getItem() == Items.DIAMOND)
                         {
-                            nbttagcompound2.func_74757_a("Trail", true);
+                            nbttagcompound2.setBoolean("Trail", true);
                         }
-                        else if (itemstack2.func_77973_b() == Items.field_151059_bz)
+                        else if (itemstack2.getItem() == Items.FIRE_CHARGE)
                         {
                             b0 = 1;
                         }
-                        else if (itemstack2.func_77973_b() == Items.field_151008_G)
+                        else if (itemstack2.getItem() == Items.FEATHER)
                         {
                             b0 = 4;
                         }
-                        else if (itemstack2.func_77973_b() == Items.field_151074_bl)
+                        else if (itemstack2.getItem() == Items.GOLD_NUGGET)
                         {
                             b0 = 2;
                         }
-                        else if (itemstack2.func_77973_b() == Items.field_151144_bL)
+                        else if (itemstack2.getItem() == Items.SKULL)
                         {
                             b0 = 3;
                         }
@@ -163,30 +168,30 @@
                     aint1[l2] = ((Integer)list.get(l2)).intValue();
                 }
 
-                nbttagcompound2.func_74783_a("Colors", aint1);
-                nbttagcompound2.func_74774_a("Type", b0);
-                nbttagcompound.func_74782_a("Explosion", nbttagcompound2);
-                this.field_92102_a.func_77982_d(nbttagcompound);
+                nbttagcompound2.setIntArray("Colors", aint1);
+                nbttagcompound2.setByte("Type", b0);
+                nbttagcompound.setTag("Explosion", nbttagcompound2);
+                this.resultItem.setTagCompound(nbttagcompound);
                 return true;
             }
             else if (j == 0 && i == 0 && l == 1 && k > 0 && k == i1)
             {
                 List<Integer> list1 = Lists.<Integer>newArrayList();
 
-                for (int i2 = 0; i2 < p_77569_1_.func_70302_i_(); ++i2)
+                for (int i2 = 0; i2 < inv.getSizeInventory(); ++i2)
                 {
-                    ItemStack itemstack1 = p_77569_1_.func_70301_a(i2);
+                    ItemStack itemstack1 = inv.getStackInSlot(i2);
 
-                    if (!itemstack1.func_190926_b())
+                    if (!itemstack1.isEmpty())
                     {
-                        if (itemstack1.func_77973_b() == Items.field_151100_aR)
+                        if (itemstack1.getItem() == Items.DYE)
                         {
-                            list1.add(Integer.valueOf(ItemDye.field_150922_c[itemstack1.func_77960_j() & 15]));
+                            list1.add(Integer.valueOf(ItemDye.DYE_COLORS[itemstack1.getMetadata() & 15]));
                         }
-                        else if (itemstack1.func_77973_b() == Items.field_151154_bQ)
+                        else if (itemstack1.getItem() == Items.FIREWORK_CHARGE)
                         {
-                            this.field_92102_a = itemstack1.func_77946_l();
-                            this.field_92102_a.func_190920_e(1);
+                            this.resultItem = itemstack1.copy();
+                            this.resultItem.setCount(1);
                         }
                     }
                 }
@@ -198,9 +203,9 @@
                     aint[j2] = ((Integer)list1.get(j2)).intValue();
                 }
 
-                if (!this.field_92102_a.func_190926_b() && this.field_92102_a.func_77942_o())
+                if (!this.resultItem.isEmpty() && this.resultItem.hasTagCompound())
                 {
-                    NBTTagCompound nbttagcompound4 = this.field_92102_a.func_77978_p().func_74775_l("Explosion");
+                    NBTTagCompound nbttagcompound4 = this.resultItem.getTagCompound().getCompoundTag("Explosion");
 
                     if (nbttagcompound4 == null)
                     {
@@ -208,7 +213,7 @@
                     }
                     else
                     {
-                        nbttagcompound4.func_74783_a("FadeColors", aint);
+                        nbttagcompound4.setIntArray("FadeColors", aint);
                         return true;
                     }
                 }
@@ -228,33 +233,30 @@
         }
     }
 
-    public ItemStack func_77572_b(InventoryCrafting p_77572_1_)
+    public ItemStack getCraftingResult(InventoryCrafting inv)
     {
-        return this.field_92102_a.func_77946_l();
+        return this.resultItem.copy();
     }
 
-    public int func_77570_a()
+    public int getRecipeSize()
     {
         return 10;
     }
 
-    public ItemStack func_77571_b()
+    public ItemStack getRecipeOutput()
     {
-        return this.field_92102_a;
+        return this.resultItem;
     }
 
-    public NonNullList<ItemStack> func_179532_b(InventoryCrafting p_179532_1_)
+    public NonNullList<ItemStack> getRemainingItems(InventoryCrafting inv)
     {
-        NonNullList<ItemStack> nonnulllist = NonNullList.<ItemStack>func_191197_a(p_179532_1_.func_70302_i_(), ItemStack.field_190927_a);
+        NonNullList<ItemStack> nonnulllist = NonNullList.<ItemStack>withSize(inv.getSizeInventory(), ItemStack.EMPTY);
 
         for (int i = 0; i < nonnulllist.size(); ++i)
         {
-            ItemStack itemstack = p_179532_1_.func_70301_a(i);
+            ItemStack itemstack = inv.getStackInSlot(i);
 
-            if (itemstack.func_77973_b().func_77634_r())
-            {
-                nonnulllist.set(i, new ItemStack(itemstack.func_77973_b().func_77668_q()));
-            }
+            nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
         }
 
         return nonnulllist;
