--- ../src-base/minecraft/net/minecraft/item/crafting/RecipesArmorDyes.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/RecipesArmorDyes.java
@@ -11,8 +11,13 @@
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
 
-public class RecipesArmorDyes implements IRecipe
+public class RecipesArmorDyes extends ShapelessRecipes implements IRecipe
 {
+    // CraftBukkit start - Delegate to new parent class with bogus info
+    public RecipesArmorDyes() {
+        super(new ItemStack(Items.field_151024_Q, 0, 0), java.util.Arrays.asList(new ItemStack(Items.field_151100_aR, 0, 5)));
+    }
+    // CraftBukkit end
     public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
     {
         ItemStack itemstack = ItemStack.field_190927_a;
@@ -147,11 +152,7 @@
         for (int i = 0; i < nonnulllist.size(); ++i)
         {
             ItemStack itemstack = p_179532_1_.func_70301_a(i);
-
-            if (itemstack.func_77973_b().func_77634_r())
-            {
-                nonnulllist.set(i, new ItemStack(itemstack.func_77973_b().func_77668_q()));
-            }
+            nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
         }
 
         return nonnulllist;
