--- ../src-base/minecraft/net/minecraft/item/crafting/ShapelessRecipes.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/ShapelessRecipes.java
@@ -6,6 +6,8 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.inventory.CraftShapelessRecipe;
 
 public class ShapelessRecipes implements IRecipe
 {
@@ -17,7 +19,20 @@
         this.field_77580_a = p_i1918_1_;
         this.field_77579_b = p_i1918_2_;
     }
-
+    // CraftBukkit start
+    @SuppressWarnings("unchecked")
+    @Override
+    public org.bukkit.inventory.ShapelessRecipe toBukkitRecipe() {
+        CraftItemStack result = CraftItemStack.asCraftMirror(this.field_77580_a);
+        CraftShapelessRecipe recipe = new CraftShapelessRecipe(result, this);
+        for (ItemStack stack : (List<ItemStack>) this.field_77579_b) {
+            if (stack != null) {
+                recipe.addIngredient(org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(stack.func_77973_b()), stack.func_77960_j());
+            }
+        }
+        return recipe;
+    }
+    // CraftBukkit end
     public ItemStack func_77571_b()
     {
         return this.field_77580_a;
@@ -30,11 +45,7 @@
         for (int i = 0; i < nonnulllist.size(); ++i)
         {
             ItemStack itemstack = p_179532_1_.func_70301_a(i);
-
-            if (itemstack.func_77973_b().func_77634_r())
-            {
-                nonnulllist.set(i, new ItemStack(itemstack.func_77973_b().func_77668_q()));
-            }
+            nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
         }
 
         return nonnulllist;
