--- ../src-base/minecraft/net/minecraft/item/crafting/ShieldRecipes.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/ShieldRecipes.java
@@ -7,6 +7,7 @@
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
+import org.bukkit.inventory.ShapelessRecipe;
 
 public class ShieldRecipes
 {
@@ -16,11 +17,14 @@
         p_185256_1_.func_180302_a(new ShieldRecipes.Decoration());
     }
 
-    public static class Decoration implements IRecipe
+    public static class Decoration extends ShapelessRecipes implements IRecipe
         {
-            private Decoration()
-            {
+
+            // CraftBukkit start - Delegate to new parent class with bogus info
+            private Decoration() {
+                super(new ItemStack(Items.field_185159_cQ, 0 ,0), java.util.Arrays.asList(new ItemStack(Items.field_179564_cE, 0, 0)));
             }
+            // CraftBukkit end
 
             public boolean func_77569_a(InventoryCrafting p_77569_1_, World p_77569_2_)
             {
