--- ../src-base/minecraft/net/minecraft/entity/item/EntityFireworkRocket.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityFireworkRocket.java
@@ -23,10 +23,11 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class EntityFireworkRocket extends Entity
 {
-    private static final DataParameter<ItemStack> field_184566_a = EntityDataManager.<ItemStack>func_187226_a(EntityFireworkRocket.class, DataSerializers.field_187196_f);
+    public static final DataParameter<ItemStack> field_184566_a = EntityDataManager.<ItemStack>func_187226_a(EntityFireworkRocket.class, DataSerializers.field_187196_f);
     private static final DataParameter<Integer> field_191512_b = EntityDataManager.<Integer>func_187226_a(EntityFireworkRocket.class, DataSerializers.field_187192_b);
     private int field_92056_a;
     public int field_92055_b;
@@ -187,8 +188,12 @@
 
         if (!this.field_70170_p.field_72995_K && this.field_92056_a > this.field_92055_b)
         {
-            this.field_70170_p.func_72960_a(this, (byte)17);
-            this.func_191510_k();
+            // CraftBukkit start
+            if (!CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+                this.field_70170_p.func_72960_a(this, (byte) 17);
+                this.func_191510_k();
+            }
+            // CraftBukkit end
             this.func_70106_y();
         }
     }
@@ -209,7 +214,9 @@
         {
             if (this.field_191513_e != null)
             {
-                this.field_191513_e.func_70097_a(DamageSource.field_191552_t, (float)(5 + nbttaglist.func_74745_c() * 2));
+                CraftEventFactory.entityDamage = this; // CraftBukkit
+                this.field_191513_e.func_70097_a(DamageSource.field_191552_t, (float) (5 + nbttaglist.func_74745_c() * 2));
+                CraftEventFactory.entityDamage = null; // CraftBukkit
             }
 
             double d0 = 5.0D;
@@ -235,7 +242,9 @@
                     if (flag)
                     {
                         float f1 = f * (float)Math.sqrt((5.0D - (double)this.func_70032_d(entitylivingbase)) / 5.0D);
+                        CraftEventFactory.entityDamage = this; // CraftBukkit
                         entitylivingbase.func_70097_a(DamageSource.field_191552_t, f1);
+                        CraftEventFactory.entityDamage = null; // CraftBukkit
                     }
                 }
             }
