--- ../src-base/minecraft/net/minecraft/entity/item/EntityMinecartContainer.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityMinecartContainer.java
@@ -1,5 +1,6 @@
 package net.minecraft.entity.item;
 
+import java.util.List;
 import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.entity.Entity;
@@ -23,39 +24,72 @@
 import net.minecraft.world.storage.loot.ILootContainer;
 import net.minecraft.world.storage.loot.LootContext;
 import net.minecraft.world.storage.loot.LootTable;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.inventory.InventoryHolder;
 
 public abstract class EntityMinecartContainer extends EntityMinecart implements ILockableContainer, ILootContainer
 {
-    private NonNullList<ItemStack> field_94113_a = NonNullList.<ItemStack>func_191197_a(36, ItemStack.field_190927_a);
-    public boolean field_94112_b = true;
-    private ResourceLocation field_184290_c;
-    private long field_184291_d;
+    protected NonNullList<ItemStack> minecartContainerItems = NonNullList.<ItemStack>withSize(36, ItemStack.EMPTY);
+    public boolean dropContentsWhenDead = true;
+    private ResourceLocation lootTable;
+    private long lootTableSeed;
 
-    public EntityMinecartContainer(World p_i1716_1_)
+    public EntityMinecartContainer(World worldIn)
     {
-        super(p_i1716_1_);
+        super(worldIn);
     }
 
-    public EntityMinecartContainer(World p_i1717_1_, double p_i1717_2_, double p_i1717_4_, double p_i1717_6_)
+    public EntityMinecartContainer(World worldIn, double x, double y, double z)
     {
-        super(p_i1717_1_, p_i1717_2_, p_i1717_4_, p_i1717_6_);
+        super(worldIn, x, y, z);
     }
+    // CraftBukkit start
+    public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+    private int maxStack = MAX_STACK;
 
-    public void func_94095_a(DamageSource p_94095_1_)
+    public List<ItemStack> getContents() {
+        return this.minecartContainerItems;
+    }
+
+    public void onOpen(CraftHumanEntity who) {
+        transaction.add(who);
+    }
+
+    public void onClose(CraftHumanEntity who) {
+        transaction.remove(who);
+    }
+
+    public List<HumanEntity> getViewers() {
+        return transaction;
+    }
+
+    public InventoryHolder getOwner() {
+        org.bukkit.entity.Entity cart = getBukkitEntity();
+        if(cart instanceof InventoryHolder) return (InventoryHolder) cart;
+        return null;
+    }
+
+    public void setMaxStackSize(int size) {
+        maxStack = size;
+    }
+    // CraftBukkit end
+    public void killMinecart(DamageSource source)
     {
-        super.func_94095_a(p_94095_1_);
+        super.killMinecart(source);
 
-        if (this.field_70170_p.func_82736_K().func_82766_b("doEntityDrops"))
+        if (this.world.getGameRules().getBoolean("doEntityDrops"))
         {
-            InventoryHelper.func_180176_a(this.field_70170_p, this, this);
+            InventoryHelper.dropInventoryItems(this.world, this, this);
         }
     }
 
-    public boolean func_191420_l()
+    public boolean isEmpty()
     {
-        for (ItemStack itemstack : this.field_94113_a)
+        for (ItemStack itemstack : this.minecartContainerItems)
         {
-            if (!itemstack.func_190926_b())
+            if (!itemstack.isEmpty())
             {
                 return false;
             }
@@ -64,230 +98,256 @@
         return true;
     }
 
-    public ItemStack func_70301_a(int p_70301_1_)
+    public ItemStack getStackInSlot(int index)
     {
-        this.func_184288_f((EntityPlayer)null);
-        return (ItemStack)this.field_94113_a.get(p_70301_1_);
+        this.addLoot((EntityPlayer)null);
+        return (ItemStack)this.minecartContainerItems.get(index);
     }
 
-    public ItemStack func_70298_a(int p_70298_1_, int p_70298_2_)
+    public ItemStack decrStackSize(int index, int count)
     {
-        this.func_184288_f((EntityPlayer)null);
-        return ItemStackHelper.func_188382_a(this.field_94113_a, p_70298_1_, p_70298_2_);
+        this.addLoot((EntityPlayer)null);
+        return ItemStackHelper.getAndSplit(this.minecartContainerItems, index, count);
     }
 
-    public ItemStack func_70304_b(int p_70304_1_)
+    public ItemStack removeStackFromSlot(int index)
     {
-        this.func_184288_f((EntityPlayer)null);
-        ItemStack itemstack = (ItemStack)this.field_94113_a.get(p_70304_1_);
+        this.addLoot((EntityPlayer)null);
+        ItemStack itemstack = (ItemStack)this.minecartContainerItems.get(index);
 
-        if (itemstack.func_190926_b())
+        if (itemstack.isEmpty())
         {
-            return ItemStack.field_190927_a;
+            return ItemStack.EMPTY;
         }
         else
         {
-            this.field_94113_a.set(p_70304_1_, ItemStack.field_190927_a);
+            this.minecartContainerItems.set(index, ItemStack.EMPTY);
             return itemstack;
         }
     }
 
-    public void func_70299_a(int p_70299_1_, ItemStack p_70299_2_)
+    public void setInventorySlotContents(int index, ItemStack stack)
     {
-        this.func_184288_f((EntityPlayer)null);
-        this.field_94113_a.set(p_70299_1_, p_70299_2_);
+        this.addLoot((EntityPlayer)null);
+        this.minecartContainerItems.set(index, stack);
 
-        if (!p_70299_2_.func_190926_b() && p_70299_2_.func_190916_E() > this.func_70297_j_())
+        if (!stack.isEmpty() && stack.getCount() > this.getInventoryStackLimit())
         {
-            p_70299_2_.func_190920_e(this.func_70297_j_());
+            stack.setCount(this.getInventoryStackLimit());
         }
     }
 
-    public void func_70296_d()
+    public void markDirty()
     {
     }
 
-    public boolean func_70300_a(EntityPlayer p_70300_1_)
+    public boolean isUsableByPlayer(EntityPlayer player)
     {
-        return this.field_70128_L ? false : p_70300_1_.func_70068_e(this) <= 64.0D;
+        return this.isDead ? false : player.getDistanceSqToEntity(this) <= 64.0D;
     }
 
-    public void func_174889_b(EntityPlayer p_174889_1_)
+    public void openInventory(EntityPlayer player)
     {
     }
 
-    public void func_174886_c(EntityPlayer p_174886_1_)
+    public void closeInventory(EntityPlayer player)
     {
     }
 
-    public boolean func_94041_b(int p_94041_1_, ItemStack p_94041_2_)
+    public boolean isItemValidForSlot(int index, ItemStack stack)
     {
         return true;
     }
 
-    public int func_70297_j_()
+    public int getInventoryStackLimit()
     {
         return 64;
     }
 
     @Nullable
-    public Entity func_184204_a(int p_184204_1_)
+    public Entity changeDimension(int dimensionIn)
     {
-        this.field_94112_b = false;
-        return super.func_184204_a(p_184204_1_);
+        this.dropContentsWhenDead = false;
+        return super.changeDimension(dimensionIn);
     }
 
-    public void func_70106_y()
+    public void setDead()
     {
-        if (this.field_94112_b)
+        if (this.dropContentsWhenDead)
         {
-            InventoryHelper.func_180176_a(this.field_70170_p, this, this);
+            InventoryHelper.dropInventoryItems(this.world, this, this);
         }
 
-        super.func_70106_y();
+        super.setDead();
     }
 
-    public void func_184174_b(boolean p_184174_1_)
+    public void setDropItemsWhenDead(boolean dropWhenDead)
     {
-        this.field_94112_b = p_184174_1_;
+        this.dropContentsWhenDead = dropWhenDead;
     }
 
-    public static void func_190574_b(DataFixer p_190574_0_, Class<?> p_190574_1_)
+    public static void addDataFixers(DataFixer p_190574_0_, Class<?> p_190574_1_)
     {
-        EntityMinecart.func_189669_a(p_190574_0_, p_190574_1_);
-        p_190574_0_.func_188258_a(FixTypes.ENTITY, new ItemStackDataLists(p_190574_1_, new String[] {"Items"}));
+        EntityMinecart.registerFixesMinecart(p_190574_0_, p_190574_1_);
+        p_190574_0_.registerWalker(FixTypes.ENTITY, new ItemStackDataLists(p_190574_1_, new String[] {"Items"}));
     }
 
-    protected void func_70014_b(NBTTagCompound p_70014_1_)
+    protected void writeEntityToNBT(NBTTagCompound compound)
     {
-        super.func_70014_b(p_70014_1_);
+        super.writeEntityToNBT(compound);
 
-        if (this.field_184290_c != null)
+        if (this.lootTable != null)
         {
-            p_70014_1_.func_74778_a("LootTable", this.field_184290_c.toString());
+            compound.setString("LootTable", this.lootTable.toString());
 
-            if (this.field_184291_d != 0L)
+            if (this.lootTableSeed != 0L)
             {
-                p_70014_1_.func_74772_a("LootTableSeed", this.field_184291_d);
+                compound.setLong("LootTableSeed", this.lootTableSeed);
             }
         }
         else
         {
-            ItemStackHelper.func_191282_a(p_70014_1_, this.field_94113_a);
+            ItemStackHelper.saveAllItems(compound, this.minecartContainerItems);
         }
     }
 
-    protected void func_70037_a(NBTTagCompound p_70037_1_)
+    protected void readEntityFromNBT(NBTTagCompound compound)
     {
-        super.func_70037_a(p_70037_1_);
-        this.field_94113_a = NonNullList.<ItemStack>func_191197_a(this.func_70302_i_(), ItemStack.field_190927_a);
+        super.readEntityFromNBT(compound);
+        this.minecartContainerItems = NonNullList.<ItemStack>withSize(this.getSizeInventory(), ItemStack.EMPTY);
 
-        if (p_70037_1_.func_150297_b("LootTable", 8))
+        if (compound.hasKey("LootTable", 8))
         {
-            this.field_184290_c = new ResourceLocation(p_70037_1_.func_74779_i("LootTable"));
-            this.field_184291_d = p_70037_1_.func_74763_f("LootTableSeed");
+            this.lootTable = new ResourceLocation(compound.getString("LootTable"));
+            this.lootTableSeed = compound.getLong("LootTableSeed");
         }
         else
         {
-            ItemStackHelper.func_191283_b(p_70037_1_, this.field_94113_a);
+            ItemStackHelper.loadAllItems(compound, this.minecartContainerItems);
         }
     }
 
-    public boolean func_184230_a(EntityPlayer p_184230_1_, EnumHand p_184230_2_)
+    public boolean processInitialInteract(EntityPlayer player, EnumHand hand)
     {
-        if (!this.field_70170_p.field_72995_K)
+        if(net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.minecart.MinecartInteractEvent(this, player, hand))) return true;
+        if (!this.world.isRemote)
         {
-            p_184230_1_.func_71007_a(this);
+            player.displayGUIChest(this);
         }
 
         return true;
     }
 
-    protected void func_94101_h()
+    protected void applyDrag()
     {
         float f = 0.98F;
 
-        if (this.field_184290_c == null)
+        if (this.lootTable == null)
         {
-            int i = 15 - Container.func_94526_b(this);
+            int i = 15 - Container.calcRedstoneFromInventory(this);
             f += (float)i * 0.001F;
         }
 
-        this.field_70159_w *= (double)f;
-        this.field_70181_x *= 0.0D;
-        this.field_70179_y *= (double)f;
+        this.motionX *= (double)f;
+        this.motionY *= 0.0D;
+        this.motionZ *= (double)f;
     }
 
-    public int func_174887_a_(int p_174887_1_)
+    public int getField(int id)
     {
         return 0;
     }
 
-    public void func_174885_b(int p_174885_1_, int p_174885_2_)
+    public void setField(int id, int value)
     {
     }
 
-    public int func_174890_g()
+    public int getFieldCount()
     {
         return 0;
     }
 
-    public boolean func_174893_q_()
+    public boolean isLocked()
     {
         return false;
     }
 
-    public void func_174892_a(LockCode p_174892_1_)
+    public void setLockCode(LockCode code)
     {
     }
 
-    public LockCode func_174891_i()
+    public LockCode getLockCode()
     {
-        return LockCode.field_180162_a;
+        return LockCode.EMPTY_CODE;
     }
 
-    public void func_184288_f(@Nullable EntityPlayer p_184288_1_)
+    public void addLoot(@Nullable EntityPlayer player)
     {
-        if (this.field_184290_c != null)
+        if (this.lootTable != null)
         {
-            LootTable loottable = this.field_70170_p.func_184146_ak().func_186521_a(this.field_184290_c);
-            this.field_184290_c = null;
+            LootTable loottable = this.world.getLootTableManager().getLootTableFromLocation(this.lootTable);
+            this.lootTable = null;
             Random random;
 
-            if (this.field_184291_d == 0L)
+            if (this.lootTableSeed == 0L)
             {
                 random = new Random();
             }
             else
             {
-                random = new Random(this.field_184291_d);
+                random = new Random(this.lootTableSeed);
             }
 
-            LootContext.Builder lootcontext$builder = new LootContext.Builder((WorldServer)this.field_70170_p);
+            LootContext.Builder lootcontext$builder = new LootContext.Builder((WorldServer)this.world);
 
-            if (p_184288_1_ != null)
+            if (player != null)
             {
-                lootcontext$builder.func_186469_a(p_184288_1_.func_184817_da());
+                lootcontext$builder.withLuck(player.getLuck());
             }
 
-            loottable.func_186460_a(this, random, lootcontext$builder.func_186471_a());
+            loottable.fillInventory(this, random, lootcontext$builder.build());
         }
     }
 
-    public void func_174888_l()
+    public net.minecraftforge.items.IItemHandler itemHandler = new net.minecraftforge.items.wrapper.InvWrapper(this);
+
+    @SuppressWarnings("unchecked")
+    @Override
+    @Nullable
+    public <T> T getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable net.minecraft.util.EnumFacing facing)
     {
-        this.func_184288_f((EntityPlayer)null);
-        this.field_94113_a.clear();
+        if (capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY)
+        {
+            return (T) itemHandler;
+        }
+        return super.getCapability(capability, facing);
     }
 
-    public void func_184289_a(ResourceLocation p_184289_1_, long p_184289_2_)
+    @Override
+    public boolean hasCapability(net.minecraftforge.common.capabilities.Capability<?> capability, @Nullable net.minecraft.util.EnumFacing facing)
     {
-        this.field_184290_c = p_184289_1_;
-        this.field_184291_d = p_184289_2_;
+        return capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY || super.hasCapability(capability, facing);
     }
 
-    public ResourceLocation func_184276_b()
+    public void clear()
     {
-        return this.field_184290_c;
+        this.addLoot((EntityPlayer)null);
+        this.minecartContainerItems.clear();
     }
+
+    public void setLootTable(ResourceLocation lootTableIn, long lootTableSeedIn)
+    {
+        this.lootTable = lootTableIn;
+        this.lootTableSeed = lootTableSeedIn;
+    }
+
+    public ResourceLocation getLootTable()
+    {
+        return this.lootTable;
+    }
+    @Override
+    public Location getLocation()
+    {
+        return getBukkitEntity().getLocation();
+    }
 }
