--- ../src-base/minecraft/net/minecraft/entity/item/EntityEnderCrystal.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityEnderCrystal.java
@@ -17,6 +17,7 @@
 import net.minecraft.world.end.DragonFightManager;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class EntityEnderCrystal extends Entity
 {
@@ -56,13 +57,15 @@
         this.field_70166_s = this.field_70161_v;
         ++this.field_70261_a;
 
-        if (!this.field_70170_p.field_72995_K)
-        {
+        if (!this.field_70170_p.field_72995_K) {
             BlockPos blockpos = new BlockPos(this);
 
-            if (this.field_70170_p.field_73011_w instanceof WorldProviderEnd && this.field_70170_p.func_180495_p(blockpos).func_177230_c() != Blocks.field_150480_ab)
-            {
-                this.field_70170_p.func_175656_a(blockpos, Blocks.field_150480_ab.func_176223_P());
+            if (this.field_70170_p.field_73011_w instanceof WorldProviderEnd && this.field_70170_p.func_180495_p(blockpos).func_177230_c() != Blocks.field_150480_ab) {
+                // CraftBukkit start
+                if (!CraftEventFactory.callBlockIgniteEvent(this.field_70170_p, blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p(), this).isCancelled()) {
+                    this.field_70170_p.func_175656_a(blockpos, Blocks.field_150480_ab.func_176223_P());
+                }
+                // CraftBukkit end            }
             }
         }
     }
