--- ../src-base/minecraft/net/minecraft/entity/item/EntityArmorStand.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityArmorStand.java
@@ -38,6 +38,13 @@
 import net.minecraft.world.WorldServer;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.CraftEquipmentSlot;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.entity.ArmorStand;
+import org.bukkit.entity.Player;
+import org.bukkit.event.player.PlayerArmorStandManipulateEvent;
+import org.bukkit.inventory.EquipmentSlot;
 
 public class EntityArmorStand extends EntityLivingBase
 {
@@ -58,7 +65,7 @@
     {
         public boolean apply(@Nullable Entity p_apply_1_)
         {
-            return p_apply_1_ instanceof EntityMinecart && ((EntityMinecart)p_apply_1_).func_184264_v() == EntityMinecart.Type.RIDEABLE;
+            return p_apply_1_ instanceof EntityMinecart && ((EntityMinecart)p_apply_1_).canBeRidden();
         }
     };
     private final NonNullList<ItemStack> field_184799_bw;
@@ -461,6 +468,21 @@
                 {
                     ItemStack itemstack2 = p_184795_3_.func_77946_l();
                     itemstack2.func_190920_e(1);
+                    // CraftBukkit start
+                    org.bukkit.inventory.ItemStack armorStandItem = CraftItemStack.asCraftMirror(p_184795_3_);
+                    org.bukkit.inventory.ItemStack playerHeldItem = CraftItemStack.asCraftMirror(itemstack);
+
+                    Player pl = (Player) p_184795_1_.getBukkitEntity();
+                    ArmorStand self = (ArmorStand) this.getBukkitEntity();
+
+                    EquipmentSlot slot = CraftEquipmentSlot.getSlot(p_184795_2_);
+                    PlayerArmorStandManipulateEvent armorStandManipulateEvent = new PlayerArmorStandManipulateEvent(pl,self,playerHeldItem,armorStandItem,slot);
+                    this.field_70170_p.getServer().getPluginManager().callEvent(armorStandManipulateEvent);
+
+                    if (armorStandManipulateEvent.isCancelled()) {
+                        return;
+                    }
+                    // CraftBukkit end
                     this.func_184201_a(p_184795_2_, itemstack2);
                 }
                 else if (!p_184795_3_.func_190926_b() && p_184795_3_.func_190916_E() > 1)
@@ -484,11 +506,16 @@
 
     public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
     {
+        // CraftBukkit start
+        if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_70097_1_, p_70097_2_)) {
+            return false;
+        }
+        // CraftBukkit end
         if (!this.field_70170_p.field_72995_K && !this.field_70128_L)
         {
             if (DamageSource.field_76380_i.equals(p_70097_1_))
             {
-                this.func_70106_y();
+                this.func_174812_G();
                 return false;
             }
             else if (!this.func_180431_b(p_70097_1_) && !this.field_175436_h && !this.func_181026_s())
@@ -496,7 +523,7 @@
                 if (p_70097_1_.func_94541_c())
                 {
                     this.func_175409_C();
-                    this.func_70106_y();
+                    this.func_174812_G();
                     return false;
                 }
                 else if (DamageSource.field_76372_a.equals(p_70097_1_))
@@ -541,7 +568,7 @@
                         {
                             this.func_190773_I();
                             this.func_175412_z();
-                            this.func_70106_y();
+                            this.func_174812_G();
                             return false;
                         }
                         else
@@ -557,7 +584,7 @@
                             {
                                 this.func_175421_A();
                                 this.func_175412_z();
-                                this.func_70106_y();
+                                this.func_174812_G();
                             }
 
                             return false;
@@ -623,7 +650,7 @@
         if (f <= 0.5F)
         {
             this.func_175409_C();
-            this.func_70106_y();
+            this.func_174812_G();
         }
         else
         {
@@ -633,7 +660,7 @@
 
     private void func_175421_A()
     {
-        Block.func_180635_a(this.field_70170_p, new BlockPos(this), new ItemStack(Items.field_179565_cj));
+        drops.add(CraftItemStack.asBukkitCopy(new ItemStack(Items.field_179565_cj))); // CraftBukkit - add to drops
         this.func_175409_C();
     }
 
@@ -647,7 +674,7 @@
 
             if (!itemstack.func_190926_b())
             {
-                Block.func_180635_a(this.field_70170_p, (new BlockPos(this)).func_177984_a(), itemstack);
+                drops.add(CraftItemStack.asBukkitCopy(itemstack)); // CraftBukkit - add to drops
                 this.field_184799_bw.set(i, ItemStack.field_190927_a);
             }
         }
