--- ../src-base/minecraft/net/minecraft/entity/item/EntityEnderCrystal.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityEnderCrystal.java
@@ -17,104 +17,107 @@
 import net.minecraft.world.end.DragonFightManager;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class EntityEnderCrystal extends Entity
 {
-    private static final DataParameter<Optional<BlockPos>> field_184521_b = EntityDataManager.<Optional<BlockPos>>func_187226_a(EntityEnderCrystal.class, DataSerializers.field_187201_k);
-    private static final DataParameter<Boolean> field_184522_c = EntityDataManager.<Boolean>func_187226_a(EntityEnderCrystal.class, DataSerializers.field_187198_h);
-    public int field_70261_a;
+    private static final DataParameter<Optional<BlockPos>> BEAM_TARGET = EntityDataManager.<Optional<BlockPos>>createKey(EntityEnderCrystal.class, DataSerializers.OPTIONAL_BLOCK_POS);
+    private static final DataParameter<Boolean> SHOW_BOTTOM = EntityDataManager.<Boolean>createKey(EntityEnderCrystal.class, DataSerializers.BOOLEAN);
+    public int innerRotation;
 
-    public EntityEnderCrystal(World p_i1698_1_)
+    public EntityEnderCrystal(World worldIn)
     {
-        super(p_i1698_1_);
-        this.field_70156_m = true;
-        this.func_70105_a(2.0F, 2.0F);
-        this.field_70261_a = this.field_70146_Z.nextInt(100000);
+        super(worldIn);
+        this.preventEntitySpawning = true;
+        this.setSize(2.0F, 2.0F);
+        this.innerRotation = this.rand.nextInt(100000);
     }
 
-    public EntityEnderCrystal(World p_i1699_1_, double p_i1699_2_, double p_i1699_4_, double p_i1699_6_)
+    public EntityEnderCrystal(World worldIn, double x, double y, double z)
     {
-        this(p_i1699_1_);
-        this.func_70107_b(p_i1699_2_, p_i1699_4_, p_i1699_6_);
+        this(worldIn);
+        this.setPosition(x, y, z);
     }
 
-    protected boolean func_70041_e_()
+    protected boolean canTriggerWalking()
     {
         return false;
     }
 
-    protected void func_70088_a()
+    protected void entityInit()
     {
-        this.func_184212_Q().func_187214_a(field_184521_b, Optional.<BlockPos>absent());
-        this.func_184212_Q().func_187214_a(field_184522_c, Boolean.valueOf(true));
+        this.getDataManager().register(BEAM_TARGET, Optional.<BlockPos>absent());
+        this.getDataManager().register(SHOW_BOTTOM, Boolean.valueOf(true));
     }
 
-    public void func_70071_h_()
+    public void onUpdate()
     {
-        this.field_70169_q = this.field_70165_t;
-        this.field_70167_r = this.field_70163_u;
-        this.field_70166_s = this.field_70161_v;
-        ++this.field_70261_a;
+        this.prevPosX = this.posX;
+        this.prevPosY = this.posY;
+        this.prevPosZ = this.posZ;
+        ++this.innerRotation;
 
-        if (!this.field_70170_p.field_72995_K)
-        {
+        if (!this.world.isRemote) {
             BlockPos blockpos = new BlockPos(this);
 
-            if (this.field_70170_p.field_73011_w instanceof WorldProviderEnd && this.field_70170_p.func_180495_p(blockpos).func_177230_c() != Blocks.field_150480_ab)
-            {
-                this.field_70170_p.func_175656_a(blockpos, Blocks.field_150480_ab.func_176223_P());
+            if (this.world.provider instanceof WorldProviderEnd && this.world.getBlockState(blockpos).getBlock() != Blocks.FIRE) {
+                // CraftBukkit start
+                if (!CraftEventFactory.callBlockIgniteEvent(this.world, blockpos.getX(), blockpos.getY(), blockpos.getZ(), this).isCancelled()) {
+                    this.world.setBlockState(blockpos, Blocks.FIRE.getDefaultState());
+                }
+                // CraftBukkit end            }
             }
         }
     }
 
-    protected void func_70014_b(NBTTagCompound p_70014_1_)
+    protected void writeEntityToNBT(NBTTagCompound compound)
     {
-        if (this.func_184518_j() != null)
+        if (this.getBeamTarget() != null)
         {
-            p_70014_1_.func_74782_a("BeamTarget", NBTUtil.func_186859_a(this.func_184518_j()));
+            compound.setTag("BeamTarget", NBTUtil.createPosTag(this.getBeamTarget()));
         }
 
-        p_70014_1_.func_74757_a("ShowBottom", this.func_184520_k());
+        compound.setBoolean("ShowBottom", this.shouldShowBottom());
     }
 
-    protected void func_70037_a(NBTTagCompound p_70037_1_)
+    protected void readEntityFromNBT(NBTTagCompound compound)
     {
-        if (p_70037_1_.func_150297_b("BeamTarget", 10))
+        if (compound.hasKey("BeamTarget", 10))
         {
-            this.func_184516_a(NBTUtil.func_186861_c(p_70037_1_.func_74775_l("BeamTarget")));
+            this.setBeamTarget(NBTUtil.getPosFromTag(compound.getCompoundTag("BeamTarget")));
         }
 
-        if (p_70037_1_.func_150297_b("ShowBottom", 1))
+        if (compound.hasKey("ShowBottom", 1))
         {
-            this.func_184517_a(p_70037_1_.func_74767_n("ShowBottom"));
+            this.setShowBottom(compound.getBoolean("ShowBottom"));
         }
     }
 
-    public boolean func_70067_L()
+    public boolean canBeCollidedWith()
     {
         return true;
     }
 
-    public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
+    public boolean attackEntityFrom(DamageSource source, float amount)
     {
-        if (this.func_180431_b(p_70097_1_))
+        if (this.isEntityInvulnerable(source))
         {
             return false;
         }
-        else if (p_70097_1_.func_76346_g() instanceof EntityDragon)
+        else if (source.getEntity() instanceof EntityDragon)
         {
             return false;
         }
         else
         {
-            if (!this.field_70128_L && !this.field_70170_p.field_72995_K)
+            if (!this.isDead && !this.world.isRemote)
             {
-                this.func_70106_y();
+                this.setDead();
 
-                if (!this.field_70170_p.field_72995_K)
+                if (!this.world.isRemote)
                 {
-                    this.field_70170_p.func_72876_a((Entity)null, this.field_70165_t, this.field_70163_u, this.field_70161_v, 6.0F, true);
-                    this.func_184519_a(p_70097_1_);
+                    this.world.createExplosion((Entity)null, this.posX, this.posY, this.posZ, 6.0F, true);
+                    this.onCrystalDestroyed(source);
                 }
             }
 
@@ -122,50 +125,50 @@
         }
     }
 
-    public void func_174812_G()
+    public void onKillCommand()
     {
-        this.func_184519_a(DamageSource.field_76377_j);
-        super.func_174812_G();
+        this.onCrystalDestroyed(DamageSource.GENERIC);
+        super.onKillCommand();
     }
 
-    private void func_184519_a(DamageSource p_184519_1_)
+    private void onCrystalDestroyed(DamageSource source)
     {
-        if (this.field_70170_p.field_73011_w instanceof WorldProviderEnd)
+        if (this.world.provider instanceof WorldProviderEnd)
         {
-            WorldProviderEnd worldproviderend = (WorldProviderEnd)this.field_70170_p.field_73011_w;
-            DragonFightManager dragonfightmanager = worldproviderend.func_186063_s();
+            WorldProviderEnd worldproviderend = (WorldProviderEnd)this.world.provider;
+            DragonFightManager dragonfightmanager = worldproviderend.getDragonFightManager();
 
             if (dragonfightmanager != null)
             {
-                dragonfightmanager.func_186090_a(this, p_184519_1_);
+                dragonfightmanager.onCrystalDestroyed(this, source);
             }
         }
     }
 
-    public void func_184516_a(@Nullable BlockPos p_184516_1_)
+    public void setBeamTarget(@Nullable BlockPos beamTarget)
     {
-        this.func_184212_Q().func_187227_b(field_184521_b, Optional.fromNullable(p_184516_1_));
+        this.getDataManager().set(BEAM_TARGET, Optional.fromNullable(beamTarget));
     }
 
     @Nullable
-    public BlockPos func_184518_j()
+    public BlockPos getBeamTarget()
     {
-        return (BlockPos)((Optional)this.func_184212_Q().func_187225_a(field_184521_b)).orNull();
+        return (BlockPos)((Optional)this.getDataManager().get(BEAM_TARGET)).orNull();
     }
 
-    public void func_184517_a(boolean p_184517_1_)
+    public void setShowBottom(boolean showBottom)
     {
-        this.func_184212_Q().func_187227_b(field_184522_c, Boolean.valueOf(p_184517_1_));
+        this.getDataManager().set(SHOW_BOTTOM, Boolean.valueOf(showBottom));
     }
 
-    public boolean func_184520_k()
+    public boolean shouldShowBottom()
     {
-        return ((Boolean)this.func_184212_Q().func_187225_a(field_184522_c)).booleanValue();
+        return ((Boolean)this.getDataManager().get(SHOW_BOTTOM)).booleanValue();
     }
 
     @SideOnly(Side.CLIENT)
-    public boolean func_70112_a(double p_70112_1_)
+    public boolean isInRangeToRenderDist(double distance)
     {
-        return super.func_70112_a(p_70112_1_) || this.func_184518_j() != null;
+        return super.isInRangeToRenderDist(distance) || this.getBeamTarget() != null;
     }
 }
