--- ../src-base/minecraft/net/minecraft/entity/item/EntityArmorStand.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityArmorStand.java
@@ -38,154 +38,161 @@
 import net.minecraft.world.WorldServer;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.CraftEquipmentSlot;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.entity.ArmorStand;
+import org.bukkit.entity.Player;
+import org.bukkit.event.player.PlayerArmorStandManipulateEvent;
+import org.bukkit.inventory.EquipmentSlot;
 
 public class EntityArmorStand extends EntityLivingBase
 {
-    private static final Rotations field_175435_a = new Rotations(0.0F, 0.0F, 0.0F);
-    private static final Rotations field_175433_b = new Rotations(0.0F, 0.0F, 0.0F);
-    private static final Rotations field_175434_c = new Rotations(-10.0F, 0.0F, -10.0F);
-    private static final Rotations field_175431_d = new Rotations(-15.0F, 0.0F, 10.0F);
-    private static final Rotations field_175432_e = new Rotations(-1.0F, 0.0F, -1.0F);
-    private static final Rotations field_175429_f = new Rotations(1.0F, 0.0F, 1.0F);
-    public static final DataParameter<Byte> field_184801_a = EntityDataManager.<Byte>func_187226_a(EntityArmorStand.class, DataSerializers.field_187191_a);
-    public static final DataParameter<Rotations> field_184802_b = EntityDataManager.<Rotations>func_187226_a(EntityArmorStand.class, DataSerializers.field_187199_i);
-    public static final DataParameter<Rotations> field_184803_c = EntityDataManager.<Rotations>func_187226_a(EntityArmorStand.class, DataSerializers.field_187199_i);
-    public static final DataParameter<Rotations> field_184804_d = EntityDataManager.<Rotations>func_187226_a(EntityArmorStand.class, DataSerializers.field_187199_i);
-    public static final DataParameter<Rotations> field_184805_e = EntityDataManager.<Rotations>func_187226_a(EntityArmorStand.class, DataSerializers.field_187199_i);
-    public static final DataParameter<Rotations> field_184806_f = EntityDataManager.<Rotations>func_187226_a(EntityArmorStand.class, DataSerializers.field_187199_i);
-    public static final DataParameter<Rotations> field_184807_g = EntityDataManager.<Rotations>func_187226_a(EntityArmorStand.class, DataSerializers.field_187199_i);
-    private static final Predicate<Entity> field_184798_bv = new Predicate<Entity>()
+    private static final Rotations DEFAULT_HEAD_ROTATION = new Rotations(0.0F, 0.0F, 0.0F);
+    private static final Rotations DEFAULT_BODY_ROTATION = new Rotations(0.0F, 0.0F, 0.0F);
+    private static final Rotations DEFAULT_LEFTARM_ROTATION = new Rotations(-10.0F, 0.0F, -10.0F);
+    private static final Rotations DEFAULT_RIGHTARM_ROTATION = new Rotations(-15.0F, 0.0F, 10.0F);
+    private static final Rotations DEFAULT_LEFTLEG_ROTATION = new Rotations(-1.0F, 0.0F, -1.0F);
+    private static final Rotations DEFAULT_RIGHTLEG_ROTATION = new Rotations(1.0F, 0.0F, 1.0F);
+    public static final DataParameter<Byte> STATUS = EntityDataManager.<Byte>createKey(EntityArmorStand.class, DataSerializers.BYTE);
+    public static final DataParameter<Rotations> HEAD_ROTATION = EntityDataManager.<Rotations>createKey(EntityArmorStand.class, DataSerializers.ROTATIONS);
+    public static final DataParameter<Rotations> BODY_ROTATION = EntityDataManager.<Rotations>createKey(EntityArmorStand.class, DataSerializers.ROTATIONS);
+    public static final DataParameter<Rotations> LEFT_ARM_ROTATION = EntityDataManager.<Rotations>createKey(EntityArmorStand.class, DataSerializers.ROTATIONS);
+    public static final DataParameter<Rotations> RIGHT_ARM_ROTATION = EntityDataManager.<Rotations>createKey(EntityArmorStand.class, DataSerializers.ROTATIONS);
+    public static final DataParameter<Rotations> LEFT_LEG_ROTATION = EntityDataManager.<Rotations>createKey(EntityArmorStand.class, DataSerializers.ROTATIONS);
+    public static final DataParameter<Rotations> RIGHT_LEG_ROTATION = EntityDataManager.<Rotations>createKey(EntityArmorStand.class, DataSerializers.ROTATIONS);
+    private static final Predicate<Entity> IS_RIDEABLE_MINECART = new Predicate<Entity>()
     {
         public boolean apply(@Nullable Entity p_apply_1_)
         {
-            return p_apply_1_ instanceof EntityMinecart && ((EntityMinecart)p_apply_1_).func_184264_v() == EntityMinecart.Type.RIDEABLE;
+            return p_apply_1_ instanceof EntityMinecart && ((EntityMinecart)p_apply_1_).canBeRidden();
         }
     };
-    private final NonNullList<ItemStack> field_184799_bw;
-    private final NonNullList<ItemStack> field_184800_bx;
-    private boolean field_175436_h;
-    public long field_175437_i;
-    private int field_175442_bg;
-    private boolean field_181028_bj;
-    public Rotations field_175443_bh;
-    public Rotations field_175444_bi;
-    public Rotations field_175438_bj;
-    public Rotations field_175439_bk;
-    public Rotations field_175440_bl;
-    public Rotations field_175441_bm;
+    private final NonNullList<ItemStack> handItems;
+    private final NonNullList<ItemStack> armorItems;
+    private boolean canInteract;
+    public long punchCooldown;
+    private int disabledSlots;
+    private boolean wasMarker;
+    public Rotations headRotation;
+    public Rotations bodyRotation;
+    public Rotations leftArmRotation;
+    public Rotations rightArmRotation;
+    public Rotations leftLegRotation;
+    public Rotations rightLegRotation;
 
-    public EntityArmorStand(World p_i45854_1_)
+    public EntityArmorStand(World worldIn)
     {
-        super(p_i45854_1_);
-        this.field_184799_bw = NonNullList.<ItemStack>func_191197_a(2, ItemStack.field_190927_a);
-        this.field_184800_bx = NonNullList.<ItemStack>func_191197_a(4, ItemStack.field_190927_a);
-        this.field_175443_bh = field_175435_a;
-        this.field_175444_bi = field_175433_b;
-        this.field_175438_bj = field_175434_c;
-        this.field_175439_bk = field_175431_d;
-        this.field_175440_bl = field_175432_e;
-        this.field_175441_bm = field_175429_f;
-        this.field_70145_X = this.func_189652_ae();
-        this.func_70105_a(0.5F, 1.975F);
+        super(worldIn);
+        this.handItems = NonNullList.<ItemStack>withSize(2, ItemStack.EMPTY);
+        this.armorItems = NonNullList.<ItemStack>withSize(4, ItemStack.EMPTY);
+        this.headRotation = DEFAULT_HEAD_ROTATION;
+        this.bodyRotation = DEFAULT_BODY_ROTATION;
+        this.leftArmRotation = DEFAULT_LEFTARM_ROTATION;
+        this.rightArmRotation = DEFAULT_RIGHTARM_ROTATION;
+        this.leftLegRotation = DEFAULT_LEFTLEG_ROTATION;
+        this.rightLegRotation = DEFAULT_RIGHTLEG_ROTATION;
+        this.noClip = this.hasNoGravity();
+        this.setSize(0.5F, 1.975F);
     }
 
-    public EntityArmorStand(World p_i45855_1_, double p_i45855_2_, double p_i45855_4_, double p_i45855_6_)
+    public EntityArmorStand(World worldIn, double posX, double posY, double posZ)
     {
-        this(p_i45855_1_);
-        this.func_70107_b(p_i45855_2_, p_i45855_4_, p_i45855_6_);
+        this(worldIn);
+        this.setPosition(posX, posY, posZ);
     }
 
-    public final void func_70105_a(float p_70105_1_, float p_70105_2_)
+    public final void setSize(float width, float height)
     {
-        double d0 = this.field_70165_t;
-        double d1 = this.field_70163_u;
-        double d2 = this.field_70161_v;
-        float f = this.func_181026_s() ? 0.0F : (this.func_70631_g_() ? 0.5F : 1.0F);
-        super.func_70105_a(p_70105_1_ * f, p_70105_2_ * f);
-        this.func_70107_b(d0, d1, d2);
+        double d0 = this.posX;
+        double d1 = this.posY;
+        double d2 = this.posZ;
+        float f = this.hasMarker() ? 0.0F : (this.isChild() ? 0.5F : 1.0F);
+        super.setSize(width * f, height * f);
+        this.setPosition(d0, d1, d2);
     }
 
-    public boolean func_70613_aW()
+    public boolean isServerWorld()
     {
-        return super.func_70613_aW() && !this.func_189652_ae();
+        return super.isServerWorld() && !this.hasNoGravity();
     }
 
-    protected void func_70088_a()
+    protected void entityInit()
     {
-        super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184801_a, Byte.valueOf((byte)0));
-        this.field_70180_af.func_187214_a(field_184802_b, field_175435_a);
-        this.field_70180_af.func_187214_a(field_184803_c, field_175433_b);
-        this.field_70180_af.func_187214_a(field_184804_d, field_175434_c);
-        this.field_70180_af.func_187214_a(field_184805_e, field_175431_d);
-        this.field_70180_af.func_187214_a(field_184806_f, field_175432_e);
-        this.field_70180_af.func_187214_a(field_184807_g, field_175429_f);
+        super.entityInit();
+        this.dataManager.register(STATUS, Byte.valueOf((byte)0));
+        this.dataManager.register(HEAD_ROTATION, DEFAULT_HEAD_ROTATION);
+        this.dataManager.register(BODY_ROTATION, DEFAULT_BODY_ROTATION);
+        this.dataManager.register(LEFT_ARM_ROTATION, DEFAULT_LEFTARM_ROTATION);
+        this.dataManager.register(RIGHT_ARM_ROTATION, DEFAULT_RIGHTARM_ROTATION);
+        this.dataManager.register(LEFT_LEG_ROTATION, DEFAULT_LEFTLEG_ROTATION);
+        this.dataManager.register(RIGHT_LEG_ROTATION, DEFAULT_RIGHTLEG_ROTATION);
     }
 
-    public Iterable<ItemStack> func_184214_aD()
+    public Iterable<ItemStack> getHeldEquipment()
     {
-        return this.field_184799_bw;
+        return this.handItems;
     }
 
-    public Iterable<ItemStack> func_184193_aE()
+    public Iterable<ItemStack> getArmorInventoryList()
     {
-        return this.field_184800_bx;
+        return this.armorItems;
     }
 
-    public ItemStack func_184582_a(EntityEquipmentSlot p_184582_1_)
+    public ItemStack getItemStackFromSlot(EntityEquipmentSlot slotIn)
     {
-        switch (p_184582_1_.func_188453_a())
+        switch (slotIn.getSlotType())
         {
             case HAND:
-                return (ItemStack)this.field_184799_bw.get(p_184582_1_.func_188454_b());
+                return (ItemStack)this.handItems.get(slotIn.getIndex());
             case ARMOR:
-                return (ItemStack)this.field_184800_bx.get(p_184582_1_.func_188454_b());
+                return (ItemStack)this.armorItems.get(slotIn.getIndex());
             default:
-                return ItemStack.field_190927_a;
+                return ItemStack.EMPTY;
         }
     }
 
-    public void func_184201_a(EntityEquipmentSlot p_184201_1_, ItemStack p_184201_2_)
+    public void setItemStackToSlot(EntityEquipmentSlot slotIn, ItemStack stack)
     {
-        switch (p_184201_1_.func_188453_a())
+        switch (slotIn.getSlotType())
         {
             case HAND:
-                this.func_184606_a_(p_184201_2_);
-                this.field_184799_bw.set(p_184201_1_.func_188454_b(), p_184201_2_);
+                this.playEquipSound(stack);
+                this.handItems.set(slotIn.getIndex(), stack);
                 break;
             case ARMOR:
-                this.func_184606_a_(p_184201_2_);
-                this.field_184800_bx.set(p_184201_1_.func_188454_b(), p_184201_2_);
+                this.playEquipSound(stack);
+                this.armorItems.set(slotIn.getIndex(), stack);
         }
     }
 
-    public boolean func_174820_d(int p_174820_1_, ItemStack p_174820_2_)
+    public boolean replaceItemInInventory(int inventorySlot, ItemStack itemStackIn)
     {
         EntityEquipmentSlot entityequipmentslot;
 
-        if (p_174820_1_ == 98)
+        if (inventorySlot == 98)
         {
             entityequipmentslot = EntityEquipmentSlot.MAINHAND;
         }
-        else if (p_174820_1_ == 99)
+        else if (inventorySlot == 99)
         {
             entityequipmentslot = EntityEquipmentSlot.OFFHAND;
         }
-        else if (p_174820_1_ == 100 + EntityEquipmentSlot.HEAD.func_188454_b())
+        else if (inventorySlot == 100 + EntityEquipmentSlot.HEAD.getIndex())
         {
             entityequipmentslot = EntityEquipmentSlot.HEAD;
         }
-        else if (p_174820_1_ == 100 + EntityEquipmentSlot.CHEST.func_188454_b())
+        else if (inventorySlot == 100 + EntityEquipmentSlot.CHEST.getIndex())
         {
             entityequipmentslot = EntityEquipmentSlot.CHEST;
         }
-        else if (p_174820_1_ == 100 + EntityEquipmentSlot.LEGS.func_188454_b())
+        else if (inventorySlot == 100 + EntityEquipmentSlot.LEGS.getIndex())
         {
             entityequipmentslot = EntityEquipmentSlot.LEGS;
         }
         else
         {
-            if (p_174820_1_ != 100 + EntityEquipmentSlot.FEET.func_188454_b())
+            if (inventorySlot != 100 + EntityEquipmentSlot.FEET.getIndex())
             {
                 return false;
             }
@@ -193,215 +200,215 @@
             entityequipmentslot = EntityEquipmentSlot.FEET;
         }
 
-        if (!p_174820_2_.func_190926_b() && !EntityLiving.func_184648_b(entityequipmentslot, p_174820_2_) && entityequipmentslot != EntityEquipmentSlot.HEAD)
+        if (!itemStackIn.isEmpty() && !EntityLiving.isItemStackInSlot(entityequipmentslot, itemStackIn) && entityequipmentslot != EntityEquipmentSlot.HEAD)
         {
             return false;
         }
         else
         {
-            this.func_184201_a(entityequipmentslot, p_174820_2_);
+            this.setItemStackToSlot(entityequipmentslot, itemStackIn);
             return true;
         }
     }
 
-    public static void func_189805_a(DataFixer p_189805_0_)
+    public static void registerFixesArmorStand(DataFixer fixer)
     {
-        p_189805_0_.func_188258_a(FixTypes.ENTITY, new ItemStackDataLists(EntityArmorStand.class, new String[] {"ArmorItems", "HandItems"}));
+        fixer.registerWalker(FixTypes.ENTITY, new ItemStackDataLists(EntityArmorStand.class, new String[] {"ArmorItems", "HandItems"}));
     }
 
-    public void func_70014_b(NBTTagCompound p_70014_1_)
+    public void writeEntityToNBT(NBTTagCompound compound)
     {
-        super.func_70014_b(p_70014_1_);
+        super.writeEntityToNBT(compound);
         NBTTagList nbttaglist = new NBTTagList();
 
-        for (ItemStack itemstack : this.field_184800_bx)
+        for (ItemStack itemstack : this.armorItems)
         {
             NBTTagCompound nbttagcompound = new NBTTagCompound();
 
-            if (!itemstack.func_190926_b())
+            if (!itemstack.isEmpty())
             {
-                itemstack.func_77955_b(nbttagcompound);
+                itemstack.writeToNBT(nbttagcompound);
             }
 
-            nbttaglist.func_74742_a(nbttagcompound);
+            nbttaglist.appendTag(nbttagcompound);
         }
 
-        p_70014_1_.func_74782_a("ArmorItems", nbttaglist);
+        compound.setTag("ArmorItems", nbttaglist);
         NBTTagList nbttaglist1 = new NBTTagList();
 
-        for (ItemStack itemstack1 : this.field_184799_bw)
+        for (ItemStack itemstack1 : this.handItems)
         {
             NBTTagCompound nbttagcompound1 = new NBTTagCompound();
 
-            if (!itemstack1.func_190926_b())
+            if (!itemstack1.isEmpty())
             {
-                itemstack1.func_77955_b(nbttagcompound1);
+                itemstack1.writeToNBT(nbttagcompound1);
             }
 
-            nbttaglist1.func_74742_a(nbttagcompound1);
+            nbttaglist1.appendTag(nbttagcompound1);
         }
 
-        p_70014_1_.func_74782_a("HandItems", nbttaglist1);
-        p_70014_1_.func_74757_a("Invisible", this.func_82150_aj());
-        p_70014_1_.func_74757_a("Small", this.func_175410_n());
-        p_70014_1_.func_74757_a("ShowArms", this.func_175402_q());
-        p_70014_1_.func_74768_a("DisabledSlots", this.field_175442_bg);
-        p_70014_1_.func_74757_a("NoBasePlate", this.func_175414_r());
+        compound.setTag("HandItems", nbttaglist1);
+        compound.setBoolean("Invisible", this.isInvisible());
+        compound.setBoolean("Small", this.isSmall());
+        compound.setBoolean("ShowArms", this.getShowArms());
+        compound.setInteger("DisabledSlots", this.disabledSlots);
+        compound.setBoolean("NoBasePlate", this.hasNoBasePlate());
 
-        if (this.func_181026_s())
+        if (this.hasMarker())
         {
-            p_70014_1_.func_74757_a("Marker", this.func_181026_s());
+            compound.setBoolean("Marker", this.hasMarker());
         }
 
-        p_70014_1_.func_74782_a("Pose", this.func_175419_y());
+        compound.setTag("Pose", this.readPoseFromNBT());
     }
 
-    public void func_70037_a(NBTTagCompound p_70037_1_)
+    public void readEntityFromNBT(NBTTagCompound compound)
     {
-        super.func_70037_a(p_70037_1_);
+        super.readEntityFromNBT(compound);
 
-        if (p_70037_1_.func_150297_b("ArmorItems", 9))
+        if (compound.hasKey("ArmorItems", 9))
         {
-            NBTTagList nbttaglist = p_70037_1_.func_150295_c("ArmorItems", 10);
+            NBTTagList nbttaglist = compound.getTagList("ArmorItems", 10);
 
-            for (int i = 0; i < this.field_184800_bx.size(); ++i)
+            for (int i = 0; i < this.armorItems.size(); ++i)
             {
-                this.field_184800_bx.set(i, new ItemStack(nbttaglist.func_150305_b(i)));
+                this.armorItems.set(i, new ItemStack(nbttaglist.getCompoundTagAt(i)));
             }
         }
 
-        if (p_70037_1_.func_150297_b("HandItems", 9))
+        if (compound.hasKey("HandItems", 9))
         {
-            NBTTagList nbttaglist1 = p_70037_1_.func_150295_c("HandItems", 10);
+            NBTTagList nbttaglist1 = compound.getTagList("HandItems", 10);
 
-            for (int j = 0; j < this.field_184799_bw.size(); ++j)
+            for (int j = 0; j < this.handItems.size(); ++j)
             {
-                this.field_184799_bw.set(j, new ItemStack(nbttaglist1.func_150305_b(j)));
+                this.handItems.set(j, new ItemStack(nbttaglist1.getCompoundTagAt(j)));
             }
         }
 
-        this.func_82142_c(p_70037_1_.func_74767_n("Invisible"));
-        this.func_175420_a(p_70037_1_.func_74767_n("Small"));
-        this.func_175413_k(p_70037_1_.func_74767_n("ShowArms"));
-        this.field_175442_bg = p_70037_1_.func_74762_e("DisabledSlots");
-        this.func_175426_l(p_70037_1_.func_74767_n("NoBasePlate"));
-        this.func_181027_m(p_70037_1_.func_74767_n("Marker"));
-        this.field_181028_bj = !this.func_181026_s();
-        this.field_70145_X = this.func_189652_ae();
-        NBTTagCompound nbttagcompound = p_70037_1_.func_74775_l("Pose");
-        this.func_175416_h(nbttagcompound);
+        this.setInvisible(compound.getBoolean("Invisible"));
+        this.setSmall(compound.getBoolean("Small"));
+        this.setShowArms(compound.getBoolean("ShowArms"));
+        this.disabledSlots = compound.getInteger("DisabledSlots");
+        this.setNoBasePlate(compound.getBoolean("NoBasePlate"));
+        this.setMarker(compound.getBoolean("Marker"));
+        this.wasMarker = !this.hasMarker();
+        this.noClip = this.hasNoGravity();
+        NBTTagCompound nbttagcompound = compound.getCompoundTag("Pose");
+        this.writePoseToNBT(nbttagcompound);
     }
 
-    private void func_175416_h(NBTTagCompound p_175416_1_)
+    private void writePoseToNBT(NBTTagCompound tagCompound)
     {
-        NBTTagList nbttaglist = p_175416_1_.func_150295_c("Head", 5);
-        this.func_175415_a(nbttaglist.func_82582_d() ? field_175435_a : new Rotations(nbttaglist));
-        NBTTagList nbttaglist1 = p_175416_1_.func_150295_c("Body", 5);
-        this.func_175424_b(nbttaglist1.func_82582_d() ? field_175433_b : new Rotations(nbttaglist1));
-        NBTTagList nbttaglist2 = p_175416_1_.func_150295_c("LeftArm", 5);
-        this.func_175405_c(nbttaglist2.func_82582_d() ? field_175434_c : new Rotations(nbttaglist2));
-        NBTTagList nbttaglist3 = p_175416_1_.func_150295_c("RightArm", 5);
-        this.func_175428_d(nbttaglist3.func_82582_d() ? field_175431_d : new Rotations(nbttaglist3));
-        NBTTagList nbttaglist4 = p_175416_1_.func_150295_c("LeftLeg", 5);
-        this.func_175417_e(nbttaglist4.func_82582_d() ? field_175432_e : new Rotations(nbttaglist4));
-        NBTTagList nbttaglist5 = p_175416_1_.func_150295_c("RightLeg", 5);
-        this.func_175427_f(nbttaglist5.func_82582_d() ? field_175429_f : new Rotations(nbttaglist5));
+        NBTTagList nbttaglist = tagCompound.getTagList("Head", 5);
+        this.setHeadRotation(nbttaglist.hasNoTags() ? DEFAULT_HEAD_ROTATION : new Rotations(nbttaglist));
+        NBTTagList nbttaglist1 = tagCompound.getTagList("Body", 5);
+        this.setBodyRotation(nbttaglist1.hasNoTags() ? DEFAULT_BODY_ROTATION : new Rotations(nbttaglist1));
+        NBTTagList nbttaglist2 = tagCompound.getTagList("LeftArm", 5);
+        this.setLeftArmRotation(nbttaglist2.hasNoTags() ? DEFAULT_LEFTARM_ROTATION : new Rotations(nbttaglist2));
+        NBTTagList nbttaglist3 = tagCompound.getTagList("RightArm", 5);
+        this.setRightArmRotation(nbttaglist3.hasNoTags() ? DEFAULT_RIGHTARM_ROTATION : new Rotations(nbttaglist3));
+        NBTTagList nbttaglist4 = tagCompound.getTagList("LeftLeg", 5);
+        this.setLeftLegRotation(nbttaglist4.hasNoTags() ? DEFAULT_LEFTLEG_ROTATION : new Rotations(nbttaglist4));
+        NBTTagList nbttaglist5 = tagCompound.getTagList("RightLeg", 5);
+        this.setRightLegRotation(nbttaglist5.hasNoTags() ? DEFAULT_RIGHTLEG_ROTATION : new Rotations(nbttaglist5));
     }
 
-    private NBTTagCompound func_175419_y()
+    private NBTTagCompound readPoseFromNBT()
     {
         NBTTagCompound nbttagcompound = new NBTTagCompound();
 
-        if (!field_175435_a.equals(this.field_175443_bh))
+        if (!DEFAULT_HEAD_ROTATION.equals(this.headRotation))
         {
-            nbttagcompound.func_74782_a("Head", this.field_175443_bh.func_179414_a());
+            nbttagcompound.setTag("Head", this.headRotation.writeToNBT());
         }
 
-        if (!field_175433_b.equals(this.field_175444_bi))
+        if (!DEFAULT_BODY_ROTATION.equals(this.bodyRotation))
         {
-            nbttagcompound.func_74782_a("Body", this.field_175444_bi.func_179414_a());
+            nbttagcompound.setTag("Body", this.bodyRotation.writeToNBT());
         }
 
-        if (!field_175434_c.equals(this.field_175438_bj))
+        if (!DEFAULT_LEFTARM_ROTATION.equals(this.leftArmRotation))
         {
-            nbttagcompound.func_74782_a("LeftArm", this.field_175438_bj.func_179414_a());
+            nbttagcompound.setTag("LeftArm", this.leftArmRotation.writeToNBT());
         }
 
-        if (!field_175431_d.equals(this.field_175439_bk))
+        if (!DEFAULT_RIGHTARM_ROTATION.equals(this.rightArmRotation))
         {
-            nbttagcompound.func_74782_a("RightArm", this.field_175439_bk.func_179414_a());
+            nbttagcompound.setTag("RightArm", this.rightArmRotation.writeToNBT());
         }
 
-        if (!field_175432_e.equals(this.field_175440_bl))
+        if (!DEFAULT_LEFTLEG_ROTATION.equals(this.leftLegRotation))
         {
-            nbttagcompound.func_74782_a("LeftLeg", this.field_175440_bl.func_179414_a());
+            nbttagcompound.setTag("LeftLeg", this.leftLegRotation.writeToNBT());
         }
 
-        if (!field_175429_f.equals(this.field_175441_bm))
+        if (!DEFAULT_RIGHTLEG_ROTATION.equals(this.rightLegRotation))
         {
-            nbttagcompound.func_74782_a("RightLeg", this.field_175441_bm.func_179414_a());
+            nbttagcompound.setTag("RightLeg", this.rightLegRotation.writeToNBT());
         }
 
         return nbttagcompound;
     }
 
-    public boolean func_70104_M()
+    public boolean canBePushed()
     {
         return false;
     }
 
-    protected void func_82167_n(Entity p_82167_1_)
+    protected void collideWithEntity(Entity entityIn)
     {
     }
 
-    protected void func_85033_bc()
+    protected void collideWithNearbyEntities()
     {
-        List<Entity> list = this.field_70170_p.func_175674_a(this, this.func_174813_aQ(), field_184798_bv);
+        List<Entity> list = this.world.getEntitiesInAABBexcluding(this, this.getEntityBoundingBox(), IS_RIDEABLE_MINECART);
 
         for (int i = 0; i < list.size(); ++i)
         {
             Entity entity = (Entity)list.get(i);
 
-            if (this.func_70068_e(entity) <= 0.2D)
+            if (this.getDistanceSqToEntity(entity) <= 0.2D)
             {
-                entity.func_70108_f(this);
+                entity.applyEntityCollision(this);
             }
         }
     }
 
-    public EnumActionResult func_184199_a(EntityPlayer p_184199_1_, Vec3d p_184199_2_, EnumHand p_184199_3_)
+    public EnumActionResult applyPlayerInteraction(EntityPlayer player, Vec3d vec, EnumHand stack)
     {
-        ItemStack itemstack = p_184199_1_.func_184586_b(p_184199_3_);
+        ItemStack itemstack = player.getHeldItem(stack);
 
-        if (!this.func_181026_s() && itemstack.func_77973_b() != Items.field_151057_cb)
+        if (!this.hasMarker() && itemstack.getItem() != Items.NAME_TAG)
         {
-            if (!this.field_70170_p.field_72995_K && !p_184199_1_.func_175149_v())
+            if (!this.world.isRemote && !player.isSpectator())
             {
-                EntityEquipmentSlot entityequipmentslot = EntityLiving.func_184640_d(itemstack);
+                EntityEquipmentSlot entityequipmentslot = EntityLiving.getSlotForItemStack(itemstack);
 
-                if (itemstack.func_190926_b())
+                if (itemstack.isEmpty())
                 {
-                    EntityEquipmentSlot entityequipmentslot1 = this.func_190772_a(p_184199_2_);
-                    EntityEquipmentSlot entityequipmentslot2 = this.func_184796_b(entityequipmentslot1) ? entityequipmentslot : entityequipmentslot1;
+                    EntityEquipmentSlot entityequipmentslot1 = this.getClickedSlot(vec);
+                    EntityEquipmentSlot entityequipmentslot2 = this.isDisabled(entityequipmentslot1) ? entityequipmentslot : entityequipmentslot1;
 
-                    if (this.func_190630_a(entityequipmentslot2))
+                    if (this.hasItemInSlot(entityequipmentslot2))
                     {
-                        this.func_184795_a(p_184199_1_, entityequipmentslot2, itemstack, p_184199_3_);
+                        this.swapItem(player, entityequipmentslot2, itemstack, stack);
                     }
                 }
                 else
                 {
-                    if (this.func_184796_b(entityequipmentslot))
+                    if (this.isDisabled(entityequipmentslot))
                     {
                         return EnumActionResult.FAIL;
                     }
 
-                    if (entityequipmentslot.func_188453_a() == EntityEquipmentSlot.Type.HAND && !this.func_175402_q())
+                    if (entityequipmentslot.getSlotType() == EntityEquipmentSlot.Type.HAND && !this.getShowArms())
                     {
                         return EnumActionResult.FAIL;
                     }
 
-                    this.func_184795_a(p_184199_1_, entityequipmentslot, itemstack, p_184199_3_);
+                    this.swapItem(player, entityequipmentslot, itemstack, stack);
                 }
 
                 return EnumActionResult.SUCCESS;
@@ -417,26 +424,26 @@
         }
     }
 
-    protected EntityEquipmentSlot func_190772_a(Vec3d p_190772_1_)
+    protected EntityEquipmentSlot getClickedSlot(Vec3d p_190772_1_)
     {
         EntityEquipmentSlot entityequipmentslot = EntityEquipmentSlot.MAINHAND;
-        boolean flag = this.func_175410_n();
-        double d0 = flag ? p_190772_1_.field_72448_b * 2.0D : p_190772_1_.field_72448_b;
+        boolean flag = this.isSmall();
+        double d0 = flag ? p_190772_1_.yCoord * 2.0D : p_190772_1_.yCoord;
         EntityEquipmentSlot entityequipmentslot1 = EntityEquipmentSlot.FEET;
 
-        if (d0 >= 0.1D && d0 < 0.1D + (flag ? 0.8D : 0.45D) && this.func_190630_a(entityequipmentslot1))
+        if (d0 >= 0.1D && d0 < 0.1D + (flag ? 0.8D : 0.45D) && this.hasItemInSlot(entityequipmentslot1))
         {
             entityequipmentslot = EntityEquipmentSlot.FEET;
         }
-        else if (d0 >= 0.9D + (flag ? 0.3D : 0.0D) && d0 < 0.9D + (flag ? 1.0D : 0.7D) && this.func_190630_a(EntityEquipmentSlot.CHEST))
+        else if (d0 >= 0.9D + (flag ? 0.3D : 0.0D) && d0 < 0.9D + (flag ? 1.0D : 0.7D) && this.hasItemInSlot(EntityEquipmentSlot.CHEST))
         {
             entityequipmentslot = EntityEquipmentSlot.CHEST;
         }
-        else if (d0 >= 0.4D && d0 < 0.4D + (flag ? 1.0D : 0.8D) && this.func_190630_a(EntityEquipmentSlot.LEGS))
+        else if (d0 >= 0.4D && d0 < 0.4D + (flag ? 1.0D : 0.8D) && this.hasItemInSlot(EntityEquipmentSlot.LEGS))
         {
             entityequipmentslot = EntityEquipmentSlot.LEGS;
         }
-        else if (d0 >= 1.6D && this.func_190630_a(EntityEquipmentSlot.HEAD))
+        else if (d0 >= 1.6D && this.hasItemInSlot(EntityEquipmentSlot.HEAD))
         {
             entityequipmentslot = EntityEquipmentSlot.HEAD;
         }
@@ -444,83 +451,103 @@
         return entityequipmentslot;
     }
 
-    private boolean func_184796_b(EntityEquipmentSlot p_184796_1_)
+    private boolean isDisabled(EntityEquipmentSlot slotIn)
     {
-        return (this.field_175442_bg & 1 << p_184796_1_.func_188452_c()) != 0;
+        return (this.disabledSlots & 1 << slotIn.getSlotIndex()) != 0;
     }
 
-    private void func_184795_a(EntityPlayer p_184795_1_, EntityEquipmentSlot p_184795_2_, ItemStack p_184795_3_, EnumHand p_184795_4_)
+    private void swapItem(EntityPlayer player, EntityEquipmentSlot p_184795_2_, ItemStack p_184795_3_, EnumHand hand)
     {
-        ItemStack itemstack = this.func_184582_a(p_184795_2_);
+        ItemStack itemstack = this.getItemStackFromSlot(p_184795_2_);
 
-        if (itemstack.func_190926_b() || (this.field_175442_bg & 1 << p_184795_2_.func_188452_c() + 8) == 0)
+        if (itemstack.isEmpty() || (this.disabledSlots & 1 << p_184795_2_.getSlotIndex() + 8) == 0)
         {
-            if (!itemstack.func_190926_b() || (this.field_175442_bg & 1 << p_184795_2_.func_188452_c() + 16) == 0)
+            if (!itemstack.isEmpty() || (this.disabledSlots & 1 << p_184795_2_.getSlotIndex() + 16) == 0)
             {
-                if (p_184795_1_.field_71075_bZ.field_75098_d && itemstack.func_190926_b() && !p_184795_3_.func_190926_b())
+                if (player.capabilities.isCreativeMode && itemstack.isEmpty() && !p_184795_3_.isEmpty())
                 {
-                    ItemStack itemstack2 = p_184795_3_.func_77946_l();
-                    itemstack2.func_190920_e(1);
-                    this.func_184201_a(p_184795_2_, itemstack2);
+                    ItemStack itemstack2 = p_184795_3_.copy();
+                    itemstack2.setCount(1);
+                    // CraftBukkit start
+                    org.bukkit.inventory.ItemStack armorStandItem = CraftItemStack.asCraftMirror(p_184795_3_);
+                    org.bukkit.inventory.ItemStack playerHeldItem = CraftItemStack.asCraftMirror(itemstack);
+
+                    Player pl = (Player) player.getBukkitEntity();
+                    ArmorStand self = (ArmorStand) this.getBukkitEntity();
+
+                    EquipmentSlot slot = CraftEquipmentSlot.getSlot(p_184795_2_);
+                    PlayerArmorStandManipulateEvent armorStandManipulateEvent = new PlayerArmorStandManipulateEvent(pl,self,playerHeldItem,armorStandItem,slot);
+                    this.world.getServer().getPluginManager().callEvent(armorStandManipulateEvent);
+
+                    if (armorStandManipulateEvent.isCancelled()) {
+                        return;
+                    }
+                    // CraftBukkit end
+                    this.setItemStackToSlot(p_184795_2_, itemstack2);
                 }
-                else if (!p_184795_3_.func_190926_b() && p_184795_3_.func_190916_E() > 1)
+                else if (!p_184795_3_.isEmpty() && p_184795_3_.getCount() > 1)
                 {
-                    if (itemstack.func_190926_b())
+                    if (itemstack.isEmpty())
                     {
-                        ItemStack itemstack1 = p_184795_3_.func_77946_l();
-                        itemstack1.func_190920_e(1);
-                        this.func_184201_a(p_184795_2_, itemstack1);
-                        p_184795_3_.func_190918_g(1);
+                        ItemStack itemstack1 = p_184795_3_.copy();
+                        itemstack1.setCount(1);
+                        this.setItemStackToSlot(p_184795_2_, itemstack1);
+                        p_184795_3_.shrink(1);
                     }
                 }
                 else
                 {
-                    this.func_184201_a(p_184795_2_, p_184795_3_);
-                    p_184795_1_.func_184611_a(p_184795_4_, itemstack);
+                    this.setItemStackToSlot(p_184795_2_, p_184795_3_);
+                    player.setHeldItem(hand, itemstack);
                 }
             }
         }
     }
 
-    public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
+    public boolean attackEntityFrom(DamageSource source, float amount)
     {
-        if (!this.field_70170_p.field_72995_K && !this.field_70128_L)
+        // CraftBukkit start
+        if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, amount)) {
+            return false;
+        }
+        // CraftBukkit end
+        if (!this.world.isRemote && !this.isDead)
         {
-            if (DamageSource.field_76380_i.equals(p_70097_1_))
+            if (DamageSource.OUT_OF_WORLD.equals(source))
             {
-                this.func_70106_y();
+                this.onKillCommand();
                 return false;
             }
-            else if (!this.func_180431_b(p_70097_1_) && !this.field_175436_h && !this.func_181026_s())
+            else if (!this.isEntityInvulnerable(source) && !this.canInteract && !this.hasMarker())
             {
-                if (p_70097_1_.func_94541_c())
+                if (source.isExplosion())
                 {
-                    this.func_175409_C();
-                    this.func_70106_y();
+                    this.dropContents();
+                    this.onKillCommand();
                     return false;
                 }
-                else if (DamageSource.field_76372_a.equals(p_70097_1_))
+                else if (DamageSource.IN_FIRE.equals(source))
                 {
-                    if (this.func_70027_ad())
+                    if (this.isBurning())
                     {
-                        this.func_175406_a(0.15F);
+                        this.damageArmorStand(0.15F);
                     }
                     else
                     {
-                        this.func_70015_d(5);
+                        this.setFire(5);
                     }
 
                     return false;
                 }
-                else if (DamageSource.field_76370_b.equals(p_70097_1_) && this.func_110143_aJ() > 0.5F)
+                else if (DamageSource.ON_FIRE.equals(source) && this.getHealth() > 0.5F)
                 {
-                    this.func_175406_a(4.0F);
+                    this.damageArmorStand(4.0F);
                     return false;
                 }
                 else
                 {
-                    boolean flag = "arrow".equals(p_70097_1_.func_76355_l());
-                    boolean flag1 = "player".equals(p_70097_1_.func_76355_l());
+                    boolean flag = "arrow".equals(source.getDamageType());
+                    boolean flag1 = "player".equals(source.getDamageType());
 
                     if (!flag1 && !flag)
                     {
@@ -528,36 +555,36 @@
                     }
                     else
                     {
-                        if (p_70097_1_.func_76364_f() instanceof EntityArrow)
+                        if (source.getSourceOfDamage() instanceof EntityArrow)
                         {
-                            p_70097_1_.func_76364_f().func_70106_y();
+                            source.getSourceOfDamage().setDead();
                         }
 
-                        if (p_70097_1_.func_76346_g() instanceof EntityPlayer && !((EntityPlayer)p_70097_1_.func_76346_g()).field_71075_bZ.field_75099_e)
+                        if (source.getEntity() instanceof EntityPlayer && !((EntityPlayer)source.getEntity()).capabilities.allowEdit)
                         {
                             return false;
                         }
-                        else if (p_70097_1_.func_180136_u())
+                        else if (source.isCreativePlayer())
                         {
-                            this.func_190773_I();
-                            this.func_175412_z();
-                            this.func_70106_y();
+                            this.playBrokenSound();
+                            this.playParticles();
+                            this.onKillCommand();
                             return false;
                         }
                         else
                         {
-                            long i = this.field_70170_p.func_82737_E();
+                            long i = this.world.getTotalWorldTime();
 
-                            if (i - this.field_175437_i > 5L && !flag)
+                            if (i - this.punchCooldown > 5L && !flag)
                             {
-                                this.field_70170_p.func_72960_a(this, (byte)32);
-                                this.field_175437_i = i;
+                                this.world.setEntityState(this, (byte)32);
+                                this.punchCooldown = i;
                             }
                             else
                             {
-                                this.func_175421_A();
-                                this.func_175412_z();
-                                this.func_70106_y();
+                                this.dropBlock();
+                                this.playParticles();
+                                this.onKillCommand();
                             }
 
                             return false;
@@ -577,26 +604,26 @@
     }
 
     @SideOnly(Side.CLIENT)
-    public void func_70103_a(byte p_70103_1_)
+    public void handleStatusUpdate(byte id)
     {
-        if (p_70103_1_ == 32)
+        if (id == 32)
         {
-            if (this.field_70170_p.field_72995_K)
+            if (this.world.isRemote)
             {
-                this.field_70170_p.func_184134_a(this.field_70165_t, this.field_70163_u, this.field_70161_v, SoundEvents.field_187707_l, this.func_184176_by(), 0.3F, 1.0F, false);
-                this.field_175437_i = this.field_70170_p.func_82737_E();
+                this.world.playSound(this.posX, this.posY, this.posZ, SoundEvents.ENTITY_ARMORSTAND_HIT, this.getSoundCategory(), 0.3F, 1.0F, false);
+                this.punchCooldown = this.world.getTotalWorldTime();
             }
         }
         else
         {
-            super.func_70103_a(p_70103_1_);
+            super.handleStatusUpdate(id);
         }
     }
 
     @SideOnly(Side.CLIENT)
-    public boolean func_70112_a(double p_70112_1_)
+    public boolean isInRangeToRenderDist(double distance)
     {
-        double d0 = this.func_174813_aQ().func_72320_b() * 4.0D;
+        double d0 = this.getEntityBoundingBox().getAverageEdgeLength() * 4.0D;
 
         if (Double.isNaN(d0) || d0 == 0.0D)
         {
@@ -604,249 +631,249 @@
         }
 
         d0 = d0 * 64.0D;
-        return p_70112_1_ < d0 * d0;
+        return distance < d0 * d0;
     }
 
-    private void func_175412_z()
+    private void playParticles()
     {
-        if (this.field_70170_p instanceof WorldServer)
+        if (this.world instanceof WorldServer)
         {
-            ((WorldServer)this.field_70170_p).func_175739_a(EnumParticleTypes.BLOCK_DUST, this.field_70165_t, this.field_70163_u + (double)this.field_70131_O / 1.5D, this.field_70161_v, 10, (double)(this.field_70130_N / 4.0F), (double)(this.field_70131_O / 4.0F), (double)(this.field_70130_N / 4.0F), 0.05D, new int[] {Block.func_176210_f(Blocks.field_150344_f.func_176223_P())});
+            ((WorldServer)this.world).spawnParticle(EnumParticleTypes.BLOCK_DUST, this.posX, this.posY + (double)this.height / 1.5D, this.posZ, 10, (double)(this.width / 4.0F), (double)(this.height / 4.0F), (double)(this.width / 4.0F), 0.05D, new int[] {Block.getStateId(Blocks.PLANKS.getDefaultState())});
         }
     }
 
-    private void func_175406_a(float p_175406_1_)
+    private void damageArmorStand(float damage)
     {
-        float f = this.func_110143_aJ();
-        f = f - p_175406_1_;
+        float f = this.getHealth();
+        f = f - damage;
 
         if (f <= 0.5F)
         {
-            this.func_175409_C();
-            this.func_70106_y();
+            this.dropContents();
+            this.onKillCommand();
         }
         else
         {
-            this.func_70606_j(f);
+            this.setHealth(f);
         }
     }
 
-    private void func_175421_A()
+    private void dropBlock()
     {
-        Block.func_180635_a(this.field_70170_p, new BlockPos(this), new ItemStack(Items.field_179565_cj));
-        this.func_175409_C();
+        drops.add(CraftItemStack.asBukkitCopy(new ItemStack(Items.ARMOR_STAND))); // CraftBukkit - add to drops
+        this.dropContents();
     }
 
-    private void func_175409_C()
+    private void dropContents()
     {
-        this.func_190773_I();
+        this.playBrokenSound();
 
-        for (int i = 0; i < this.field_184799_bw.size(); ++i)
+        for (int i = 0; i < this.handItems.size(); ++i)
         {
-            ItemStack itemstack = (ItemStack)this.field_184799_bw.get(i);
+            ItemStack itemstack = (ItemStack)this.handItems.get(i);
 
-            if (!itemstack.func_190926_b())
+            if (!itemstack.isEmpty())
             {
-                Block.func_180635_a(this.field_70170_p, (new BlockPos(this)).func_177984_a(), itemstack);
-                this.field_184799_bw.set(i, ItemStack.field_190927_a);
+                drops.add(CraftItemStack.asBukkitCopy(itemstack)); // CraftBukkit - add to drops
+                this.handItems.set(i, ItemStack.EMPTY);
             }
         }
 
-        for (int j = 0; j < this.field_184800_bx.size(); ++j)
+        for (int j = 0; j < this.armorItems.size(); ++j)
         {
-            ItemStack itemstack1 = (ItemStack)this.field_184800_bx.get(j);
+            ItemStack itemstack1 = (ItemStack)this.armorItems.get(j);
 
-            if (!itemstack1.func_190926_b())
+            if (!itemstack1.isEmpty())
             {
-                Block.func_180635_a(this.field_70170_p, (new BlockPos(this)).func_177984_a(), itemstack1);
-                this.field_184800_bx.set(j, ItemStack.field_190927_a);
+                Block.spawnAsEntity(this.world, (new BlockPos(this)).up(), itemstack1);
+                this.armorItems.set(j, ItemStack.EMPTY);
             }
         }
     }
 
-    private void func_190773_I()
+    private void playBrokenSound()
     {
-        this.field_70170_p.func_184148_a((EntityPlayer)null, this.field_70165_t, this.field_70163_u, this.field_70161_v, SoundEvents.field_187701_j, this.func_184176_by(), 1.0F, 1.0F);
+        this.world.playSound((EntityPlayer)null, this.posX, this.posY, this.posZ, SoundEvents.ENTITY_ARMORSTAND_BREAK, this.getSoundCategory(), 1.0F, 1.0F);
     }
 
-    protected float func_110146_f(float p_110146_1_, float p_110146_2_)
+    protected float updateDistance(float p_110146_1_, float p_110146_2_)
     {
-        this.field_70760_ar = this.field_70126_B;
-        this.field_70761_aq = this.field_70177_z;
+        this.prevRenderYawOffset = this.prevRotationYaw;
+        this.renderYawOffset = this.rotationYaw;
         return 0.0F;
     }
 
-    public float func_70047_e()
+    public float getEyeHeight()
     {
-        return this.func_70631_g_() ? this.field_70131_O * 0.5F : this.field_70131_O * 0.9F;
+        return this.isChild() ? this.height * 0.5F : this.height * 0.9F;
     }
 
-    public double func_70033_W()
+    public double getYOffset()
     {
-        return this.func_181026_s() ? 0.0D : 0.10000000149011612D;
+        return this.hasMarker() ? 0.0D : 0.10000000149011612D;
     }
 
-    public void func_70612_e(float p_70612_1_, float p_70612_2_)
+    public void moveEntityWithHeading(float strafe, float forward)
     {
-        if (!this.func_189652_ae())
+        if (!this.hasNoGravity())
         {
-            super.func_70612_e(p_70612_1_, p_70612_2_);
+            super.moveEntityWithHeading(strafe, forward);
         }
     }
 
-    public void func_181013_g(float p_181013_1_)
+    public void setRenderYawOffset(float offset)
     {
-        this.field_70760_ar = this.field_70126_B = p_181013_1_;
-        this.field_70758_at = this.field_70759_as = p_181013_1_;
+        this.prevRenderYawOffset = this.prevRotationYaw = offset;
+        this.prevRotationYawHead = this.rotationYawHead = offset;
     }
 
-    public void func_70034_d(float p_70034_1_)
+    public void setRotationYawHead(float rotation)
     {
-        this.field_70760_ar = this.field_70126_B = p_70034_1_;
-        this.field_70758_at = this.field_70759_as = p_70034_1_;
+        this.prevRenderYawOffset = this.prevRotationYaw = rotation;
+        this.prevRotationYawHead = this.rotationYawHead = rotation;
     }
 
-    public void func_70071_h_()
+    public void onUpdate()
     {
-        super.func_70071_h_();
-        Rotations rotations = (Rotations)this.field_70180_af.func_187225_a(field_184802_b);
+        super.onUpdate();
+        Rotations rotations = (Rotations)this.dataManager.get(HEAD_ROTATION);
 
-        if (!this.field_175443_bh.equals(rotations))
+        if (!this.headRotation.equals(rotations))
         {
-            this.func_175415_a(rotations);
+            this.setHeadRotation(rotations);
         }
 
-        Rotations rotations1 = (Rotations)this.field_70180_af.func_187225_a(field_184803_c);
+        Rotations rotations1 = (Rotations)this.dataManager.get(BODY_ROTATION);
 
-        if (!this.field_175444_bi.equals(rotations1))
+        if (!this.bodyRotation.equals(rotations1))
         {
-            this.func_175424_b(rotations1);
+            this.setBodyRotation(rotations1);
         }
 
-        Rotations rotations2 = (Rotations)this.field_70180_af.func_187225_a(field_184804_d);
+        Rotations rotations2 = (Rotations)this.dataManager.get(LEFT_ARM_ROTATION);
 
-        if (!this.field_175438_bj.equals(rotations2))
+        if (!this.leftArmRotation.equals(rotations2))
         {
-            this.func_175405_c(rotations2);
+            this.setLeftArmRotation(rotations2);
         }
 
-        Rotations rotations3 = (Rotations)this.field_70180_af.func_187225_a(field_184805_e);
+        Rotations rotations3 = (Rotations)this.dataManager.get(RIGHT_ARM_ROTATION);
 
-        if (!this.field_175439_bk.equals(rotations3))
+        if (!this.rightArmRotation.equals(rotations3))
         {
-            this.func_175428_d(rotations3);
+            this.setRightArmRotation(rotations3);
         }
 
-        Rotations rotations4 = (Rotations)this.field_70180_af.func_187225_a(field_184806_f);
+        Rotations rotations4 = (Rotations)this.dataManager.get(LEFT_LEG_ROTATION);
 
-        if (!this.field_175440_bl.equals(rotations4))
+        if (!this.leftLegRotation.equals(rotations4))
         {
-            this.func_175417_e(rotations4);
+            this.setLeftLegRotation(rotations4);
         }
 
-        Rotations rotations5 = (Rotations)this.field_70180_af.func_187225_a(field_184807_g);
+        Rotations rotations5 = (Rotations)this.dataManager.get(RIGHT_LEG_ROTATION);
 
-        if (!this.field_175441_bm.equals(rotations5))
+        if (!this.rightLegRotation.equals(rotations5))
         {
-            this.func_175427_f(rotations5);
+            this.setRightLegRotation(rotations5);
         }
 
-        boolean flag = this.func_181026_s();
+        boolean flag = this.hasMarker();
 
-        if (this.field_181028_bj != flag)
+        if (this.wasMarker != flag)
         {
-            this.func_181550_a(flag);
-            this.field_70156_m = !flag;
-            this.field_181028_bj = flag;
+            this.updateBoundingBox(flag);
+            this.preventEntitySpawning = !flag;
+            this.wasMarker = flag;
         }
     }
 
-    private void func_181550_a(boolean p_181550_1_)
+    private void updateBoundingBox(boolean p_181550_1_)
     {
         if (p_181550_1_)
         {
-            this.func_70105_a(0.0F, 0.0F);
+            this.setSize(0.0F, 0.0F);
         }
         else
         {
-            this.func_70105_a(0.5F, 1.975F);
+            this.setSize(0.5F, 1.975F);
         }
     }
 
-    protected void func_175135_B()
+    protected void updatePotionMetadata()
     {
-        this.func_82142_c(this.field_175436_h);
+        this.setInvisible(this.canInteract);
     }
 
-    public void func_82142_c(boolean p_82142_1_)
+    public void setInvisible(boolean invisible)
     {
-        this.field_175436_h = p_82142_1_;
-        super.func_82142_c(p_82142_1_);
+        this.canInteract = invisible;
+        super.setInvisible(invisible);
     }
 
-    public boolean func_70631_g_()
+    public boolean isChild()
     {
-        return this.func_175410_n();
+        return this.isSmall();
     }
 
-    public void func_174812_G()
+    public void onKillCommand()
     {
-        this.func_70106_y();
+        this.setDead();
     }
 
-    public boolean func_180427_aV()
+    public boolean isImmuneToExplosions()
     {
-        return this.func_82150_aj();
+        return this.isInvisible();
     }
 
-    public EnumPushReaction func_184192_z()
+    public EnumPushReaction getPushReaction()
     {
-        return this.func_181026_s() ? EnumPushReaction.IGNORE : super.func_184192_z();
+        return this.hasMarker() ? EnumPushReaction.IGNORE : super.getPushReaction();
     }
 
-    public void func_175420_a(boolean p_175420_1_)
+    public void setSmall(boolean small)
     {
-        this.field_70180_af.func_187227_b(field_184801_a, Byte.valueOf(this.func_184797_a(((Byte)this.field_70180_af.func_187225_a(field_184801_a)).byteValue(), 1, p_175420_1_)));
-        this.func_70105_a(0.5F, 1.975F);
+        this.dataManager.set(STATUS, Byte.valueOf(this.setBit(((Byte)this.dataManager.get(STATUS)).byteValue(), 1, small)));
+        this.setSize(0.5F, 1.975F);
     }
 
-    public boolean func_175410_n()
+    public boolean isSmall()
     {
-        return (((Byte)this.field_70180_af.func_187225_a(field_184801_a)).byteValue() & 1) != 0;
+        return (((Byte)this.dataManager.get(STATUS)).byteValue() & 1) != 0;
     }
 
-    public void func_175413_k(boolean p_175413_1_)
+    public void setShowArms(boolean showArms)
     {
-        this.field_70180_af.func_187227_b(field_184801_a, Byte.valueOf(this.func_184797_a(((Byte)this.field_70180_af.func_187225_a(field_184801_a)).byteValue(), 4, p_175413_1_)));
+        this.dataManager.set(STATUS, Byte.valueOf(this.setBit(((Byte)this.dataManager.get(STATUS)).byteValue(), 4, showArms)));
     }
 
-    public boolean func_175402_q()
+    public boolean getShowArms()
     {
-        return (((Byte)this.field_70180_af.func_187225_a(field_184801_a)).byteValue() & 4) != 0;
+        return (((Byte)this.dataManager.get(STATUS)).byteValue() & 4) != 0;
     }
 
-    public void func_175426_l(boolean p_175426_1_)
+    public void setNoBasePlate(boolean noBasePlate)
     {
-        this.field_70180_af.func_187227_b(field_184801_a, Byte.valueOf(this.func_184797_a(((Byte)this.field_70180_af.func_187225_a(field_184801_a)).byteValue(), 8, p_175426_1_)));
+        this.dataManager.set(STATUS, Byte.valueOf(this.setBit(((Byte)this.dataManager.get(STATUS)).byteValue(), 8, noBasePlate)));
     }
 
-    public boolean func_175414_r()
+    public boolean hasNoBasePlate()
     {
-        return (((Byte)this.field_70180_af.func_187225_a(field_184801_a)).byteValue() & 8) != 0;
+        return (((Byte)this.dataManager.get(STATUS)).byteValue() & 8) != 0;
     }
 
-    public void func_181027_m(boolean p_181027_1_)
+    public void setMarker(boolean marker)
     {
-        this.field_70180_af.func_187227_b(field_184801_a, Byte.valueOf(this.func_184797_a(((Byte)this.field_70180_af.func_187225_a(field_184801_a)).byteValue(), 16, p_181027_1_)));
-        this.func_70105_a(0.5F, 1.975F);
+        this.dataManager.set(STATUS, Byte.valueOf(this.setBit(((Byte)this.dataManager.get(STATUS)).byteValue(), 16, marker)));
+        this.setSize(0.5F, 1.975F);
     }
 
-    public boolean func_181026_s()
+    public boolean hasMarker()
     {
-        return (((Byte)this.field_70180_af.func_187225_a(field_184801_a)).byteValue() & 16) != 0;
+        return (((Byte)this.dataManager.get(STATUS)).byteValue() & 16) != 0;
     }
 
-    private byte func_184797_a(byte p_184797_1_, int p_184797_2_, boolean p_184797_3_)
+    private byte setBit(byte p_184797_1_, int p_184797_2_, boolean p_184797_3_)
     {
         if (p_184797_3_)
         {
@@ -860,123 +887,123 @@
         return p_184797_1_;
     }
 
-    public void func_175415_a(Rotations p_175415_1_)
+    public void setHeadRotation(Rotations vec)
     {
-        this.field_175443_bh = p_175415_1_;
-        this.field_70180_af.func_187227_b(field_184802_b, p_175415_1_);
+        this.headRotation = vec;
+        this.dataManager.set(HEAD_ROTATION, vec);
     }
 
-    public void func_175424_b(Rotations p_175424_1_)
+    public void setBodyRotation(Rotations vec)
     {
-        this.field_175444_bi = p_175424_1_;
-        this.field_70180_af.func_187227_b(field_184803_c, p_175424_1_);
+        this.bodyRotation = vec;
+        this.dataManager.set(BODY_ROTATION, vec);
     }
 
-    public void func_175405_c(Rotations p_175405_1_)
+    public void setLeftArmRotation(Rotations vec)
     {
-        this.field_175438_bj = p_175405_1_;
-        this.field_70180_af.func_187227_b(field_184804_d, p_175405_1_);
+        this.leftArmRotation = vec;
+        this.dataManager.set(LEFT_ARM_ROTATION, vec);
     }
 
-    public void func_175428_d(Rotations p_175428_1_)
+    public void setRightArmRotation(Rotations vec)
     {
-        this.field_175439_bk = p_175428_1_;
-        this.field_70180_af.func_187227_b(field_184805_e, p_175428_1_);
+        this.rightArmRotation = vec;
+        this.dataManager.set(RIGHT_ARM_ROTATION, vec);
     }
 
-    public void func_175417_e(Rotations p_175417_1_)
+    public void setLeftLegRotation(Rotations vec)
     {
-        this.field_175440_bl = p_175417_1_;
-        this.field_70180_af.func_187227_b(field_184806_f, p_175417_1_);
+        this.leftLegRotation = vec;
+        this.dataManager.set(LEFT_LEG_ROTATION, vec);
     }
 
-    public void func_175427_f(Rotations p_175427_1_)
+    public void setRightLegRotation(Rotations vec)
     {
-        this.field_175441_bm = p_175427_1_;
-        this.field_70180_af.func_187227_b(field_184807_g, p_175427_1_);
+        this.rightLegRotation = vec;
+        this.dataManager.set(RIGHT_LEG_ROTATION, vec);
     }
 
-    public Rotations func_175418_s()
+    public Rotations getHeadRotation()
     {
-        return this.field_175443_bh;
+        return this.headRotation;
     }
 
-    public Rotations func_175408_t()
+    public Rotations getBodyRotation()
     {
-        return this.field_175444_bi;
+        return this.bodyRotation;
     }
 
     @SideOnly(Side.CLIENT)
-    public Rotations func_175404_u()
+    public Rotations getLeftArmRotation()
     {
-        return this.field_175438_bj;
+        return this.leftArmRotation;
     }
 
     @SideOnly(Side.CLIENT)
-    public Rotations func_175411_v()
+    public Rotations getRightArmRotation()
     {
-        return this.field_175439_bk;
+        return this.rightArmRotation;
     }
 
     @SideOnly(Side.CLIENT)
-    public Rotations func_175403_w()
+    public Rotations getLeftLegRotation()
     {
-        return this.field_175440_bl;
+        return this.leftLegRotation;
     }
 
     @SideOnly(Side.CLIENT)
-    public Rotations func_175407_x()
+    public Rotations getRightLegRotation()
     {
-        return this.field_175441_bm;
+        return this.rightLegRotation;
     }
 
-    public boolean func_70067_L()
+    public boolean canBeCollidedWith()
     {
-        return super.func_70067_L() && !this.func_181026_s();
+        return super.canBeCollidedWith() && !this.hasMarker();
     }
 
-    public EnumHandSide func_184591_cq()
+    public EnumHandSide getPrimaryHand()
     {
         return EnumHandSide.RIGHT;
     }
 
-    protected SoundEvent func_184588_d(int p_184588_1_)
+    protected SoundEvent getFallSound(int heightIn)
     {
-        return SoundEvents.field_187704_k;
+        return SoundEvents.ENTITY_ARMORSTAND_FALL;
     }
 
     @Nullable
-    protected SoundEvent func_184601_bQ()
+    protected SoundEvent getHurtSound()
     {
-        return SoundEvents.field_187707_l;
+        return SoundEvents.ENTITY_ARMORSTAND_HIT;
     }
 
     @Nullable
-    protected SoundEvent func_184615_bR()
+    protected SoundEvent getDeathSound()
     {
-        return SoundEvents.field_187701_j;
+        return SoundEvents.ENTITY_ARMORSTAND_BREAK;
     }
 
-    public void func_70077_a(EntityLightningBolt p_70077_1_)
+    public void onStruckByLightning(EntityLightningBolt lightningBolt)
     {
     }
 
-    public boolean func_184603_cC()
+    public boolean canBeHitWithPotion()
     {
         return false;
     }
 
-    public void func_184206_a(DataParameter<?> p_184206_1_)
+    public void notifyDataManagerChange(DataParameter<?> key)
     {
-        if (field_184801_a.equals(p_184206_1_))
+        if (STATUS.equals(key))
         {
-            this.func_70105_a(0.5F, 1.975F);
+            this.setSize(0.5F, 1.975F);
         }
 
-        super.func_184206_a(p_184206_1_);
+        super.notifyDataManagerChange(key);
     }
 
-    public boolean func_190631_cK()
+    public boolean attackable()
     {
         return false;
     }
