--- ../src-base/minecraft/net/minecraft/entity/item/EntityFallingBlock.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityFallingBlock.java
@@ -29,325 +29,324 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class EntityFallingBlock extends Entity
 {
-    private IBlockState field_175132_d;
-    public int field_145812_b;
-    public boolean field_145813_c = true;
-    private boolean field_145808_f;
-    public boolean field_145809_g;
-    private int field_145815_h = 40;
-    private float field_145816_i = 2.0F;
-    public NBTTagCompound field_145810_d;
-    protected static final DataParameter<BlockPos> field_184532_d = EntityDataManager.<BlockPos>func_187226_a(EntityFallingBlock.class, DataSerializers.field_187200_j);
+    private IBlockState fallTile;
+    public int fallTime;
+    public boolean shouldDropItem = true;
+    private boolean dontSetBlock;
+    public boolean hurtEntities;
+    private int fallHurtMax = 40;
+    private float fallHurtAmount = 2.0F;
+    public NBTTagCompound tileEntityData;
+    protected static final DataParameter<BlockPos> ORIGIN = EntityDataManager.<BlockPos>createKey(EntityFallingBlock.class, DataSerializers.BLOCK_POS);
 
-    public EntityFallingBlock(World p_i1706_1_)
+    public EntityFallingBlock(World worldIn)
     {
-        super(p_i1706_1_);
+        super(worldIn);
     }
 
-    public EntityFallingBlock(World p_i45848_1_, double p_i45848_2_, double p_i45848_4_, double p_i45848_6_, IBlockState p_i45848_8_)
+    public EntityFallingBlock(World worldIn, double x, double y, double z, IBlockState fallingBlockState)
     {
-        super(p_i45848_1_);
-        this.field_175132_d = p_i45848_8_;
-        this.field_70156_m = true;
-        this.func_70105_a(0.98F, 0.98F);
-        this.func_70107_b(p_i45848_2_, p_i45848_4_ + (double)((1.0F - this.field_70131_O) / 2.0F), p_i45848_6_);
-        this.field_70159_w = 0.0D;
-        this.field_70181_x = 0.0D;
-        this.field_70179_y = 0.0D;
-        this.field_70169_q = p_i45848_2_;
-        this.field_70167_r = p_i45848_4_;
-        this.field_70166_s = p_i45848_6_;
-        this.func_184530_a(new BlockPos(this));
+        super(worldIn);
+        this.fallTile = fallingBlockState;
+        this.preventEntitySpawning = true;
+        this.setSize(0.98F, 0.98F);
+        this.setPosition(x, y + (double)((1.0F - this.height) / 2.0F), z);
+        this.motionX = 0.0D;
+        this.motionY = 0.0D;
+        this.motionZ = 0.0D;
+        this.prevPosX = x;
+        this.prevPosY = y;
+        this.prevPosZ = z;
+        this.setOrigin(new BlockPos(this));
     }
 
-    public void func_184530_a(BlockPos p_184530_1_)
+    public void setOrigin(BlockPos p_184530_1_)
     {
-        this.field_70180_af.func_187227_b(field_184532_d, p_184530_1_);
+        this.dataManager.set(ORIGIN, p_184530_1_);
     }
 
     @SideOnly(Side.CLIENT)
-    public BlockPos func_184531_j()
+    public BlockPos getOrigin()
     {
-        return (BlockPos)this.field_70180_af.func_187225_a(field_184532_d);
+        return (BlockPos)this.dataManager.get(ORIGIN);
     }
 
-    protected boolean func_70041_e_()
+    protected boolean canTriggerWalking()
     {
         return false;
     }
 
-    protected void func_70088_a()
+    protected void entityInit()
     {
-        this.field_70180_af.func_187214_a(field_184532_d, BlockPos.field_177992_a);
+        this.dataManager.register(ORIGIN, BlockPos.ORIGIN);
     }
 
-    public boolean func_70067_L()
+    public boolean canBeCollidedWith()
     {
-        return !this.field_70128_L;
+        return !this.isDead;
     }
 
-    public void func_70071_h_()
+    public void onUpdate()
     {
-        Block block = this.field_175132_d.func_177230_c();
+        Block block = this.fallTile.getBlock();
 
-        if (this.field_175132_d.func_185904_a() == Material.field_151579_a)
+        if (this.fallTile.getMaterial() == Material.AIR)
         {
-            this.func_70106_y();
+            this.setDead();
         }
         else
         {
-            this.field_70169_q = this.field_70165_t;
-            this.field_70167_r = this.field_70163_u;
-            this.field_70166_s = this.field_70161_v;
+            this.prevPosX = this.posX;
+            this.prevPosY = this.posY;
+            this.prevPosZ = this.posZ;
 
-            if (this.field_145812_b++ == 0)
+            if (this.fallTime++ == 0)
             {
                 BlockPos blockpos = new BlockPos(this);
 
-                if (this.field_70170_p.func_180495_p(blockpos).func_177230_c() == block)
-                {
-                    this.field_70170_p.func_175698_g(blockpos);
-                }
-                else if (!this.field_70170_p.field_72995_K)
-                {
-                    this.func_70106_y();
+                if (this.world.getBlockState(blockpos).getBlock() == block && !CraftEventFactory.callEntityChangeBlockEvent(this, blockpos, Blocks.AIR, 0).isCancelled()) {
+                    this.world.setBlockToAir(blockpos);
+                } else if (!this.world.isRemote) {
+                    this.setDead();
                     return;
                 }
             }
 
-            if (!this.func_189652_ae())
+            if (!this.hasNoGravity())
             {
-                this.field_70181_x -= 0.03999999910593033D;
+                this.motionY -= 0.03999999910593033D;
             }
 
-            this.func_70091_d(MoverType.SELF, this.field_70159_w, this.field_70181_x, this.field_70179_y);
-            this.field_70159_w *= 0.9800000190734863D;
-            this.field_70181_x *= 0.9800000190734863D;
-            this.field_70179_y *= 0.9800000190734863D;
+            this.move(MoverType.SELF, this.motionX, this.motionY, this.motionZ);
+            this.motionX *= 0.9800000190734863D;
+            this.motionY *= 0.9800000190734863D;
+            this.motionZ *= 0.9800000190734863D;
 
-            if (!this.field_70170_p.field_72995_K)
+            if (!this.world.isRemote)
             {
                 BlockPos blockpos1 = new BlockPos(this);
 
-                if (this.field_70122_E)
+                if (this.onGround)
                 {
-                    IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos1);
+                    IBlockState iblockstate = this.world.getBlockState(blockpos1);
 
-                    if (BlockFalling.func_185759_i(this.field_70170_p.func_180495_p(new BlockPos(this.field_70165_t, this.field_70163_u - 0.009999999776482582D, this.field_70161_v))))
+                    if (this.world.isAirBlock(new BlockPos(this.posX, this.posY - 0.009999999776482582D, this.posZ))) //Forge: Don't indent below.
+                    if (BlockFalling.canFallThrough(this.world.getBlockState(new BlockPos(this.posX, this.posY - 0.009999999776482582D, this.posZ))))
                     {
-                        this.field_70122_E = false;
-                        return;
+                        this.onGround = false;
+                        //return;
                     }
 
-                    this.field_70159_w *= 0.699999988079071D;
-                    this.field_70179_y *= 0.699999988079071D;
-                    this.field_70181_x *= -0.5D;
+                    this.motionX *= 0.699999988079071D;
+                    this.motionZ *= 0.699999988079071D;
+                    this.motionY *= -0.5D;
 
-                    if (iblockstate.func_177230_c() != Blocks.field_180384_M)
+                    if (iblockstate.getBlock() != Blocks.PISTON_EXTENSION)
                     {
-                        this.func_70106_y();
+                        this.setDead();
 
-                        if (!this.field_145808_f)
+                        if (!this.dontSetBlock)
                         {
-                            if (this.field_70170_p.func_190527_a(block, blockpos1, true, EnumFacing.UP, (Entity)null) && !BlockFalling.func_185759_i(this.field_70170_p.func_180495_p(blockpos1.func_177977_b())) && this.field_70170_p.func_180501_a(blockpos1, this.field_175132_d, 3))
-                            {
+                            // CraftBukkit start
+                            if (this.world.mayPlace(block, blockpos1, true, EnumFacing.UP, (Entity) null) && !BlockFalling.canFallThrough(this.world.getBlockState(blockpos1.down()))) {
+                                if (CraftEventFactory.callEntityChangeBlockEvent(this, blockpos1, this.fallTile.getBlock(), this.fallTile.getBlock().getMetaFromState(this.fallTile)).isCancelled()) {
+                                    return;
+                                }
+                                this.world.setBlockState(blockpos1, this.fallTile, 3);
+                                // CraftBukkit end
                                 if (block instanceof BlockFalling)
                                 {
-                                    ((BlockFalling)block).func_176502_a_(this.field_70170_p, blockpos1);
+                                    ((BlockFalling)block).onEndFalling(this.world, blockpos1);
                                 }
 
-                                if (this.field_145810_d != null && block instanceof ITileEntityProvider)
+                                if (this.tileEntityData != null && block.hasTileEntity(this.fallTile))
                                 {
-                                    TileEntity tileentity = this.field_70170_p.func_175625_s(blockpos1);
+                                    TileEntity tileentity = this.world.getTileEntity(blockpos1);
 
                                     if (tileentity != null)
                                     {
-                                        NBTTagCompound nbttagcompound = tileentity.func_189515_b(new NBTTagCompound());
+                                        NBTTagCompound nbttagcompound = tileentity.writeToNBT(new NBTTagCompound());
 
-                                        for (String s : this.field_145810_d.func_150296_c())
+                                        for (String s : this.tileEntityData.getKeySet())
                                         {
-                                            NBTBase nbtbase = this.field_145810_d.func_74781_a(s);
+                                            NBTBase nbtbase = this.tileEntityData.getTag(s);
 
                                             if (!"x".equals(s) && !"y".equals(s) && !"z".equals(s))
                                             {
-                                                nbttagcompound.func_74782_a(s, nbtbase.func_74737_b());
+                                                nbttagcompound.setTag(s, nbtbase.copy());
                                             }
                                         }
 
-                                        tileentity.func_145839_a(nbttagcompound);
-                                        tileentity.func_70296_d();
+                                        tileentity.readFromNBT(nbttagcompound);
+                                        tileentity.markDirty();
                                     }
                                 }
                             }
-                            else if (this.field_145813_c && this.field_70170_p.func_82736_K().func_82766_b("doEntityDrops"))
+                            else if (this.shouldDropItem && this.world.getGameRules().getBoolean("doEntityDrops"))
                             {
-                                this.func_70099_a(new ItemStack(block, 1, block.func_180651_a(this.field_175132_d)), 0.0F);
+                                this.entityDropItem(new ItemStack(block, 1, block.damageDropped(this.fallTile)), 0.0F);
                             }
                         }
                         else if (block instanceof BlockFalling)
                         {
-                            ((BlockFalling)block).func_190974_b(this.field_70170_p, blockpos1);
+                            ((BlockFalling)block).onBroken(this.world, blockpos1);
                         }
                     }
                 }
-                else if (this.field_145812_b > 100 && !this.field_70170_p.field_72995_K && (blockpos1.func_177956_o() < 1 || blockpos1.func_177956_o() > 256) || this.field_145812_b > 600)
+                else if (this.fallTime > 100 && !this.world.isRemote && (blockpos1.getY() < 1 || blockpos1.getY() > 256) || this.fallTime > 600)
                 {
-                    if (this.field_145813_c && this.field_70170_p.func_82736_K().func_82766_b("doEntityDrops"))
+                    if (this.shouldDropItem && this.world.getGameRules().getBoolean("doEntityDrops"))
                     {
-                        this.func_70099_a(new ItemStack(block, 1, block.func_180651_a(this.field_175132_d)), 0.0F);
+                        this.entityDropItem(new ItemStack(block, 1, block.damageDropped(this.fallTile)), 0.0F);
                     }
 
-                    this.func_70106_y();
+                    this.setDead();
                 }
             }
         }
     }
 
-    public void func_180430_e(float p_180430_1_, float p_180430_2_)
-    {
-        Block block = this.field_175132_d.func_177230_c();
+    public void fall(float distance, float damageMultiplier) {
+        Block block = this.fallTile.getBlock();
 
-        if (this.field_145809_g)
-        {
-            int i = MathHelper.func_76123_f(p_180430_1_ - 1.0F);
+        if (this.hurtEntities) {
+            int i = MathHelper.ceil(distance - 1.0F);
 
-            if (i > 0)
-            {
-                List<Entity> list = Lists.newArrayList(this.field_70170_p.func_72839_b(this, this.func_174813_aQ()));
-                boolean flag = block == Blocks.field_150467_bQ;
-                DamageSource damagesource = flag ? DamageSource.field_82728_o : DamageSource.field_82729_p;
+            if (i > 0) {
+                List<Entity> list = Lists.newArrayList(this.world.getEntitiesWithinAABBExcludingEntity(this, this.getEntityBoundingBox()));
+                boolean flag = block == Blocks.ANVIL;
+                DamageSource damagesource = flag ? DamageSource.ANVIL : DamageSource.FALLING_BLOCK;
 
-                for (Entity entity : list)
-                {
-                    entity.func_70097_a(damagesource, (float)Math.min(MathHelper.func_76141_d((float)i * this.field_145816_i), this.field_145815_h));
-                }
+                for (Entity entity : list) {
 
-                if (flag && (double)this.field_70146_Z.nextFloat() < 0.05000000074505806D + (double)i * 0.05D)
-                {
-                    int j = ((Integer)this.field_175132_d.func_177229_b(BlockAnvil.field_176505_b)).intValue();
-                    ++j;
+                    CraftEventFactory.entityDamage = this; // CraftBukkit
+                    entity.attackEntityFrom(damagesource, (float) Math.min(MathHelper.floor((float) i * this.fallHurtAmount), this.fallHurtMax));
+                    CraftEventFactory.entityDamage = null; // CraftBukkit                }
 
-                    if (j > 2)
-                    {
-                        this.field_145808_f = true;
+                    if (flag && (double) this.rand.nextFloat() < 0.05000000074505806D + (double) i * 0.05D) {
+                        int j = ((Integer) this.fallTile.getValue(BlockAnvil.DAMAGE)).intValue();
+                        ++j;
+
+                        if (j > 2) {
+                            this.dontSetBlock = true;
+                        } else {
+                            this.fallTile = this.fallTile.withProperty(BlockAnvil.DAMAGE, Integer.valueOf(j));
+                        }
                     }
-                    else
-                    {
-                        this.field_175132_d = this.field_175132_d.func_177226_a(BlockAnvil.field_176505_b, Integer.valueOf(j));
-                    }
                 }
             }
         }
     }
 
-    public static void func_189741_a(DataFixer p_189741_0_)
+    public static void registerFixesFallingBlock(DataFixer fixer)
     {
     }
 
-    protected void func_70014_b(NBTTagCompound p_70014_1_)
+    protected void writeEntityToNBT(NBTTagCompound compound)
     {
-        Block block = this.field_175132_d != null ? this.field_175132_d.func_177230_c() : Blocks.field_150350_a;
-        ResourceLocation resourcelocation = (ResourceLocation)Block.field_149771_c.func_177774_c(block);
-        p_70014_1_.func_74778_a("Block", resourcelocation == null ? "" : resourcelocation.toString());
-        p_70014_1_.func_74774_a("Data", (byte)block.func_176201_c(this.field_175132_d));
-        p_70014_1_.func_74768_a("Time", this.field_145812_b);
-        p_70014_1_.func_74757_a("DropItem", this.field_145813_c);
-        p_70014_1_.func_74757_a("HurtEntities", this.field_145809_g);
-        p_70014_1_.func_74776_a("FallHurtAmount", this.field_145816_i);
-        p_70014_1_.func_74768_a("FallHurtMax", this.field_145815_h);
+        Block block = this.fallTile != null ? this.fallTile.getBlock() : Blocks.AIR;
+        ResourceLocation resourcelocation = (ResourceLocation)Block.REGISTRY.getNameForObject(block);
+        compound.setString("Block", resourcelocation == null ? "" : resourcelocation.toString());
+        compound.setByte("Data", (byte)block.getMetaFromState(this.fallTile));
+        compound.setInteger("Time", this.fallTime);
+        compound.setBoolean("DropItem", this.shouldDropItem);
+        compound.setBoolean("HurtEntities", this.hurtEntities);
+        compound.setFloat("FallHurtAmount", this.fallHurtAmount);
+        compound.setInteger("FallHurtMax", this.fallHurtMax);
 
-        if (this.field_145810_d != null)
+        if (this.tileEntityData != null)
         {
-            p_70014_1_.func_74782_a("TileEntityData", this.field_145810_d);
+            compound.setTag("TileEntityData", this.tileEntityData);
         }
     }
 
-    protected void func_70037_a(NBTTagCompound p_70037_1_)
+    protected void readEntityFromNBT(NBTTagCompound compound)
     {
-        int i = p_70037_1_.func_74771_c("Data") & 255;
+        int i = compound.getByte("Data") & 255;
 
-        if (p_70037_1_.func_150297_b("Block", 8))
+        if (compound.hasKey("Block", 8))
         {
-            this.field_175132_d = Block.func_149684_b(p_70037_1_.func_74779_i("Block")).func_176203_a(i);
+            this.fallTile = Block.getBlockFromName(compound.getString("Block")).getStateFromMeta(i);
         }
-        else if (p_70037_1_.func_150297_b("TileID", 99))
+        else if (compound.hasKey("TileID", 99))
         {
-            this.field_175132_d = Block.func_149729_e(p_70037_1_.func_74762_e("TileID")).func_176203_a(i);
+            this.fallTile = Block.getBlockById(compound.getInteger("TileID")).getStateFromMeta(i);
         }
         else
         {
-            this.field_175132_d = Block.func_149729_e(p_70037_1_.func_74771_c("Tile") & 255).func_176203_a(i);
+            this.fallTile = Block.getBlockById(compound.getByte("Tile") & 255).getStateFromMeta(i);
         }
 
-        this.field_145812_b = p_70037_1_.func_74762_e("Time");
-        Block block = this.field_175132_d.func_177230_c();
+        this.fallTime = compound.getInteger("Time");
+        Block block = this.fallTile.getBlock();
 
-        if (p_70037_1_.func_150297_b("HurtEntities", 99))
+        if (compound.hasKey("HurtEntities", 99))
         {
-            this.field_145809_g = p_70037_1_.func_74767_n("HurtEntities");
-            this.field_145816_i = p_70037_1_.func_74760_g("FallHurtAmount");
-            this.field_145815_h = p_70037_1_.func_74762_e("FallHurtMax");
+            this.hurtEntities = compound.getBoolean("HurtEntities");
+            this.fallHurtAmount = compound.getFloat("FallHurtAmount");
+            this.fallHurtMax = compound.getInteger("FallHurtMax");
         }
-        else if (block == Blocks.field_150467_bQ)
+        else if (block == Blocks.ANVIL)
         {
-            this.field_145809_g = true;
+            this.hurtEntities = true;
         }
 
-        if (p_70037_1_.func_150297_b("DropItem", 99))
+        if (compound.hasKey("DropItem", 99))
         {
-            this.field_145813_c = p_70037_1_.func_74767_n("DropItem");
+            this.shouldDropItem = compound.getBoolean("DropItem");
         }
 
-        if (p_70037_1_.func_150297_b("TileEntityData", 10))
+        if (compound.hasKey("TileEntityData", 10))
         {
-            this.field_145810_d = p_70037_1_.func_74775_l("TileEntityData");
+            this.tileEntityData = compound.getCompoundTag("TileEntityData");
         }
 
-        if (block == null || block.func_176223_P().func_185904_a() == Material.field_151579_a)
+        if (block == null || block.getDefaultState().getMaterial() == Material.AIR)
         {
-            this.field_175132_d = Blocks.field_150354_m.func_176223_P();
+            this.fallTile = Blocks.SAND.getDefaultState();
         }
     }
 
-    public void func_145806_a(boolean p_145806_1_)
+    public void setHurtEntities(boolean p_145806_1_)
     {
-        this.field_145809_g = p_145806_1_;
+        this.hurtEntities = p_145806_1_;
     }
 
-    public void func_85029_a(CrashReportCategory p_85029_1_)
+    public void addEntityCrashInfo(CrashReportCategory category)
     {
-        super.func_85029_a(p_85029_1_);
+        super.addEntityCrashInfo(category);
 
-        if (this.field_175132_d != null)
+        if (this.fallTile != null)
         {
-            Block block = this.field_175132_d.func_177230_c();
-            p_85029_1_.func_71507_a("Immitating block ID", Integer.valueOf(Block.func_149682_b(block)));
-            p_85029_1_.func_71507_a("Immitating block data", Integer.valueOf(block.func_176201_c(this.field_175132_d)));
+            Block block = this.fallTile.getBlock();
+            category.addCrashSection("Immitating block ID", Integer.valueOf(Block.getIdFromBlock(block)));
+            category.addCrashSection("Immitating block data", Integer.valueOf(block.getMetaFromState(this.fallTile)));
         }
     }
 
     @SideOnly(Side.CLIENT)
-    public World func_145807_e()
+    public World getWorldObj()
     {
-        return this.field_70170_p;
+        return this.world;
     }
 
     @SideOnly(Side.CLIENT)
-    public boolean func_90999_ad()
+    public boolean canRenderOnFire()
     {
         return false;
     }
 
     @Nullable
-    public IBlockState func_175131_l()
+    public IBlockState getBlock()
     {
-        return this.field_175132_d;
+        return this.fallTile;
     }
 
-    public boolean func_184213_bq()
+    public boolean ignoreItemEntityData()
     {
         return true;
     }
