--- ../src-base/minecraft/net/minecraft/entity/item/EntityMinecartHopper.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityMinecartHopper.java
@@ -21,185 +21,186 @@
 
 public class EntityMinecartHopper extends EntityMinecartContainer implements IHopper
 {
-    private boolean field_96113_a = true;
-    private int field_98044_b = -1;
-    private final BlockPos field_174900_c = BlockPos.field_177992_a;
+    private boolean isBlocked = true;
+    private int transferTicker = -1;
+    private final BlockPos lastPosition = BlockPos.ORIGIN;
 
-    public EntityMinecartHopper(World p_i1720_1_)
+    public EntityMinecartHopper(World worldIn)
     {
-        super(p_i1720_1_);
+        super(worldIn);
     }
 
-    public EntityMinecartHopper(World p_i1721_1_, double p_i1721_2_, double p_i1721_4_, double p_i1721_6_)
+    public EntityMinecartHopper(World worldIn, double x, double y, double z)
     {
-        super(p_i1721_1_, p_i1721_2_, p_i1721_4_, p_i1721_6_);
+        super(worldIn, x, y, z);
     }
 
-    public EntityMinecart.Type func_184264_v()
+    public EntityMinecart.Type getType()
     {
         return EntityMinecart.Type.HOPPER;
     }
 
-    public IBlockState func_180457_u()
+    public IBlockState getDefaultDisplayTile()
     {
-        return Blocks.field_150438_bZ.func_176223_P();
+        return Blocks.HOPPER.getDefaultState();
     }
 
-    public int func_94085_r()
+    public int getDefaultDisplayTileOffset()
     {
         return 1;
     }
 
-    public int func_70302_i_()
+    public int getSizeInventory()
     {
         return 5;
     }
 
-    public boolean func_184230_a(EntityPlayer p_184230_1_, EnumHand p_184230_2_)
+    public boolean processInitialInteract(EntityPlayer player, EnumHand hand)
     {
-        if (!this.field_70170_p.field_72995_K)
+        if(net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.minecart.MinecartInteractEvent(this, player, hand))) return true;
+        if (!this.world.isRemote)
         {
-            p_184230_1_.func_71007_a(this);
+            player.displayGUIChest(this);
         }
 
         return true;
     }
 
-    public void func_96095_a(int p_96095_1_, int p_96095_2_, int p_96095_3_, boolean p_96095_4_)
+    public void onActivatorRailPass(int x, int y, int z, boolean receivingPower)
     {
-        boolean flag = !p_96095_4_;
+        boolean flag = !receivingPower;
 
-        if (flag != this.func_96111_ay())
+        if (flag != this.getBlocked())
         {
-            this.func_96110_f(flag);
+            this.setBlocked(flag);
         }
     }
 
-    public boolean func_96111_ay()
+    public boolean getBlocked()
     {
-        return this.field_96113_a;
+        return this.isBlocked;
     }
 
-    public void func_96110_f(boolean p_96110_1_)
+    public void setBlocked(boolean p_96110_1_)
     {
-        this.field_96113_a = p_96110_1_;
+        this.isBlocked = p_96110_1_;
     }
 
-    public World func_145831_w()
+    public World getWorld()
     {
-        return this.field_70170_p;
+        return this.world;
     }
 
-    public double func_96107_aA()
+    public double getXPos()
     {
-        return this.field_70165_t;
+        return this.posX;
     }
 
-    public double func_96109_aB()
+    public double getYPos()
     {
-        return this.field_70163_u + 0.5D;
+        return this.posY + 0.5D;
     }
 
-    public double func_96108_aC()
+    public double getZPos()
     {
-        return this.field_70161_v;
+        return this.posZ;
     }
 
-    public void func_70071_h_()
+    public void onUpdate()
     {
-        super.func_70071_h_();
+        super.onUpdate();
 
-        if (!this.field_70170_p.field_72995_K && this.func_70089_S() && this.func_96111_ay())
+        if (!this.world.isRemote && this.isEntityAlive() && this.getBlocked())
         {
             BlockPos blockpos = new BlockPos(this);
 
-            if (blockpos.equals(this.field_174900_c))
+            if (blockpos.equals(this.lastPosition))
             {
-                --this.field_98044_b;
+                --this.transferTicker;
             }
             else
             {
-                this.func_98042_n(0);
+                this.setTransferTicker(0);
             }
 
-            if (!this.func_98043_aE())
+            if (!this.canTransfer())
             {
-                this.func_98042_n(0);
+                this.setTransferTicker(0);
 
-                if (this.func_96112_aD())
+                if (this.captureDroppedItems())
                 {
-                    this.func_98042_n(4);
-                    this.func_70296_d();
+                    this.setTransferTicker(4);
+                    this.markDirty();
                 }
             }
         }
     }
 
-    public boolean func_96112_aD()
+    public boolean captureDroppedItems()
     {
-        if (TileEntityHopper.func_145891_a(this))
+        if (TileEntityHopper.captureDroppedItems(this))
         {
             return true;
         }
         else
         {
-            List<EntityItem> list = this.field_70170_p.<EntityItem>func_175647_a(EntityItem.class, this.func_174813_aQ().func_72314_b(0.25D, 0.0D, 0.25D), EntitySelectors.field_94557_a);
+            List<EntityItem> list = this.world.<EntityItem>getEntitiesWithinAABB(EntityItem.class, this.getEntityBoundingBox().expand(0.25D, 0.0D, 0.25D), EntitySelectors.IS_ALIVE);
 
             if (!list.isEmpty())
             {
-                TileEntityHopper.func_145898_a((IInventory)null, this, (EntityItem)list.get(0));
+                TileEntityHopper.putDropInInventoryAllSlots((IInventory)null, this, (EntityItem)list.get(0));
             }
 
             return false;
         }
     }
 
-    public void func_94095_a(DamageSource p_94095_1_)
+    public void killMinecart(DamageSource source)
     {
-        super.func_94095_a(p_94095_1_);
+        super.killMinecart(source);
 
-        if (this.field_70170_p.func_82736_K().func_82766_b("doEntityDrops"))
+        if (this.world.getGameRules().getBoolean("doEntityDrops"))
         {
-            this.func_145778_a(Item.func_150898_a(Blocks.field_150438_bZ), 1, 0.0F);
+            this.dropItemWithOffset(Item.getItemFromBlock(Blocks.HOPPER), 1, 0.0F);
         }
     }
 
-    public static void func_189682_a(DataFixer p_189682_0_)
+    public static void registerFixesMinecartHopper(DataFixer fixer)
     {
-        EntityMinecartContainer.func_190574_b(p_189682_0_, EntityMinecartHopper.class);
+        EntityMinecartContainer.addDataFixers(fixer, EntityMinecartHopper.class);
     }
 
-    protected void func_70014_b(NBTTagCompound p_70014_1_)
+    protected void writeEntityToNBT(NBTTagCompound compound)
     {
-        super.func_70014_b(p_70014_1_);
-        p_70014_1_.func_74768_a("TransferCooldown", this.field_98044_b);
-        p_70014_1_.func_74757_a("Enabled", this.field_96113_a);
+        super.writeEntityToNBT(compound);
+        compound.setInteger("TransferCooldown", this.transferTicker);
+        compound.setBoolean("Enabled", this.isBlocked);
     }
 
-    protected void func_70037_a(NBTTagCompound p_70037_1_)
+    protected void readEntityFromNBT(NBTTagCompound compound)
     {
-        super.func_70037_a(p_70037_1_);
-        this.field_98044_b = p_70037_1_.func_74762_e("TransferCooldown");
-        this.field_96113_a = p_70037_1_.func_74764_b("Enabled") ? p_70037_1_.func_74767_n("Enabled") : true;
+        super.readEntityFromNBT(compound);
+        this.transferTicker = compound.getInteger("TransferCooldown");
+        this.isBlocked = compound.hasKey("Enabled") ? compound.getBoolean("Enabled") : true;
     }
 
-    public void func_98042_n(int p_98042_1_)
+    public void setTransferTicker(int p_98042_1_)
     {
-        this.field_98044_b = p_98042_1_;
+        this.transferTicker = p_98042_1_;
     }
 
-    public boolean func_98043_aE()
+    public boolean canTransfer()
     {
-        return this.field_98044_b > 0;
+        return this.transferTicker > 0;
     }
 
-    public String func_174875_k()
+    public String getGuiID()
     {
         return "minecraft:hopper";
     }
 
-    public Container func_174876_a(InventoryPlayer p_174876_1_, EntityPlayer p_174876_2_)
+    public Container createContainer(InventoryPlayer playerInventory, EntityPlayer playerIn)
     {
-        return new ContainerHopper(p_174876_1_, this, p_174876_2_);
+        return new ContainerHopper(playerInventory, this, playerIn);
     }
 }
