--- ../src-base/minecraft/net/minecraft/entity/item/EntityTNTPrimed.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityTNTPrimed.java
@@ -17,7 +17,8 @@
     @Nullable
     private EntityLivingBase field_94084_b;
     private int field_70516_a;
-
+    public float yield = 4; // CraftBukkit - add field
+    public boolean isIncendiary = false; // CraftBukkit - add field
     public EntityTNTPrimed(World p_i1729_1_)
     {
         super(p_i1729_1_);
