--- ../src-base/minecraft/net/minecraft/entity/boss/EntityWither.java
+++ ../src-work/minecraft/net/minecraft/entity/boss/EntityWither.java
@@ -48,8 +48,10 @@
 import net.minecraft.world.BossInfoServer;
 import net.minecraft.world.EnumDifficulty;
 import net.minecraft.world.World;
+import net.minecraftforge.event.ForgeEventFactory;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class EntityWither extends EntityMob implements IRangedAttackMob
 {
@@ -383,9 +385,13 @@
                                 IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos);
                                 Block block = iblockstate.func_177230_c();
 
-                                if (iblockstate.func_185904_a() != Material.field_151579_a && func_181033_a(block))
+                                if (!block.isAir(iblockstate, this.field_70170_p, blockpos) && block.canEntityDestroy(iblockstate, field_70170_p, blockpos, this) && ForgeEventFactory.onEntityDestroyBlock(this, blockpos, iblockstate))
                                 {
-                                    flag = this.field_70170_p.func_175655_b(blockpos, true) || flag;
+// CraftBukkit start                //CraftBukkit start
+                                    if (CraftEventFactory.callEntityChangeBlockEvent(this, blockpos, Blocks.field_150350_a, 0).isCancelled()) {
+                                        continue;
+                                    }
+                                    // CraftBukkit end
                                 }
                             }
                         }
