--- ../src-base/minecraft/net/minecraft/item/ItemBucket.java
+++ ../src-work/minecraft/net/minecraft/item/ItemBucket.java
@@ -21,6 +21,10 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.RayTraceResult;
 import net.minecraft.world.World;
+import net.minecraftforge.event.ForgeEventFactory;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.player.PlayerBucketEmptyEvent;
+import org.bukkit.event.player.PlayerBucketFillEvent;
 
 public class ItemBucket extends Item
 {
@@ -38,6 +42,8 @@
         boolean flag = this.field_77876_a == Blocks.field_150350_a;
         ItemStack itemstack = p_77659_2_.func_184586_b(p_77659_3_);
         RayTraceResult raytraceresult = this.func_77621_a(p_77659_1_, p_77659_2_, flag);
+        ActionResult<ItemStack> ret = ForgeEventFactory.onBucketUse(p_77659_2_, p_77659_1_, itemstack, raytraceresult);
+        if (ret != null) return ret;
 
         if (raytraceresult == null)
         {
@@ -68,6 +74,13 @@
 
                     if (material == Material.field_151586_h && ((Integer)iblockstate.func_177229_b(BlockLiquid.field_176367_b)).intValue() == 0)
                     {
+                        // CraftBukkit start
+                        PlayerBucketFillEvent event = CraftEventFactory.callPlayerBucketFillEvent(p_77659_2_, blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p(), null, itemstack, Items.field_151131_as);
+
+                        if (event.isCancelled()) {
+                            return new ActionResult(EnumActionResult.FAIL, itemstack);
+                        }
+                        // CraftBukkit end
                         p_77659_1_.func_180501_a(blockpos, Blocks.field_150350_a.func_176223_P(), 11);
                         p_77659_2_.func_71029_a(StatList.func_188057_b(this));
                         p_77659_2_.func_184185_a(SoundEvents.field_187630_M, 1.0F, 1.0F);
@@ -75,6 +88,13 @@
                     }
                     else if (material == Material.field_151587_i && ((Integer)iblockstate.func_177229_b(BlockLiquid.field_176367_b)).intValue() == 0)
                     {
+                        // CraftBukkit start
+                        PlayerBucketFillEvent event = CraftEventFactory.callPlayerBucketFillEvent(p_77659_2_, blockpos.func_177958_n(), blockpos.func_177956_o(), blockpos.func_177952_p(), null, itemstack, Items.field_151129_at);
+
+                        if (event.isCancelled()) {
+                            return new ActionResult(EnumActionResult.FAIL, itemstack);
+                        }
+                        // CraftBukkit end
                         p_77659_2_.func_184185_a(SoundEvents.field_187633_N, 1.0F, 1.0F);
                         p_77659_1_.func_180501_a(blockpos, Blocks.field_150350_a.func_176223_P(), 11);
                         p_77659_2_.func_71029_a(StatList.func_188057_b(this));
@@ -153,6 +173,15 @@
             }
             else
             {
+                // CraftBukkit start
+                if (p_180616_1_ != null) {
+                    PlayerBucketEmptyEvent event = CraftEventFactory.callPlayerBucketEmptyEvent(p_180616_1_, p_180616_3_.func_177958_n(), p_180616_3_.func_177956_o(), p_180616_3_.func_177952_p(), null, p_180616_1_.field_71071_by.func_70448_g());
+                    if (event.isCancelled()) {
+                        // TODO: inventory not updated
+                        return false;
+                    }
+                }
+                // CraftBukkit end
                 if (p_180616_2_.field_73011_w.func_177500_n() && this.field_77876_a == Blocks.field_150358_i)
                 {
                     int l = p_180616_3_.func_177958_n();
@@ -181,4 +210,16 @@
             }
         }
     }
+
+    @Override
+    public net.minecraftforge.common.capabilities.ICapabilityProvider initCapabilities(ItemStack stack, @Nullable net.minecraft.nbt.NBTTagCompound nbt) {
+        if (this.getClass() == ItemBucket.class)
+        {
+            return new net.minecraftforge.fluids.capability.wrappers.FluidBucketWrapper(stack);
+        }
+        else
+        {
+            return super.initCapabilities(stack, nbt);
+        }
+    }
 }
