--- ../src-base/minecraft/net/minecraft/world/Teleporter.java
+++ ../src-work/minecraft/net/minecraft/world/Teleporter.java
@@ -17,31 +17,31 @@
 
 public class Teleporter
 {
-    protected final WorldServer field_85192_a;
-    protected final Random field_77187_a;
-    protected final Long2ObjectMap<Teleporter.PortalPosition> field_85191_c = new Long2ObjectOpenHashMap(4096);
+    protected final WorldServer worldServerInstance;
+    protected final Random random;
+    protected final Long2ObjectMap<Teleporter.PortalPosition> destinationCoordinateCache = new Long2ObjectOpenHashMap(4096);
 
-    public Teleporter(WorldServer p_i1963_1_)
+    public Teleporter(WorldServer worldIn)
     {
-        this.field_85192_a = p_i1963_1_;
-        this.field_77187_a = new Random(p_i1963_1_.func_72905_C());
+        this.worldServerInstance = worldIn;
+        this.random = new Random(worldIn.getSeed());
     }
 
-    public void func_180266_a(Entity p_180266_1_, float p_180266_2_)
+    public void placeInPortal(Entity entityIn, float rotationYaw)
     {
-        if (this.field_85192_a.field_73011_w.func_186058_p().func_186068_a() != 1)
+        if (this.worldServerInstance.provider.getDimensionType().getId() != 1)
         {
-            if (!this.func_180620_b(p_180266_1_, p_180266_2_))
+            if (!this.placeInExistingPortal(entityIn, rotationYaw))
             {
-                this.func_85188_a(p_180266_1_);
-                this.func_180620_b(p_180266_1_, p_180266_2_);
+                this.makePortal(entityIn);
+                this.placeInExistingPortal(entityIn, rotationYaw);
             }
         }
         else
         {
-            int i = MathHelper.func_76128_c(p_180266_1_.field_70165_t);
-            int j = MathHelper.func_76128_c(p_180266_1_.field_70163_u) - 1;
-            int k = MathHelper.func_76128_c(p_180266_1_.field_70161_v);
+            int i = MathHelper.floor(entityIn.posX);
+            int j = MathHelper.floor(entityIn.posY) - 1;
+            int k = MathHelper.floor(entityIn.posZ);
             int l = 1;
             int i1 = 0;
 
@@ -55,39 +55,39 @@
                         int j2 = j + l1;
                         int k2 = k + k1 * 0 - j1 * 1;
                         boolean flag = l1 < 0;
-                        this.field_85192_a.func_175656_a(new BlockPos(i2, j2, k2), flag ? Blocks.field_150343_Z.func_176223_P() : Blocks.field_150350_a.func_176223_P());
+                        this.worldServerInstance.setBlockState(new BlockPos(i2, j2, k2), flag ? Blocks.OBSIDIAN.getDefaultState() : Blocks.AIR.getDefaultState());
                     }
                 }
             }
 
-            p_180266_1_.func_70012_b((double)i, (double)j, (double)k, p_180266_1_.field_70177_z, 0.0F);
-            p_180266_1_.field_70159_w = 0.0D;
-            p_180266_1_.field_70181_x = 0.0D;
-            p_180266_1_.field_70179_y = 0.0D;
+            entityIn.setLocationAndAngles((double)i, (double)j, (double)k, entityIn.rotationYaw, 0.0F);
+            entityIn.motionX = 0.0D;
+            entityIn.motionY = 0.0D;
+            entityIn.motionZ = 0.0D;
         }
     }
 
-    public boolean func_180620_b(Entity p_180620_1_, float p_180620_2_)
+    public boolean placeInExistingPortal(Entity entityIn, float rotationYaw)
     {
         int i = 128;
         double d0 = -1.0D;
-        int j = MathHelper.func_76128_c(p_180620_1_.field_70165_t);
-        int k = MathHelper.func_76128_c(p_180620_1_.field_70161_v);
+        int j = MathHelper.floor(entityIn.posX);
+        int k = MathHelper.floor(entityIn.posZ);
         boolean flag = true;
-        BlockPos blockpos = BlockPos.field_177992_a;
-        long l = ChunkPos.func_77272_a(j, k);
+        BlockPos blockpos = BlockPos.ORIGIN;
+        long l = ChunkPos.asLong(j, k);
 
-        if (this.field_85191_c.containsKey(l))
+        if (this.destinationCoordinateCache.containsKey(l))
         {
-            Teleporter.PortalPosition teleporter$portalposition = (Teleporter.PortalPosition)this.field_85191_c.get(l);
+            Teleporter.PortalPosition teleporter$portalposition = (Teleporter.PortalPosition)this.destinationCoordinateCache.get(l);
             d0 = 0.0D;
             blockpos = teleporter$portalposition;
-            teleporter$portalposition.field_85087_d = this.field_85192_a.func_82737_E();
+            teleporter$portalposition.lastUpdateTime = this.worldServerInstance.getTotalWorldTime();
             flag = false;
         }
         else
         {
-            BlockPos blockpos3 = new BlockPos(p_180620_1_);
+            BlockPos blockpos3 = new BlockPos(entityIn);
 
             for (int i1 = -128; i1 <= 128; ++i1)
             {
@@ -95,18 +95,18 @@
 
                 for (int j1 = -128; j1 <= 128; ++j1)
                 {
-                    for (BlockPos blockpos1 = blockpos3.func_177982_a(i1, this.field_85192_a.func_72940_L() - 1 - blockpos3.func_177956_o(), j1); blockpos1.func_177956_o() >= 0; blockpos1 = blockpos2)
+                    for (BlockPos blockpos1 = blockpos3.add(i1, this.worldServerInstance.getActualHeight() - 1 - blockpos3.getY(), j1); blockpos1.getY() >= 0; blockpos1 = blockpos2)
                     {
-                        blockpos2 = blockpos1.func_177977_b();
+                        blockpos2 = blockpos1.down();
 
-                        if (this.field_85192_a.func_180495_p(blockpos1).func_177230_c() == Blocks.field_150427_aO)
+                        if (this.worldServerInstance.getBlockState(blockpos1).getBlock() == Blocks.PORTAL)
                         {
-                            for (blockpos2 = blockpos1.func_177977_b(); this.field_85192_a.func_180495_p(blockpos2).func_177230_c() == Blocks.field_150427_aO; blockpos2 = blockpos2.func_177977_b())
+                            for (blockpos2 = blockpos1.down(); this.worldServerInstance.getBlockState(blockpos2).getBlock() == Blocks.PORTAL; blockpos2 = blockpos2.down())
                             {
                                 blockpos1 = blockpos2;
                             }
 
-                            double d1 = blockpos1.func_177951_i(blockpos3);
+                            double d1 = blockpos1.distanceSq(blockpos3);
 
                             if (d0 < 0.0D || d1 < d0)
                             {
@@ -123,28 +123,28 @@
         {
             if (flag)
             {
-                this.field_85191_c.put(l, new Teleporter.PortalPosition(blockpos, this.field_85192_a.func_82737_E()));
+                this.destinationCoordinateCache.put(l, new Teleporter.PortalPosition(blockpos, this.worldServerInstance.getTotalWorldTime()));
             }
 
-            double d5 = (double)blockpos.func_177958_n() + 0.5D;
-            double d7 = (double)blockpos.func_177952_p() + 0.5D;
-            BlockPattern.PatternHelper blockpattern$patternhelper = Blocks.field_150427_aO.func_181089_f(this.field_85192_a, blockpos);
-            boolean flag1 = blockpattern$patternhelper.func_177669_b().func_176746_e().func_176743_c() == EnumFacing.AxisDirection.NEGATIVE;
-            double d2 = blockpattern$patternhelper.func_177669_b().func_176740_k() == EnumFacing.Axis.X ? (double)blockpattern$patternhelper.func_181117_a().func_177952_p() : (double)blockpattern$patternhelper.func_181117_a().func_177958_n();
-            double d6 = (double)(blockpattern$patternhelper.func_181117_a().func_177956_o() + 1) - p_180620_1_.func_181014_aG().field_72448_b * (double)blockpattern$patternhelper.func_181119_e();
+            double d5 = (double)blockpos.getX() + 0.5D;
+            double d7 = (double)blockpos.getZ() + 0.5D;
+            BlockPattern.PatternHelper blockpattern$patternhelper = Blocks.PORTAL.createPatternHelper(this.worldServerInstance, blockpos);
+            boolean flag1 = blockpattern$patternhelper.getForwards().rotateY().getAxisDirection() == EnumFacing.AxisDirection.NEGATIVE;
+            double d2 = blockpattern$patternhelper.getForwards().getAxis() == EnumFacing.Axis.X ? (double)blockpattern$patternhelper.getFrontTopLeft().getZ() : (double)blockpattern$patternhelper.getFrontTopLeft().getX();
+            double d6 = (double)(blockpattern$patternhelper.getFrontTopLeft().getY() + 1) - entityIn.getLastPortalVec().yCoord * (double)blockpattern$patternhelper.getHeight();
 
             if (flag1)
             {
                 ++d2;
             }
 
-            if (blockpattern$patternhelper.func_177669_b().func_176740_k() == EnumFacing.Axis.X)
+            if (blockpattern$patternhelper.getForwards().getAxis() == EnumFacing.Axis.X)
             {
-                d7 = d2 + (1.0D - p_180620_1_.func_181014_aG().field_72450_a) * (double)blockpattern$patternhelper.func_181118_d() * (double)blockpattern$patternhelper.func_177669_b().func_176746_e().func_176743_c().func_179524_a();
+                d7 = d2 + (1.0D - entityIn.getLastPortalVec().xCoord) * (double)blockpattern$patternhelper.getWidth() * (double)blockpattern$patternhelper.getForwards().rotateY().getAxisDirection().getOffset();
             }
             else
             {
-                d5 = d2 + (1.0D - p_180620_1_.func_181014_aG().field_72450_a) * (double)blockpattern$patternhelper.func_181118_d() * (double)blockpattern$patternhelper.func_177669_b().func_176746_e().func_176743_c().func_179524_a();
+                d5 = d2 + (1.0D - entityIn.getLastPortalVec().xCoord) * (double)blockpattern$patternhelper.getWidth() * (double)blockpattern$patternhelper.getForwards().rotateY().getAxisDirection().getOffset();
             }
 
             float f = 0.0F;
@@ -152,17 +152,17 @@
             float f2 = 0.0F;
             float f3 = 0.0F;
 
-            if (blockpattern$patternhelper.func_177669_b().func_176734_d() == p_180620_1_.func_181012_aH())
+            if (blockpattern$patternhelper.getForwards().getOpposite() == entityIn.getTeleportDirection())
             {
                 f = 1.0F;
                 f1 = 1.0F;
             }
-            else if (blockpattern$patternhelper.func_177669_b().func_176734_d() == p_180620_1_.func_181012_aH().func_176734_d())
+            else if (blockpattern$patternhelper.getForwards().getOpposite() == entityIn.getTeleportDirection().getOpposite())
             {
                 f = -1.0F;
                 f1 = -1.0F;
             }
-            else if (blockpattern$patternhelper.func_177669_b().func_176734_d() == p_180620_1_.func_181012_aH().func_176746_e())
+            else if (blockpattern$patternhelper.getForwards().getOpposite() == entityIn.getTeleportDirection().rotateY())
             {
                 f2 = 1.0F;
                 f3 = -1.0F;
@@ -173,19 +173,19 @@
                 f3 = 1.0F;
             }
 
-            double d3 = p_180620_1_.field_70159_w;
-            double d4 = p_180620_1_.field_70179_y;
-            p_180620_1_.field_70159_w = d3 * (double)f + d4 * (double)f3;
-            p_180620_1_.field_70179_y = d3 * (double)f2 + d4 * (double)f1;
-            p_180620_1_.field_70177_z = p_180620_2_ - (float)(p_180620_1_.func_181012_aH().func_176734_d().func_176736_b() * 90) + (float)(blockpattern$patternhelper.func_177669_b().func_176736_b() * 90);
+            double d3 = entityIn.motionX;
+            double d4 = entityIn.motionZ;
+            entityIn.motionX = d3 * (double)f + d4 * (double)f3;
+            entityIn.motionZ = d3 * (double)f2 + d4 * (double)f1;
+            entityIn.rotationYaw = rotationYaw - (float)(entityIn.getTeleportDirection().getOpposite().getHorizontalIndex() * 90) + (float)(blockpattern$patternhelper.getForwards().getHorizontalIndex() * 90);
 
-            if (p_180620_1_ instanceof EntityPlayerMP)
+            if (entityIn instanceof EntityPlayerMP)
             {
-                ((EntityPlayerMP)p_180620_1_).field_71135_a.func_147364_a(d5, d6, d7, p_180620_1_.field_70177_z, p_180620_1_.field_70125_A);
+                ((EntityPlayerMP)entityIn).connection.setPlayerLocation(d5, d6, d7, entityIn.rotationYaw, entityIn.rotationPitch);
             }
             else
             {
-                p_180620_1_.func_70012_b(d5, d6, d7, p_180620_1_.field_70177_z, p_180620_1_.field_70125_A);
+                entityIn.setLocationAndAngles(d5, d6, d7, entityIn.rotationYaw, entityIn.rotationPitch);
             }
 
             return true;
@@ -196,34 +196,34 @@
         }
     }
 
-    public boolean func_85188_a(Entity p_85188_1_)
+    public boolean makePortal(Entity entityIn)
     {
         int i = 16;
         double d0 = -1.0D;
-        int j = MathHelper.func_76128_c(p_85188_1_.field_70165_t);
-        int k = MathHelper.func_76128_c(p_85188_1_.field_70163_u);
-        int l = MathHelper.func_76128_c(p_85188_1_.field_70161_v);
+        int j = MathHelper.floor(entityIn.posX);
+        int k = MathHelper.floor(entityIn.posY);
+        int l = MathHelper.floor(entityIn.posZ);
         int i1 = j;
         int j1 = k;
         int k1 = l;
         int l1 = 0;
-        int i2 = this.field_77187_a.nextInt(4);
+        int i2 = this.random.nextInt(4);
         BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
 
         for (int j2 = j - 16; j2 <= j + 16; ++j2)
         {
-            double d1 = (double)j2 + 0.5D - p_85188_1_.field_70165_t;
+            double d1 = (double)j2 + 0.5D - entityIn.posX;
 
             for (int l2 = l - 16; l2 <= l + 16; ++l2)
             {
-                double d2 = (double)l2 + 0.5D - p_85188_1_.field_70161_v;
+                double d2 = (double)l2 + 0.5D - entityIn.posZ;
                 label146:
 
-                for (int j3 = this.field_85192_a.func_72940_L() - 1; j3 >= 0; --j3)
+                for (int j3 = this.worldServerInstance.getActualHeight() - 1; j3 >= 0; --j3)
                 {
-                    if (this.field_85192_a.func_175623_d(blockpos$mutableblockpos.func_181079_c(j2, j3, l2)))
+                    if (this.worldServerInstance.isAirBlock(blockpos$mutableblockpos.setPos(j2, j3, l2)))
                     {
-                        while (j3 > 0 && this.field_85192_a.func_175623_d(blockpos$mutableblockpos.func_181079_c(j2, j3 - 1, l2)))
+                        while (j3 > 0 && this.worldServerInstance.isAirBlock(blockpos$mutableblockpos.setPos(j2, j3 - 1, l2)))
                         {
                             --j3;
                         }
@@ -248,9 +248,9 @@
                                         int i5 = j2 + (k4 - 1) * l3 + j4 * i4;
                                         int j5 = j3 + l4;
                                         int k5 = l2 + (k4 - 1) * i4 - j4 * l3;
-                                        blockpos$mutableblockpos.func_181079_c(i5, j5, k5);
+                                        blockpos$mutableblockpos.setPos(i5, j5, k5);
 
-                                        if (l4 < 0 && !this.field_85192_a.func_180495_p(blockpos$mutableblockpos).func_185904_a().func_76220_a() || l4 >= 0 && !this.field_85192_a.func_175623_d(blockpos$mutableblockpos))
+                                        if (l4 < 0 && !this.worldServerInstance.getBlockState(blockpos$mutableblockpos).getMaterial().isSolid() || l4 >= 0 && !this.worldServerInstance.isAirBlock(blockpos$mutableblockpos))
                                         {
                                             continue label146;
                                         }
@@ -258,7 +258,7 @@
                                 }
                             }
 
-                            double d5 = (double)j3 + 0.5D - p_85188_1_.field_70163_u;
+                            double d5 = (double)j3 + 0.5D - entityIn.posY;
                             double d7 = d1 * d1 + d5 * d5 + d2 * d2;
 
                             if (d0 < 0.0D || d7 < d0)
@@ -279,18 +279,18 @@
         {
             for (int l5 = j - 16; l5 <= j + 16; ++l5)
             {
-                double d3 = (double)l5 + 0.5D - p_85188_1_.field_70165_t;
+                double d3 = (double)l5 + 0.5D - entityIn.posX;
 
                 for (int j6 = l - 16; j6 <= l + 16; ++j6)
                 {
-                    double d4 = (double)j6 + 0.5D - p_85188_1_.field_70161_v;
+                    double d4 = (double)j6 + 0.5D - entityIn.posZ;
                     label567:
 
-                    for (int i7 = this.field_85192_a.func_72940_L() - 1; i7 >= 0; --i7)
+                    for (int i7 = this.worldServerInstance.getActualHeight() - 1; i7 >= 0; --i7)
                     {
-                        if (this.field_85192_a.func_175623_d(blockpos$mutableblockpos.func_181079_c(l5, i7, j6)))
+                        if (this.worldServerInstance.isAirBlock(blockpos$mutableblockpos.setPos(l5, i7, j6)))
                         {
-                            while (i7 > 0 && this.field_85192_a.func_175623_d(blockpos$mutableblockpos.func_181079_c(l5, i7 - 1, j6)))
+                            while (i7 > 0 && this.worldServerInstance.isAirBlock(blockpos$mutableblockpos.setPos(l5, i7 - 1, j6)))
                             {
                                 --i7;
                             }
@@ -307,16 +307,16 @@
                                         int j12 = l5 + (j10 - 1) * j8;
                                         int i13 = i7 + j11;
                                         int j13 = j6 + (j10 - 1) * j9;
-                                        blockpos$mutableblockpos.func_181079_c(j12, i13, j13);
+                                        blockpos$mutableblockpos.setPos(j12, i13, j13);
 
-                                        if (j11 < 0 && !this.field_85192_a.func_180495_p(blockpos$mutableblockpos).func_185904_a().func_76220_a() || j11 >= 0 && !this.field_85192_a.func_175623_d(blockpos$mutableblockpos))
+                                        if (j11 < 0 && !this.worldServerInstance.getBlockState(blockpos$mutableblockpos).getMaterial().isSolid() || j11 >= 0 && !this.worldServerInstance.isAirBlock(blockpos$mutableblockpos))
                                         {
                                             continue label567;
                                         }
                                     }
                                 }
 
-                                double d6 = (double)i7 + 0.5D - p_85188_1_.field_70163_u;
+                                double d6 = (double)i7 + 0.5D - entityIn.posY;
                                 double d8 = d3 * d3 + d6 * d6 + d4 * d4;
 
                                 if (d0 < 0.0D || d8 < d0)
@@ -348,7 +348,7 @@
 
         if (d0 < 0.0D)
         {
-            j1 = MathHelper.func_76125_a(j1, 70, this.field_85192_a.func_72940_L() - 10);
+            j1 = MathHelper.clamp(j1, 70, this.worldServerInstance.getActualHeight() - 10);
             k2 = j1;
 
             for (int j7 = -1; j7 <= 1; ++j7)
@@ -361,13 +361,13 @@
                         int k10 = k2 + k8;
                         int k11 = k6 + (l7 - 1) * i3 - j7 * l6;
                         boolean flag = k8 < 0;
-                        this.field_85192_a.func_175656_a(new BlockPos(k9, k10, k11), flag ? Blocks.field_150343_Z.func_176223_P() : Blocks.field_150350_a.func_176223_P());
+                        this.worldServerInstance.setBlockState(new BlockPos(k9, k10, k11), flag ? Blocks.OBSIDIAN.getDefaultState() : Blocks.AIR.getDefaultState());
                     }
                 }
             }
         }
 
-        IBlockState iblockstate = Blocks.field_150427_aO.func_176223_P().func_177226_a(BlockPortal.field_176550_a, l6 == 0 ? EnumFacing.Axis.Z : EnumFacing.Axis.X);
+        IBlockState iblockstate = Blocks.PORTAL.getDefaultState().withProperty(BlockPortal.AXIS, l6 == 0 ? EnumFacing.Axis.Z : EnumFacing.Axis.X);
 
         for (int i8 = 0; i8 < 4; ++i8)
         {
@@ -379,7 +379,7 @@
                     int l11 = k2 + l9;
                     int k12 = k6 + (l8 - 1) * i3;
                     boolean flag1 = l8 == 0 || l8 == 3 || l9 == -1 || l9 == 3;
-                    this.field_85192_a.func_180501_a(new BlockPos(l10, l11, k12), flag1 ? Blocks.field_150343_Z.func_176223_P() : iblockstate, 2);
+                    this.worldServerInstance.setBlockState(new BlockPos(l10, l11, k12), flag1 ? Blocks.OBSIDIAN.getDefaultState() : iblockstate, 2);
                 }
             }
 
@@ -391,7 +391,7 @@
                     int i12 = k2 + i10;
                     int l12 = k6 + (i9 - 1) * i3;
                     BlockPos blockpos = new BlockPos(i11, i12, l12);
-                    this.field_85192_a.func_175685_c(blockpos, this.field_85192_a.func_180495_p(blockpos).func_177230_c(), false);
+                    this.worldServerInstance.notifyNeighborsOfStateChange(blockpos, this.worldServerInstance.getBlockState(blockpos).getBlock(), false);
                 }
             }
         }
@@ -399,18 +399,18 @@
         return true;
     }
 
-    public void func_85189_a(long p_85189_1_)
+    public void removeStalePortalLocations(long worldTime)
     {
-        if (p_85189_1_ % 100L == 0L)
+        if (worldTime % 100L == 0L)
         {
-            long i = p_85189_1_ - 300L;
-            ObjectIterator<Teleporter.PortalPosition> objectiterator = this.field_85191_c.values().iterator();
+            long i = worldTime - 300L;
+            ObjectIterator<Teleporter.PortalPosition> objectiterator = this.destinationCoordinateCache.values().iterator();
 
             while (objectiterator.hasNext())
             {
                 Teleporter.PortalPosition teleporter$portalposition = (Teleporter.PortalPosition)objectiterator.next();
 
-                if (teleporter$portalposition == null || teleporter$portalposition.field_85087_d < i)
+                if (teleporter$portalposition == null || teleporter$portalposition.lastUpdateTime < i)
                 {
                     objectiterator.remove();
                 }
@@ -420,12 +420,310 @@
 
     public class PortalPosition extends BlockPos
     {
-        public long field_85087_d;
+        public long lastUpdateTime;
 
-        public PortalPosition(BlockPos p_i45747_2_, long p_i45747_3_)
+        public PortalPosition(BlockPos pos, long lastUpdate)
         {
-            super(p_i45747_2_.func_177958_n(), p_i45747_2_.func_177956_o(), p_i45747_2_.func_177952_p());
-            this.field_85087_d = p_i45747_3_;
+            super(pos.getX(), pos.getY(), pos.getZ());
+            this.lastUpdateTime = lastUpdate;
         }
     }
+
+    public BlockPos findPortal(double x, double y, double z, int radius) {
+        if (this.worldServerInstance.getWorld().getEnvironment() == org.bukkit.World.Environment.THE_END) {
+            return this.findEndPortal(this.worldServerInstance.provider.getSpawnCoordinate());
+        }
+        // CraftBukkit end
+        double d0 = -1.0D;
+        // CraftBukkit start
+        int i = MathHelper.floor(x);
+        int j = MathHelper.floor(z);
+        // CraftBukkit end
+        boolean flag1 = true;
+        Object object = BlockPos.ORIGIN;
+        long k = ChunkPos.asLong(i, j);
+
+        if (this.destinationCoordinateCache.containsKey(k)) {
+            PortalPosition portaltravelagent_chunkcoordinatesportal = (PortalPosition) this.destinationCoordinateCache.get(k);
+
+            d0 = 0.0D;
+            object = portaltravelagent_chunkcoordinatesportal;
+            portaltravelagent_chunkcoordinatesportal.lastUpdateTime = this.worldServerInstance.getTotalWorldTime();
+            flag1 = false;
+        } else {
+            BlockPos blockposition = new BlockPos(x, y, z); // CraftBukkit
+
+            for (int l = -radius; l <= radius; ++l) {
+                BlockPos blockposition1;
+
+                for (int i1 = -radius; i1 <= radius; ++i1) {
+                    for (BlockPos blockposition2 = blockposition.add(l, this.worldServerInstance.getActualHeight() - 1 - blockposition.getY(), i1); blockposition2.getY() >= 0; blockposition2 = blockposition1) {
+                        blockposition1 = blockposition2.down();
+                        if (this.worldServerInstance.getBlockState(blockposition2).getBlock() == Blocks.PORTAL) {
+                            for (blockposition1 = blockposition2.down(); this.worldServerInstance.getBlockState(blockposition1).getBlock() == Blocks.PORTAL; blockposition1 = blockposition1.down()) {
+                                blockposition2 = blockposition1;
+                            }
+
+                            double d1 = blockposition2.distanceSq(blockposition);
+
+                            if (d0 < 0.0D || d1 < d0) {
+                                d0 = d1;
+                                object = blockposition2;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        if (d0 >= 0.0D) {
+            if (flag1) {
+                this.destinationCoordinateCache.put(k, new PortalPosition((BlockPos) object, this.worldServerInstance.getTotalWorldTime()));
+            }
+            // CraftBukkit start - Move entity teleportation logic into exit
+            return (BlockPos) object;
+        } else {
+            return null;
+        }
+    }
+    public boolean createPortal(double x, double y, double z, int b0) {
+        if (this.worldServerInstance.getWorld().getEnvironment() == org.bukkit.World.Environment.THE_END) {
+            createEndPortal(x, y, z);
+            return true;
+        }
+        // CraftBukkit end
+        double d0 = -1.0D;
+        // CraftBukkit start
+        int i = MathHelper.floor(x);
+        int j = MathHelper.floor(y);
+        int k = MathHelper.floor(z);
+        // CraftBukkit end
+        int l = i;
+        int i1 = j;
+        int j1 = k;
+        int k1 = 0;
+        int l1 = this.random.nextInt(4);
+        BlockPos.MutableBlockPos blockposition_mutableblockposition = new BlockPos.MutableBlockPos();
+
+        int i2;
+        double d1;
+        int j2;
+        double d2;
+        int k2;
+        int l2;
+        int i3;
+        int j3;
+        int k3;
+        int l3;
+        int i4;
+        int j4;
+        int k4;
+        double d3;
+        double d4;
+
+        for (i2 = i - 16; i2 <= i + 16; ++i2) {
+            d1 = (double) i2 + 0.5D - x; // CraftBukkit
+
+            for (j2 = k - 16; j2 <= k + 16; ++j2) {
+                d2 = (double) j2 + 0.5D - z; // CraftBukkit
+
+                label271:
+                for (k2 = this.worldServerInstance.getActualHeight() - 1; k2 >= 0; --k2) {
+                    if (this.worldServerInstance.isAirBlock(blockposition_mutableblockposition.setPos(i2, k2, j2))) {
+                        while (k2 > 0 && this.worldServerInstance.isAirBlock(blockposition_mutableblockposition.setPos(i2, k2 - 1, j2))) {
+                            --k2;
+                        }
+
+                        for (l2 = l1; l2 < l1 + 4; ++l2) {
+                            i3 = l2 % 2;
+                            j3 = 1 - i3;
+                            if (l2 % 4 >= 2) {
+                                i3 = -i3;
+                                j3 = -j3;
+                            }
+
+                            for (k3 = 0; k3 < 3; ++k3) {
+                                for (l3 = 0; l3 < 4; ++l3) {
+                                    for (i4 = -1; i4 < 4; ++i4) {
+                                        j4 = i2 + (l3 - 1) * i3 + k3 * j3;
+                                        k4 = k2 + i4;
+                                        int l4 = j2 + (l3 - 1) * j3 - k3 * i3;
+
+                                        blockposition_mutableblockposition.setPos(j4, k4, l4);
+                                        if (i4 < 0 && !this.worldServerInstance.getBlockState(blockposition_mutableblockposition).getMaterial().isSolid() || i4 >= 0 && !this.worldServerInstance.isAirBlock(blockposition_mutableblockposition)) {
+                                            continue label271;
+                                        }
+                                    }
+                                }
+                            }
+
+                            d3 = (double) k2 + 0.5D - y; // CraftBukkit
+                            d4 = d1 * d1 + d3 * d3 + d2 * d2;
+                            if (d0 < 0.0D || d4 < d0) {
+                                d0 = d4;
+                                l = i2;
+                                i1 = k2;
+                                j1 = j2;
+                                k1 = l2 % 4;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        if (d0 < 0.0D) {
+            for (i2 = i - 16; i2 <= i + 16; ++i2) {
+                d1 = (double) i2 + 0.5D - x; // CraftBukkit
+
+                for (j2 = k - 16; j2 <= k + 16; ++j2) {
+                    d2 = (double) j2 + 0.5D - z; // CraftBukkit
+
+                    label219:
+                    for (k2 = this.worldServerInstance.getActualHeight() - 1; k2 >= 0; --k2) {
+                        if (this.worldServerInstance.isAirBlock(blockposition_mutableblockposition.setPos(i2, k2, j2))) {
+                            while (k2 > 0 && this.worldServerInstance.isAirBlock(blockposition_mutableblockposition.setPos(i2, k2 - 1, j2))) {
+                                --k2;
+                            }
+
+                            for (l2 = l1; l2 < l1 + 2; ++l2) {
+                                i3 = l2 % 2;
+                                j3 = 1 - i3;
+
+                                for (k3 = 0; k3 < 4; ++k3) {
+                                    for (l3 = -1; l3 < 4; ++l3) {
+                                        i4 = i2 + (k3 - 1) * i3;
+                                        j4 = k2 + l3;
+                                        k4 = j2 + (k3 - 1) * j3;
+                                        blockposition_mutableblockposition.setPos(i4, j4, k4);
+                                        if (l3 < 0 && !this.worldServerInstance.getBlockState(blockposition_mutableblockposition).getMaterial().isSolid() || l3 >= 0 && !this.worldServerInstance.isAirBlock(blockposition_mutableblockposition)) {
+                                            continue label219;
+                                        }
+                                    }
+                                }
+
+                                d3 = (double) k2 + 0.5D - y; // CraftBukkit
+                                d4 = d1 * d1 + d3 * d3 + d2 * d2;
+                                if (d0 < 0.0D || d4 < d0) {
+                                    d0 = d4;
+                                    l = i2;
+                                    i1 = k2;
+                                    j1 = j2;
+                                    k1 = l2 % 2;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        int i5 = l;
+        int j5 = i1;
+
+        j2 = j1;
+        int k5 = k1 % 2;
+        int l5 = 1 - k5;
+
+        if (k1 % 4 >= 2) {
+            k5 = -k5;
+            l5 = -l5;
+        }
+
+        if (d0 < 0.0D) {
+            i1 = MathHelper.clamp(i1, 70, this.worldServerInstance.getActualHeight() - 10);
+            j5 = i1;
+
+            for (k2 = -1; k2 <= 1; ++k2) {
+                for (l2 = 1; l2 < 3; ++l2) {
+                    for (i3 = -1; i3 < 3; ++i3) {
+                        j3 = i5 + (l2 - 1) * k5 + k2 * l5;
+                        k3 = j5 + i3;
+                        l3 = j2 + (l2 - 1) * l5 - k2 * k5;
+                        boolean flag1 = i3 < 0;
+
+                        this.worldServerInstance.setBlockState(new BlockPos(j3, k3, l3), flag1 ? Blocks.OBSIDIAN.getDefaultState() : Blocks.AIR.getDefaultState());
+                    }
+                }
+            }
+        }
+
+        IBlockState iblockdata = Blocks.PORTAL.getDefaultState().withProperty(BlockPortal.AXIS, k5 == 0 ? EnumFacing.Axis.Z : EnumFacing.Axis.X);
+
+        for (l2 = 0; l2 < 4; ++l2) {
+            for (i3 = 0; i3 < 4; ++i3) {
+                for (j3 = -1; j3 < 4; ++j3) {
+                    k3 = i5 + (i3 - 1) * k5;
+                    l3 = j5 + j3;
+                    i4 = j2 + (i3 - 1) * l5;
+                    boolean flag2 = i3 == 0 || i3 == 3 || j3 == -1 || j3 == 3;
+
+                    this.worldServerInstance.setBlockState(new BlockPos(k3, l3, i4), flag2 ? Blocks.OBSIDIAN.getDefaultState() : iblockdata, 2);
+                }
+            }
+
+            for (i3 = 0; i3 < 4; ++i3) {
+                for (j3 = -1; j3 < 4; ++j3) {
+                    k3 = i5 + (i3 - 1) * k5;
+                    l3 = j5 + j3;
+                    i4 = j2 + (i3 - 1) * l5;
+                    BlockPos blockposition = new BlockPos(k3, l3, i4);
+
+                    this.worldServerInstance.notifyNeighborsOfStateChange(blockposition, this.worldServerInstance.getBlockState(blockposition).getBlock(), false);
+                }
+            }
+        }
+
+        return true;
+    }
+    // use logic based on creation to verify end portal
+    private BlockPos findEndPortal(BlockPos portal) {
+        int i = portal.getX();
+        int j = portal.getY() - 1;
+        int k = portal.getZ();
+        byte b0 = 1;
+        byte b1 = 0;
+
+        for (int l = -2; l <= 2; ++l) {
+            for (int i1 = -2; i1 <= 2; ++i1) {
+                for (int j1 = -1; j1 < 3; ++j1) {
+                    int k1 = i + i1 * b0 + l * b1;
+                    int l1 = j + j1;
+                    int i2 = k + i1 * b1 - l * b0;
+                    boolean flag = j1 < 0;
+
+                    if (this.worldServerInstance.getBlockState(new BlockPos(k1, l1, i2)).getBlock() != (flag ? Blocks.OBSIDIAN : Blocks.AIR)) {
+                        return null;
+                    }
+                }
+            }
+        }
+        return new BlockPos(i, j, k);
+    }
+    // CraftBukkit end
+    // Split out from original a(Entity, double, double, double, float) method in order to enable being called from createPortal
+    private BlockPos createEndPortal(double x, double y, double z) {
+        int i = MathHelper.floor(x);
+        int j = MathHelper.floor(y) - 1;
+        int k = MathHelper.floor(z);
+        // CraftBukkit end
+        byte b0 = 1;
+        byte b1 = 0;
+
+        for (int l = -2; l <= 2; ++l) {
+            for (int i1 = -2; i1 <= 2; ++i1) {
+                for (int j1 = -1; j1 < 3; ++j1) {
+                    int k1 = i + i1 * 1 + l * 0;
+                    int l1 = j + j1;
+                    int i2 = k + i1 * 0 - l * 1;
+                    boolean flag2 = j1 < 0;
+
+                    this.worldServerInstance.setBlockState(new BlockPos(k1, l1, i2), flag2 ? Blocks.OBSIDIAN.getDefaultState() : Blocks.AIR.getDefaultState());
+                }
+            }
+        }
+
+        // CraftBukkit start
+        return new BlockPos(i, k, k);
+    }
 }
