--- ../src-base/minecraft/net/minecraft/block/BlockTripWireHook.java
+++ ../src-work/minecraft/net/minecraft/block/BlockTripWireHook.java
@@ -26,6 +26,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class BlockTripWireHook extends Block
 {
@@ -79,14 +80,14 @@
 
     public boolean func_176198_a(World p_176198_1_, BlockPos p_176198_2_, EnumFacing p_176198_3_)
     {
-        return p_176198_3_.func_176740_k().func_176722_c() && p_176198_1_.func_180495_p(p_176198_2_.func_177972_a(p_176198_3_.func_176734_d())).func_185915_l();
+        return p_176198_3_.func_176740_k().func_176722_c() && p_176198_1_.func_180495_p(p_176198_2_.func_177972_a(p_176198_3_.func_176734_d())).isSideSolid(p_176198_1_, p_176198_2_.func_177972_a(p_176198_3_.func_176734_d()), p_176198_3_);
     }
 
     public boolean func_176196_c(World p_176196_1_, BlockPos p_176196_2_)
     {
         for (EnumFacing enumfacing : EnumFacing.Plane.HORIZONTAL)
         {
-            if (p_176196_1_.func_180495_p(p_176196_2_.func_177972_a(enumfacing)).func_185915_l())
+            if (p_176196_1_.func_180495_p(p_176196_2_.func_177972_a(enumfacing)).isSideSolid(p_176196_1_, p_176196_2_.func_177972_a(enumfacing), enumfacing.func_176734_d()))
             {
                 return true;
             }
@@ -120,7 +121,7 @@
             {
                 EnumFacing enumfacing = (EnumFacing)p_189540_1_.func_177229_b(field_176264_a);
 
-                if (!p_189540_2_.func_180495_p(p_189540_3_.func_177972_a(enumfacing.func_176734_d())).func_185915_l())
+                if (!p_189540_2_.func_180495_p(p_189540_3_.func_177972_a(enumfacing.func_176734_d())).isSideSolid(p_189540_2_, p_189540_3_.func_177972_a(enumfacing.func_176734_d()), enumfacing))
                 {
                     this.func_176226_b(p_189540_2_, p_189540_3_, p_189540_1_, 0);
                     p_189540_2_.func_175698_g(p_189540_3_);
@@ -191,7 +192,16 @@
             this.func_176262_b(p_176260_1_, blockpos1, enumfacing1);
             this.func_180694_a(p_176260_1_, blockpos1, flag2, flag3, flag, flag1);
         }
+        // CraftBukkit start
+        org.bukkit.block.Block block = p_176260_1_.getWorld().getBlockAt(p_176260_2_.func_177958_n(), p_176260_2_.func_177956_o(), p_176260_2_.func_177952_p());
 
+        BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, 15, 0);
+        p_176260_1_.getServer().getPluginManager().callEvent(eventRedstone);
+
+        if (eventRedstone.getNewCurrent() > 0) {
+            return;
+        }
+        // CraftBukkit end
         this.func_180694_a(p_176260_1_, p_176260_2_, flag2, flag3, flag, flag1);
 
         if (!p_176260_4_)
