--- ../src-base/minecraft/net/minecraft/block/BlockCauldron.java
+++ ../src-work/minecraft/net/minecraft/block/BlockCauldron.java
@@ -30,6 +30,7 @@
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
+import org.bukkit.event.block.CauldronLevelChangeEvent;
 
 public class BlockCauldron extends Block
 {
@@ -77,7 +78,13 @@
 
         if (!p_180634_1_.field_72995_K && p_180634_4_.func_70027_ad() && i > 0 && p_180634_4_.func_174813_aQ().field_72338_b <= (double)f)
         {
+            // CraftBukkit start
+            if (!this.changeLevel(p_180634_1_, p_180634_2_, p_180634_3_, i - 1, p_180634_4_, CauldronLevelChangeEvent.ChangeReason.EXTINGUISH)) {
+                return;
+            }
             p_180634_4_.func_70066_B();
+            // this.a(world, blockposition, iblockdata, i - 1);
+            // CraftBukkit end
             this.func_176590_a(p_180634_1_, p_180634_2_, p_180634_3_, i - 1);
         }
     }
@@ -99,13 +106,17 @@
             {
                 if (i < 3 && !p_180639_1_.field_72995_K)
                 {
-                    if (!p_180639_4_.field_71075_bZ.field_75098_d)
-                    {
+                    // CraftBukkit start
+                    if (!this.changeLevel(p_180639_1_, p_180639_2_, p_180639_3_, 3, p_180639_4_, CauldronLevelChangeEvent.ChangeReason.BUCKET_EMPTY)) {
+                        return true;
+                    }
+                    if (!p_180639_4_.field_71075_bZ.field_75098_d) {
                         p_180639_4_.func_184611_a(p_180639_5_, new ItemStack(Items.field_151133_ar));
                     }
 
                     p_180639_4_.func_71029_a(StatList.field_188077_K);
-                    this.func_176590_a(p_180639_1_, p_180639_2_, p_180639_3_, 3);
+                    // this.a(world, blockposition, iblockdata, 3);
+                    // CraftBukkit end
                     p_180639_1_.func_184133_a((EntityPlayer)null, p_180639_2_, SoundEvents.field_187624_K, SoundCategory.BLOCKS, 1.0F, 1.0F);
                 }
 
@@ -115,22 +126,22 @@
             {
                 if (i == 3 && !p_180639_1_.field_72995_K)
                 {
-                    if (!p_180639_4_.field_71075_bZ.field_75098_d)
-                    {
+                    // CraftBukkit start
+                    if (!this.changeLevel(p_180639_1_, p_180639_2_, p_180639_3_, 0, p_180639_4_, CauldronLevelChangeEvent.ChangeReason.BUCKET_FILL)) {
+                        return true;
+                    }
+                    if (!p_180639_4_.field_71075_bZ.field_75098_d) {
                         itemstack.func_190918_g(1);
-
-                        if (itemstack.func_190926_b())
-                        {
+                        if (itemstack.func_190926_b()) {
                             p_180639_4_.func_184611_a(p_180639_5_, new ItemStack(Items.field_151131_as));
-                        }
-                        else if (!p_180639_4_.field_71071_by.func_70441_a(new ItemStack(Items.field_151131_as)))
-                        {
+                        } else if (!p_180639_4_.field_71071_by.func_70441_a(new ItemStack(Items.field_151131_as))) {
                             p_180639_4_.func_71019_a(new ItemStack(Items.field_151131_as), false);
                         }
                     }
 
                     p_180639_4_.func_71029_a(StatList.field_188078_L);
-                    this.func_176590_a(p_180639_1_, p_180639_2_, p_180639_3_, 0);
+                    // this.a(world, blockposition, iblockdata, 0);
+                    // CraftBukkit end
                     p_180639_1_.func_184133_a((EntityPlayer)null, p_180639_2_, SoundEvents.field_187630_M, SoundCategory.BLOCKS, 1.0F, 1.0F);
                 }
 
@@ -140,28 +151,26 @@
             {
                 if (i > 0 && !p_180639_1_.field_72995_K)
                 {
-                    if (!p_180639_4_.field_71075_bZ.field_75098_d)
-                    {
-                        ItemStack itemstack3 = PotionUtils.func_185188_a(new ItemStack(Items.field_151068_bn), PotionTypes.field_185230_b);
+                    // CraftBukkit start
+                    if (!this.changeLevel(p_180639_1_, p_180639_2_, p_180639_3_, i - 1, p_180639_4_, CauldronLevelChangeEvent.ChangeReason.BOTTLE_FILL)) {
+                        return true;
+                    }
+                    if (!p_180639_4_.field_71075_bZ.field_75098_d) {
+                        ItemStack itemstack1 = PotionUtils.func_185188_a(new ItemStack(Items.field_151068_bn), PotionTypes.field_185230_b);
                         p_180639_4_.func_71029_a(StatList.field_188078_L);
                         itemstack.func_190918_g(1);
-
-                        if (itemstack.func_190926_b())
-                        {
-                            p_180639_4_.func_184611_a(p_180639_5_, itemstack3);
+                        if (itemstack.func_190926_b()) {
+                            p_180639_4_.func_184611_a(p_180639_5_, itemstack1);
+                        } else if (!p_180639_4_.field_71071_by.func_70441_a(itemstack1)) {
+                            p_180639_4_.func_71019_a(itemstack1, false);
+                        } else if (p_180639_4_ instanceof EntityPlayerMP) {
+                            ((EntityPlayerMP) p_180639_4_).func_71120_a(p_180639_4_.field_71069_bz);
                         }
-                        else if (!p_180639_4_.field_71071_by.func_70441_a(itemstack3))
-                        {
-                            p_180639_4_.func_71019_a(itemstack3, false);
-                        }
-                        else if (p_180639_4_ instanceof EntityPlayerMP)
-                        {
-                            ((EntityPlayerMP)p_180639_4_).func_71120_a(p_180639_4_.field_71069_bz);
-                        }
                     }
 
-                    p_180639_1_.func_184133_a((EntityPlayer)null, p_180639_2_, SoundEvents.field_187615_H, SoundCategory.BLOCKS, 1.0F, 1.0F);
-                    this.func_176590_a(p_180639_1_, p_180639_2_, p_180639_3_, i - 1);
+                    p_180639_1_.func_184133_a((EntityPlayer) null, p_180639_2_, SoundEvents.field_187615_H, SoundCategory.BLOCKS, 1.0F, 1.0F);
+                    // this.a(world, blockposition, iblockdata, i - 1);
+                    // CraftBukkit end
                 }
 
                 return true;
@@ -170,20 +179,22 @@
             {
                 if (i < 3 && !p_180639_1_.field_72995_K)
                 {
-                    if (!p_180639_4_.field_71075_bZ.field_75098_d)
-                    {
-                        ItemStack itemstack2 = new ItemStack(Items.field_151069_bo);
+                    // CraftBukkit start
+                    if (!this.changeLevel(p_180639_1_, p_180639_2_, p_180639_3_, i + 1, p_180639_4_, CauldronLevelChangeEvent.ChangeReason.BOTTLE_EMPTY)) {
+                        return true;
+                    }
+                    if (!p_180639_4_.field_71075_bZ.field_75098_d) {
+                        ItemStack itemstack1 = new ItemStack(Items.field_151069_bo);
                         p_180639_4_.func_71029_a(StatList.field_188078_L);
-                        p_180639_4_.func_184611_a(p_180639_5_, itemstack2);
-
-                        if (p_180639_4_ instanceof EntityPlayerMP)
-                        {
-                            ((EntityPlayerMP)p_180639_4_).func_71120_a(p_180639_4_.field_71069_bz);
+                        p_180639_4_.func_184611_a(p_180639_5_, itemstack1);
+                        if (p_180639_4_ instanceof EntityPlayerMP) {
+                            ((EntityPlayerMP) p_180639_4_).func_71120_a(p_180639_4_.field_71069_bz);
                         }
                     }
 
-                    p_180639_1_.func_184133_a((EntityPlayer)null, p_180639_2_, SoundEvents.field_191241_J, SoundCategory.BLOCKS, 1.0F, 1.0F);
-                    this.func_176590_a(p_180639_1_, p_180639_2_, p_180639_3_, i + 1);
+                    p_180639_1_.func_184133_a((EntityPlayer) null, p_180639_2_, SoundEvents.field_191241_J, SoundCategory.BLOCKS, 1.0F, 1.0F);
+                    // this.a(world, blockposition, iblockdata, i + 1);
+                    // CraftBukkit end
                 }
 
                 return true;
@@ -196,8 +207,13 @@
 
                     if (itemarmor.func_82812_d() == ItemArmor.ArmorMaterial.LEATHER && itemarmor.func_82816_b_(itemstack) && !p_180639_1_.field_72995_K)
                     {
+                        // CraftBukkit start
+                        if (!this.changeLevel(p_180639_1_, p_180639_2_, p_180639_3_, i - 1, p_180639_4_, CauldronLevelChangeEvent.ChangeReason.ARMOR_WASH)) {
+                            return true;
+                        }
                         itemarmor.func_82815_c(itemstack);
-                        this.func_176590_a(p_180639_1_, p_180639_2_, p_180639_3_, i - 1);
+                        // this.a(world, blockposition, iblockdata, i - 1);
+                        // CraftBukkit end
                         p_180639_4_.func_71029_a(StatList.field_188079_M);
                         return true;
                     }
@@ -207,15 +223,18 @@
                 {
                     if (TileEntityBanner.func_175113_c(itemstack) > 0 && !p_180639_1_.field_72995_K)
                     {
+                        // CraftBukkit start
+                        if (!this.changeLevel(p_180639_1_, p_180639_2_, p_180639_3_, i - 1, p_180639_4_, CauldronLevelChangeEvent.ChangeReason.BANNER_WASH)) {
+                            return true;
+                        }
                         ItemStack itemstack1 = itemstack.func_77946_l();
                         itemstack1.func_190920_e(1);
                         TileEntityBanner.func_175117_e(itemstack1);
                         p_180639_4_.func_71029_a(StatList.field_188080_N);
-
-                        if (!p_180639_4_.field_71075_bZ.field_75098_d)
-                        {
+                        if (!p_180639_4_.field_71075_bZ.field_75098_d) {
                             itemstack.func_190918_g(1);
-                            this.func_176590_a(p_180639_1_, p_180639_2_, p_180639_3_, i - 1);
+                            // this.a(world, blockposition, iblockdata, i - 1);
+                            // CraftBukkit end
                         }
 
                         if (itemstack.func_190926_b())
@@ -242,12 +261,27 @@
         }
     }
 
-    public void func_176590_a(World p_176590_1_, BlockPos p_176590_2_, IBlockState p_176590_3_, int p_176590_4_)
-    {
-        p_176590_1_.func_180501_a(p_176590_2_, p_176590_3_.func_177226_a(field_176591_a, Integer.valueOf(MathHelper.func_76125_a(p_176590_4_, 0, 3))), 2);
-        p_176590_1_.func_175666_e(p_176590_2_, this);
+    // CraftBukkit start
+    public void func_176590_a(World p_176590_1_, BlockPos p_176590_2_, IBlockState p_176590_3_, int p_176590_4_) {
+        this.changeLevel(p_176590_1_, p_176590_2_, p_176590_3_, p_176590_4_, null, CauldronLevelChangeEvent.ChangeReason.UNKNOWN);
     }
 
+    private boolean changeLevel(World world, BlockPos blockposition, IBlockState iblockdata, int i, Entity entity, CauldronLevelChangeEvent.ChangeReason reason) {
+        int newLevel = Integer.valueOf(MathHelper.func_76125_a(i, 0, 3));
+        CauldronLevelChangeEvent event = new CauldronLevelChangeEvent(
+                world.getWorld().getBlockAt(blockposition.func_177958_n(), blockposition.func_177956_o(), blockposition.func_177952_p()),
+                (entity == null) ? null : entity.getBukkitEntity(), reason, iblockdata.func_177229_b(BlockCauldron.field_176591_a), newLevel
+        );
+        world.getServer().getPluginManager().callEvent(event);
+        if (event.isCancelled()) {
+            return false;
+        }
+        world.func_180501_a(blockposition, iblockdata.func_177226_a(BlockCauldron.field_176591_a, newLevel), 2);
+        world.func_175666_e(blockposition, this);
+        return true;
+        // CraftBukkit end
+    }
+
     public void func_176224_k(World p_176224_1_, BlockPos p_176224_2_)
     {
         if (p_176224_1_.field_73012_v.nextInt(20) == 1)
