--- ../src-base/minecraft/net/minecraft/block/BlockButton.java
+++ ../src-work/minecraft/net/minecraft/block/BlockButton.java
@@ -21,6 +21,8 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
+import org.bukkit.event.block.BlockRedstoneEvent;
+import org.bukkit.event.entity.EntityInteractEvent;
 
 public abstract class BlockButton extends BlockDirectional
 {
@@ -90,7 +92,7 @@
     protected static boolean func_181088_a(World p_181088_0_, BlockPos p_181088_1_, EnumFacing p_181088_2_)
     {
         BlockPos blockpos = p_181088_1_.func_177972_a(p_181088_2_);
-        return p_181088_2_ == EnumFacing.DOWN ? p_181088_0_.func_180495_p(blockpos).func_185896_q() : p_181088_0_.func_180495_p(blockpos).func_185915_l();
+        return p_181088_0_.func_180495_p(blockpos).isSideSolid(p_181088_0_, blockpos, p_181088_2_.func_176734_d());
     }
 
     public IBlockState func_180642_a(World p_180642_1_, BlockPos p_180642_2_, EnumFacing p_180642_3_, float p_180642_4_, float p_180642_5_, float p_180642_6_, int p_180642_7_, EntityLivingBase p_180642_8_)
@@ -152,6 +154,19 @@
         }
         else
         {
+            // CraftBukkit start
+            boolean powered = ((Boolean) p_180639_3_.func_177229_b(field_176584_b));
+            org.bukkit.block.Block block = p_180639_1_.getWorld().getBlockAt(p_180639_2_.func_177958_n(), p_180639_2_.func_177956_o(), p_180639_2_.func_177952_p());
+            int old = (powered) ? 15 : 0;
+            int current = (!powered) ? 15 : 0;
+
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, old, current);
+            p_180639_1_.getServer().getPluginManager().callEvent(eventRedstone);
+
+            if ((eventRedstone.getNewCurrent() > 0) != (!powered)) {
+                return true;
+            }
+            // CraftBukkit end
             p_180639_1_.func_180501_a(p_180639_2_, p_180639_3_.func_177226_a(field_176584_b, Boolean.valueOf(true)), 3);
             p_180639_1_.func_175704_b(p_180639_2_, p_180639_2_);
             this.func_185615_a(p_180639_4_, p_180639_1_, p_180639_2_);
@@ -202,10 +217,21 @@
             {
                 if (this.field_150047_a)
                 {
+
                     this.func_185616_e(p_180650_3_, p_180650_1_, p_180650_2_);
                 }
                 else
                 {
+                    // CraftBukkit start
+                    org.bukkit.block.Block block = p_180650_1_.getWorld().getBlockAt(p_180650_2_.func_177958_n(), p_180650_2_.func_177956_o(), p_180650_2_.func_177952_p());
+
+                    BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, 15, 0);
+                    p_180650_1_.getServer().getPluginManager().callEvent(eventRedstone);
+
+                    if (eventRedstone.getNewCurrent() > 0) {
+                        return;
+                    }
+                    // CraftBukkit end
                     p_180650_1_.func_175656_a(p_180650_2_, p_180650_3_.func_177226_a(field_176584_b, Boolean.valueOf(false)));
                     this.func_176582_b(p_180650_1_, p_180650_2_, (EnumFacing)p_180650_3_.func_177229_b(field_176387_N));
                     this.func_185617_b(p_180650_1_, p_180650_2_);
@@ -234,9 +260,42 @@
         List <? extends Entity > list = p_185616_2_.<Entity>func_72872_a(EntityArrow.class, p_185616_1_.func_185900_c(p_185616_2_, p_185616_3_).func_186670_a(p_185616_3_));
         boolean flag = !list.isEmpty();
         boolean flag1 = ((Boolean)p_185616_1_.func_177229_b(field_176584_b)).booleanValue();
+        // CraftBukkit start - Call interact event when arrows turn on wooden buttons
+        if (flag1 != flag && flag) {
+            org.bukkit.block.Block block = p_185616_2_.getWorld().getBlockAt(p_185616_3_.func_177958_n(), p_185616_3_.func_177956_o(), p_185616_3_.func_177952_p());
+            boolean allowed = false;
 
+            // If all of the events are cancelled block the button press, else allow
+            for (Object object : list) {
+                if (object != null) {
+                    EntityInteractEvent event = new EntityInteractEvent(((Entity) object).getBukkitEntity(), block);
+                    p_185616_2_.getServer().getPluginManager().callEvent(event);
+
+                    if (!event.isCancelled()) {
+                        allowed = true;
+                        break;
+                    }
+                }
+            }
+
+            if (!allowed) {
+                return;
+            }
+        }
+        // CraftBukkit end
+
         if (flag && !flag1)
         {
+            // CraftBukkit start
+            org.bukkit.block.Block block = p_185616_2_.getWorld().getBlockAt(p_185616_3_.func_177958_n(), p_185616_3_.func_177956_o(), p_185616_3_.func_177952_p());
+
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, 0, 15);
+            p_185616_2_.getServer().getPluginManager().callEvent(eventRedstone);
+
+            if (eventRedstone.getNewCurrent() <= 0) {
+                return;
+            }
+            // CraftBukkit end
             p_185616_2_.func_175656_a(p_185616_3_, p_185616_1_.func_177226_a(field_176584_b, Boolean.valueOf(true)));
             this.func_176582_b(p_185616_2_, p_185616_3_, (EnumFacing)p_185616_1_.func_177229_b(field_176387_N));
             p_185616_2_.func_175704_b(p_185616_3_, p_185616_3_);
@@ -245,6 +304,16 @@
 
         if (!flag && flag1)
         {
+            // CraftBukkit start
+            org.bukkit.block.Block block = p_185616_2_.getWorld().getBlockAt(p_185616_3_.func_177958_n(), p_185616_3_.func_177956_o(), p_185616_3_.func_177952_p());
+
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, 15, 0);
+            p_185616_2_.getServer().getPluginManager().callEvent(eventRedstone);
+
+            if (eventRedstone.getNewCurrent() > 0) {
+                return;
+            }
+            // CraftBukkit end
             p_185616_2_.func_175656_a(p_185616_3_, p_185616_1_.func_177226_a(field_176584_b, Boolean.valueOf(false)));
             this.func_176582_b(p_185616_2_, p_185616_3_, (EnumFacing)p_185616_1_.func_177229_b(field_176387_N));
             p_185616_2_.func_175704_b(p_185616_3_, p_185616_3_);
