--- ../src-base/minecraft/net/minecraft/world/storage/ISaveHandler.java
+++ ../src-work/minecraft/net/minecraft/world/storage/ISaveHandler.java
@@ -1,6 +1,7 @@
 package net.minecraft.world.storage;
 
 import java.io.File;
+import java.util.UUID;
 import javax.annotation.Nullable;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.world.MinecraftException;
@@ -30,4 +31,5 @@
     File func_75758_b(String p_75758_1_);
 
     TemplateManager func_186340_h();
+    public UUID getUUID();
 }
