--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityNote.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityNote.java
@@ -6,62 +6,71 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.block.NotePlayEvent;
 
 public class TileEntityNote extends TileEntity
 {
-    public byte field_145879_a;
-    public boolean field_145880_i;
+    public byte note;
+    public boolean previousRedstoneState;
 
-    public NBTTagCompound func_189515_b(NBTTagCompound p_189515_1_)
+    public NBTTagCompound writeToNBT(NBTTagCompound compound)
     {
-        super.func_189515_b(p_189515_1_);
-        p_189515_1_.func_74774_a("note", this.field_145879_a);
-        p_189515_1_.func_74757_a("powered", this.field_145880_i);
-        return p_189515_1_;
+        super.writeToNBT(compound);
+        compound.setByte("note", this.note);
+        compound.setBoolean("powered", this.previousRedstoneState);
+        return compound;
     }
 
-    public void func_145839_a(NBTTagCompound p_145839_1_)
+    public void readFromNBT(NBTTagCompound compound)
     {
-        super.func_145839_a(p_145839_1_);
-        this.field_145879_a = p_145839_1_.func_74771_c("note");
-        this.field_145879_a = (byte)MathHelper.func_76125_a(this.field_145879_a, 0, 24);
-        this.field_145880_i = p_145839_1_.func_74767_n("powered");
+        super.readFromNBT(compound);
+        this.note = compound.getByte("note");
+        this.note = (byte)MathHelper.clamp(this.note, 0, 24);
+        this.previousRedstoneState = compound.getBoolean("powered");
     }
 
-    public void func_145877_a()
+    public void changePitch()
     {
-        this.field_145879_a = (byte)((this.field_145879_a + 1) % 25);
-        this.func_70296_d();
+        byte old = note;
+        this.note = (byte)((this.note + 1) % 25);
+        if (!net.minecraftforge.common.ForgeHooks.onNoteChange(this, old)) return;
+        this.markDirty();
     }
 
-    public void func_175108_a(World p_175108_1_, BlockPos p_175108_2_)
+    public void triggerNote(World worldIn, BlockPos posIn)
     {
-        if (p_175108_1_.func_180495_p(p_175108_2_.func_177984_a()).func_185904_a() == Material.field_151579_a)
+        if (worldIn.getBlockState(posIn.up()).getMaterial() == Material.AIR)
         {
-            Material material = p_175108_1_.func_180495_p(p_175108_2_.func_177977_b()).func_185904_a();
+            Material material = worldIn.getBlockState(posIn.down()).getMaterial();
             int i = 0;
 
-            if (material == Material.field_151576_e)
+            if (material == Material.ROCK)
             {
                 i = 1;
             }
 
-            if (material == Material.field_151595_p)
+            if (material == Material.SAND)
             {
                 i = 2;
             }
 
-            if (material == Material.field_151592_s)
+            if (material == Material.GLASS)
             {
                 i = 3;
             }
 
-            if (material == Material.field_151575_d)
+            if (material == Material.WOOD)
             {
                 i = 4;
             }
-
-            p_175108_1_.func_175641_c(p_175108_2_, Blocks.field_150323_B, i, this.field_145879_a);
+            // CraftBukkit start
+            NotePlayEvent event = CraftEventFactory.callNotePlayEvent(this.world, posIn.getX(), posIn.getY(), posIn.getZ(), (byte)i, this.note);
+            if (!event.isCancelled()) {
+                world.addBlockEvent(posIn, Blocks.NOTEBLOCK, event.getInstrument().getType(), event.getNote().getId());
+            }
+            // CraftBukkit end
+            worldIn.addBlockEvent(posIn, Blocks.NOTEBLOCK, i, this.note);
         }
     }
 }
