--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityBrewingStand.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityBrewingStand.java
@@ -1,6 +1,8 @@
 package net.minecraft.tileentity;
 
 import java.util.Arrays;
+import java.util.List;
+
 import net.minecraft.block.BlockBrewingStand;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.player.EntityPlayer;
@@ -15,6 +17,7 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.potion.PotionHelper;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.EnumFacing;
 import net.minecraft.util.ITickable;
 import net.minecraft.util.NonNullList;
@@ -22,44 +25,76 @@
 import net.minecraft.util.datafix.FixTypes;
 import net.minecraft.util.datafix.walkers.ItemStackDataLists;
 import net.minecraft.util.math.BlockPos;
+import org.bukkit.craftbukkit.CraftChunk;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.event.inventory.BrewEvent;
+import org.bukkit.event.inventory.BrewingStandFuelEvent;
 
 public class TileEntityBrewingStand extends TileEntityLockable implements ITickable, ISidedInventory
 {
-    private static final int[] field_145941_a = new int[] {3};
-    private static final int[] field_184277_f = new int[] {0, 1, 2, 3};
-    private static final int[] field_145947_i = new int[] {0, 1, 2, 4};
-    private NonNullList<ItemStack> field_145945_j = NonNullList.<ItemStack>func_191197_a(5, ItemStack.field_190927_a);
-    private int field_145946_k;
-    private boolean[] field_145943_l;
-    private Item field_145944_m;
-    private String field_145942_n;
-    private int field_184278_m;
+    private static final int[] SLOTS_FOR_UP = new int[] {3};
+    private static final int[] SLOTS_FOR_DOWN = new int[] {0, 1, 2, 3};
+    private static final int[] OUTPUT_SLOTS = new int[] {0, 1, 2, 4};
+    private NonNullList<ItemStack> brewingItemStacks = NonNullList.<ItemStack>withSize(5, ItemStack.EMPTY);
+    private int brewTime;
+    private boolean[] filledSlots;
+    private Item ingredientID;
+    private String customName;
+    private int fuel;
+    // CraftBukkit start - add fields and methods
+    private int lastTick = MinecraftServer.currentTick;
+    public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+    private int maxStack = 64;
 
-    public String func_70005_c_()
+    public void onOpen(CraftHumanEntity who) {
+        transaction.add(who);
+    }
+
+    public void onClose(CraftHumanEntity who) {
+        transaction.remove(who);
+    }
+
+    public List<HumanEntity> getViewers() {
+        return transaction;
+    }
+
+    public List<ItemStack> getContents() {
+        return this.brewingItemStacks;
+    }
+
+    public void setMaxStackSize(int size) {
+        maxStack = size;
+    }
+    // CraftBukkit end
+    public String getName()
     {
-        return this.func_145818_k_() ? this.field_145942_n : "container.brewing";
+        return this.hasCustomName() ? this.customName : "container.brewing";
     }
 
-    public boolean func_145818_k_()
+    public boolean hasCustomName()
     {
-        return this.field_145942_n != null && !this.field_145942_n.isEmpty();
+        return this.customName != null && !this.customName.isEmpty();
     }
 
-    public void func_145937_a(String p_145937_1_)
+    public void setName(String name)
     {
-        this.field_145942_n = p_145937_1_;
+        this.customName = name;
     }
 
-    public int func_70302_i_()
+    public int getSizeInventory()
     {
-        return this.field_145945_j.size();
+        return this.brewingItemStacks.size();
     }
 
-    public boolean func_191420_l()
+    public boolean isEmpty()
     {
-        for (ItemStack itemstack : this.field_145945_j)
+        for (ItemStack itemstack : this.brewingItemStacks)
         {
-            if (!itemstack.func_190926_b())
+            if (!itemstack.isEmpty())
             {
                 return false;
             }
@@ -68,81 +103,92 @@
         return true;
     }
 
-    public void func_73660_a()
+    public void update()
     {
-        ItemStack itemstack = (ItemStack)this.field_145945_j.get(4);
+        ItemStack itemstack = (ItemStack)this.brewingItemStacks.get(4);
 
-        if (this.field_184278_m <= 0 && itemstack.func_77973_b() == Items.field_151065_br)
+        if (this.fuel <= 0 && itemstack.getItem() == Items.BLAZE_POWDER)
         {
-            this.field_184278_m = 20;
-            itemstack.func_190918_g(1);
-            this.func_70296_d();
+            // CraftBukkit start
+            BrewingStandFuelEvent event = new BrewingStandFuelEvent(world.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), CraftItemStack.asCraftMirror(itemstack), 20);
+            this.world.getServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+                return;
+            }
+
+            this.fuel = event.getFuelPower(); // PAIL fuelLevel
+            if (this.fuel > 0 && event.isConsuming()) {
+                itemstack.shrink(1);
+            }
+            // CraftBukkit end
+            this.markDirty();
         }
 
-        boolean flag = this.func_145934_k();
-        boolean flag1 = this.field_145946_k > 0;
-        ItemStack itemstack1 = (ItemStack)this.field_145945_j.get(3);
+        boolean flag = this.canBrew();
+        boolean flag1 = this.brewTime > 0;
+        ItemStack itemstack1 = (ItemStack)this.brewingItemStacks.get(3);
 
         if (flag1)
         {
-            --this.field_145946_k;
-            boolean flag2 = this.field_145946_k == 0;
+            --this.brewTime;
+            boolean flag2 = this.brewTime == 0;
 
             if (flag2 && flag)
             {
-                this.func_145940_l();
-                this.func_70296_d();
+                this.brewPotions();
+                this.markDirty();
             }
             else if (!flag)
             {
-                this.field_145946_k = 0;
-                this.func_70296_d();
+                this.brewTime = 0;
+                this.markDirty();
             }
-            else if (this.field_145944_m != itemstack1.func_77973_b())
+            else if (this.ingredientID != itemstack1.getItem())
             {
-                this.field_145946_k = 0;
-                this.func_70296_d();
+                this.brewTime = 0;
+                this.markDirty();
             }
         }
-        else if (flag && this.field_184278_m > 0)
+        else if (flag && this.fuel > 0)
         {
-            --this.field_184278_m;
-            this.field_145946_k = 400;
-            this.field_145944_m = itemstack1.func_77973_b();
-            this.func_70296_d();
+            --this.fuel;
+            this.brewTime = 400;
+            this.ingredientID = itemstack1.getItem();
+            this.markDirty();
         }
 
-        if (!this.field_145850_b.field_72995_K)
+        if (!this.world.isRemote)
         {
-            boolean[] aboolean = this.func_174902_m();
+            boolean[] aboolean = this.createFilledSlotsArray();
 
-            if (!Arrays.equals(aboolean, this.field_145943_l))
+            if (!Arrays.equals(aboolean, this.filledSlots))
             {
-                this.field_145943_l = aboolean;
-                IBlockState iblockstate = this.field_145850_b.func_180495_p(this.func_174877_v());
+                this.filledSlots = aboolean;
+                IBlockState iblockstate = this.world.getBlockState(this.getPos());
 
-                if (!(iblockstate.func_177230_c() instanceof BlockBrewingStand))
+                if (!(iblockstate.getBlock() instanceof BlockBrewingStand))
                 {
                     return;
                 }
 
-                for (int i = 0; i < BlockBrewingStand.field_176451_a.length; ++i)
+                for (int i = 0; i < BlockBrewingStand.HAS_BOTTLE.length; ++i)
                 {
-                    iblockstate = iblockstate.func_177226_a(BlockBrewingStand.field_176451_a[i], Boolean.valueOf(aboolean[i]));
+                    iblockstate = iblockstate.withProperty(BlockBrewingStand.HAS_BOTTLE[i], Boolean.valueOf(aboolean[i]));
                 }
 
-                this.field_145850_b.func_180501_a(this.field_174879_c, iblockstate, 2);
+                this.world.setBlockState(this.pos, iblockstate, 2);
             }
         }
     }
 
-    public boolean[] func_174902_m()
+    public boolean[] createFilledSlotsArray()
     {
         boolean[] aboolean = new boolean[3];
 
         for (int i = 0; i < 3; ++i)
         {
-            if (!((ItemStack)this.field_145945_j.get(i)).func_190926_b())
+            if (!((ItemStack)this.brewingItemStacks.get(i)).isEmpty())
             {
                 aboolean[i] = true;
             }
@@ -151,15 +197,16 @@
         return aboolean;
     }
 
-    private boolean func_145934_k()
+    private boolean canBrew()
     {
-        ItemStack itemstack = (ItemStack)this.field_145945_j.get(3);
+        if (1 == 1) return net.minecraftforge.common.brewing.BrewingRecipeRegistry.canBrew(brewingItemStacks, brewingItemStacks.get(3), OUTPUT_SLOTS); // divert to VanillaBrewingRegistry
+        ItemStack itemstack = (ItemStack)this.brewingItemStacks.get(3);
 
-        if (itemstack.func_190926_b())
+        if (itemstack.isEmpty())
         {
             return false;
         }
-        else if (!PotionHelper.func_185205_a(itemstack))
+        else if (!PotionHelper.isReagent(itemstack))
         {
             return false;
         }
@@ -167,9 +214,9 @@
         {
             for (int i = 0; i < 3; ++i)
             {
-                ItemStack itemstack1 = (ItemStack)this.field_145945_j.get(i);
+                ItemStack itemstack1 = (ItemStack)this.brewingItemStacks.get(i);
 
-                if (!itemstack1.func_190926_b() && PotionHelper.func_185208_a(itemstack1, itemstack))
+                if (!itemstack1.isEmpty() && PotionHelper.hasConversions(itemstack1, itemstack))
                 {
                     return true;
                 }
@@ -179,182 +226,209 @@
         }
     }
 
-    private void func_145940_l()
+    private void brewPotions()
     {
-        ItemStack itemstack = (ItemStack)this.field_145945_j.get(3);
-
-        for (int i = 0; i < 3; ++i)
-        {
-            this.field_145945_j.set(i, PotionHelper.func_185212_d(itemstack, (ItemStack)this.field_145945_j.get(i)));
+        if (net.minecraftforge.event.ForgeEventFactory.onPotionAttemptBrew(brewingItemStacks)) return;
+        // CraftBukkit start
+        if (getOwner() != null) {
+            BrewEvent event = new BrewEvent(world.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), (org.bukkit.inventory.BrewerInventory) this.getOwner().getInventory(), this.fuel);
+            org.bukkit.Bukkit.getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                return;
+            }
         }
+        // CraftBukkit end
+        ItemStack itemstack = (ItemStack)this.brewingItemStacks.get(3);
 
-        itemstack.func_190918_g(1);
-        BlockPos blockpos = this.func_174877_v();
+        net.minecraftforge.common.brewing.BrewingRecipeRegistry.brewPotions(brewingItemStacks, brewingItemStacks.get(3), OUTPUT_SLOTS);
 
-        if (itemstack.func_77973_b().func_77634_r())
+        itemstack.shrink(1);
+        BlockPos blockpos = this.getPos();
+
+        if (itemstack.getItem().hasContainerItem(itemstack))
         {
-            ItemStack itemstack1 = new ItemStack(itemstack.func_77973_b().func_77668_q());
+            ItemStack itemstack1 = itemstack.getItem().getContainerItem(itemstack);
 
-            if (itemstack.func_190926_b())
+            if (itemstack.isEmpty())
             {
                 itemstack = itemstack1;
             }
             else
             {
-                InventoryHelper.func_180173_a(this.field_145850_b, (double)blockpos.func_177958_n(), (double)blockpos.func_177956_o(), (double)blockpos.func_177952_p(), itemstack1);
+                InventoryHelper.spawnItemStack(this.world, (double)blockpos.getX(), (double)blockpos.getY(), (double)blockpos.getZ(), itemstack1);
             }
         }
-
-        this.field_145945_j.set(3, itemstack);
-        this.field_145850_b.func_175718_b(1035, blockpos, 0);
+        this.brewingItemStacks.set(3, itemstack);
+        this.world.playEvent(1035, blockpos, 0);
+        net.minecraftforge.event.ForgeEventFactory.onPotionBrewed(brewingItemStacks);
     }
 
-    public static void func_189675_a(DataFixer p_189675_0_)
+    public static void registerFixesBrewingStand(DataFixer fixer)
     {
-        p_189675_0_.func_188258_a(FixTypes.BLOCK_ENTITY, new ItemStackDataLists(TileEntityBrewingStand.class, new String[] {"Items"}));
+        fixer.registerWalker(FixTypes.BLOCK_ENTITY, new ItemStackDataLists(TileEntityBrewingStand.class, new String[] {"Items"}));
     }
 
-    public void func_145839_a(NBTTagCompound p_145839_1_)
+    public void readFromNBT(NBTTagCompound compound)
     {
-        super.func_145839_a(p_145839_1_);
-        this.field_145945_j = NonNullList.<ItemStack>func_191197_a(this.func_70302_i_(), ItemStack.field_190927_a);
-        ItemStackHelper.func_191283_b(p_145839_1_, this.field_145945_j);
-        this.field_145946_k = p_145839_1_.func_74765_d("BrewTime");
+        super.readFromNBT(compound);
+        this.brewingItemStacks = NonNullList.<ItemStack>withSize(this.getSizeInventory(), ItemStack.EMPTY);
+        ItemStackHelper.loadAllItems(compound, this.brewingItemStacks);
+        this.brewTime = compound.getShort("BrewTime");
 
-        if (p_145839_1_.func_150297_b("CustomName", 8))
+        if (compound.hasKey("CustomName", 8))
         {
-            this.field_145942_n = p_145839_1_.func_74779_i("CustomName");
+            this.customName = compound.getString("CustomName");
         }
 
-        this.field_184278_m = p_145839_1_.func_74771_c("Fuel");
+        this.fuel = compound.getByte("Fuel");
     }
 
-    public NBTTagCompound func_189515_b(NBTTagCompound p_189515_1_)
+    public NBTTagCompound writeToNBT(NBTTagCompound compound)
     {
-        super.func_189515_b(p_189515_1_);
-        p_189515_1_.func_74777_a("BrewTime", (short)this.field_145946_k);
-        ItemStackHelper.func_191282_a(p_189515_1_, this.field_145945_j);
+        super.writeToNBT(compound);
+        compound.setShort("BrewTime", (short)this.brewTime);
+        ItemStackHelper.saveAllItems(compound, this.brewingItemStacks);
 
-        if (this.func_145818_k_())
+        if (this.hasCustomName())
         {
-            p_189515_1_.func_74778_a("CustomName", this.field_145942_n);
+            compound.setString("CustomName", this.customName);
         }
 
-        p_189515_1_.func_74774_a("Fuel", (byte)this.field_184278_m);
-        return p_189515_1_;
+        compound.setByte("Fuel", (byte)this.fuel);
+        return compound;
     }
 
-    public ItemStack func_70301_a(int p_70301_1_)
+    public ItemStack getStackInSlot(int index)
     {
-        return p_70301_1_ >= 0 && p_70301_1_ < this.field_145945_j.size() ? (ItemStack)this.field_145945_j.get(p_70301_1_) : ItemStack.field_190927_a;
+        return index >= 0 && index < this.brewingItemStacks.size() ? (ItemStack)this.brewingItemStacks.get(index) : ItemStack.EMPTY;
     }
 
-    public ItemStack func_70298_a(int p_70298_1_, int p_70298_2_)
+    public ItemStack decrStackSize(int index, int count)
     {
-        return ItemStackHelper.func_188382_a(this.field_145945_j, p_70298_1_, p_70298_2_);
+        return ItemStackHelper.getAndSplit(this.brewingItemStacks, index, count);
     }
 
-    public ItemStack func_70304_b(int p_70304_1_)
+    public ItemStack removeStackFromSlot(int index)
     {
-        return ItemStackHelper.func_188383_a(this.field_145945_j, p_70304_1_);
+        return ItemStackHelper.getAndRemove(this.brewingItemStacks, index);
     }
 
-    public void func_70299_a(int p_70299_1_, ItemStack p_70299_2_)
+    public void setInventorySlotContents(int index, ItemStack stack)
     {
-        if (p_70299_1_ >= 0 && p_70299_1_ < this.field_145945_j.size())
+        if (index >= 0 && index < this.brewingItemStacks.size())
         {
-            this.field_145945_j.set(p_70299_1_, p_70299_2_);
+            this.brewingItemStacks.set(index, stack);
         }
     }
 
-    public int func_70297_j_()
+    public int getInventoryStackLimit()
     {
         return 64;
     }
 
-    public boolean func_70300_a(EntityPlayer p_70300_1_)
+    public boolean isUsableByPlayer(EntityPlayer player)
     {
-        return this.field_145850_b.func_175625_s(this.field_174879_c) != this ? false : p_70300_1_.func_70092_e((double)this.field_174879_c.func_177958_n() + 0.5D, (double)this.field_174879_c.func_177956_o() + 0.5D, (double)this.field_174879_c.func_177952_p() + 0.5D) <= 64.0D;
+        return this.world.getTileEntity(this.pos) != this ? false : player.getDistanceSq((double)this.pos.getX() + 0.5D, (double)this.pos.getY() + 0.5D, (double)this.pos.getZ() + 0.5D) <= 64.0D;
     }
 
-    public void func_174889_b(EntityPlayer p_174889_1_)
+    public void openInventory(EntityPlayer player)
     {
     }
 
-    public void func_174886_c(EntityPlayer p_174886_1_)
+    public void closeInventory(EntityPlayer player)
     {
     }
 
-    public boolean func_94041_b(int p_94041_1_, ItemStack p_94041_2_)
+    public boolean isItemValidForSlot(int index, ItemStack stack)
     {
-        if (p_94041_1_ == 3)
+        if (index == 3)
         {
-            return PotionHelper.func_185205_a(p_94041_2_);
+            return net.minecraftforge.common.brewing.BrewingRecipeRegistry.isValidIngredient(stack);
         }
         else
         {
-            Item item = p_94041_2_.func_77973_b();
-            return p_94041_1_ == 4 ? item == Items.field_151065_br : (item == Items.field_151068_bn || item == Items.field_185155_bH || item == Items.field_185156_bI || item == Items.field_151069_bo) && this.func_70301_a(p_94041_1_) == ItemStack.field_190927_a;
+            Item item = stack.getItem();
+            return index == 4 ? item == Items.BLAZE_POWDER : net.minecraftforge.common.brewing.BrewingRecipeRegistry.isValidInput(stack) && this.getStackInSlot(index).isEmpty(); // FORGE - Fix MC-111753 Identity comparing with ItemStack.EMPTY
         }
     }
 
-    public int[] func_180463_a(EnumFacing p_180463_1_)
+    public int[] getSlotsForFace(EnumFacing side)
     {
-        return p_180463_1_ == EnumFacing.UP ? field_145941_a : (p_180463_1_ == EnumFacing.DOWN ? field_184277_f : field_145947_i);
+        return side == EnumFacing.UP ? SLOTS_FOR_UP : (side == EnumFacing.DOWN ? SLOTS_FOR_DOWN : OUTPUT_SLOTS);
     }
 
-    public boolean func_180462_a(int p_180462_1_, ItemStack p_180462_2_, EnumFacing p_180462_3_)
+    public boolean canInsertItem(int index, ItemStack itemStackIn, EnumFacing direction)
     {
-        return this.func_94041_b(p_180462_1_, p_180462_2_);
+        return this.isItemValidForSlot(index, itemStackIn);
     }
 
-    public boolean func_180461_b(int p_180461_1_, ItemStack p_180461_2_, EnumFacing p_180461_3_)
+    public boolean canExtractItem(int index, ItemStack stack, EnumFacing direction)
     {
-        return p_180461_1_ == 3 ? p_180461_2_.func_77973_b() == Items.field_151069_bo : true;
+        return index == 3 ? stack.getItem() == Items.GLASS_BOTTLE : true;
     }
 
-    public String func_174875_k()
+    public String getGuiID()
     {
         return "minecraft:brewing_stand";
     }
 
-    public Container func_174876_a(InventoryPlayer p_174876_1_, EntityPlayer p_174876_2_)
+    public Container createContainer(InventoryPlayer playerInventory, EntityPlayer playerIn)
     {
-        return new ContainerBrewingStand(p_174876_1_, this);
+        return new ContainerBrewingStand(playerInventory, this);
     }
 
-    public int func_174887_a_(int p_174887_1_)
+    public int getField(int id)
     {
-        switch (p_174887_1_)
+        switch (id)
         {
             case 0:
-                return this.field_145946_k;
+                return this.brewTime;
             case 1:
-                return this.field_184278_m;
+                return this.fuel;
             default:
                 return 0;
         }
     }
 
-    public void func_174885_b(int p_174885_1_, int p_174885_2_)
+    public void setField(int id, int value)
     {
-        switch (p_174885_1_)
+        switch (id)
         {
             case 0:
-                this.field_145946_k = p_174885_2_;
+                this.brewTime = value;
                 break;
             case 1:
-                this.field_184278_m = p_174885_2_;
+                this.fuel = value;
         }
     }
 
-    public int func_174890_g()
+    net.minecraftforge.items.IItemHandler handlerInput = new net.minecraftforge.items.wrapper.SidedInvWrapper(this, net.minecraft.util.EnumFacing.UP);
+    net.minecraftforge.items.IItemHandler handlerOutput = new net.minecraftforge.items.wrapper.SidedInvWrapper(this, net.minecraft.util.EnumFacing.DOWN);
+    net.minecraftforge.items.IItemHandler handlerSides = new net.minecraftforge.items.wrapper.SidedInvWrapper(this, net.minecraft.util.EnumFacing.NORTH);
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T> T getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @javax.annotation.Nullable net.minecraft.util.EnumFacing facing)
     {
+        if (facing != null && capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY)
+        {
+            if (facing == EnumFacing.UP)
+                return (T) handlerInput;
+            else if (facing == EnumFacing.DOWN)
+                return (T) handlerOutput;
+            else
+                return (T) handlerSides;
+        }
+        return super.getCapability(capability, facing);
+    }
+
+    public int getFieldCount()
+    {
         return 2;
     }
 
-    public void func_174888_l()
+    public void clear()
     {
-        this.field_145945_j.clear();
+        this.brewingItemStacks.clear();
     }
 }
