--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityLockableLoot.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityLockableLoot.java
@@ -15,16 +15,16 @@
 
 public abstract class TileEntityLockableLoot extends TileEntityLockable implements ILootContainer
 {
-    protected ResourceLocation field_184284_m;
-    protected long field_184285_n;
-    protected String field_190577_o;
+    protected ResourceLocation lootTable;
+    protected long lootTableSeed;
+    protected String customName;
 
-    protected boolean func_184283_b(NBTTagCompound p_184283_1_)
+    protected boolean checkLootAndRead(NBTTagCompound compound)
     {
-        if (p_184283_1_.func_150297_b("LootTable", 8))
+        if (compound.hasKey("LootTable", 8))
         {
-            this.field_184284_m = new ResourceLocation(p_184283_1_.func_74779_i("LootTable"));
-            this.field_184285_n = p_184283_1_.func_74763_f("LootTableSeed");
+            this.lootTable = new ResourceLocation(compound.getString("LootTable"));
+            this.lootTableSeed = compound.getLong("LootTableSeed");
             return true;
         }
         else
@@ -33,15 +33,15 @@
         }
     }
 
-    protected boolean func_184282_c(NBTTagCompound p_184282_1_)
+    protected boolean checkLootAndWrite(NBTTagCompound compound)
     {
-        if (this.field_184284_m != null)
+        if (this.lootTable != null)
         {
-            p_184282_1_.func_74778_a("LootTable", this.field_184284_m.toString());
+            compound.setString("LootTable", this.lootTable.toString());
 
-            if (this.field_184285_n != 0L)
+            if (this.lootTableSeed != 0L)
             {
-                p_184282_1_.func_74772_a("LootTableSeed", this.field_184285_n);
+                compound.setLong("LootTableSeed", this.lootTableSeed);
             }
 
             return true;
@@ -52,130 +52,130 @@
         }
     }
 
-    public void func_184281_d(@Nullable EntityPlayer p_184281_1_)
+    public void fillWithLoot(@Nullable EntityPlayer player)
     {
-        if (this.field_184284_m != null)
+        if (this.lootTable != null)
         {
-            LootTable loottable = this.field_145850_b.func_184146_ak().func_186521_a(this.field_184284_m);
-            this.field_184284_m = null;
+            LootTable loottable = this.world.getLootTableManager().getLootTableFromLocation(this.lootTable);
+            this.lootTable = null;
             Random random;
 
-            if (this.field_184285_n == 0L)
+            if (this.lootTableSeed == 0L)
             {
                 random = new Random();
             }
             else
             {
-                random = new Random(this.field_184285_n);
+                random = new Random(this.lootTableSeed);
             }
 
-            LootContext.Builder lootcontext$builder = new LootContext.Builder((WorldServer)this.field_145850_b);
+            LootContext.Builder lootcontext$builder = new LootContext.Builder((WorldServer)this.world);
 
-            if (p_184281_1_ != null)
+            if (player != null)
             {
-                lootcontext$builder.func_186469_a(p_184281_1_.func_184817_da());
+                lootcontext$builder.withLuck(player.getLuck());
             }
 
-            loottable.func_186460_a(this, random, lootcontext$builder.func_186471_a());
+            loottable.fillInventory(this, random, lootcontext$builder.build());
         }
     }
 
-    public ResourceLocation func_184276_b()
+    public ResourceLocation getLootTable()
     {
-        return this.field_184284_m;
+        return this.lootTable;
     }
 
-    public void func_189404_a(ResourceLocation p_189404_1_, long p_189404_2_)
+    public void setLootTable(ResourceLocation p_189404_1_, long p_189404_2_)
     {
-        this.field_184284_m = p_189404_1_;
-        this.field_184285_n = p_189404_2_;
+        this.lootTable = p_189404_1_;
+        this.lootTableSeed = p_189404_2_;
     }
 
-    public boolean func_145818_k_()
+    public boolean hasCustomName()
     {
-        return this.field_190577_o != null && !this.field_190577_o.isEmpty();
+        return this.customName != null && !this.customName.isEmpty();
     }
 
-    public void func_190575_a(String p_190575_1_)
+    public void setCustomName(String p_190575_1_)
     {
-        this.field_190577_o = p_190575_1_;
+        this.customName = p_190575_1_;
     }
 
-    public ItemStack func_70301_a(int p_70301_1_)
+    public ItemStack getStackInSlot(int index)
     {
-        this.func_184281_d((EntityPlayer)null);
-        return (ItemStack)this.func_190576_q().get(p_70301_1_);
+        this.fillWithLoot((EntityPlayer)null);
+        return (ItemStack)this.getItems().get(index);
     }
 
-    public ItemStack func_70298_a(int p_70298_1_, int p_70298_2_)
+    public ItemStack decrStackSize(int index, int count)
     {
-        this.func_184281_d((EntityPlayer)null);
-        ItemStack itemstack = ItemStackHelper.func_188382_a(this.func_190576_q(), p_70298_1_, p_70298_2_);
+        this.fillWithLoot((EntityPlayer)null);
+        ItemStack itemstack = ItemStackHelper.getAndSplit(this.getItems(), index, count);
 
-        if (!itemstack.func_190926_b())
+        if (!itemstack.isEmpty())
         {
-            this.func_70296_d();
+            this.markDirty();
         }
 
         return itemstack;
     }
 
-    public ItemStack func_70304_b(int p_70304_1_)
+    public ItemStack removeStackFromSlot(int index)
     {
-        this.func_184281_d((EntityPlayer)null);
-        return ItemStackHelper.func_188383_a(this.func_190576_q(), p_70304_1_);
+        this.fillWithLoot((EntityPlayer)null);
+        return ItemStackHelper.getAndRemove(this.getItems(), index);
     }
 
-    public void func_70299_a(int p_70299_1_, @Nullable ItemStack p_70299_2_)
+    public void setInventorySlotContents(int index, @Nullable ItemStack stack)
     {
-        this.func_184281_d((EntityPlayer)null);
-        this.func_190576_q().set(p_70299_1_, p_70299_2_);
+        this.fillWithLoot((EntityPlayer)null);
+        this.getItems().set(index, stack);
 
-        if (p_70299_2_.func_190916_E() > this.func_70297_j_())
+        if (stack.getCount() > this.getInventoryStackLimit())
         {
-            p_70299_2_.func_190920_e(this.func_70297_j_());
+            stack.setCount(this.getInventoryStackLimit());
         }
 
-        this.func_70296_d();
+        this.markDirty();
     }
 
-    public boolean func_70300_a(EntityPlayer p_70300_1_)
+    public boolean isUsableByPlayer(EntityPlayer player)
     {
-        return this.field_145850_b.func_175625_s(this.field_174879_c) != this ? false : p_70300_1_.func_70092_e((double)this.field_174879_c.func_177958_n() + 0.5D, (double)this.field_174879_c.func_177956_o() + 0.5D, (double)this.field_174879_c.func_177952_p() + 0.5D) <= 64.0D;
+        return this.world.getTileEntity(this.pos) != this ? false : player.getDistanceSq((double)this.pos.getX() + 0.5D, (double)this.pos.getY() + 0.5D, (double)this.pos.getZ() + 0.5D) <= 64.0D;
     }
 
-    public void func_174889_b(EntityPlayer p_174889_1_)
+    public void openInventory(EntityPlayer player)
     {
     }
 
-    public void func_174886_c(EntityPlayer p_174886_1_)
+    public void closeInventory(EntityPlayer player)
     {
     }
 
-    public boolean func_94041_b(int p_94041_1_, ItemStack p_94041_2_)
+    public boolean isItemValidForSlot(int index, ItemStack stack)
     {
         return true;
     }
 
-    public int func_174887_a_(int p_174887_1_)
+    public int getField(int id)
     {
         return 0;
     }
 
-    public void func_174885_b(int p_174885_1_, int p_174885_2_)
+    public void setField(int id, int value)
     {
     }
 
-    public int func_174890_g()
+    public int getFieldCount()
     {
         return 0;
     }
 
-    public void func_174888_l()
+    public void clear()
     {
-        this.func_184281_d((EntityPlayer)null);
-        this.func_190576_q().clear();
+        this.fillWithLoot((EntityPlayer)null);
+        this.getItems().clear();
     }
 
-    protected abstract NonNullList<ItemStack> func_190576_q();
+    protected abstract NonNullList<ItemStack> getItems();
 }
