--- ../src-base/minecraft/net/minecraft/tileentity/TileEntityShulkerBox.java
+++ ../src-work/minecraft/net/minecraft/tileentity/TileEntityShulkerBox.java
@@ -29,106 +29,131 @@
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
 
 public class TileEntityShulkerBox extends TileEntityLockableLoot implements ITickable, ISidedInventory
 {
-    private static final int[] field_190595_a = new int[27];
-    private NonNullList<ItemStack> field_190596_f;
-    private boolean field_190597_g;
-    private int field_190598_h;
-    private TileEntityShulkerBox.AnimationStatus field_190599_i;
-    private float field_190600_j;
-    private float field_190601_k;
-    private EnumDyeColor field_190602_l;
-    private boolean field_190594_p;
+    private static final int[] SLOTS = new int[27];
+    private NonNullList<ItemStack> items;
+    private boolean hasBeenCleared;
+    private int openCount;
+    private TileEntityShulkerBox.AnimationStatus animationStatus;
+    private float progress;
+    private float progressOld;
+    private EnumDyeColor color;
+    private boolean destroyedByCreativePlayer;
+    // CraftBukkit start - add fields and methods
+    public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+    private int maxStack = 64;
 
+    public List<ItemStack> getContents() {
+        return this.items;
+    }
+
+    public void onOpen(CraftHumanEntity who) {
+        transaction.add(who);
+    }
+
+    public void onClose(CraftHumanEntity who) {
+        transaction.remove(who);
+    }
+
+    public List<HumanEntity> getViewers() {
+        return transaction;
+    }
+
+    public void setMaxStackSize(int size) {
+        maxStack = size;
+    }
+    // CraftBukkit end
     public TileEntityShulkerBox()
     {
         this((EnumDyeColor)null);
     }
 
-    public TileEntityShulkerBox(@Nullable EnumDyeColor p_i47242_1_)
+    public TileEntityShulkerBox(@Nullable EnumDyeColor colorIn)
     {
-        this.field_190596_f = NonNullList.<ItemStack>func_191197_a(27, ItemStack.field_190927_a);
-        this.field_190599_i = TileEntityShulkerBox.AnimationStatus.CLOSED;
-        this.field_190602_l = p_i47242_1_;
+        this.items = NonNullList.<ItemStack>withSize(27, ItemStack.EMPTY);
+        this.animationStatus = TileEntityShulkerBox.AnimationStatus.CLOSED;
+        this.color = colorIn;
     }
 
-    public void func_73660_a()
+    public void update()
     {
-        this.func_190583_o();
+        this.updateAnimation();
 
-        if (this.field_190599_i == TileEntityShulkerBox.AnimationStatus.OPENING || this.field_190599_i == TileEntityShulkerBox.AnimationStatus.CLOSING)
+        if (this.animationStatus == TileEntityShulkerBox.AnimationStatus.OPENING || this.animationStatus == TileEntityShulkerBox.AnimationStatus.CLOSING)
         {
-            this.func_190589_G();
+            this.moveCollidedEntities();
         }
     }
 
-    protected void func_190583_o()
+    protected void updateAnimation()
     {
-        this.field_190601_k = this.field_190600_j;
+        this.progressOld = this.progress;
 
-        switch (this.field_190599_i)
+        switch (this.animationStatus)
         {
             case CLOSED:
-                this.field_190600_j = 0.0F;
+                this.progress = 0.0F;
                 break;
             case OPENING:
-                this.field_190600_j += 0.1F;
+                this.progress += 0.1F;
 
-                if (this.field_190600_j >= 1.0F)
+                if (this.progress >= 1.0F)
                 {
-                    this.func_190589_G();
-                    this.field_190599_i = TileEntityShulkerBox.AnimationStatus.OPENED;
-                    this.field_190600_j = 1.0F;
+                    this.moveCollidedEntities();
+                    this.animationStatus = TileEntityShulkerBox.AnimationStatus.OPENED;
+                    this.progress = 1.0F;
                 }
 
                 break;
             case CLOSING:
-                this.field_190600_j -= 0.1F;
+                this.progress -= 0.1F;
 
-                if (this.field_190600_j <= 0.0F)
+                if (this.progress <= 0.0F)
                 {
-                    this.field_190599_i = TileEntityShulkerBox.AnimationStatus.CLOSED;
-                    this.field_190600_j = 0.0F;
+                    this.animationStatus = TileEntityShulkerBox.AnimationStatus.CLOSED;
+                    this.progress = 0.0F;
                 }
 
                 break;
             case OPENED:
-                this.field_190600_j = 1.0F;
+                this.progress = 1.0F;
         }
     }
 
-    public TileEntityShulkerBox.AnimationStatus func_190591_p()
+    public TileEntityShulkerBox.AnimationStatus getAnimationStatus()
     {
-        return this.field_190599_i;
+        return this.animationStatus;
     }
 
-    public AxisAlignedBB func_190584_a(IBlockState p_190584_1_)
+    public AxisAlignedBB getBoundingBox(IBlockState p_190584_1_)
     {
-        return this.func_190587_b((EnumFacing)p_190584_1_.func_177229_b(BlockShulkerBox.field_190957_a));
+        return this.getBoundingBox((EnumFacing)p_190584_1_.getValue(BlockShulkerBox.FACING));
     }
 
-    public AxisAlignedBB func_190587_b(EnumFacing p_190587_1_)
+    public AxisAlignedBB getBoundingBox(EnumFacing p_190587_1_)
     {
-        return Block.field_185505_j.func_72321_a((double)(0.5F * this.func_190585_a(1.0F) * (float)p_190587_1_.func_82601_c()), (double)(0.5F * this.func_190585_a(1.0F) * (float)p_190587_1_.func_96559_d()), (double)(0.5F * this.func_190585_a(1.0F) * (float)p_190587_1_.func_82599_e()));
+        return Block.FULL_BLOCK_AABB.addCoord((double)(0.5F * this.getProgress(1.0F) * (float)p_190587_1_.getFrontOffsetX()), (double)(0.5F * this.getProgress(1.0F) * (float)p_190587_1_.getFrontOffsetY()), (double)(0.5F * this.getProgress(1.0F) * (float)p_190587_1_.getFrontOffsetZ()));
     }
 
-    private AxisAlignedBB func_190588_c(EnumFacing p_190588_1_)
+    private AxisAlignedBB getTopBoundingBox(EnumFacing p_190588_1_)
     {
-        EnumFacing enumfacing = p_190588_1_.func_176734_d();
-        return this.func_190587_b(p_190588_1_).func_191195_a((double)enumfacing.func_82601_c(), (double)enumfacing.func_96559_d(), (double)enumfacing.func_82599_e());
+        EnumFacing enumfacing = p_190588_1_.getOpposite();
+        return this.getBoundingBox(p_190588_1_).contract((double)enumfacing.getFrontOffsetX(), (double)enumfacing.getFrontOffsetY(), (double)enumfacing.getFrontOffsetZ());
     }
 
-    private void func_190589_G()
+    private void moveCollidedEntities()
     {
-        IBlockState iblockstate = this.field_145850_b.func_180495_p(this.func_174877_v());
+        IBlockState iblockstate = this.world.getBlockState(this.getPos());
 
-        if (iblockstate.func_177230_c() instanceof BlockShulkerBox)
+        if (iblockstate.getBlock() instanceof BlockShulkerBox)
         {
-            EnumFacing enumfacing = (EnumFacing)iblockstate.func_177229_b(BlockShulkerBox.field_190957_a);
-            AxisAlignedBB axisalignedbb = this.func_190588_c(enumfacing).func_186670_a(this.field_174879_c);
-            List<Entity> list = this.field_145850_b.func_72839_b((Entity)null, axisalignedbb);
+            EnumFacing enumfacing = (EnumFacing)iblockstate.getValue(BlockShulkerBox.FACING);
+            AxisAlignedBB axisalignedbb = this.getTopBoundingBox(enumfacing).offset(this.pos);
+            List<Entity> list = this.world.getEntitiesWithinAABBExcludingEntity((Entity)null, axisalignedbb);
 
             if (!list.isEmpty())
             {
@@ -136,206 +161,206 @@
                 {
                     Entity entity = (Entity)list.get(i);
 
-                    if (entity.func_184192_z() != EnumPushReaction.IGNORE)
+                    if (entity.getPushReaction() != EnumPushReaction.IGNORE)
                     {
                         double d0 = 0.0D;
                         double d1 = 0.0D;
                         double d2 = 0.0D;
-                        AxisAlignedBB axisalignedbb1 = entity.func_174813_aQ();
+                        AxisAlignedBB axisalignedbb1 = entity.getEntityBoundingBox();
 
-                        switch (enumfacing.func_176740_k())
+                        switch (enumfacing.getAxis())
                         {
                             case X:
 
-                                if (enumfacing.func_176743_c() == EnumFacing.AxisDirection.POSITIVE)
+                                if (enumfacing.getAxisDirection() == EnumFacing.AxisDirection.POSITIVE)
                                 {
-                                    d0 = axisalignedbb.field_72336_d - axisalignedbb1.field_72340_a;
+                                    d0 = axisalignedbb.maxX - axisalignedbb1.minX;
                                 }
                                 else
                                 {
-                                    d0 = axisalignedbb1.field_72336_d - axisalignedbb.field_72340_a;
+                                    d0 = axisalignedbb1.maxX - axisalignedbb.minX;
                                 }
 
                                 d0 = d0 + 0.01D;
                                 break;
                             case Y:
 
-                                if (enumfacing.func_176743_c() == EnumFacing.AxisDirection.POSITIVE)
+                                if (enumfacing.getAxisDirection() == EnumFacing.AxisDirection.POSITIVE)
                                 {
-                                    d1 = axisalignedbb.field_72337_e - axisalignedbb1.field_72338_b;
+                                    d1 = axisalignedbb.maxY - axisalignedbb1.minY;
                                 }
                                 else
                                 {
-                                    d1 = axisalignedbb1.field_72337_e - axisalignedbb.field_72338_b;
+                                    d1 = axisalignedbb1.maxY - axisalignedbb.minY;
                                 }
 
                                 d1 = d1 + 0.01D;
                                 break;
                             case Z:
 
-                                if (enumfacing.func_176743_c() == EnumFacing.AxisDirection.POSITIVE)
+                                if (enumfacing.getAxisDirection() == EnumFacing.AxisDirection.POSITIVE)
                                 {
-                                    d2 = axisalignedbb.field_72334_f - axisalignedbb1.field_72339_c;
+                                    d2 = axisalignedbb.maxZ - axisalignedbb1.minZ;
                                 }
                                 else
                                 {
-                                    d2 = axisalignedbb1.field_72334_f - axisalignedbb.field_72339_c;
+                                    d2 = axisalignedbb1.maxZ - axisalignedbb.minZ;
                                 }
 
                                 d2 = d2 + 0.01D;
                         }
 
-                        entity.func_70091_d(MoverType.SHULKER_BOX, d0 * (double)enumfacing.func_82601_c(), d1 * (double)enumfacing.func_96559_d(), d2 * (double)enumfacing.func_82599_e());
+                        entity.move(MoverType.SHULKER_BOX, d0 * (double)enumfacing.getFrontOffsetX(), d1 * (double)enumfacing.getFrontOffsetY(), d2 * (double)enumfacing.getFrontOffsetZ());
                     }
                 }
             }
         }
     }
 
-    public int func_70302_i_()
+    public int getSizeInventory()
     {
-        return this.field_190596_f.size();
+        return this.items.size();
     }
 
-    public int func_70297_j_()
+    public int getInventoryStackLimit()
     {
         return 64;
     }
 
-    public boolean func_145842_c(int p_145842_1_, int p_145842_2_)
+    public boolean receiveClientEvent(int id, int type)
     {
-        if (p_145842_1_ == 1)
+        if (id == 1)
         {
-            this.field_190598_h = p_145842_2_;
+            this.openCount = type;
 
-            if (p_145842_2_ == 0)
+            if (type == 0)
             {
-                this.field_190599_i = TileEntityShulkerBox.AnimationStatus.CLOSING;
+                this.animationStatus = TileEntityShulkerBox.AnimationStatus.CLOSING;
             }
 
-            if (p_145842_2_ == 1)
+            if (type == 1)
             {
-                this.field_190599_i = TileEntityShulkerBox.AnimationStatus.OPENING;
+                this.animationStatus = TileEntityShulkerBox.AnimationStatus.OPENING;
             }
 
             return true;
         }
         else
         {
-            return super.func_145842_c(p_145842_1_, p_145842_2_);
+            return super.receiveClientEvent(id, type);
         }
     }
 
-    public void func_174889_b(EntityPlayer p_174889_1_)
+    public void openInventory(EntityPlayer player)
     {
-        if (!p_174889_1_.func_175149_v())
+        if (!player.isSpectator())
         {
-            if (this.field_190598_h < 0)
+            if (this.openCount < 0)
             {
-                this.field_190598_h = 0;
+                this.openCount = 0;
             }
 
-            ++this.field_190598_h;
-            this.field_145850_b.func_175641_c(this.field_174879_c, this.func_145838_q(), 1, this.field_190598_h);
+            ++this.openCount;
+            this.world.addBlockEvent(this.pos, this.getBlockType(), 1, this.openCount);
 
-            if (this.field_190598_h == 1)
+            if (this.openCount == 1)
             {
-                this.field_145850_b.func_184133_a((EntityPlayer)null, this.field_174879_c, SoundEvents.field_191262_fB, SoundCategory.BLOCKS, 0.5F, this.field_145850_b.field_73012_v.nextFloat() * 0.1F + 0.9F);
+                this.world.playSound((EntityPlayer)null, this.pos, SoundEvents.BLOCK_SHULKER_BOX_OPEN, SoundCategory.BLOCKS, 0.5F, this.world.rand.nextFloat() * 0.1F + 0.9F);
             }
         }
     }
 
-    public void func_174886_c(EntityPlayer p_174886_1_)
+    public void closeInventory(EntityPlayer player)
     {
-        if (!p_174886_1_.func_175149_v())
+        if (!player.isSpectator())
         {
-            --this.field_190598_h;
-            this.field_145850_b.func_175641_c(this.field_174879_c, this.func_145838_q(), 1, this.field_190598_h);
+            --this.openCount;
+            this.world.addBlockEvent(this.pos, this.getBlockType(), 1, this.openCount);
 
-            if (this.field_190598_h <= 0)
+            if (this.openCount <= 0)
             {
-                this.field_145850_b.func_184133_a((EntityPlayer)null, this.field_174879_c, SoundEvents.field_191261_fA, SoundCategory.BLOCKS, 0.5F, this.field_145850_b.field_73012_v.nextFloat() * 0.1F + 0.9F);
+                this.world.playSound((EntityPlayer)null, this.pos, SoundEvents.BLOCK_SHULKER_BOX_CLOSE, SoundCategory.BLOCKS, 0.5F, this.world.rand.nextFloat() * 0.1F + 0.9F);
             }
         }
     }
 
-    public Container func_174876_a(InventoryPlayer p_174876_1_, EntityPlayer p_174876_2_)
+    public Container createContainer(InventoryPlayer playerInventory, EntityPlayer playerIn)
     {
-        return new ContainerShulkerBox(p_174876_1_, this, p_174876_2_);
+        return new ContainerShulkerBox(playerInventory, this, playerIn);
     }
 
-    public String func_174875_k()
+    public String getGuiID()
     {
         return "minecraft:shulker_box";
     }
 
-    public String func_70005_c_()
+    public String getName()
     {
-        return this.func_145818_k_() ? this.field_190577_o : "container.shulkerBox";
+        return this.hasCustomName() ? this.customName : "container.shulkerBox";
     }
 
-    public static void func_190593_a(DataFixer p_190593_0_)
+    public static void registerFixesShulkerBox(DataFixer p_190593_0_)
     {
-        p_190593_0_.func_188258_a(FixTypes.BLOCK_ENTITY, new ItemStackDataLists(TileEntityShulkerBox.class, new String[] {"Items"}));
+        p_190593_0_.registerWalker(FixTypes.BLOCK_ENTITY, new ItemStackDataLists(TileEntityShulkerBox.class, new String[] {"Items"}));
     }
 
-    public void func_145839_a(NBTTagCompound p_145839_1_)
+    public void readFromNBT(NBTTagCompound compound)
     {
-        super.func_145839_a(p_145839_1_);
-        this.func_190586_e(p_145839_1_);
+        super.readFromNBT(compound);
+        this.loadFromNbt(compound);
     }
 
-    public NBTTagCompound func_189515_b(NBTTagCompound p_189515_1_)
+    public NBTTagCompound writeToNBT(NBTTagCompound compound)
     {
-        super.func_189515_b(p_189515_1_);
-        return this.func_190580_f(p_189515_1_);
+        super.writeToNBT(compound);
+        return this.saveToNbt(compound);
     }
 
-    public void func_190586_e(NBTTagCompound p_190586_1_)
+    public void loadFromNbt(NBTTagCompound compound)
     {
-        this.field_190596_f = NonNullList.<ItemStack>func_191197_a(this.func_70302_i_(), ItemStack.field_190927_a);
+        this.items = NonNullList.<ItemStack>withSize(this.getSizeInventory(), ItemStack.EMPTY);
 
-        if (!this.func_184283_b(p_190586_1_) && p_190586_1_.func_150297_b("Items", 9))
+        if (!this.checkLootAndRead(compound) && compound.hasKey("Items", 9))
         {
-            ItemStackHelper.func_191283_b(p_190586_1_, this.field_190596_f);
+            ItemStackHelper.loadAllItems(compound, this.items);
         }
 
-        if (p_190586_1_.func_150297_b("CustomName", 8))
+        if (compound.hasKey("CustomName", 8))
         {
-            this.field_190577_o = p_190586_1_.func_74779_i("CustomName");
+            this.customName = compound.getString("CustomName");
         }
     }
 
-    public NBTTagCompound func_190580_f(NBTTagCompound p_190580_1_)
+    public NBTTagCompound saveToNbt(NBTTagCompound compound)
     {
-        if (!this.func_184282_c(p_190580_1_))
+        if (!this.checkLootAndWrite(compound))
         {
-            ItemStackHelper.func_191281_a(p_190580_1_, this.field_190596_f, false);
+            ItemStackHelper.saveAllItems(compound, this.items, false);
         }
 
-        if (this.func_145818_k_())
+        if (this.hasCustomName())
         {
-            p_190580_1_.func_74778_a("CustomName", this.field_190577_o);
+            compound.setString("CustomName", this.customName);
         }
 
-        if (!p_190580_1_.func_74764_b("Lock") && this.func_174893_q_())
+        if (!compound.hasKey("Lock") && this.isLocked())
         {
-            this.func_174891_i().func_180157_a(p_190580_1_);
+            this.getLockCode().toNBT(compound);
         }
 
-        return p_190580_1_;
+        return compound;
     }
 
-    protected NonNullList<ItemStack> func_190576_q()
+    protected NonNullList<ItemStack> getItems()
     {
-        return this.field_190596_f;
+        return this.items;
     }
 
-    public boolean func_191420_l()
+    public boolean isEmpty()
     {
-        for (ItemStack itemstack : this.field_190596_f)
+        for (ItemStack itemstack : this.items)
         {
-            if (!itemstack.func_190926_b())
+            if (!itemstack.isEmpty())
             {
                 return false;
             }
@@ -344,72 +369,72 @@
         return true;
     }
 
-    public int[] func_180463_a(EnumFacing p_180463_1_)
+    public int[] getSlotsForFace(EnumFacing side)
     {
-        return field_190595_a;
+        return SLOTS;
     }
 
-    public boolean func_180462_a(int p_180462_1_, ItemStack p_180462_2_, EnumFacing p_180462_3_)
+    public boolean canInsertItem(int index, ItemStack itemStackIn, EnumFacing direction)
     {
-        return !(Block.func_149634_a(p_180462_2_.func_77973_b()) instanceof BlockShulkerBox);
+        return !(Block.getBlockFromItem(itemStackIn.getItem()) instanceof BlockShulkerBox);
     }
 
-    public boolean func_180461_b(int p_180461_1_, ItemStack p_180461_2_, EnumFacing p_180461_3_)
+    public boolean canExtractItem(int index, ItemStack stack, EnumFacing direction)
     {
         return true;
     }
 
-    public void func_174888_l()
+    public void clear()
     {
-        this.field_190597_g = true;
-        super.func_174888_l();
+        this.hasBeenCleared = true;
+        super.clear();
     }
 
-    public boolean func_190590_r()
+    public boolean isCleared()
     {
-        return this.field_190597_g;
+        return this.hasBeenCleared;
     }
 
-    public float func_190585_a(float p_190585_1_)
+    public float getProgress(float p_190585_1_)
     {
-        return this.field_190601_k + (this.field_190600_j - this.field_190601_k) * p_190585_1_;
+        return this.progressOld + (this.progress - this.progressOld) * p_190585_1_;
     }
 
     @SideOnly(Side.CLIENT)
-    public EnumDyeColor func_190592_s()
+    public EnumDyeColor getColor()
     {
-        if (this.field_190602_l == null)
+        if (this.color == null)
         {
-            this.field_190602_l = BlockShulkerBox.func_190954_c(this.func_145838_q());
+            this.color = BlockShulkerBox.getColorFromBlock(this.getBlockType());
         }
 
-        return this.field_190602_l;
+        return this.color;
     }
 
     @Nullable
-    public SPacketUpdateTileEntity func_189518_D_()
+    public SPacketUpdateTileEntity getUpdatePacket()
     {
-        return new SPacketUpdateTileEntity(this.field_174879_c, 10, this.func_189517_E_());
+        return new SPacketUpdateTileEntity(this.pos, 10, this.getUpdateTag());
     }
 
-    public boolean func_190581_E()
+    public boolean isDestroyedByCreativePlayer()
     {
-        return this.field_190594_p;
+        return this.destroyedByCreativePlayer;
     }
 
-    public void func_190579_a(boolean p_190579_1_)
+    public void setDestroyedByCreativePlayer(boolean p_190579_1_)
     {
-        this.field_190594_p = p_190579_1_;
+        this.destroyedByCreativePlayer = p_190579_1_;
     }
 
-    public boolean func_190582_F()
+    public boolean shouldDrop()
     {
-        return !this.func_190581_E() || !this.func_191420_l() || this.func_145818_k_() || this.field_184284_m != null;
+        return !this.isDestroyedByCreativePlayer() || !this.isEmpty() || this.hasCustomName() || this.lootTable != null;
     }
 
     static
     {
-        for (int i = 0; i < field_190595_a.length; field_190595_a[i] = i++)
+        for (int i = 0; i < SLOTS.length; SLOTS[i] = i++)
         {
             ;
         }
@@ -422,4 +447,9 @@
         OPENED,
         CLOSING;
     }
+
+    protected net.minecraftforge.items.IItemHandler createUnSidedHandler()
+    {
+        return new net.minecraftforge.items.wrapper.SidedInvWrapper(this, EnumFacing.UP);
+    }
 }
