--- ../src-base/minecraft/net/minecraft/client/gui/inventory/GuiContainerCreative.java
+++ ../src-work/minecraft/net/minecraft/client/gui/inventory/GuiContainerCreative.java
@@ -35,329 +35,352 @@
 import net.minecraft.util.text.TextFormatting;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.inventory.InventoryView;
 import org.lwjgl.input.Keyboard;
 import org.lwjgl.input.Mouse;
 
 @SideOnly(Side.CLIENT)
 public class GuiContainerCreative extends InventoryEffectRenderer
 {
-    private static final ResourceLocation field_147061_u = new ResourceLocation("textures/gui/container/creative_inventory/tabs.png");
-    private static final InventoryBasic field_147060_v = new InventoryBasic("tmp", true, 45);
-    private static int field_147058_w = CreativeTabs.field_78030_b.func_78021_a();
-    private float field_147067_x;
-    private boolean field_147066_y;
-    private boolean field_147065_z;
-    private GuiTextField field_147062_A;
-    private List<Slot> field_147063_B;
-    private Slot field_147064_C;
-    private boolean field_147057_D;
-    private CreativeCrafting field_147059_E;
+    private static final ResourceLocation CREATIVE_INVENTORY_TABS = new ResourceLocation("textures/gui/container/creative_inventory/tabs.png");
+    private static final InventoryBasic basicInventory = new InventoryBasic("tmp", true, 45);
+    private static int selectedTabIndex = CreativeTabs.BUILDING_BLOCKS.getTabIndex();
+    private float currentScroll;
+    private boolean isScrolling;
+    private boolean wasClicking;
+    private GuiTextField searchField;
+    private List<Slot> originalSlots;
+    private Slot destroyItemSlot;
+    private boolean clearSearch;
+    private CreativeCrafting listener;
+    private static int tabPage = 0;
+    private int maxPages = 0;
 
-    public GuiContainerCreative(EntityPlayer p_i1088_1_)
+    public GuiContainerCreative(EntityPlayer player)
     {
-        super(new GuiContainerCreative.ContainerCreative(p_i1088_1_));
-        p_i1088_1_.field_71070_bA = this.field_147002_h;
-        this.field_146291_p = true;
-        this.field_147000_g = 136;
-        this.field_146999_f = 195;
+        super(new GuiContainerCreative.ContainerCreative(player));
+        player.openContainer = this.inventorySlots;
+        this.allowUserInput = true;
+        this.ySize = 136;
+        this.xSize = 195;
     }
 
-    public void func_73876_c()
+    public void updateScreen()
     {
-        if (!this.field_146297_k.field_71442_b.func_78758_h())
+        if (!this.mc.playerController.isInCreativeMode())
         {
-            this.field_146297_k.func_147108_a(new GuiInventory(this.field_146297_k.field_71439_g));
+            this.mc.displayGuiScreen(new GuiInventory(this.mc.player));
         }
     }
 
-    protected void func_184098_a(@Nullable Slot p_184098_1_, int p_184098_2_, int p_184098_3_, ClickType p_184098_4_)
+    protected void handleMouseClick(@Nullable Slot slotIn, int slotId, int mouseButton, ClickType type)
     {
-        this.field_147057_D = true;
-        boolean flag = p_184098_4_ == ClickType.QUICK_MOVE;
-        p_184098_4_ = p_184098_2_ == -999 && p_184098_4_ == ClickType.PICKUP ? ClickType.THROW : p_184098_4_;
+        this.clearSearch = true;
+        boolean flag = type == ClickType.QUICK_MOVE;
+        type = slotId == -999 && type == ClickType.PICKUP ? ClickType.THROW : type;
 
-        if (p_184098_1_ == null && field_147058_w != CreativeTabs.field_78036_m.func_78021_a() && p_184098_4_ != ClickType.QUICK_CRAFT)
+        if (slotIn == null && selectedTabIndex != CreativeTabs.INVENTORY.getTabIndex() && type != ClickType.QUICK_CRAFT)
         {
-            InventoryPlayer inventoryplayer1 = this.field_146297_k.field_71439_g.field_71071_by;
+            InventoryPlayer inventoryplayer1 = this.mc.player.inventory;
 
-            if (!inventoryplayer1.func_70445_o().func_190926_b())
+            if (!inventoryplayer1.getItemStack().isEmpty())
             {
-                if (p_184098_3_ == 0)
+                if (mouseButton == 0)
                 {
-                    this.field_146297_k.field_71439_g.func_71019_a(inventoryplayer1.func_70445_o(), true);
-                    this.field_146297_k.field_71442_b.func_78752_a(inventoryplayer1.func_70445_o());
-                    inventoryplayer1.func_70437_b(ItemStack.field_190927_a);
+                    this.mc.player.dropItem(inventoryplayer1.getItemStack(), true);
+                    this.mc.playerController.sendPacketDropItem(inventoryplayer1.getItemStack());
+                    inventoryplayer1.setItemStack(ItemStack.EMPTY);
                 }
 
-                if (p_184098_3_ == 1)
+                if (mouseButton == 1)
                 {
-                    ItemStack itemstack6 = inventoryplayer1.func_70445_o().func_77979_a(1);
-                    this.field_146297_k.field_71439_g.func_71019_a(itemstack6, true);
-                    this.field_146297_k.field_71442_b.func_78752_a(itemstack6);
+                    ItemStack itemstack6 = inventoryplayer1.getItemStack().splitStack(1);
+                    this.mc.player.dropItem(itemstack6, true);
+                    this.mc.playerController.sendPacketDropItem(itemstack6);
                 }
             }
         }
-        else if (p_184098_1_ == this.field_147064_C && flag)
+        else if (slotIn == this.destroyItemSlot && flag)
         {
-            for (int j = 0; j < this.field_146297_k.field_71439_g.field_71069_bz.func_75138_a().size(); ++j)
+            for (int j = 0; j < this.mc.player.inventoryContainer.getInventory().size(); ++j)
             {
-                this.field_146297_k.field_71442_b.func_78761_a(ItemStack.field_190927_a, j);
+                this.mc.playerController.sendSlotPacket(ItemStack.EMPTY, j);
             }
         }
-        else if (field_147058_w == CreativeTabs.field_78036_m.func_78021_a())
+        else if (selectedTabIndex == CreativeTabs.INVENTORY.getTabIndex())
         {
-            if (p_184098_1_ == this.field_147064_C)
+            if (slotIn == this.destroyItemSlot)
             {
-                this.field_146297_k.field_71439_g.field_71071_by.func_70437_b(ItemStack.field_190927_a);
+                this.mc.player.inventory.setItemStack(ItemStack.EMPTY);
             }
-            else if (p_184098_4_ == ClickType.THROW && p_184098_1_ != null && p_184098_1_.func_75216_d())
+            else if (type == ClickType.THROW && slotIn != null && slotIn.getHasStack())
             {
-                ItemStack itemstack = p_184098_1_.func_75209_a(p_184098_3_ == 0 ? 1 : p_184098_1_.func_75211_c().func_77976_d());
-                ItemStack itemstack1 = p_184098_1_.func_75211_c();
-                this.field_146297_k.field_71439_g.func_71019_a(itemstack, true);
-                this.field_146297_k.field_71442_b.func_78752_a(itemstack);
-                this.field_146297_k.field_71442_b.func_78761_a(itemstack1, ((GuiContainerCreative.CreativeSlot)p_184098_1_).field_148332_b.field_75222_d);
+                ItemStack itemstack = slotIn.decrStackSize(mouseButton == 0 ? 1 : slotIn.getStack().getMaxStackSize());
+                ItemStack itemstack1 = slotIn.getStack();
+                this.mc.player.dropItem(itemstack, true);
+                this.mc.playerController.sendPacketDropItem(itemstack);
+                this.mc.playerController.sendSlotPacket(itemstack1, ((GuiContainerCreative.CreativeSlot)slotIn).slot.slotNumber);
             }
-            else if (p_184098_4_ == ClickType.THROW && !this.field_146297_k.field_71439_g.field_71071_by.func_70445_o().func_190926_b())
+            else if (type == ClickType.THROW && !this.mc.player.inventory.getItemStack().isEmpty())
             {
-                this.field_146297_k.field_71439_g.func_71019_a(this.field_146297_k.field_71439_g.field_71071_by.func_70445_o(), true);
-                this.field_146297_k.field_71442_b.func_78752_a(this.field_146297_k.field_71439_g.field_71071_by.func_70445_o());
-                this.field_146297_k.field_71439_g.field_71071_by.func_70437_b(ItemStack.field_190927_a);
+                this.mc.player.dropItem(this.mc.player.inventory.getItemStack(), true);
+                this.mc.playerController.sendPacketDropItem(this.mc.player.inventory.getItemStack());
+                this.mc.player.inventory.setItemStack(ItemStack.EMPTY);
             }
             else
             {
-                this.field_146297_k.field_71439_g.field_71069_bz.func_184996_a(p_184098_1_ == null ? p_184098_2_ : ((GuiContainerCreative.CreativeSlot)p_184098_1_).field_148332_b.field_75222_d, p_184098_3_, p_184098_4_, this.field_146297_k.field_71439_g);
-                this.field_146297_k.field_71439_g.field_71069_bz.func_75142_b();
+                this.mc.player.inventoryContainer.slotClick(slotIn == null ? slotId : ((GuiContainerCreative.CreativeSlot)slotIn).slot.slotNumber, mouseButton, type, this.mc.player);
+                this.mc.player.inventoryContainer.detectAndSendChanges();
             }
         }
-        else if (p_184098_4_ != ClickType.QUICK_CRAFT && p_184098_1_.field_75224_c == field_147060_v)
+        else if (type != ClickType.QUICK_CRAFT && slotIn.inventory == basicInventory)
         {
-            InventoryPlayer inventoryplayer = this.field_146297_k.field_71439_g.field_71071_by;
-            ItemStack itemstack5 = inventoryplayer.func_70445_o();
-            ItemStack itemstack7 = p_184098_1_.func_75211_c();
+            InventoryPlayer inventoryplayer = this.mc.player.inventory;
+            ItemStack itemstack5 = inventoryplayer.getItemStack();
+            ItemStack itemstack7 = slotIn.getStack();
 
-            if (p_184098_4_ == ClickType.SWAP)
+            if (type == ClickType.SWAP)
             {
-                if (!itemstack7.func_190926_b() && p_184098_3_ >= 0 && p_184098_3_ < 9)
+                if (!itemstack7.isEmpty() && mouseButton >= 0 && mouseButton < 9)
                 {
-                    ItemStack itemstack10 = itemstack7.func_77946_l();
-                    itemstack10.func_190920_e(itemstack10.func_77976_d());
-                    this.field_146297_k.field_71439_g.field_71071_by.func_70299_a(p_184098_3_, itemstack10);
-                    this.field_146297_k.field_71439_g.field_71069_bz.func_75142_b();
+                    ItemStack itemstack10 = itemstack7.copy();
+                    itemstack10.setCount(itemstack10.getMaxStackSize());
+                    this.mc.player.inventory.setInventorySlotContents(mouseButton, itemstack10);
+                    this.mc.player.inventoryContainer.detectAndSendChanges();
                 }
 
                 return;
             }
 
-            if (p_184098_4_ == ClickType.CLONE)
+            if (type == ClickType.CLONE)
             {
-                if (inventoryplayer.func_70445_o().func_190926_b() && p_184098_1_.func_75216_d())
+                if (inventoryplayer.getItemStack().isEmpty() && slotIn.getHasStack())
                 {
-                    ItemStack itemstack9 = p_184098_1_.func_75211_c().func_77946_l();
-                    itemstack9.func_190920_e(itemstack9.func_77976_d());
-                    inventoryplayer.func_70437_b(itemstack9);
+                    ItemStack itemstack9 = slotIn.getStack().copy();
+                    itemstack9.setCount(itemstack9.getMaxStackSize());
+                    inventoryplayer.setItemStack(itemstack9);
                 }
 
                 return;
             }
 
-            if (p_184098_4_ == ClickType.THROW)
+            if (type == ClickType.THROW)
             {
-                if (!itemstack7.func_190926_b())
+                if (!itemstack7.isEmpty())
                 {
-                    ItemStack itemstack8 = itemstack7.func_77946_l();
-                    itemstack8.func_190920_e(p_184098_3_ == 0 ? 1 : itemstack8.func_77976_d());
-                    this.field_146297_k.field_71439_g.func_71019_a(itemstack8, true);
-                    this.field_146297_k.field_71442_b.func_78752_a(itemstack8);
+                    ItemStack itemstack8 = itemstack7.copy();
+                    itemstack8.setCount(mouseButton == 0 ? 1 : itemstack8.getMaxStackSize());
+                    this.mc.player.dropItem(itemstack8, true);
+                    this.mc.playerController.sendPacketDropItem(itemstack8);
                 }
 
                 return;
             }
 
-            if (!itemstack5.func_190926_b() && !itemstack7.func_190926_b() && itemstack5.func_77969_a(itemstack7) && ItemStack.func_77970_a(itemstack5, itemstack7))
+            if (!itemstack5.isEmpty() && !itemstack7.isEmpty() && itemstack5.isItemEqual(itemstack7) && ItemStack.areItemStackTagsEqual(itemstack5, itemstack7))
             {
-                if (p_184098_3_ == 0)
+                if (mouseButton == 0)
                 {
                     if (flag)
                     {
-                        itemstack5.func_190920_e(itemstack5.func_77976_d());
+                        itemstack5.setCount(itemstack5.getMaxStackSize());
                     }
-                    else if (itemstack5.func_190916_E() < itemstack5.func_77976_d())
+                    else if (itemstack5.getCount() < itemstack5.getMaxStackSize())
                     {
-                        itemstack5.func_190917_f(1);
+                        itemstack5.grow(1);
                     }
                 }
-                else if (itemstack5.func_190916_E() == 1)
+                else if (itemstack5.getCount() == 1)
                 {
-                    inventoryplayer.func_70437_b(ItemStack.field_190927_a);
+                    inventoryplayer.setItemStack(ItemStack.EMPTY);
                 }
                 else
                 {
-                    itemstack5.func_190918_g(1);
+                    itemstack5.shrink(1);
                 }
             }
-            else if (!itemstack7.func_190926_b() && itemstack5.func_190926_b())
+            else if (!itemstack7.isEmpty() && itemstack5.isEmpty())
             {
-                inventoryplayer.func_70437_b(itemstack7.func_77946_l());
-                itemstack5 = inventoryplayer.func_70445_o();
+                inventoryplayer.setItemStack(itemstack7.copy());
+                itemstack5 = inventoryplayer.getItemStack();
 
                 if (flag)
                 {
-                    itemstack5.func_190920_e(itemstack5.func_77976_d());
+                    itemstack5.setCount(itemstack5.getMaxStackSize());
                 }
             }
             else
             {
-                inventoryplayer.func_70437_b(ItemStack.field_190927_a);
+                inventoryplayer.setItemStack(ItemStack.EMPTY);
             }
         }
-        else if (this.field_147002_h != null)
+        else if (this.inventorySlots != null)
         {
-            ItemStack itemstack3 = p_184098_1_ == null ? ItemStack.field_190927_a : this.field_147002_h.func_75139_a(p_184098_1_.field_75222_d).func_75211_c();
-            this.field_147002_h.func_184996_a(p_184098_1_ == null ? p_184098_2_ : p_184098_1_.field_75222_d, p_184098_3_, p_184098_4_, this.field_146297_k.field_71439_g);
+            ItemStack itemstack3 = slotIn == null ? ItemStack.EMPTY : this.inventorySlots.getSlot(slotIn.slotNumber).getStack();
+            this.inventorySlots.slotClick(slotIn == null ? slotId : slotIn.slotNumber, mouseButton, type, this.mc.player);
 
-            if (Container.func_94532_c(p_184098_3_) == 2)
+            if (Container.getDragEvent(mouseButton) == 2)
             {
                 for (int k = 0; k < 9; ++k)
                 {
-                    this.field_146297_k.field_71442_b.func_78761_a(this.field_147002_h.func_75139_a(45 + k).func_75211_c(), 36 + k);
+                    this.mc.playerController.sendSlotPacket(this.inventorySlots.getSlot(45 + k).getStack(), 36 + k);
                 }
             }
-            else if (p_184098_1_ != null)
+            else if (slotIn != null)
             {
-                ItemStack itemstack4 = this.field_147002_h.func_75139_a(p_184098_1_.field_75222_d).func_75211_c();
-                this.field_146297_k.field_71442_b.func_78761_a(itemstack4, p_184098_1_.field_75222_d - this.field_147002_h.field_75151_b.size() + 9 + 36);
-                int i = 45 + p_184098_3_;
+                ItemStack itemstack4 = this.inventorySlots.getSlot(slotIn.slotNumber).getStack();
+                this.mc.playerController.sendSlotPacket(itemstack4, slotIn.slotNumber - this.inventorySlots.inventorySlots.size() + 9 + 36);
+                int i = 45 + mouseButton;
 
-                if (p_184098_4_ == ClickType.SWAP)
+                if (type == ClickType.SWAP)
                 {
-                    this.field_146297_k.field_71442_b.func_78761_a(itemstack3, i - this.field_147002_h.field_75151_b.size() + 9 + 36);
+                    this.mc.playerController.sendSlotPacket(itemstack3, i - this.inventorySlots.inventorySlots.size() + 9 + 36);
                 }
-                else if (p_184098_4_ == ClickType.THROW && itemstack3 != ItemStack.field_190927_a)
+                else if (type == ClickType.THROW && itemstack3 != ItemStack.EMPTY)
                 {
-                    ItemStack itemstack2 = itemstack3.func_77946_l();
-                    itemstack2.func_190920_e(p_184098_3_ == 0 ? 1 : itemstack2.func_77976_d());
-                    this.field_146297_k.field_71439_g.func_71019_a(itemstack2, true);
-                    this.field_146297_k.field_71442_b.func_78752_a(itemstack2);
+                    ItemStack itemstack2 = itemstack3.copy();
+                    itemstack2.setCount(mouseButton == 0 ? 1 : itemstack2.getMaxStackSize());
+                    this.mc.player.dropItem(itemstack2, true);
+                    this.mc.playerController.sendPacketDropItem(itemstack2);
                 }
             }
         }
     }
 
-    protected void func_175378_g()
+    protected void updateActivePotionEffects()
     {
-        int i = this.field_147003_i;
-        super.func_175378_g();
+        int i = this.guiLeft;
+        super.updateActivePotionEffects();
 
-        if (this.field_147062_A != null && this.field_147003_i != i)
+        if (this.searchField != null && this.guiLeft != i)
         {
-            this.field_147062_A.field_146209_f = this.field_147003_i + 82;
+            this.searchField.xPosition = this.guiLeft + 82;
         }
     }
 
-    public void func_73866_w_()
+    public void initGui()
     {
-        if (this.field_146297_k.field_71442_b.func_78758_h())
+        if (this.mc.playerController.isInCreativeMode())
         {
-            super.func_73866_w_();
-            this.field_146292_n.clear();
+            super.initGui();
+            this.buttonList.clear();
             Keyboard.enableRepeatEvents(true);
-            this.field_147062_A = new GuiTextField(0, this.field_146289_q, this.field_147003_i + 82, this.field_147009_r + 6, 89, this.field_146289_q.field_78288_b);
-            this.field_147062_A.func_146203_f(15);
-            this.field_147062_A.func_146185_a(false);
-            this.field_147062_A.func_146189_e(false);
-            this.field_147062_A.func_146193_g(16777215);
-            int i = field_147058_w;
-            field_147058_w = -1;
-            this.func_147050_b(CreativeTabs.field_78032_a[i]);
-            this.field_147059_E = new CreativeCrafting(this.field_146297_k);
-            this.field_146297_k.field_71439_g.field_71069_bz.func_75132_a(this.field_147059_E);
+            this.searchField = new GuiTextField(0, this.fontRendererObj, this.guiLeft + 82, this.guiTop + 6, 89, this.fontRendererObj.FONT_HEIGHT);
+            this.searchField.setMaxStringLength(15);
+            this.searchField.setEnableBackgroundDrawing(false);
+            this.searchField.setVisible(false);
+            this.searchField.setTextColor(16777215);
+            int i = selectedTabIndex;
+            selectedTabIndex = -1;
+            this.setCurrentCreativeTab(CreativeTabs.CREATIVE_TAB_ARRAY[i]);
+            this.listener = new CreativeCrafting(this.mc);
+            this.mc.player.inventoryContainer.addListener(this.listener);
+            int tabCount = CreativeTabs.CREATIVE_TAB_ARRAY.length;
+            if (tabCount > 12)
+            {
+                buttonList.add(new GuiButton(101, guiLeft,              guiTop - 50, 20, 20, "<"));
+                buttonList.add(new GuiButton(102, guiLeft + xSize - 20, guiTop - 50, 20, 20, ">"));
+                maxPages = (int) Math.ceil((tabCount - 12) / 10D);
+            }
         }
         else
         {
-            this.field_146297_k.func_147108_a(new GuiInventory(this.field_146297_k.field_71439_g));
+            this.mc.displayGuiScreen(new GuiInventory(this.mc.player));
         }
     }
 
-    public void func_146281_b()
+    public void onGuiClosed()
     {
-        super.func_146281_b();
+        super.onGuiClosed();
 
-        if (this.field_146297_k.field_71439_g != null && this.field_146297_k.field_71439_g.field_71071_by != null)
+        if (this.mc.player != null && this.mc.player.inventory != null)
         {
-            this.field_146297_k.field_71439_g.field_71069_bz.func_82847_b(this.field_147059_E);
+            this.mc.player.inventoryContainer.removeListener(this.listener);
         }
 
         Keyboard.enableRepeatEvents(false);
     }
 
-    protected void func_73869_a(char p_73869_1_, int p_73869_2_) throws IOException
+    protected void keyTyped(char typedChar, int keyCode) throws IOException
     {
-        if (field_147058_w != CreativeTabs.field_78027_g.func_78021_a())
+        if (!CreativeTabs.CREATIVE_TAB_ARRAY[selectedTabIndex].hasSearchBar())
         {
-            if (GameSettings.func_100015_a(this.field_146297_k.field_71474_y.field_74310_D))
+            if (GameSettings.isKeyDown(this.mc.gameSettings.keyBindChat))
             {
-                this.func_147050_b(CreativeTabs.field_78027_g);
+                this.setCurrentCreativeTab(CreativeTabs.SEARCH);
             }
             else
             {
-                super.func_73869_a(p_73869_1_, p_73869_2_);
+                super.keyTyped(typedChar, keyCode);
             }
         }
         else
         {
-            if (this.field_147057_D)
+            if (this.clearSearch)
             {
-                this.field_147057_D = false;
-                this.field_147062_A.func_146180_a("");
+                this.clearSearch = false;
+                this.searchField.setText("");
             }
 
-            if (!this.func_146983_a(p_73869_2_))
+            if (!this.checkHotbarKeys(keyCode))
             {
-                if (this.field_147062_A.func_146201_a(p_73869_1_, p_73869_2_))
+                if (this.searchField.textboxKeyTyped(typedChar, keyCode))
                 {
-                    this.func_147053_i();
+                    this.updateCreativeSearch();
                 }
                 else
                 {
-                    super.func_73869_a(p_73869_1_, p_73869_2_);
+                    super.keyTyped(typedChar, keyCode);
                 }
             }
         }
     }
 
-    private void func_147053_i()
+    private void updateCreativeSearch()
     {
-        GuiContainerCreative.ContainerCreative guicontainercreative$containercreative = (GuiContainerCreative.ContainerCreative)this.field_147002_h;
-        guicontainercreative$containercreative.field_148330_a.clear();
+        GuiContainerCreative.ContainerCreative guicontainercreative$containercreative = (GuiContainerCreative.ContainerCreative)this.inventorySlots;
+        guicontainercreative$containercreative.itemList.clear();
 
-        for (Item item : Item.field_150901_e)
+        CreativeTabs tab = CreativeTabs.CREATIVE_TAB_ARRAY[selectedTabIndex];
+        if (tab.hasSearchBar() && tab != CreativeTabs.SEARCH)
         {
-            if (item != null && item.func_77640_w() != null)
+            tab.displayAllRelevantItems(guicontainercreative$containercreative.itemList);
+            updateFilteredItems(guicontainercreative$containercreative);
+            return;
+        }
+
+        for (Item item : Item.REGISTRY)
+        {
+            if (item != null && item.getCreativeTab() != null)
             {
-                item.func_150895_a(item, (CreativeTabs)null, guicontainercreative$containercreative.field_148330_a);
+                item.getSubItems(item, (CreativeTabs)null, guicontainercreative$containercreative.itemList);
             }
         }
+        updateFilteredItems(guicontainercreative$containercreative);
+    }
 
-        for (Enchantment enchantment : Enchantment.field_185264_b)
+    //split from above for custom search tabs
+    private void updateFilteredItems(GuiContainerCreative.ContainerCreative guicontainercreative$containercreative)
+    {
+        if (CreativeTabs.CREATIVE_TAB_ARRAY[selectedTabIndex] == CreativeTabs.SEARCH) // FORGE: Only add enchanted books to the regular search
+        for (Enchantment enchantment : Enchantment.REGISTRY)
         {
-            if (enchantment != null && enchantment.field_77351_y != null)
+            if (enchantment != null && enchantment.type != null)
             {
-                Items.field_151134_bR.func_92113_a(enchantment, guicontainercreative$containercreative.field_148330_a);
+                Items.ENCHANTED_BOOK.getAll(enchantment, guicontainercreative$containercreative.itemList);
             }
         }
+        Iterator<ItemStack> iterator = guicontainercreative$containercreative.itemList.iterator();
+        String s1 = this.searchField.getText().toLowerCase(Locale.ROOT);
 
-        Iterator<ItemStack> iterator = guicontainercreative$containercreative.field_148330_a.iterator();
-        String s1 = this.field_147062_A.func_146179_b().toLowerCase(Locale.ROOT);
-
         while (iterator.hasNext())
         {
             ItemStack itemstack = (ItemStack)iterator.next();
             boolean flag = false;
 
-            for (String s : itemstack.func_82840_a(this.field_146297_k.field_71439_g, this.field_146297_k.field_71474_y.field_82882_x))
+            for (String s : itemstack.getTooltip(this.mc.player, this.mc.gameSettings.advancedItemTooltips))
             {
-                if (TextFormatting.func_110646_a(s).toLowerCase(Locale.ROOT).contains(s1))
+                if (TextFormatting.getTextWithoutFormattingCodes(s).toLowerCase(Locale.ROOT).contains(s1))
                 {
                     flag = true;
                     break;
@@ -370,165 +393,169 @@
             }
         }
 
-        this.field_147067_x = 0.0F;
-        guicontainercreative$containercreative.func_148329_a(0.0F);
+        this.currentScroll = 0.0F;
+        guicontainercreative$containercreative.scrollTo(0.0F);
     }
 
-    protected void func_146979_b(int p_146979_1_, int p_146979_2_)
+    protected void drawGuiContainerForegroundLayer(int mouseX, int mouseY)
     {
-        CreativeTabs creativetabs = CreativeTabs.field_78032_a[field_147058_w];
+        CreativeTabs creativetabs = CreativeTabs.CREATIVE_TAB_ARRAY[selectedTabIndex];
 
-        if (creativetabs.func_78019_g())
+        if (creativetabs != null && creativetabs.drawInForegroundOfTab())
         {
-            GlStateManager.func_179084_k();
-            this.field_146289_q.func_78276_b(I18n.func_135052_a(creativetabs.func_78024_c(), new Object[0]), 8, 6, 4210752);
+            GlStateManager.disableBlend();
+            this.fontRendererObj.drawString(I18n.format(creativetabs.getTranslatedTabLabel(), new Object[0]), 8, 6, 4210752);
         }
     }
 
-    protected void func_73864_a(int p_73864_1_, int p_73864_2_, int p_73864_3_) throws IOException
+    protected void mouseClicked(int mouseX, int mouseY, int mouseButton) throws IOException
     {
-        if (p_73864_3_ == 0)
+        if (mouseButton == 0)
         {
-            int i = p_73864_1_ - this.field_147003_i;
-            int j = p_73864_2_ - this.field_147009_r;
+            int i = mouseX - this.guiLeft;
+            int j = mouseY - this.guiTop;
 
-            for (CreativeTabs creativetabs : CreativeTabs.field_78032_a)
+            for (CreativeTabs creativetabs : CreativeTabs.CREATIVE_TAB_ARRAY)
             {
-                if (this.func_147049_a(creativetabs, i, j))
+                if (this.isMouseOverTab(creativetabs, i, j))
                 {
                     return;
                 }
             }
         }
 
-        super.func_73864_a(p_73864_1_, p_73864_2_, p_73864_3_);
+        super.mouseClicked(mouseX, mouseY, mouseButton);
     }
 
-    protected void func_146286_b(int p_146286_1_, int p_146286_2_, int p_146286_3_)
+    protected void mouseReleased(int mouseX, int mouseY, int state)
     {
-        if (p_146286_3_ == 0)
+        if (state == 0)
         {
-            int i = p_146286_1_ - this.field_147003_i;
-            int j = p_146286_2_ - this.field_147009_r;
+            int i = mouseX - this.guiLeft;
+            int j = mouseY - this.guiTop;
 
-            for (CreativeTabs creativetabs : CreativeTabs.field_78032_a)
+            for (CreativeTabs creativetabs : CreativeTabs.CREATIVE_TAB_ARRAY)
             {
-                if (this.func_147049_a(creativetabs, i, j))
+                if (creativetabs != null && this.isMouseOverTab(creativetabs, i, j))
                 {
-                    this.func_147050_b(creativetabs);
+                    this.setCurrentCreativeTab(creativetabs);
                     return;
                 }
             }
         }
 
-        super.func_146286_b(p_146286_1_, p_146286_2_, p_146286_3_);
+        super.mouseReleased(mouseX, mouseY, state);
     }
 
-    private boolean func_147055_p()
+    private boolean needsScrollBars()
     {
-        return field_147058_w != CreativeTabs.field_78036_m.func_78021_a() && CreativeTabs.field_78032_a[field_147058_w].func_78017_i() && ((GuiContainerCreative.ContainerCreative)this.field_147002_h).func_148328_e();
+        if (CreativeTabs.CREATIVE_TAB_ARRAY[selectedTabIndex] == null) return false;
+        return selectedTabIndex != CreativeTabs.INVENTORY.getTabIndex() && CreativeTabs.CREATIVE_TAB_ARRAY[selectedTabIndex].shouldHidePlayerInventory() && ((GuiContainerCreative.ContainerCreative)this.inventorySlots).canScroll();
     }
 
-    private void func_147050_b(CreativeTabs p_147050_1_)
+    private void setCurrentCreativeTab(CreativeTabs tab)
     {
-        int i = field_147058_w;
-        field_147058_w = p_147050_1_.func_78021_a();
-        GuiContainerCreative.ContainerCreative guicontainercreative$containercreative = (GuiContainerCreative.ContainerCreative)this.field_147002_h;
-        this.field_147008_s.clear();
-        guicontainercreative$containercreative.field_148330_a.clear();
-        p_147050_1_.func_78018_a(guicontainercreative$containercreative.field_148330_a);
+        if (tab == null) return;
+        int i = selectedTabIndex;
+        selectedTabIndex = tab.getTabIndex();
+        GuiContainerCreative.ContainerCreative guicontainercreative$containercreative = (GuiContainerCreative.ContainerCreative)this.inventorySlots;
+        this.dragSplittingSlots.clear();
+        guicontainercreative$containercreative.itemList.clear();
+        tab.displayAllRelevantItems(guicontainercreative$containercreative.itemList);
 
-        if (p_147050_1_ == CreativeTabs.field_78036_m)
+        if (tab == CreativeTabs.INVENTORY)
         {
-            Container container = this.field_146297_k.field_71439_g.field_71069_bz;
+            Container container = this.mc.player.inventoryContainer;
 
-            if (this.field_147063_B == null)
+            if (this.originalSlots == null)
             {
-                this.field_147063_B = guicontainercreative$containercreative.field_75151_b;
+                this.originalSlots = guicontainercreative$containercreative.inventorySlots;
             }
 
-            guicontainercreative$containercreative.field_75151_b = Lists.<Slot>newArrayList();
+            guicontainercreative$containercreative.inventorySlots = Lists.<Slot>newArrayList();
 
-            for (int j = 0; j < container.field_75151_b.size(); ++j)
+            for (int j = 0; j < container.inventorySlots.size(); ++j)
             {
-                Slot slot = new GuiContainerCreative.CreativeSlot((Slot)container.field_75151_b.get(j), j);
-                guicontainercreative$containercreative.field_75151_b.add(slot);
+                Slot slot = new GuiContainerCreative.CreativeSlot((Slot)container.inventorySlots.get(j), j);
+                guicontainercreative$containercreative.inventorySlots.add(slot);
 
                 if (j >= 5 && j < 9)
                 {
                     int j1 = j - 5;
                     int k1 = j1 / 2;
                     int l1 = j1 % 2;
-                    slot.field_75223_e = 54 + k1 * 54;
-                    slot.field_75221_f = 6 + l1 * 27;
+                    slot.xPos = 54 + k1 * 54;
+                    slot.yPos = 6 + l1 * 27;
                 }
                 else if (j >= 0 && j < 5)
                 {
-                    slot.field_75223_e = -2000;
-                    slot.field_75221_f = -2000;
+                    slot.xPos = -2000;
+                    slot.yPos = -2000;
                 }
                 else if (j == 45)
                 {
-                    slot.field_75223_e = 35;
-                    slot.field_75221_f = 20;
+                    slot.xPos = 35;
+                    slot.yPos = 20;
                 }
-                else if (j < container.field_75151_b.size())
+                else if (j < container.inventorySlots.size())
                 {
                     int k = j - 9;
                     int l = k % 9;
                     int i1 = k / 9;
-                    slot.field_75223_e = 9 + l * 18;
+                    slot.xPos = 9 + l * 18;
 
                     if (j >= 36)
                     {
-                        slot.field_75221_f = 112;
+                        slot.yPos = 112;
                     }
                     else
                     {
-                        slot.field_75221_f = 54 + i1 * 18;
+                        slot.yPos = 54 + i1 * 18;
                     }
                 }
             }
 
-            this.field_147064_C = new Slot(field_147060_v, 0, 173, 112);
-            guicontainercreative$containercreative.field_75151_b.add(this.field_147064_C);
+            this.destroyItemSlot = new Slot(basicInventory, 0, 173, 112);
+            guicontainercreative$containercreative.inventorySlots.add(this.destroyItemSlot);
         }
-        else if (i == CreativeTabs.field_78036_m.func_78021_a())
+        else if (i == CreativeTabs.INVENTORY.getTabIndex())
         {
-            guicontainercreative$containercreative.field_75151_b = this.field_147063_B;
-            this.field_147063_B = null;
+            guicontainercreative$containercreative.inventorySlots = this.originalSlots;
+            this.originalSlots = null;
         }
 
-        if (this.field_147062_A != null)
+        if (this.searchField != null)
         {
-            if (p_147050_1_ == CreativeTabs.field_78027_g)
+            if (tab.hasSearchBar())
             {
-                this.field_147062_A.func_146189_e(true);
-                this.field_147062_A.func_146205_d(false);
-                this.field_147062_A.func_146195_b(true);
-                this.field_147062_A.func_146180_a("");
-                this.func_147053_i();
+                this.searchField.setVisible(true);
+                this.searchField.setCanLoseFocus(false);
+                this.searchField.setFocused(true);
+                this.searchField.setText("");
+                this.searchField.width = tab.getSearchbarWidth();
+                this.searchField.xPosition = this.guiLeft + (82 /*default left*/ + 89 /*default width*/) - this.searchField.width;
+                this.updateCreativeSearch();
             }
             else
             {
-                this.field_147062_A.func_146189_e(false);
-                this.field_147062_A.func_146205_d(true);
-                this.field_147062_A.func_146195_b(false);
+                this.searchField.setVisible(false);
+                this.searchField.setCanLoseFocus(true);
+                this.searchField.setFocused(false);
             }
         }
 
-        this.field_147067_x = 0.0F;
-        guicontainercreative$containercreative.func_148329_a(0.0F);
+        this.currentScroll = 0.0F;
+        guicontainercreative$containercreative.scrollTo(0.0F);
     }
 
-    public void func_146274_d() throws IOException
+    public void handleMouseInput() throws IOException
     {
-        super.func_146274_d();
+        super.handleMouseInput();
         int i = Mouse.getEventDWheel();
 
-        if (i != 0 && this.func_147055_p())
+        if (i != 0 && this.needsScrollBars())
         {
-            int j = (((GuiContainerCreative.ContainerCreative)this.field_147002_h).field_148330_a.size() + 9 - 1) / 9 - 5;
+            int j = (((GuiContainerCreative.ContainerCreative)this.inventorySlots).itemList.size() + 9 - 1) / 9 - 5;
 
             if (i > 0)
             {
@@ -540,78 +567,101 @@
                 i = -1;
             }
 
-            this.field_147067_x = (float)((double)this.field_147067_x - (double)i / (double)j);
-            this.field_147067_x = MathHelper.func_76131_a(this.field_147067_x, 0.0F, 1.0F);
-            ((GuiContainerCreative.ContainerCreative)this.field_147002_h).func_148329_a(this.field_147067_x);
+            this.currentScroll = (float)((double)this.currentScroll - (double)i / (double)j);
+            this.currentScroll = MathHelper.clamp(this.currentScroll, 0.0F, 1.0F);
+            ((GuiContainerCreative.ContainerCreative)this.inventorySlots).scrollTo(this.currentScroll);
         }
     }
 
-    public void func_73863_a(int p_73863_1_, int p_73863_2_, float p_73863_3_)
+    public void drawScreen(int mouseX, int mouseY, float partialTicks)
     {
         boolean flag = Mouse.isButtonDown(0);
-        int i = this.field_147003_i;
-        int j = this.field_147009_r;
+        int i = this.guiLeft;
+        int j = this.guiTop;
         int k = i + 175;
         int l = j + 18;
         int i1 = k + 14;
         int j1 = l + 112;
 
-        if (!this.field_147065_z && flag && p_73863_1_ >= k && p_73863_2_ >= l && p_73863_1_ < i1 && p_73863_2_ < j1)
+        if (!this.wasClicking && flag && mouseX >= k && mouseY >= l && mouseX < i1 && mouseY < j1)
         {
-            this.field_147066_y = this.func_147055_p();
+            this.isScrolling = this.needsScrollBars();
         }
 
         if (!flag)
         {
-            this.field_147066_y = false;
+            this.isScrolling = false;
         }
 
-        this.field_147065_z = flag;
+        this.wasClicking = flag;
 
-        if (this.field_147066_y)
+        if (this.isScrolling)
         {
-            this.field_147067_x = ((float)(p_73863_2_ - l) - 7.5F) / ((float)(j1 - l) - 15.0F);
-            this.field_147067_x = MathHelper.func_76131_a(this.field_147067_x, 0.0F, 1.0F);
-            ((GuiContainerCreative.ContainerCreative)this.field_147002_h).func_148329_a(this.field_147067_x);
+            this.currentScroll = ((float)(mouseY - l) - 7.5F) / ((float)(j1 - l) - 15.0F);
+            this.currentScroll = MathHelper.clamp(this.currentScroll, 0.0F, 1.0F);
+            ((GuiContainerCreative.ContainerCreative)this.inventorySlots).scrollTo(this.currentScroll);
         }
 
-        super.func_73863_a(p_73863_1_, p_73863_2_, p_73863_3_);
+        super.drawScreen(mouseX, mouseY, partialTicks);
+        int start = tabPage * 10;
+        int end = Math.min(CreativeTabs.CREATIVE_TAB_ARRAY.length, ((tabPage + 1) * 10) + 2);
+        if (tabPage != 0) start += 2;
+        boolean rendered = false;
 
-        for (CreativeTabs creativetabs : CreativeTabs.field_78032_a)
+        for (CreativeTabs creativetabs : java.util.Arrays.copyOfRange(CreativeTabs.CREATIVE_TAB_ARRAY,start,end))
         {
-            if (this.func_147052_b(creativetabs, p_73863_1_, p_73863_2_))
+            if (creativetabs == null) continue;
+            if (this.renderCreativeInventoryHoveringText(creativetabs, mouseX, mouseY))
             {
+                rendered = true;
                 break;
             }
         }
 
-        if (this.field_147064_C != null && field_147058_w == CreativeTabs.field_78036_m.func_78021_a() && this.func_146978_c(this.field_147064_C.field_75223_e, this.field_147064_C.field_75221_f, 16, 16, p_73863_1_, p_73863_2_))
+        if (!rendered && !renderCreativeInventoryHoveringText(CreativeTabs.SEARCH, mouseX, mouseY))
         {
-            this.func_146279_a(I18n.func_135052_a("inventory.binSlot", new Object[0]), p_73863_1_, p_73863_2_);
+            renderCreativeInventoryHoveringText(CreativeTabs.INVENTORY, mouseX, mouseY);
         }
 
-        GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, 1.0F);
-        GlStateManager.func_179140_f();
+        if (this.destroyItemSlot != null && selectedTabIndex == CreativeTabs.INVENTORY.getTabIndex() && this.isPointInRegion(this.destroyItemSlot.xPos, this.destroyItemSlot.yPos, 16, 16, mouseX, mouseY))
+        {
+            this.drawCreativeTabHoveringText(I18n.format("inventory.binSlot", new Object[0]), mouseX, mouseY);
+        }
+
+        if (maxPages != 0)
+        {
+            String page = String.format("%d / %d", tabPage + 1, maxPages + 1);
+            int width = fontRendererObj.getStringWidth(page);
+            GlStateManager.disableLighting();
+            this.zLevel = 300.0F;
+            itemRender.zLevel = 300.0F;
+            fontRendererObj.drawString(page, guiLeft + (xSize / 2) - (width / 2), guiTop - 44, -1);
+            this.zLevel = 0.0F;
+            itemRender.zLevel = 0.0F;
+        }
+
+        GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
+        GlStateManager.disableLighting();
     }
 
-    protected void func_146285_a(ItemStack p_146285_1_, int p_146285_2_, int p_146285_3_)
+    protected void renderToolTip(ItemStack stack, int x, int y)
     {
-        if (field_147058_w == CreativeTabs.field_78027_g.func_78021_a())
+        if (selectedTabIndex == CreativeTabs.SEARCH.getTabIndex())
         {
-            List<String> list = p_146285_1_.func_82840_a(this.field_146297_k.field_71439_g, this.field_146297_k.field_71474_y.field_82882_x);
-            CreativeTabs creativetabs = p_146285_1_.func_77973_b().func_77640_w();
+            List<String> list = stack.getTooltip(this.mc.player, this.mc.gameSettings.advancedItemTooltips);
+            CreativeTabs creativetabs = stack.getItem().getCreativeTab();
 
-            if (creativetabs == null && p_146285_1_.func_77973_b() == Items.field_151134_bR)
+            if (creativetabs == null && stack.getItem() == Items.ENCHANTED_BOOK)
             {
-                Map<Enchantment, Integer> map = EnchantmentHelper.func_82781_a(p_146285_1_);
+                Map<Enchantment, Integer> map = EnchantmentHelper.getEnchantments(stack);
 
                 if (map.size() == 1)
                 {
                     Enchantment enchantment = (Enchantment)map.keySet().iterator().next();
 
-                    for (CreativeTabs creativetabs1 : CreativeTabs.field_78032_a)
+                    for (CreativeTabs creativetabs1 : CreativeTabs.CREATIVE_TAB_ARRAY)
                     {
-                        if (creativetabs1.func_111226_a(enchantment.field_77351_y))
+                        if (creativetabs1.hasRelevantEnchantmentType(enchantment.type))
                         {
                             creativetabs = creativetabs1;
                             break;
@@ -622,14 +672,14 @@
 
             if (creativetabs != null)
             {
-                list.add(1, "" + TextFormatting.BOLD + TextFormatting.BLUE + I18n.func_135052_a(creativetabs.func_78024_c(), new Object[0]));
+                list.add(1, "" + TextFormatting.BOLD + TextFormatting.BLUE + I18n.format(creativetabs.getTranslatedTabLabel(), new Object[0]));
             }
 
             for (int i = 0; i < list.size(); ++i)
             {
                 if (i == 0)
                 {
-                    list.set(i, p_146285_1_.func_77953_t().field_77937_e + (String)list.get(i));
+                    list.set(i, stack.getRarity().rarityColor + (String)list.get(i));
                 }
                 else
                 {
@@ -637,106 +687,143 @@
                 }
             }
 
-            this.func_146283_a(list, p_146285_2_, p_146285_3_);
+            net.minecraftforge.fml.client.config.GuiUtils.preItemToolTip(stack);
+            this.drawHoveringText(list, x, y);
+            net.minecraftforge.fml.client.config.GuiUtils.postItemToolTip();
         }
         else
         {
-            super.func_146285_a(p_146285_1_, p_146285_2_, p_146285_3_);
+            super.renderToolTip(stack, x, y);
         }
     }
 
-    protected void func_146976_a(float p_146976_1_, int p_146976_2_, int p_146976_3_)
+    protected void drawGuiContainerBackgroundLayer(float partialTicks, int mouseX, int mouseY)
     {
-        GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, 1.0F);
-        RenderHelper.func_74520_c();
-        CreativeTabs creativetabs = CreativeTabs.field_78032_a[field_147058_w];
+        GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
+        RenderHelper.enableGUIStandardItemLighting();
+        CreativeTabs creativetabs = CreativeTabs.CREATIVE_TAB_ARRAY[selectedTabIndex];
 
-        for (CreativeTabs creativetabs1 : CreativeTabs.field_78032_a)
+        int start = tabPage * 10;
+        int end = Math.min(CreativeTabs.CREATIVE_TAB_ARRAY.length, ((tabPage + 1) * 10 + 2));
+        if (tabPage != 0) start += 2;
+
+        for (CreativeTabs creativetabs1 : java.util.Arrays.copyOfRange(CreativeTabs.CREATIVE_TAB_ARRAY,start,end))
         {
-            this.field_146297_k.func_110434_K().func_110577_a(field_147061_u);
+            this.mc.getTextureManager().bindTexture(CREATIVE_INVENTORY_TABS);
 
-            if (creativetabs1.func_78021_a() != field_147058_w)
+            if (creativetabs1 == null) continue;
+            if (creativetabs1.getTabIndex() != selectedTabIndex)
             {
-                this.func_147051_a(creativetabs1);
+                this.drawTab(creativetabs1);
             }
         }
 
-        this.field_146297_k.func_110434_K().func_110577_a(new ResourceLocation("textures/gui/container/creative_inventory/tab_" + creativetabs.func_78015_f()));
-        this.func_73729_b(this.field_147003_i, this.field_147009_r, 0, 0, this.field_146999_f, this.field_147000_g);
-        this.field_147062_A.func_146194_f();
-        GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, 1.0F);
-        int i = this.field_147003_i + 175;
-        int j = this.field_147009_r + 18;
+        if (tabPage != 0)
+        {
+            if (creativetabs != CreativeTabs.SEARCH)
+            {
+                this.mc.getTextureManager().bindTexture(CREATIVE_INVENTORY_TABS);
+                drawTab(CreativeTabs.SEARCH);
+            }
+            if (creativetabs != CreativeTabs.INVENTORY)
+            {
+                this.mc.getTextureManager().bindTexture(CREATIVE_INVENTORY_TABS);
+                drawTab(CreativeTabs.INVENTORY);
+            }
+        }
+
+        this.mc.getTextureManager().bindTexture(new ResourceLocation("textures/gui/container/creative_inventory/tab_" + creativetabs.getBackgroundImageName()));
+        this.drawTexturedModalRect(this.guiLeft, this.guiTop, 0, 0, this.xSize, this.ySize);
+        this.searchField.drawTextBox();
+        GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
+        int i = this.guiLeft + 175;
+        int j = this.guiTop + 18;
         int k = j + 112;
-        this.field_146297_k.func_110434_K().func_110577_a(field_147061_u);
+        this.mc.getTextureManager().bindTexture(CREATIVE_INVENTORY_TABS);
 
-        if (creativetabs.func_78017_i())
+        if (creativetabs.shouldHidePlayerInventory())
         {
-            this.func_73729_b(i, j + (int)((float)(k - j - 17) * this.field_147067_x), 232 + (this.func_147055_p() ? 0 : 12), 0, 12, 15);
+            this.drawTexturedModalRect(i, j + (int)((float)(k - j - 17) * this.currentScroll), 232 + (this.needsScrollBars() ? 0 : 12), 0, 12, 15);
         }
 
-        this.func_147051_a(creativetabs);
+        if (creativetabs == null || creativetabs.getTabPage() != tabPage)
+        {
+            if (creativetabs != CreativeTabs.SEARCH && creativetabs != CreativeTabs.INVENTORY)
+            {
+                return;
+            }
+        }
 
-        if (creativetabs == CreativeTabs.field_78036_m)
+        this.drawTab(creativetabs);
+
+        if (creativetabs == CreativeTabs.INVENTORY)
         {
-            GuiInventory.func_147046_a(this.field_147003_i + 88, this.field_147009_r + 45, 20, (float)(this.field_147003_i + 88 - p_146976_2_), (float)(this.field_147009_r + 45 - 30 - p_146976_3_), this.field_146297_k.field_71439_g);
+            GuiInventory.drawEntityOnScreen(this.guiLeft + 88, this.guiTop + 45, 20, (float)(this.guiLeft + 88 - mouseX), (float)(this.guiTop + 45 - 30 - mouseY), this.mc.player);
         }
     }
 
-    protected boolean func_147049_a(CreativeTabs p_147049_1_, int p_147049_2_, int p_147049_3_)
+    protected boolean isMouseOverTab(CreativeTabs tab, int mouseX, int mouseY)
     {
-        int i = p_147049_1_.func_78020_k();
+        if (tab.getTabPage() != tabPage)
+        {
+            if (tab != CreativeTabs.SEARCH && tab != CreativeTabs.INVENTORY)
+            {
+                return false;
+            }
+        }
+
+        int i = tab.getTabColumn();
         int j = 28 * i;
         int k = 0;
 
         if (i == 5)
         {
-            j = this.field_146999_f - 28 + 2;
+            j = this.xSize - 28 + 2;
         }
         else if (i > 0)
         {
             j += i;
         }
 
-        if (p_147049_1_.func_78023_l())
+        if (tab.isTabInFirstRow())
         {
             k = k - 32;
         }
         else
         {
-            k = k + this.field_147000_g;
+            k = k + this.ySize;
         }
 
-        return p_147049_2_ >= j && p_147049_2_ <= j + 28 && p_147049_3_ >= k && p_147049_3_ <= k + 32;
+        return mouseX >= j && mouseX <= j + 28 && mouseY >= k && mouseY <= k + 32;
     }
 
-    protected boolean func_147052_b(CreativeTabs p_147052_1_, int p_147052_2_, int p_147052_3_)
+    protected boolean renderCreativeInventoryHoveringText(CreativeTabs tab, int mouseX, int mouseY)
     {
-        int i = p_147052_1_.func_78020_k();
+        int i = tab.getTabColumn();
         int j = 28 * i;
         int k = 0;
 
         if (i == 5)
         {
-            j = this.field_146999_f - 28 + 2;
+            j = this.xSize - 28 + 2;
         }
         else if (i > 0)
         {
             j += i;
         }
 
-        if (p_147052_1_.func_78023_l())
+        if (tab.isTabInFirstRow())
         {
             k = k - 32;
         }
         else
         {
-            k = k + this.field_147000_g;
+            k = k + this.ySize;
         }
 
-        if (this.func_146978_c(j + 3, k + 3, 23, 27, p_147052_2_, p_147052_3_))
+        if (this.isPointInRegion(j + 3, k + 3, 23, 27, mouseX, mouseY))
         {
-            this.func_146279_a(I18n.func_135052_a(p_147052_1_.func_78024_c(), new Object[0]), p_147052_2_, p_147052_3_);
+            this.drawCreativeTabHoveringText(I18n.format(tab.getTranslatedTabLabel(), new Object[0]), mouseX, mouseY);
             return true;
         }
         else
@@ -745,15 +832,15 @@
         }
     }
 
-    protected void func_147051_a(CreativeTabs p_147051_1_)
+    protected void drawTab(CreativeTabs tab)
     {
-        boolean flag = p_147051_1_.func_78021_a() == field_147058_w;
-        boolean flag1 = p_147051_1_.func_78023_l();
-        int i = p_147051_1_.func_78020_k();
+        boolean flag = tab.getTabIndex() == selectedTabIndex;
+        boolean flag1 = tab.isTabInFirstRow();
+        int i = tab.getTabColumn();
         int j = i * 28;
         int k = 0;
-        int l = this.field_147003_i + 28 * i;
-        int i1 = this.field_147009_r;
+        int l = this.guiLeft + 28 * i;
+        int i1 = this.guiTop;
         int j1 = 32;
 
         if (flag)
@@ -763,7 +850,7 @@
 
         if (i == 5)
         {
-            l = this.field_147003_i + this.field_146999_f - 28;
+            l = this.guiLeft + this.xSize - 28;
         }
         else if (i > 0)
         {
@@ -777,76 +864,87 @@
         else
         {
             k += 64;
-            i1 = i1 + (this.field_147000_g - 4);
+            i1 = i1 + (this.ySize - 4);
         }
 
-        GlStateManager.func_179140_f();
-        this.func_73729_b(l, i1, j, k, 28, 32);
-        this.field_73735_i = 100.0F;
-        this.field_146296_j.field_77023_b = 100.0F;
+        GlStateManager.disableLighting();
+        GlStateManager.color(1F, 1F, 1F); //Forge: Reset color in case Items change it.
+        GlStateManager.enableBlend(); //Forge: Make sure blend is enabled else tabs show a white border.
+        this.drawTexturedModalRect(l, i1, j, k, 28, 32);
+        this.zLevel = 100.0F;
+        this.itemRender.zLevel = 100.0F;
         l = l + 6;
         i1 = i1 + 8 + (flag1 ? 1 : -1);
-        GlStateManager.func_179145_e();
-        GlStateManager.func_179091_B();
-        ItemStack itemstack = p_147051_1_.func_151244_d();
-        this.field_146296_j.func_180450_b(itemstack, l, i1);
-        this.field_146296_j.func_175030_a(this.field_146289_q, itemstack, l, i1);
-        GlStateManager.func_179140_f();
-        this.field_146296_j.field_77023_b = 0.0F;
-        this.field_73735_i = 0.0F;
+        GlStateManager.enableLighting();
+        GlStateManager.enableRescaleNormal();
+        ItemStack itemstack = tab.getIconItemStack();
+        this.itemRender.renderItemAndEffectIntoGUI(itemstack, l, i1);
+        this.itemRender.renderItemOverlays(this.fontRendererObj, itemstack, l, i1);
+        GlStateManager.disableLighting();
+        this.itemRender.zLevel = 0.0F;
+        this.zLevel = 0.0F;
     }
 
-    protected void func_146284_a(GuiButton p_146284_1_) throws IOException
+    protected void actionPerformed(GuiButton button) throws IOException
     {
-        if (p_146284_1_.field_146127_k == 0)
+        if (button.id == 0)
         {
-            this.field_146297_k.func_147108_a(new GuiAchievements(this, this.field_146297_k.field_71439_g.func_146107_m()));
+            this.mc.displayGuiScreen(new GuiAchievements(this, this.mc.player.getStatFileWriter()));
         }
 
-        if (p_146284_1_.field_146127_k == 1)
+        if (button.id == 1)
         {
-            this.field_146297_k.func_147108_a(new GuiStats(this, this.field_146297_k.field_71439_g.func_146107_m()));
+            this.mc.displayGuiScreen(new GuiStats(this, this.mc.player.getStatFileWriter()));
         }
+
+        if (button.id == 101)
+        {
+            tabPage = Math.max(tabPage - 1, 0);
+        }
+        else if (button.id == 102)
+        {
+            tabPage = Math.min(tabPage + 1, maxPages);
+        }
     }
 
-    public int func_147056_g()
+    public int getSelectedTabIndex()
     {
-        return field_147058_w;
+        return selectedTabIndex;
     }
 
     @SideOnly(Side.CLIENT)
     static class ContainerCreative extends Container
         {
-            public NonNullList<ItemStack> field_148330_a = NonNullList.<ItemStack>func_191196_a();
+            public NonNullList<ItemStack> itemList = NonNullList.<ItemStack>create();
 
-            public ContainerCreative(EntityPlayer p_i1086_1_)
+            public ContainerCreative(EntityPlayer player)
             {
-                InventoryPlayer inventoryplayer = p_i1086_1_.field_71071_by;
+                InventoryPlayer inventoryplayer = player.inventory;
 
                 for (int i = 0; i < 5; ++i)
                 {
                     for (int j = 0; j < 9; ++j)
                     {
-                        this.func_75146_a(new Slot(GuiContainerCreative.field_147060_v, i * 9 + j, 9 + j * 18, 18 + i * 18));
+                        this.addSlotToContainer(new Slot(GuiContainerCreative.basicInventory, i * 9 + j, 9 + j * 18, 18 + i * 18));
                     }
                 }
 
                 for (int k = 0; k < 9; ++k)
                 {
-                    this.func_75146_a(new Slot(inventoryplayer, k, 9 + k * 18, 112));
+                    this.addSlotToContainer(new Slot(inventoryplayer, k, 9 + k * 18, 112));
                 }
 
-                this.func_148329_a(0.0F);
+                this.scrollTo(0.0F);
             }
 
-            public boolean func_75145_c(EntityPlayer p_75145_1_)
+            public boolean canInteractWith(EntityPlayer playerIn)
             {
                 return true;
             }
 
-            public void func_148329_a(float p_148329_1_)
+            public void scrollTo(float p_148329_1_)
             {
-                int i = (this.field_148330_a.size() + 9 - 1) / 9 - 5;
+                int i = (this.itemList.size() + 9 - 1) / 9 - 5;
                 int j = (int)((double)(p_148329_1_ * (float)i) + 0.5D);
 
                 if (j < 0)
@@ -860,129 +958,166 @@
                     {
                         int i1 = l + (k + j) * 9;
 
-                        if (i1 >= 0 && i1 < this.field_148330_a.size())
+                        if (i1 >= 0 && i1 < this.itemList.size())
                         {
-                            GuiContainerCreative.field_147060_v.func_70299_a(l + k * 9, (ItemStack)this.field_148330_a.get(i1));
+                            GuiContainerCreative.basicInventory.setInventorySlotContents(l + k * 9, (ItemStack)this.itemList.get(i1));
                         }
                         else
                         {
-                            GuiContainerCreative.field_147060_v.func_70299_a(l + k * 9, ItemStack.field_190927_a);
+                            GuiContainerCreative.basicInventory.setInventorySlotContents(l + k * 9, ItemStack.EMPTY);
                         }
                     }
                 }
             }
 
-            public boolean func_148328_e()
+            public boolean canScroll()
             {
-                return this.field_148330_a.size() > 45;
+                return this.itemList.size() > 45;
             }
 
-            protected void func_75133_b(int p_75133_1_, int p_75133_2_, boolean p_75133_3_, EntityPlayer p_75133_4_)
+            protected void retrySlotClick(int slotId, int clickedButton, boolean mode, EntityPlayer playerIn)
             {
             }
 
-            public ItemStack func_82846_b(EntityPlayer p_82846_1_, int p_82846_2_)
+            @Override
+            public InventoryView getBukkitView() {
+                return null;
+            }
+
+            public ItemStack transferStackInSlot(EntityPlayer playerIn, int index)
             {
-                if (p_82846_2_ >= this.field_75151_b.size() - 9 && p_82846_2_ < this.field_75151_b.size())
+                if (index >= this.inventorySlots.size() - 9 && index < this.inventorySlots.size())
                 {
-                    Slot slot = (Slot)this.field_75151_b.get(p_82846_2_);
+                    Slot slot = (Slot)this.inventorySlots.get(index);
 
-                    if (slot != null && slot.func_75216_d())
+                    if (slot != null && slot.getHasStack())
                     {
-                        slot.func_75215_d(ItemStack.field_190927_a);
+                        slot.putStack(ItemStack.EMPTY);
                     }
                 }
 
-                return ItemStack.field_190927_a;
+                return ItemStack.EMPTY;
             }
 
-            public boolean func_94530_a(ItemStack p_94530_1_, Slot p_94530_2_)
+            public boolean canMergeSlot(ItemStack stack, Slot slotIn)
             {
-                return p_94530_2_.field_75221_f > 90;
+                return slotIn.yPos > 90;
             }
 
-            public boolean func_94531_b(Slot p_94531_1_)
+            public boolean canDragIntoSlot(Slot slotIn)
             {
-                return p_94531_1_.field_75224_c instanceof InventoryPlayer || p_94531_1_.field_75221_f > 90 && p_94531_1_.field_75223_e <= 162;
+                return slotIn.inventory instanceof InventoryPlayer || slotIn.yPos > 90 && slotIn.xPos <= 162;
             }
         }
 
     @SideOnly(Side.CLIENT)
     class CreativeSlot extends Slot
     {
-        private final Slot field_148332_b;
+        private final Slot slot;
 
         public CreativeSlot(Slot p_i46313_2_, int p_i46313_3_)
         {
-            super(p_i46313_2_.field_75224_c, p_i46313_3_, 0, 0);
-            this.field_148332_b = p_i46313_2_;
+            super(p_i46313_2_.inventory, p_i46313_3_, 0, 0);
+            this.slot = p_i46313_2_;
         }
 
-        public ItemStack func_190901_a(EntityPlayer p_190901_1_, ItemStack p_190901_2_)
+        public ItemStack onTake(EntityPlayer thePlayer, ItemStack stack)
         {
-            this.field_148332_b.func_190901_a(p_190901_1_, p_190901_2_);
-            return p_190901_2_;
+            this.slot.onTake(thePlayer, stack);
+            return stack;
         }
 
-        public boolean func_75214_a(ItemStack p_75214_1_)
+        public boolean isItemValid(ItemStack stack)
         {
-            return this.field_148332_b.func_75214_a(p_75214_1_);
+            return this.slot.isItemValid(stack);
         }
 
-        public ItemStack func_75211_c()
+        public ItemStack getStack()
         {
-            return this.field_148332_b.func_75211_c();
+            return this.slot.getStack();
         }
 
-        public boolean func_75216_d()
+        public boolean getHasStack()
         {
-            return this.field_148332_b.func_75216_d();
+            return this.slot.getHasStack();
         }
 
-        public void func_75215_d(ItemStack p_75215_1_)
+        public void putStack(ItemStack stack)
         {
-            this.field_148332_b.func_75215_d(p_75215_1_);
+            this.slot.putStack(stack);
         }
 
-        public void func_75218_e()
+        public void onSlotChanged()
         {
-            this.field_148332_b.func_75218_e();
+            this.slot.onSlotChanged();
         }
 
-        public int func_75219_a()
+        public int getSlotStackLimit()
         {
-            return this.field_148332_b.func_75219_a();
+            return this.slot.getSlotStackLimit();
         }
 
-        public int func_178170_b(ItemStack p_178170_1_)
+        public int getItemStackLimit(ItemStack stack)
         {
-            return this.field_148332_b.func_178170_b(p_178170_1_);
+            return this.slot.getItemStackLimit(stack);
         }
 
         @Nullable
-        public String func_178171_c()
+        public String getSlotTexture()
         {
-            return this.field_148332_b.func_178171_c();
+            return this.slot.getSlotTexture();
         }
 
-        public ItemStack func_75209_a(int p_75209_1_)
+        public ItemStack decrStackSize(int amount)
         {
-            return this.field_148332_b.func_75209_a(p_75209_1_);
+            return this.slot.decrStackSize(amount);
         }
 
-        public boolean func_75217_a(IInventory p_75217_1_, int p_75217_2_)
+        public boolean isHere(IInventory inv, int slotIn)
         {
-            return this.field_148332_b.func_75217_a(p_75217_1_, p_75217_2_);
+            return this.slot.isHere(inv, slotIn);
         }
 
-        public boolean func_111238_b()
+        public boolean canBeHovered()
         {
-            return this.field_148332_b.func_111238_b();
+            return this.slot.canBeHovered();
         }
 
-        public boolean func_82869_a(EntityPlayer p_82869_1_)
+        public boolean canTakeStack(EntityPlayer playerIn)
         {
-            return this.field_148332_b.func_82869_a(p_82869_1_);
+            return this.slot.canTakeStack(playerIn);
         }
+
+        /*========================================= FORGE START =====================================*/
+        public net.minecraft.util.ResourceLocation getBackgroundLocation()
+        {
+            return this.slot.getBackgroundLocation();
+        }
+
+        public void setBackgroundLocation(net.minecraft.util.ResourceLocation texture)
+        {
+            this.slot.setBackgroundLocation(texture);
+        }
+
+        public void setBackgroundName(String name)
+        {
+            this.slot.setBackgroundName(name);
+        }
+
+        public net.minecraft.client.renderer.texture.TextureAtlasSprite getBackgroundSprite()
+        {
+            return this.slot.getBackgroundSprite();
+        }
+
+        public int getSlotIndex()
+        {
+            return this.slot.getSlotIndex();
+        }
+
+        public boolean isSameInventory(Slot other)
+        {
+            return this.slot.isSameInventory(other);
+        }
+        /*========================================= FORGE END =====================================*/
     }
 }
