--- ../src-base/minecraft/net/minecraft/entity/monster/AbstractSkeleton.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/AbstractSkeleton.java
@@ -44,245 +44,263 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.event.entity.EntityCombustEvent;
 
 public abstract class AbstractSkeleton extends EntityMob implements IRangedAttackMob
 {
-    private static final DataParameter<Boolean> field_184728_b = EntityDataManager.<Boolean>func_187226_a(AbstractSkeleton.class, DataSerializers.field_187198_h);
-    private final EntityAIAttackRangedBow field_85037_d = new EntityAIAttackRangedBow(this, 1.0D, 20, 15.0F);
-    private final EntityAIAttackMelee field_85038_e = new EntityAIAttackMelee(this, 1.2D, false)
+    private static final DataParameter<Boolean> SWINGING_ARMS = EntityDataManager.<Boolean>createKey(AbstractSkeleton.class, DataSerializers.BOOLEAN);
+    private final EntityAIAttackRangedBow aiArrowAttack = new EntityAIAttackRangedBow(this, 1.0D, 20, 15.0F);
+    private final EntityAIAttackMelee aiAttackOnCollide = new EntityAIAttackMelee(this, 1.2D, false)
     {
-        public void func_75251_c()
+        public void resetTask()
         {
-            super.func_75251_c();
-            AbstractSkeleton.this.func_184724_a(false);
+            super.resetTask();
+            AbstractSkeleton.this.setSwingingArms(false);
         }
-        public void func_75249_e()
+        public void startExecuting()
         {
-            super.func_75249_e();
-            AbstractSkeleton.this.func_184724_a(true);
+            super.startExecuting();
+            AbstractSkeleton.this.setSwingingArms(true);
         }
     };
 
-    public AbstractSkeleton(World p_i47289_1_)
+    public AbstractSkeleton(World worldIn)
     {
-        super(p_i47289_1_);
-        this.func_70105_a(0.6F, 1.99F);
-        this.func_85036_m();
+        super(worldIn);
+        this.setSize(0.6F, 1.99F);
+        this.setCombatTask();
     }
 
-    protected void func_184651_r()
+    protected void initEntityAI()
     {
-        this.field_70714_bg.func_75776_a(1, new EntityAISwimming(this));
-        this.field_70714_bg.func_75776_a(2, new EntityAIRestrictSun(this));
-        this.field_70714_bg.func_75776_a(3, new EntityAIFleeSun(this, 1.0D));
-        this.field_70714_bg.func_75776_a(3, new EntityAIAvoidEntity(this, EntityWolf.class, 6.0F, 1.0D, 1.2D));
-        this.field_70714_bg.func_75776_a(5, new EntityAIWanderAvoidWater(this, 1.0D));
-        this.field_70714_bg.func_75776_a(6, new EntityAIWatchClosest(this, EntityPlayer.class, 8.0F));
-        this.field_70714_bg.func_75776_a(6, new EntityAILookIdle(this));
-        this.field_70715_bh.func_75776_a(1, new EntityAIHurtByTarget(this, false, new Class[0]));
-        this.field_70715_bh.func_75776_a(2, new EntityAINearestAttackableTarget(this, EntityPlayer.class, true));
-        this.field_70715_bh.func_75776_a(3, new EntityAINearestAttackableTarget(this, EntityIronGolem.class, true));
+        this.tasks.addTask(1, new EntityAISwimming(this));
+        this.tasks.addTask(2, new EntityAIRestrictSun(this));
+        this.tasks.addTask(3, new EntityAIFleeSun(this, 1.0D));
+        this.tasks.addTask(3, new EntityAIAvoidEntity(this, EntityWolf.class, 6.0F, 1.0D, 1.2D));
+        this.tasks.addTask(5, new EntityAIWanderAvoidWater(this, 1.0D));
+        this.tasks.addTask(6, new EntityAIWatchClosest(this, EntityPlayer.class, 8.0F));
+        this.tasks.addTask(6, new EntityAILookIdle(this));
+        this.targetTasks.addTask(1, new EntityAIHurtByTarget(this, false, new Class[0]));
+        this.targetTasks.addTask(2, new EntityAINearestAttackableTarget(this, EntityPlayer.class, true));
+        this.targetTasks.addTask(3, new EntityAINearestAttackableTarget(this, EntityIronGolem.class, true));
     }
 
-    protected void func_110147_ax()
+    protected void applyEntityAttributes()
     {
-        super.func_110147_ax();
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.25D);
+        super.applyEntityAttributes();
+        this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).setBaseValue(0.25D);
     }
 
-    protected void func_70088_a()
+    protected void entityInit()
     {
-        super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184728_b, Boolean.valueOf(false));
+        super.entityInit();
+        this.dataManager.register(SWINGING_ARMS, Boolean.valueOf(false));
     }
 
-    protected void func_180429_a(BlockPos p_180429_1_, Block p_180429_2_)
+    protected void playStepSound(BlockPos pos, Block blockIn)
     {
-        this.func_184185_a(this.func_190727_o(), 0.15F, 1.0F);
+        this.playSound(this.getStepSound(), 0.15F, 1.0F);
     }
 
-    protected abstract SoundEvent func_190727_o();
+    protected abstract SoundEvent getStepSound();
 
-    public EnumCreatureAttribute func_70668_bt()
+    public EnumCreatureAttribute getCreatureAttribute()
     {
         return EnumCreatureAttribute.UNDEAD;
     }
 
-    public void func_70636_d()
+    public void onLivingUpdate()
     {
-        if (this.field_70170_p.func_72935_r() && !this.field_70170_p.field_72995_K)
+        if (this.world.isDaytime() && !this.world.isRemote)
         {
-            float f = this.func_70013_c(1.0F);
-            BlockPos blockpos = this.func_184187_bx() instanceof EntityBoat ? (new BlockPos(this.field_70165_t, (double)Math.round(this.field_70163_u), this.field_70161_v)).func_177984_a() : new BlockPos(this.field_70165_t, (double)Math.round(this.field_70163_u), this.field_70161_v);
+            float f = this.getBrightness(1.0F);
+            BlockPos blockpos = this.getRidingEntity() instanceof EntityBoat ? (new BlockPos(this.posX, (double)Math.round(this.posY), this.posZ)).up() : new BlockPos(this.posX, (double)Math.round(this.posY), this.posZ);
 
-            if (f > 0.5F && this.field_70146_Z.nextFloat() * 30.0F < (f - 0.4F) * 2.0F && this.field_70170_p.func_175678_i(blockpos))
+            if (f > 0.5F && this.rand.nextFloat() * 30.0F < (f - 0.4F) * 2.0F && this.world.canSeeSky(blockpos))
             {
                 boolean flag = true;
-                ItemStack itemstack = this.func_184582_a(EntityEquipmentSlot.HEAD);
+                ItemStack itemstack = this.getItemStackFromSlot(EntityEquipmentSlot.HEAD);
 
-                if (!itemstack.func_190926_b())
+                if (!itemstack.isEmpty())
                 {
-                    if (itemstack.func_77984_f())
+                    if (itemstack.isItemStackDamageable())
                     {
-                        itemstack.func_77964_b(itemstack.func_77952_i() + this.field_70146_Z.nextInt(2));
+                        itemstack.setItemDamage(itemstack.getItemDamage() + this.rand.nextInt(2));
 
-                        if (itemstack.func_77952_i() >= itemstack.func_77958_k())
+                        if (itemstack.getItemDamage() >= itemstack.getMaxDamage())
                         {
-                            this.func_70669_a(itemstack);
-                            this.func_184201_a(EntityEquipmentSlot.HEAD, ItemStack.field_190927_a);
+                            this.renderBrokenItemStack(itemstack);
+                            this.setItemStackToSlot(EntityEquipmentSlot.HEAD, ItemStack.EMPTY);
                         }
                     }
 
                     flag = false;
                 }
 
-                if (flag)
-                {
-                    this.func_70015_d(8);
+                if (flag) {
+                    // CraftBukkit start
+                    EntityCombustEvent event = new EntityCombustEvent(this.getBukkitEntity(), 8);
+                    this.world.getServer().getPluginManager().callEvent(event);
+
+                    if (!event.isCancelled()) {
+                        this.setFire(event.getDuration());
+                    }
+                    //CraftBukkit end
                 }
             }
         }
 
-        super.func_70636_d();
+        super.onLivingUpdate();
     }
 
-    public void func_70098_U()
+    public void updateRidden()
     {
-        super.func_70098_U();
+        super.updateRidden();
 
-        if (this.func_184187_bx() instanceof EntityCreature)
+        if (this.getRidingEntity() instanceof EntityCreature)
         {
-            EntityCreature entitycreature = (EntityCreature)this.func_184187_bx();
-            this.field_70761_aq = entitycreature.field_70761_aq;
+            EntityCreature entitycreature = (EntityCreature)this.getRidingEntity();
+            this.renderYawOffset = entitycreature.renderYawOffset;
         }
     }
 
-    public void func_70645_a(DamageSource p_70645_1_)
+    public void onDeath(DamageSource cause)
     {
-        super.func_70645_a(p_70645_1_);
+        super.onDeath(cause);
 
-        if (p_70645_1_.func_76364_f() instanceof EntityArrow && p_70645_1_.func_76346_g() instanceof EntityPlayer)
+        if (cause.getSourceOfDamage() instanceof EntityArrow && cause.getEntity() instanceof EntityPlayer)
         {
-            EntityPlayer entityplayer = (EntityPlayer)p_70645_1_.func_76346_g();
-            double d0 = entityplayer.field_70165_t - this.field_70165_t;
-            double d1 = entityplayer.field_70161_v - this.field_70161_v;
+            EntityPlayer entityplayer = (EntityPlayer)cause.getEntity();
+            double d0 = entityplayer.posX - this.posX;
+            double d1 = entityplayer.posZ - this.posZ;
 
             if (d0 * d0 + d1 * d1 >= 2500.0D)
             {
-                entityplayer.func_71029_a(AchievementList.field_187994_v);
+                entityplayer.addStat(AchievementList.SNIPE_SKELETON);
             }
         }
     }
 
-    protected void func_180481_a(DifficultyInstance p_180481_1_)
+    protected void setEquipmentBasedOnDifficulty(DifficultyInstance difficulty)
     {
-        super.func_180481_a(p_180481_1_);
-        this.func_184201_a(EntityEquipmentSlot.MAINHAND, new ItemStack(Items.field_151031_f));
+        super.setEquipmentBasedOnDifficulty(difficulty);
+        this.setItemStackToSlot(EntityEquipmentSlot.MAINHAND, new ItemStack(Items.BOW));
     }
 
     @Nullable
-    public IEntityLivingData func_180482_a(DifficultyInstance p_180482_1_, @Nullable IEntityLivingData p_180482_2_)
+    public IEntityLivingData onInitialSpawn(DifficultyInstance difficulty, @Nullable IEntityLivingData livingdata)
     {
-        p_180482_2_ = super.func_180482_a(p_180482_1_, p_180482_2_);
-        this.func_180481_a(p_180482_1_);
-        this.func_180483_b(p_180482_1_);
-        this.func_85036_m();
-        this.func_98053_h(this.field_70146_Z.nextFloat() < 0.55F * p_180482_1_.func_180170_c());
+        livingdata = super.onInitialSpawn(difficulty, livingdata);
+        this.setEquipmentBasedOnDifficulty(difficulty);
+        this.setEnchantmentBasedOnDifficulty(difficulty);
+        this.setCombatTask();
+        this.setCanPickUpLoot(this.rand.nextFloat() < 0.55F * difficulty.getClampedAdditionalDifficulty());
 
-        if (this.func_184582_a(EntityEquipmentSlot.HEAD).func_190926_b())
+        if (this.getItemStackFromSlot(EntityEquipmentSlot.HEAD).isEmpty())
         {
-            Calendar calendar = this.field_70170_p.func_83015_S();
+            Calendar calendar = this.world.getCurrentDate();
 
-            if (calendar.get(2) + 1 == 10 && calendar.get(5) == 31 && this.field_70146_Z.nextFloat() < 0.25F)
+            if (calendar.get(2) + 1 == 10 && calendar.get(5) == 31 && this.rand.nextFloat() < 0.25F)
             {
-                this.func_184201_a(EntityEquipmentSlot.HEAD, new ItemStack(this.field_70146_Z.nextFloat() < 0.1F ? Blocks.field_150428_aP : Blocks.field_150423_aK));
-                this.field_184655_bs[EntityEquipmentSlot.HEAD.func_188454_b()] = 0.0F;
+                this.setItemStackToSlot(EntityEquipmentSlot.HEAD, new ItemStack(this.rand.nextFloat() < 0.1F ? Blocks.LIT_PUMPKIN : Blocks.PUMPKIN));
+                this.inventoryArmorDropChances[EntityEquipmentSlot.HEAD.getIndex()] = 0.0F;
             }
         }
 
-        return p_180482_2_;
+        return livingdata;
     }
 
-    public void func_85036_m()
+    public void setCombatTask()
     {
-        if (this.field_70170_p != null && !this.field_70170_p.field_72995_K)
+        if (this.world != null && !this.world.isRemote)
         {
-            this.field_70714_bg.func_85156_a(this.field_85038_e);
-            this.field_70714_bg.func_85156_a(this.field_85037_d);
-            ItemStack itemstack = this.func_184614_ca();
+            this.tasks.removeTask(this.aiAttackOnCollide);
+            this.tasks.removeTask(this.aiArrowAttack);
+            ItemStack itemstack = this.getHeldItemMainhand();
 
-            if (itemstack.func_77973_b() == Items.field_151031_f)
+            if (itemstack.getItem() == Items.BOW)
             {
                 int i = 20;
 
-                if (this.field_70170_p.func_175659_aa() != EnumDifficulty.HARD)
+                if (this.world.getDifficulty() != EnumDifficulty.HARD)
                 {
                     i = 40;
                 }
 
-                this.field_85037_d.func_189428_b(i);
-                this.field_70714_bg.func_75776_a(4, this.field_85037_d);
+                this.aiArrowAttack.setAttackCooldown(i);
+                this.tasks.addTask(4, this.aiArrowAttack);
             }
             else
             {
-                this.field_70714_bg.func_75776_a(4, this.field_85038_e);
+                this.tasks.addTask(4, this.aiAttackOnCollide);
             }
         }
     }
 
-    public void func_82196_d(EntityLivingBase p_82196_1_, float p_82196_2_)
+    public void attackEntityWithRangedAttack(EntityLivingBase target, float distanceFactor)
     {
-        EntityArrow entityarrow = this.func_190726_a(p_82196_2_);
-        double d0 = p_82196_1_.field_70165_t - this.field_70165_t;
-        double d1 = p_82196_1_.func_174813_aQ().field_72338_b + (double)(p_82196_1_.field_70131_O / 3.0F) - entityarrow.field_70163_u;
-        double d2 = p_82196_1_.field_70161_v - this.field_70161_v;
-        double d3 = (double)MathHelper.func_76133_a(d0 * d0 + d2 * d2);
-        entityarrow.func_70186_c(d0, d1 + d3 * 0.20000000298023224D, d2, 1.6F, (float)(14 - this.field_70170_p.func_175659_aa().func_151525_a() * 4));
-        this.func_184185_a(SoundEvents.field_187866_fi, 1.0F, 1.0F / (this.func_70681_au().nextFloat() * 0.4F + 0.8F));
-        this.field_70170_p.func_72838_d(entityarrow);
+        EntityArrow entityarrow = this.getArrow(distanceFactor);
+        double d0 = target.posX - this.posX;
+        double d1 = target.getEntityBoundingBox().minY + (double)(target.height / 3.0F) - entityarrow.posY;
+        double d2 = target.posZ - this.posZ;
+        double d3 = (double)MathHelper.sqrt(d0 * d0 + d2 * d2);
+        entityarrow.setThrowableHeading(d0, d1 + d3 * 0.20000000298023224D, d2, 1.6F, (float)(14 - this.world.getDifficulty().getDifficultyId() * 4));
+        // CraftBukkit start
+        org.bukkit.event.entity.EntityShootBowEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityShootBowEvent(this, this.getHeldItemMainhand(), entityarrow, 0.8F);
+        if (event.isCancelled()) {
+            event.getProjectile().remove();
+            return;
+        }
+
+        if (event.getProjectile() == entityarrow.getBukkitEntity()) {
+            world.spawnEntity(entityarrow);
+        }
+        // CraftBukkit end
+        this.playSound(SoundEvents.ENTITY_SKELETON_SHOOT, 1.0F, 1.0F / (this.getRNG().nextFloat() * 0.4F + 0.8F));
+        //this.world.spawnEntity(entityarrow);
     }
 
-    protected EntityArrow func_190726_a(float p_190726_1_)
+    protected EntityArrow getArrow(float p_190726_1_)
     {
-        EntityTippedArrow entitytippedarrow = new EntityTippedArrow(this.field_70170_p, this);
-        entitytippedarrow.func_190547_a(this, p_190726_1_);
+        EntityTippedArrow entitytippedarrow = new EntityTippedArrow(this.world, this);
+        entitytippedarrow.setEnchantmentEffectsFromEntity(this, p_190726_1_);
         return entitytippedarrow;
     }
 
-    public void func_70037_a(NBTTagCompound p_70037_1_)
+    public void readEntityFromNBT(NBTTagCompound compound)
     {
-        super.func_70037_a(p_70037_1_);
-        this.func_85036_m();
+        super.readEntityFromNBT(compound);
+        this.setCombatTask();
     }
 
-    public void func_184201_a(EntityEquipmentSlot p_184201_1_, ItemStack p_184201_2_)
+    public void setItemStackToSlot(EntityEquipmentSlot slotIn, ItemStack stack)
     {
-        super.func_184201_a(p_184201_1_, p_184201_2_);
+        super.setItemStackToSlot(slotIn, stack);
 
-        if (!this.field_70170_p.field_72995_K && p_184201_1_ == EntityEquipmentSlot.MAINHAND)
+        if (!this.world.isRemote && slotIn == EntityEquipmentSlot.MAINHAND)
         {
-            this.func_85036_m();
+            this.setCombatTask();
         }
     }
 
-    public float func_70047_e()
+    public float getEyeHeight()
     {
         return 1.74F;
     }
 
-    public double func_70033_W()
+    public double getYOffset()
     {
         return -0.6D;
     }
 
     @SideOnly(Side.CLIENT)
-    public boolean func_184725_db()
+    public boolean isSwingingArms()
     {
-        return ((Boolean)this.field_70180_af.func_187225_a(field_184728_b)).booleanValue();
+        return ((Boolean)this.dataManager.get(SWINGING_ARMS)).booleanValue();
     }
 
-    public void func_184724_a(boolean p_184724_1_)
+    public void setSwingingArms(boolean swingingArms)
     {
-        this.field_70180_af.func_187227_b(field_184728_b, Boolean.valueOf(p_184724_1_));
+        this.dataManager.set(SWINGING_ARMS, Boolean.valueOf(swingingArms));
     }
 }
