--- ../src-base/minecraft/net/minecraft/entity/monster/EntityZombie.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntityZombie.java
@@ -51,162 +51,165 @@
 import net.minecraft.world.storage.loot.LootTableList;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
+import org.bukkit.event.entity.EntityCombustEvent;
 
 public class EntityZombie extends EntityMob
 {
-    protected static final IAttribute field_110186_bp = (new RangedAttribute((IAttribute)null, "zombie.spawnReinforcements", 0.0D, 0.0D, 1.0D)).func_111117_a("Spawn Reinforcements Chance");
-    private static final UUID field_110187_bq = UUID.fromString("B9766B59-9566-4402-BC1F-2EE2A276D836");
-    private static final AttributeModifier field_110188_br = new AttributeModifier(field_110187_bq, "Baby speed boost", 0.5D, 1);
-    private static final DataParameter<Boolean> field_184737_bv = EntityDataManager.<Boolean>func_187226_a(EntityZombie.class, DataSerializers.field_187198_h);
-    private static final DataParameter<Integer> field_184738_bw = EntityDataManager.<Integer>func_187226_a(EntityZombie.class, DataSerializers.field_187192_b);
-    private static final DataParameter<Boolean> field_184740_by = EntityDataManager.<Boolean>func_187226_a(EntityZombie.class, DataSerializers.field_187198_h);
-    private final EntityAIBreakDoor field_146075_bs = new EntityAIBreakDoor(this);
-    private boolean field_146076_bu;
-    private float field_146074_bv = -1.0F;
-    private float field_146073_bw;
+    protected static final IAttribute SPAWN_REINFORCEMENTS_CHANCE = (new RangedAttribute((IAttribute)null, "zombie.spawnReinforcements", 0.0D, 0.0D, 1.0D)).setDescription("Spawn Reinforcements Chance");
+    private static final UUID BABY_SPEED_BOOST_ID = UUID.fromString("B9766B59-9566-4402-BC1F-2EE2A276D836");
+    private static final AttributeModifier BABY_SPEED_BOOST = new AttributeModifier(BABY_SPEED_BOOST_ID, "Baby speed boost", 0.5D, 1);
+    private static final DataParameter<Boolean> IS_CHILD = EntityDataManager.<Boolean>createKey(EntityZombie.class, DataSerializers.BOOLEAN);
+    private static final DataParameter<Integer> VILLAGER_TYPE = EntityDataManager.<Integer>createKey(EntityZombie.class, DataSerializers.VARINT);
+    private static final DataParameter<Boolean> ARMS_RAISED = EntityDataManager.<Boolean>createKey(EntityZombie.class, DataSerializers.BOOLEAN);
+    private final EntityAIBreakDoor breakDoor = new EntityAIBreakDoor(this);
+    private boolean isBreakDoorsTaskSet;
+    private float zombieWidth = -1.0F;
+    private float zombieHeight;
 
-    public EntityZombie(World p_i1745_1_)
+    public EntityZombie(World worldIn)
     {
-        super(p_i1745_1_);
-        this.func_70105_a(0.6F, 1.95F);
+        super(worldIn);
+        this.setSize(0.6F, 1.95F);
     }
 
-    protected void func_184651_r()
+    protected void initEntityAI()
     {
-        this.field_70714_bg.func_75776_a(0, new EntityAISwimming(this));
-        this.field_70714_bg.func_75776_a(2, new EntityAIZombieAttack(this, 1.0D, false));
-        this.field_70714_bg.func_75776_a(5, new EntityAIMoveTowardsRestriction(this, 1.0D));
-        this.field_70714_bg.func_75776_a(7, new EntityAIWanderAvoidWater(this, 1.0D));
-        this.field_70714_bg.func_75776_a(8, new EntityAIWatchClosest(this, EntityPlayer.class, 8.0F));
-        this.field_70714_bg.func_75776_a(8, new EntityAILookIdle(this));
-        this.func_175456_n();
+        this.tasks.addTask(0, new EntityAISwimming(this));
+        this.tasks.addTask(2, new EntityAIZombieAttack(this, 1.0D, false));
+        this.tasks.addTask(5, new EntityAIMoveTowardsRestriction(this, 1.0D));
+        this.tasks.addTask(7, new EntityAIWanderAvoidWater(this, 1.0D));
+        this.tasks.addTask(8, new EntityAIWatchClosest(this, EntityPlayer.class, 8.0F));
+        this.tasks.addTask(8, new EntityAILookIdle(this));
+        this.applyEntityAI();
     }
 
-    protected void func_175456_n()
+    protected void applyEntityAI()
     {
-        this.field_70714_bg.func_75776_a(6, new EntityAIMoveThroughVillage(this, 1.0D, false));
-        this.field_70715_bh.func_75776_a(1, new EntityAIHurtByTarget(this, true, new Class[] {EntityPigZombie.class}));
-        this.field_70715_bh.func_75776_a(2, new EntityAINearestAttackableTarget(this, EntityPlayer.class, true));
-        this.field_70715_bh.func_75776_a(3, new EntityAINearestAttackableTarget(this, EntityVillager.class, false));
-        this.field_70715_bh.func_75776_a(3, new EntityAINearestAttackableTarget(this, EntityIronGolem.class, true));
+        this.tasks.addTask(6, new EntityAIMoveThroughVillage(this, 1.0D, false));
+        this.targetTasks.addTask(1, new EntityAIHurtByTarget(this, true, new Class[] {EntityPigZombie.class}));
+        this.targetTasks.addTask(2, new EntityAINearestAttackableTarget(this, EntityPlayer.class, true));
+        this.targetTasks.addTask(3, new EntityAINearestAttackableTarget(this, EntityVillager.class, false));
+        this.targetTasks.addTask(3, new EntityAINearestAttackableTarget(this, EntityIronGolem.class, true));
     }
 
-    protected void func_110147_ax()
+    protected void applyEntityAttributes()
     {
-        super.func_110147_ax();
-        this.func_110148_a(SharedMonsterAttributes.field_111265_b).func_111128_a(35.0D);
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.23000000417232513D);
-        this.func_110148_a(SharedMonsterAttributes.field_111264_e).func_111128_a(3.0D);
-        this.func_110148_a(SharedMonsterAttributes.field_188791_g).func_111128_a(2.0D);
-        this.func_110140_aT().func_111150_b(field_110186_bp).func_111128_a(this.field_70146_Z.nextDouble() * 0.10000000149011612D);
+        super.applyEntityAttributes();
+        this.getEntityAttribute(SharedMonsterAttributes.FOLLOW_RANGE).setBaseValue(35.0D);
+        this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).setBaseValue(0.23000000417232513D);
+        this.getEntityAttribute(SharedMonsterAttributes.ATTACK_DAMAGE).setBaseValue(3.0D);
+        this.getEntityAttribute(SharedMonsterAttributes.ARMOR).setBaseValue(2.0D);
+        this.getAttributeMap().registerAttribute(SPAWN_REINFORCEMENTS_CHANCE).setBaseValue(this.rand.nextDouble() * net.minecraftforge.common.ForgeModContainer.zombieSummonBaseChance);
     }
 
-    protected void func_70088_a()
+    protected void entityInit()
     {
-        super.func_70088_a();
-        this.func_184212_Q().func_187214_a(field_184737_bv, Boolean.valueOf(false));
-        this.func_184212_Q().func_187214_a(field_184738_bw, Integer.valueOf(0));
-        this.func_184212_Q().func_187214_a(field_184740_by, Boolean.valueOf(false));
+        super.entityInit();
+        this.getDataManager().register(IS_CHILD, Boolean.valueOf(false));
+        this.getDataManager().register(VILLAGER_TYPE, Integer.valueOf(0));
+        this.getDataManager().register(ARMS_RAISED, Boolean.valueOf(false));
     }
 
-    public void func_184733_a(boolean p_184733_1_)
+    public void setArmsRaised(boolean armsRaised)
     {
-        this.func_184212_Q().func_187227_b(field_184740_by, Boolean.valueOf(p_184733_1_));
+        this.getDataManager().set(ARMS_RAISED, Boolean.valueOf(armsRaised));
     }
 
     @SideOnly(Side.CLIENT)
-    public boolean func_184734_db()
+    public boolean isArmsRaised()
     {
-        return ((Boolean)this.func_184212_Q().func_187225_a(field_184740_by)).booleanValue();
+        return ((Boolean)this.getDataManager().get(ARMS_RAISED)).booleanValue();
     }
 
-    public boolean func_146072_bX()
+    public boolean isBreakDoorsTaskSet()
     {
-        return this.field_146076_bu;
+        return this.isBreakDoorsTaskSet;
     }
 
-    public void func_146070_a(boolean p_146070_1_)
+    public void setBreakDoorsAItask(boolean enabled)
     {
-        if (this.field_146076_bu != p_146070_1_)
+        if (this.isBreakDoorsTaskSet != enabled)
         {
-            this.field_146076_bu = p_146070_1_;
-            ((PathNavigateGround)this.func_70661_as()).func_179688_b(p_146070_1_);
+            this.isBreakDoorsTaskSet = enabled;
+            ((PathNavigateGround)this.getNavigator()).setBreakDoors(enabled);
 
-            if (p_146070_1_)
+            if (enabled)
             {
-                this.field_70714_bg.func_75776_a(1, this.field_146075_bs);
+                this.tasks.addTask(1, this.breakDoor);
             }
             else
             {
-                this.field_70714_bg.func_85156_a(this.field_146075_bs);
+                this.tasks.removeTask(this.breakDoor);
             }
         }
     }
 
-    public boolean func_70631_g_()
+    public boolean isChild()
     {
-        return ((Boolean)this.func_184212_Q().func_187225_a(field_184737_bv)).booleanValue();
+        return ((Boolean)this.getDataManager().get(IS_CHILD)).booleanValue();
     }
 
-    protected int func_70693_a(EntityPlayer p_70693_1_)
+    protected int getExperiencePoints(EntityPlayer player)
     {
-        if (this.func_70631_g_())
+        if (this.isChild())
         {
-            this.field_70728_aV = (int)((float)this.field_70728_aV * 2.5F);
+            this.experienceValue = (int)((float)this.experienceValue * 2.5F);
         }
 
-        return super.func_70693_a(p_70693_1_);
+        return super.getExperiencePoints(player);
     }
 
-    public void func_82227_f(boolean p_82227_1_)
+    public void setChild(boolean childZombie)
     {
-        this.func_184212_Q().func_187227_b(field_184737_bv, Boolean.valueOf(p_82227_1_));
+        this.getDataManager().set(IS_CHILD, Boolean.valueOf(childZombie));
 
-        if (this.field_70170_p != null && !this.field_70170_p.field_72995_K)
+        if (this.world != null && !this.world.isRemote)
         {
-            IAttributeInstance iattributeinstance = this.func_110148_a(SharedMonsterAttributes.field_111263_d);
-            iattributeinstance.func_111124_b(field_110188_br);
+            IAttributeInstance iattributeinstance = this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED);
+            iattributeinstance.removeModifier(BABY_SPEED_BOOST);
 
-            if (p_82227_1_)
+            if (childZombie)
             {
-                iattributeinstance.func_111121_a(field_110188_br);
+                iattributeinstance.applyModifier(BABY_SPEED_BOOST);
             }
         }
 
-        this.func_146071_k(p_82227_1_);
+        this.setChildSize(childZombie);
     }
 
-    public void func_184206_a(DataParameter<?> p_184206_1_)
+    public void notifyDataManagerChange(DataParameter<?> key)
     {
-        if (field_184737_bv.equals(p_184206_1_))
+        if (IS_CHILD.equals(key))
         {
-            this.func_146071_k(this.func_70631_g_());
+            this.setChildSize(this.isChild());
         }
 
-        super.func_184206_a(p_184206_1_);
+        super.notifyDataManagerChange(key);
     }
 
-    public void func_70636_d()
+    public void onLivingUpdate()
     {
-        if (this.field_70170_p.func_72935_r() && !this.field_70170_p.field_72995_K && !this.func_70631_g_() && this.func_190730_o())
+        if (this.world.isDaytime() && !this.world.isRemote && !this.isChild() && this.shouldBurnInDay())
         {
-            float f = this.func_70013_c(1.0F);
+            float f = this.getBrightness(1.0F);
 
-            if (f > 0.5F && this.field_70146_Z.nextFloat() * 30.0F < (f - 0.4F) * 2.0F && this.field_70170_p.func_175678_i(new BlockPos(this.field_70165_t, this.field_70163_u + (double)this.func_70047_e(), this.field_70161_v)))
+            if (f > 0.5F && this.rand.nextFloat() * 30.0F < (f - 0.4F) * 2.0F && this.world.canSeeSky(new BlockPos(this.posX, this.posY + (double)this.getEyeHeight(), this.posZ)))
             {
                 boolean flag = true;
-                ItemStack itemstack = this.func_184582_a(EntityEquipmentSlot.HEAD);
+                ItemStack itemstack = this.getItemStackFromSlot(EntityEquipmentSlot.HEAD);
 
-                if (!itemstack.func_190926_b())
+                if (!itemstack.isEmpty())
                 {
-                    if (itemstack.func_77984_f())
+                    if (itemstack.isItemStackDamageable())
                     {
-                        itemstack.func_77964_b(itemstack.func_77952_i() + this.field_70146_Z.nextInt(2));
+                        itemstack.setItemDamage(itemstack.getItemDamage() + this.rand.nextInt(2));
 
-                        if (itemstack.func_77952_i() >= itemstack.func_77958_k())
+                        if (itemstack.getItemDamage() >= itemstack.getMaxDamage())
                         {
-                            this.func_70669_a(itemstack);
-                            this.func_184201_a(EntityEquipmentSlot.HEAD, ItemStack.field_190927_a);
+                            this.renderBrokenItemStack(itemstack);
+                            this.setItemStackToSlot(EntityEquipmentSlot.HEAD, ItemStack.EMPTY);
                         }
                     }
 
@@ -215,54 +218,73 @@
 
                 if (flag)
                 {
-                    this.func_70015_d(8);
+                    // CraftBukkit start
+                    EntityCombustEvent event = new EntityCombustEvent(this.getBukkitEntity(), 8);
+                    this.world.getServer().getPluginManager().callEvent(event);
+
+                    if (!event.isCancelled()) {
+                        this.setFire(event.getDuration());
+                    }
+                    // CraftBukkit end
                 }
             }
         }
 
-        super.func_70636_d();
+        super.onLivingUpdate();
     }
 
-    protected boolean func_190730_o()
+    protected boolean shouldBurnInDay()
     {
         return true;
     }
 
-    public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
+    public boolean attackEntityFrom(DamageSource source, float amount)
     {
-        if (super.func_70097_a(p_70097_1_, p_70097_2_))
+        if (super.attackEntityFrom(source, amount))
         {
-            EntityLivingBase entitylivingbase = this.func_70638_az();
+            EntityLivingBase entitylivingbase = this.getAttackTarget();
 
-            if (entitylivingbase == null && p_70097_1_.func_76346_g() instanceof EntityLivingBase)
+            if (entitylivingbase == null && source.getEntity() instanceof EntityLivingBase)
             {
-                entitylivingbase = (EntityLivingBase)p_70097_1_.func_76346_g();
+                entitylivingbase = (EntityLivingBase)source.getEntity();
             }
 
-            if (entitylivingbase != null && this.field_70170_p.func_175659_aa() == EnumDifficulty.HARD && (double)this.field_70146_Z.nextFloat() < this.func_110148_a(field_110186_bp).func_111126_e() && this.field_70170_p.func_82736_K().func_82766_b("doMobSpawning"))
+            int i = MathHelper.floor(this.posX);
+            int j = MathHelper.floor(this.posY);
+            int k = MathHelper.floor(this.posZ);
+            net.minecraftforge.event.entity.living.ZombieEvent.SummonAidEvent summonAid = net.minecraftforge.event.ForgeEventFactory.fireZombieSummonAid(this, world, i, j, k, entitylivingbase, this.getEntityAttribute(SPAWN_REINFORCEMENTS_CHANCE).getAttributeValue());
+            if (summonAid.getResult() == net.minecraftforge.fml.common.eventhandler.Event.Result.DENY) return true;
+
+            if (summonAid.getResult() == net.minecraftforge.fml.common.eventhandler.Event.Result.ALLOW  ||
+                entitylivingbase != null && this.world.getDifficulty() == EnumDifficulty.HARD && (double)this.rand.nextFloat() < this.getEntityAttribute(SPAWN_REINFORCEMENTS_CHANCE).getAttributeValue() && this.world.getGameRules().getBoolean("doMobSpawning"))
             {
-                int i = MathHelper.func_76128_c(this.field_70165_t);
-                int j = MathHelper.func_76128_c(this.field_70163_u);
-                int k = MathHelper.func_76128_c(this.field_70161_v);
-                EntityZombie entityzombie = new EntityZombie(this.field_70170_p);
+                EntityZombie entityzombie;
+                if (summonAid.getCustomSummonedAid() != null && summonAid.getResult() == net.minecraftforge.fml.common.eventhandler.Event.Result.ALLOW)
+                {
+                    entityzombie = summonAid.getCustomSummonedAid();
+                }
+                else
+                {
+                    entityzombie = new EntityZombie(this.world);
+                }
 
                 for (int l = 0; l < 50; ++l)
                 {
-                    int i1 = i + MathHelper.func_76136_a(this.field_70146_Z, 7, 40) * MathHelper.func_76136_a(this.field_70146_Z, -1, 1);
-                    int j1 = j + MathHelper.func_76136_a(this.field_70146_Z, 7, 40) * MathHelper.func_76136_a(this.field_70146_Z, -1, 1);
-                    int k1 = k + MathHelper.func_76136_a(this.field_70146_Z, 7, 40) * MathHelper.func_76136_a(this.field_70146_Z, -1, 1);
+                    int i1 = i + MathHelper.getInt(this.rand, 7, 40) * MathHelper.getInt(this.rand, -1, 1);
+                    int j1 = j + MathHelper.getInt(this.rand, 7, 40) * MathHelper.getInt(this.rand, -1, 1);
+                    int k1 = k + MathHelper.getInt(this.rand, 7, 40) * MathHelper.getInt(this.rand, -1, 1);
 
-                    if (this.field_70170_p.func_180495_p(new BlockPos(i1, j1 - 1, k1)).func_185896_q() && this.field_70170_p.func_175671_l(new BlockPos(i1, j1, k1)) < 10)
+                    if (this.world.getBlockState(new BlockPos(i1, j1 - 1, k1)).isSideSolid(this.world, new BlockPos(i1, j1 - 1, k1), net.minecraft.util.EnumFacing.UP) && this.world.getLightFromNeighbors(new BlockPos(i1, j1, k1)) < 10)
                     {
-                        entityzombie.func_70107_b((double)i1, (double)j1, (double)k1);
+                        entityzombie.setPosition((double)i1, (double)j1, (double)k1);
 
-                        if (!this.field_70170_p.func_175636_b((double)i1, (double)j1, (double)k1, 7.0D) && this.field_70170_p.func_72917_a(entityzombie.func_174813_aQ(), entityzombie) && this.field_70170_p.func_184144_a(entityzombie, entityzombie.func_174813_aQ()).isEmpty() && !this.field_70170_p.func_72953_d(entityzombie.func_174813_aQ()))
+                        if (!this.world.isAnyPlayerWithinRangeAt((double)i1, (double)j1, (double)k1, 7.0D) && this.world.checkNoEntityCollision(entityzombie.getEntityBoundingBox(), entityzombie) && this.world.getCollisionBoxes(entityzombie, entityzombie.getEntityBoundingBox()).isEmpty() && !this.world.containsAnyLiquid(entityzombie.getEntityBoundingBox()))
                         {
-                            this.field_70170_p.func_72838_d(entityzombie);
-                            entityzombie.func_70624_b(entitylivingbase);
-                            entityzombie.func_180482_a(this.field_70170_p.func_175649_E(new BlockPos(entityzombie)), (IEntityLivingData)null);
-                            this.func_110148_a(field_110186_bp).func_111121_a(new AttributeModifier("Zombie reinforcement caller charge", -0.05000000074505806D, 0));
-                            entityzombie.func_110148_a(field_110186_bp).func_111121_a(new AttributeModifier("Zombie reinforcement callee charge", -0.05000000074505806D, 0));
+                            this.world.spawnEntity(entityzombie);
+                            if (entitylivingbase != null) entityzombie.setAttackTarget(entitylivingbase);
+                            entityzombie.onInitialSpawn(this.world.getDifficultyForLocation(new BlockPos(entityzombie)), (IEntityLivingData)null);
+                            this.getEntityAttribute(SPAWN_REINFORCEMENTS_CHANCE).applyModifier(new AttributeModifier("Zombie reinforcement caller charge", -0.05000000074505806D, 0));
+                            entityzombie.getEntityAttribute(SPAWN_REINFORCEMENTS_CHANCE).applyModifier(new AttributeModifier("Zombie reinforcement callee charge", -0.05000000074505806D, 0));
                             break;
                         }
                     }
@@ -277,143 +299,150 @@
         }
     }
 
-    public boolean func_70652_k(Entity p_70652_1_)
+    public boolean attackEntityAsMob(Entity entityIn)
     {
-        boolean flag = super.func_70652_k(p_70652_1_);
+        boolean flag = super.attackEntityAsMob(entityIn);
 
         if (flag)
         {
-            float f = this.field_70170_p.func_175649_E(new BlockPos(this)).func_180168_b();
+            float f = this.world.getDifficultyForLocation(new BlockPos(this)).getAdditionalDifficulty();
 
-            if (this.func_184614_ca().func_190926_b() && this.func_70027_ad() && this.field_70146_Z.nextFloat() < f * 0.3F)
+            if (this.getHeldItemMainhand().isEmpty() && this.isBurning() && this.rand.nextFloat() < f * 0.3F)
             {
-                p_70652_1_.func_70015_d(2 * (int)f);
+                // CraftBukkit start
+                EntityCombustByEntityEvent event = new EntityCombustByEntityEvent(this.getBukkitEntity(), entityIn.getBukkitEntity(), 2 * (int) f); // PAIL: fixme
+                this.world.getServer().getPluginManager().callEvent(event);
+
+                if (!event.isCancelled()) {
+                    entityIn.setFire(event.getDuration());
+                }
+                // CraftBukkit end
             }
         }
 
         return flag;
     }
 
-    protected SoundEvent func_184639_G()
+    protected SoundEvent getAmbientSound()
     {
-        return SoundEvents.field_187899_gZ;
+        return SoundEvents.ENTITY_ZOMBIE_AMBIENT;
     }
 
-    protected SoundEvent func_184601_bQ()
+    protected SoundEvent getHurtSound()
     {
-        return SoundEvents.field_187934_hh;
+        return SoundEvents.ENTITY_ZOMBIE_HURT;
     }
 
-    protected SoundEvent func_184615_bR()
+    protected SoundEvent getDeathSound()
     {
-        return SoundEvents.field_187930_hd;
+        return SoundEvents.ENTITY_ZOMBIE_DEATH;
     }
 
-    protected SoundEvent func_190731_di()
+    protected SoundEvent getStepSound()
     {
-        return SoundEvents.field_187939_hm;
+        return SoundEvents.ENTITY_ZOMBIE_STEP;
     }
 
-    protected void func_180429_a(BlockPos p_180429_1_, Block p_180429_2_)
+    protected void playStepSound(BlockPos pos, Block blockIn)
     {
-        this.func_184185_a(this.func_190731_di(), 0.15F, 1.0F);
+        this.playSound(this.getStepSound(), 0.15F, 1.0F);
     }
 
-    public EnumCreatureAttribute func_70668_bt()
+    public EnumCreatureAttribute getCreatureAttribute()
     {
         return EnumCreatureAttribute.UNDEAD;
     }
 
     @Nullable
-    protected ResourceLocation func_184647_J()
+    protected ResourceLocation getLootTable()
     {
-        return LootTableList.field_186383_ah;
+        return LootTableList.ENTITIES_ZOMBIE;
     }
 
-    protected void func_180481_a(DifficultyInstance p_180481_1_)
+    protected void setEquipmentBasedOnDifficulty(DifficultyInstance difficulty)
     {
-        super.func_180481_a(p_180481_1_);
+        super.setEquipmentBasedOnDifficulty(difficulty);
 
-        if (this.field_70146_Z.nextFloat() < (this.field_70170_p.func_175659_aa() == EnumDifficulty.HARD ? 0.05F : 0.01F))
+        if (this.rand.nextFloat() < (this.world.getDifficulty() == EnumDifficulty.HARD ? 0.05F : 0.01F))
         {
-            int i = this.field_70146_Z.nextInt(3);
+            int i = this.rand.nextInt(3);
 
             if (i == 0)
             {
-                this.func_184201_a(EntityEquipmentSlot.MAINHAND, new ItemStack(Items.field_151040_l));
+                this.setItemStackToSlot(EntityEquipmentSlot.MAINHAND, new ItemStack(Items.IRON_SWORD));
             }
             else
             {
-                this.func_184201_a(EntityEquipmentSlot.MAINHAND, new ItemStack(Items.field_151037_a));
+                this.setItemStackToSlot(EntityEquipmentSlot.MAINHAND, new ItemStack(Items.IRON_SHOVEL));
             }
         }
     }
 
-    public static void func_189779_d(DataFixer p_189779_0_)
+    public static void registerFixesZombie(DataFixer fixer)
     {
-        EntityLiving.func_189752_a(p_189779_0_, EntityZombie.class);
+        EntityLiving.registerFixesMob(fixer, EntityZombie.class);
     }
 
-    public void func_70014_b(NBTTagCompound p_70014_1_)
+    public void writeEntityToNBT(NBTTagCompound compound)
     {
-        super.func_70014_b(p_70014_1_);
+        super.writeEntityToNBT(compound);
 
-        if (this.func_70631_g_())
+        if (this.isChild())
         {
-            p_70014_1_.func_74757_a("IsBaby", true);
+            compound.setBoolean("IsBaby", true);
         }
 
-        p_70014_1_.func_74757_a("CanBreakDoors", this.func_146072_bX());
+        compound.setBoolean("CanBreakDoors", this.isBreakDoorsTaskSet());
     }
 
-    public void func_70037_a(NBTTagCompound p_70037_1_)
+    public void readEntityFromNBT(NBTTagCompound compound)
     {
-        super.func_70037_a(p_70037_1_);
+        super.readEntityFromNBT(compound);
 
-        if (p_70037_1_.func_74767_n("IsBaby"))
+        if (compound.getBoolean("IsBaby"))
         {
-            this.func_82227_f(true);
+            this.setChild(true);
         }
 
-        this.func_146070_a(p_70037_1_.func_74767_n("CanBreakDoors"));
+        this.setBreakDoorsAItask(compound.getBoolean("CanBreakDoors"));
     }
 
-    public void func_70074_a(EntityLivingBase p_70074_1_)
+    public void onKillEntity(EntityLivingBase entityLivingIn)
     {
-        super.func_70074_a(p_70074_1_);
+        super.onKillEntity(entityLivingIn);
 
-        if ((this.field_70170_p.func_175659_aa() == EnumDifficulty.NORMAL || this.field_70170_p.func_175659_aa() == EnumDifficulty.HARD) && p_70074_1_ instanceof EntityVillager)
+        if ((this.world.getDifficulty() == EnumDifficulty.NORMAL || this.world.getDifficulty() == EnumDifficulty.HARD) && entityLivingIn instanceof EntityVillager)
         {
-            if (this.field_70170_p.func_175659_aa() != EnumDifficulty.HARD && this.field_70146_Z.nextBoolean())
+            if (this.world.getDifficulty() != EnumDifficulty.HARD && this.rand.nextBoolean())
             {
                 return;
             }
 
-            EntityVillager entityvillager = (EntityVillager)p_70074_1_;
-            EntityZombieVillager entityzombievillager = new EntityZombieVillager(this.field_70170_p);
-            entityzombievillager.func_82149_j(entityvillager);
-            this.field_70170_p.func_72900_e(entityvillager);
-            entityzombievillager.func_180482_a(this.field_70170_p.func_175649_E(new BlockPos(entityzombievillager)), new EntityZombie.GroupData(false));
-            entityzombievillager.func_190733_a(entityvillager.func_70946_n());
-            entityzombievillager.func_82227_f(entityvillager.func_70631_g_());
-            entityzombievillager.func_94061_f(entityvillager.func_175446_cd());
+            EntityVillager entityvillager = (EntityVillager)entityLivingIn;
+            EntityZombieVillager entityzombievillager = new EntityZombieVillager(this.world);
+            entityzombievillager.copyLocationAndAnglesFrom(entityvillager);
+            this.world.removeEntity(entityvillager);
+            entityzombievillager.onInitialSpawn(this.world.getDifficultyForLocation(new BlockPos(entityzombievillager)), new EntityZombie.GroupData(false));
+            entityzombievillager.setProfession(entityvillager.getProfession());
+            entityzombievillager.setChild(entityvillager.isChild());
+            entityzombievillager.setNoAI(entityvillager.isAIDisabled());
 
-            if (entityvillager.func_145818_k_())
+            if (entityvillager.hasCustomName())
             {
-                entityzombievillager.func_96094_a(entityvillager.func_95999_t());
-                entityzombievillager.func_174805_g(entityvillager.func_174833_aM());
+                entityzombievillager.setCustomNameTag(entityvillager.getCustomNameTag());
+                entityzombievillager.setAlwaysRenderNameTag(entityvillager.getAlwaysRenderNameTag());
             }
 
-            this.field_70170_p.func_72838_d(entityzombievillager);
-            this.field_70170_p.func_180498_a((EntityPlayer)null, 1026, new BlockPos(this), 0);
+            this.world.addEntity(entityzombievillager, CreatureSpawnEvent.SpawnReason.INFECTION); // CraftBukkit - add SpawnReason
+            this.world.playEvent((EntityPlayer)null, 1026, new BlockPos(this), 0);
         }
     }
 
-    public float func_70047_e()
+    public float getEyeHeight()
     {
         float f = 1.74F;
 
-        if (this.func_70631_g_())
+        if (this.isChild())
         {
             f = (float)((double)f - 0.81D);
         }
@@ -421,147 +450,147 @@
         return f;
     }
 
-    protected boolean func_175448_a(ItemStack p_175448_1_)
+    protected boolean canEquipItem(ItemStack stack)
     {
-        return p_175448_1_.func_77973_b() == Items.field_151110_aK && this.func_70631_g_() && this.func_184218_aH() ? false : super.func_175448_a(p_175448_1_);
+        return stack.getItem() == Items.EGG && this.isChild() && this.isRiding() ? false : super.canEquipItem(stack);
     }
 
     @Nullable
-    public IEntityLivingData func_180482_a(DifficultyInstance p_180482_1_, @Nullable IEntityLivingData p_180482_2_)
+    public IEntityLivingData onInitialSpawn(DifficultyInstance difficulty, @Nullable IEntityLivingData livingdata)
     {
-        p_180482_2_ = super.func_180482_a(p_180482_1_, p_180482_2_);
-        float f = p_180482_1_.func_180170_c();
-        this.func_98053_h(this.field_70146_Z.nextFloat() < 0.55F * f);
+        livingdata = super.onInitialSpawn(difficulty, livingdata);
+        float f = difficulty.getClampedAdditionalDifficulty();
+        this.setCanPickUpLoot(this.rand.nextFloat() < 0.55F * f);
 
-        if (p_180482_2_ == null)
+        if (livingdata == null)
         {
-            p_180482_2_ = new EntityZombie.GroupData(this.field_70170_p.field_73012_v.nextFloat() < 0.05F);
+            livingdata = new EntityZombie.GroupData(this.world.rand.nextFloat() < net.minecraftforge.common.ForgeModContainer.zombieBabyChance);
         }
 
-        if (p_180482_2_ instanceof EntityZombie.GroupData)
+        if (livingdata instanceof EntityZombie.GroupData)
         {
-            EntityZombie.GroupData entityzombie$groupdata = (EntityZombie.GroupData)p_180482_2_;
+            EntityZombie.GroupData entityzombie$groupdata = (EntityZombie.GroupData)livingdata;
 
-            if (entityzombie$groupdata.field_142048_a)
+            if (entityzombie$groupdata.isChild)
             {
-                this.func_82227_f(true);
+                this.setChild(true);
 
-                if ((double)this.field_70170_p.field_73012_v.nextFloat() < 0.05D)
+                if ((double)this.world.rand.nextFloat() < 0.05D)
                 {
-                    List<EntityChicken> list = this.field_70170_p.<EntityChicken>func_175647_a(EntityChicken.class, this.func_174813_aQ().func_72314_b(5.0D, 3.0D, 5.0D), EntitySelectors.field_152785_b);
+                    List<EntityChicken> list = this.world.<EntityChicken>getEntitiesWithinAABB(EntityChicken.class, this.getEntityBoundingBox().expand(5.0D, 3.0D, 5.0D), EntitySelectors.IS_STANDALONE);
 
                     if (!list.isEmpty())
                     {
                         EntityChicken entitychicken = (EntityChicken)list.get(0);
-                        entitychicken.func_152117_i(true);
-                        this.func_184220_m(entitychicken);
+                        entitychicken.setChickenJockey(true);
+                        this.startRiding(entitychicken);
                     }
                 }
-                else if ((double)this.field_70170_p.field_73012_v.nextFloat() < 0.05D)
+                else if ((double)this.world.rand.nextFloat() < 0.05D)
                 {
-                    EntityChicken entitychicken1 = new EntityChicken(this.field_70170_p);
-                    entitychicken1.func_70012_b(this.field_70165_t, this.field_70163_u, this.field_70161_v, this.field_70177_z, 0.0F);
-                    entitychicken1.func_180482_a(p_180482_1_, (IEntityLivingData)null);
-                    entitychicken1.func_152117_i(true);
-                    this.field_70170_p.func_72838_d(entitychicken1);
-                    this.func_184220_m(entitychicken1);
+                    EntityChicken entitychicken1 = new EntityChicken(this.world);
+                    entitychicken1.setLocationAndAngles(this.posX, this.posY, this.posZ, this.rotationYaw, 0.0F);
+                    entitychicken1.onInitialSpawn(difficulty, (IEntityLivingData)null);
+                    entitychicken1.setChickenJockey(true);
+                    this.world.addEntity(entitychicken1, CreatureSpawnEvent.SpawnReason.MOUNT); // CraftBukkit
+                    this.startRiding(entitychicken1);
                 }
             }
         }
 
-        this.func_146070_a(this.field_70146_Z.nextFloat() < f * 0.1F);
-        this.func_180481_a(p_180482_1_);
-        this.func_180483_b(p_180482_1_);
+        this.setBreakDoorsAItask(this.rand.nextFloat() < f * 0.1F);
+        this.setEquipmentBasedOnDifficulty(difficulty);
+        this.setEnchantmentBasedOnDifficulty(difficulty);
 
-        if (this.func_184582_a(EntityEquipmentSlot.HEAD).func_190926_b())
+        if (this.getItemStackFromSlot(EntityEquipmentSlot.HEAD).isEmpty())
         {
-            Calendar calendar = this.field_70170_p.func_83015_S();
+            Calendar calendar = this.world.getCurrentDate();
 
-            if (calendar.get(2) + 1 == 10 && calendar.get(5) == 31 && this.field_70146_Z.nextFloat() < 0.25F)
+            if (calendar.get(2) + 1 == 10 && calendar.get(5) == 31 && this.rand.nextFloat() < 0.25F)
             {
-                this.func_184201_a(EntityEquipmentSlot.HEAD, new ItemStack(this.field_70146_Z.nextFloat() < 0.1F ? Blocks.field_150428_aP : Blocks.field_150423_aK));
-                this.field_184655_bs[EntityEquipmentSlot.HEAD.func_188454_b()] = 0.0F;
+                this.setItemStackToSlot(EntityEquipmentSlot.HEAD, new ItemStack(this.rand.nextFloat() < 0.1F ? Blocks.LIT_PUMPKIN : Blocks.PUMPKIN));
+                this.inventoryArmorDropChances[EntityEquipmentSlot.HEAD.getIndex()] = 0.0F;
             }
         }
 
-        this.func_110148_a(SharedMonsterAttributes.field_111266_c).func_111121_a(new AttributeModifier("Random spawn bonus", this.field_70146_Z.nextDouble() * 0.05000000074505806D, 0));
-        double d0 = this.field_70146_Z.nextDouble() * 1.5D * (double)f;
+        this.getEntityAttribute(SharedMonsterAttributes.KNOCKBACK_RESISTANCE).applyModifier(new AttributeModifier("Random spawn bonus", this.rand.nextDouble() * 0.05000000074505806D, 0));
+        double d0 = this.rand.nextDouble() * 1.5D * (double)f;
 
         if (d0 > 1.0D)
         {
-            this.func_110148_a(SharedMonsterAttributes.field_111265_b).func_111121_a(new AttributeModifier("Random zombie-spawn bonus", d0, 2));
+            this.getEntityAttribute(SharedMonsterAttributes.FOLLOW_RANGE).applyModifier(new AttributeModifier("Random zombie-spawn bonus", d0, 2));
         }
 
-        if (this.field_70146_Z.nextFloat() < f * 0.05F)
+        if (this.rand.nextFloat() < f * 0.05F)
         {
-            this.func_110148_a(field_110186_bp).func_111121_a(new AttributeModifier("Leader zombie bonus", this.field_70146_Z.nextDouble() * 0.25D + 0.5D, 0));
-            this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111121_a(new AttributeModifier("Leader zombie bonus", this.field_70146_Z.nextDouble() * 3.0D + 1.0D, 2));
-            this.func_146070_a(true);
+            this.getEntityAttribute(SPAWN_REINFORCEMENTS_CHANCE).applyModifier(new AttributeModifier("Leader zombie bonus", this.rand.nextDouble() * 0.25D + 0.5D, 0));
+            this.getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).applyModifier(new AttributeModifier("Leader zombie bonus", this.rand.nextDouble() * 3.0D + 1.0D, 2));
+            this.setBreakDoorsAItask(true);
         }
 
-        return p_180482_2_;
+        return livingdata;
     }
 
-    public void func_146071_k(boolean p_146071_1_)
+    public void setChildSize(boolean isChild)
     {
-        this.func_146069_a(p_146071_1_ ? 0.5F : 1.0F);
+        this.multiplySize(isChild ? 0.5F : 1.0F);
     }
 
-    public final void func_70105_a(float p_70105_1_, float p_70105_2_)
+    public final void setSize(float width, float height)
     {
-        boolean flag = this.field_146074_bv > 0.0F && this.field_146073_bw > 0.0F;
-        this.field_146074_bv = p_70105_1_;
-        this.field_146073_bw = p_70105_2_;
+        boolean flag = this.zombieWidth > 0.0F && this.zombieHeight > 0.0F;
+        this.zombieWidth = width;
+        this.zombieHeight = height;
 
         if (!flag)
         {
-            this.func_146069_a(1.0F);
+            this.multiplySize(1.0F);
         }
     }
 
-    protected final void func_146069_a(float p_146069_1_)
+    protected final void multiplySize(float size)
     {
-        super.func_70105_a(this.field_146074_bv * p_146069_1_, this.field_146073_bw * p_146069_1_);
+        super.setSize(this.zombieWidth * size, this.zombieHeight * size);
     }
 
-    public double func_70033_W()
+    public double getYOffset()
     {
-        return this.func_70631_g_() ? 0.0D : -0.45D;
+        return this.isChild() ? 0.0D : -0.45D;
     }
 
-    public void func_70645_a(DamageSource p_70645_1_)
+    public void onDeath(DamageSource cause)
     {
-        super.func_70645_a(p_70645_1_);
+        super.onDeath(cause);
 
-        if (p_70645_1_.func_76346_g() instanceof EntityCreeper)
+        if (cause.getEntity() instanceof EntityCreeper)
         {
-            EntityCreeper entitycreeper = (EntityCreeper)p_70645_1_.func_76346_g();
+            EntityCreeper entitycreeper = (EntityCreeper)cause.getEntity();
 
-            if (entitycreeper.func_70830_n() && entitycreeper.func_70650_aV())
+            if (entitycreeper.getPowered() && entitycreeper.isAIEnabled())
             {
-                entitycreeper.func_175493_co();
-                ItemStack itemstack = this.func_190732_dj();
+                entitycreeper.incrementDroppedSkulls();
+                ItemStack itemstack = this.getSkullDrop();
 
-                if (!itemstack.func_190926_b())
+                if (!itemstack.isEmpty())
                 {
-                    this.func_70099_a(itemstack, 0.0F);
+                    this.entityDropItem(itemstack, 0.0F);
                 }
             }
         }
     }
 
-    protected ItemStack func_190732_dj()
+    protected ItemStack getSkullDrop()
     {
-        return new ItemStack(Items.field_151144_bL, 1, 2);
+        return new ItemStack(Items.SKULL, 1, 2);
     }
 
     class GroupData implements IEntityLivingData
     {
-        public boolean field_142048_a;
+        public boolean isChild;
 
         private GroupData(boolean p_i47328_2_)
         {
-            this.field_142048_a = p_i47328_2_;
+            this.isChild = p_i47328_2_;
         }
     }
 }
