--- ../src-base/minecraft/net/minecraft/entity/monster/EntitySilverfish.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntitySilverfish.java
@@ -26,126 +26,127 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 import net.minecraft.world.storage.loot.LootTableList;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class EntitySilverfish extends EntityMob
 {
-    private EntitySilverfish.AISummonSilverfish field_175460_b;
+    private EntitySilverfish.AISummonSilverfish summonSilverfish;
 
-    public EntitySilverfish(World p_i1740_1_)
+    public EntitySilverfish(World worldIn)
     {
-        super(p_i1740_1_);
-        this.func_70105_a(0.4F, 0.3F);
+        super(worldIn);
+        this.setSize(0.4F, 0.3F);
     }
 
-    public static void func_189767_b(DataFixer p_189767_0_)
+    public static void registerFixesSilverfish(DataFixer fixer)
     {
-        EntityLiving.func_189752_a(p_189767_0_, EntitySilverfish.class);
+        EntityLiving.registerFixesMob(fixer, EntitySilverfish.class);
     }
 
-    protected void func_184651_r()
+    protected void initEntityAI()
     {
-        this.field_175460_b = new EntitySilverfish.AISummonSilverfish(this);
-        this.field_70714_bg.func_75776_a(1, new EntityAISwimming(this));
-        this.field_70714_bg.func_75776_a(3, this.field_175460_b);
-        this.field_70714_bg.func_75776_a(4, new EntityAIAttackMelee(this, 1.0D, false));
-        this.field_70714_bg.func_75776_a(5, new EntitySilverfish.AIHideInStone(this));
-        this.field_70715_bh.func_75776_a(1, new EntityAIHurtByTarget(this, true, new Class[0]));
-        this.field_70715_bh.func_75776_a(2, new EntityAINearestAttackableTarget(this, EntityPlayer.class, true));
+        this.summonSilverfish = new EntitySilverfish.AISummonSilverfish(this);
+        this.tasks.addTask(1, new EntityAISwimming(this));
+        this.tasks.addTask(3, this.summonSilverfish);
+        this.tasks.addTask(4, new EntityAIAttackMelee(this, 1.0D, false));
+        this.tasks.addTask(5, new EntitySilverfish.AIHideInStone(this));
+        this.targetTasks.addTask(1, new EntityAIHurtByTarget(this, true, new Class[0]));
+        this.targetTasks.addTask(2, new EntityAINearestAttackableTarget(this, EntityPlayer.class, true));
     }
 
-    public double func_70033_W()
+    public double getYOffset()
     {
         return 0.1D;
     }
 
-    public float func_70047_e()
+    public float getEyeHeight()
     {
         return 0.1F;
     }
 
-    protected void func_110147_ax()
+    protected void applyEntityAttributes()
     {
-        super.func_110147_ax();
-        this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(8.0D);
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.25D);
-        this.func_110148_a(SharedMonsterAttributes.field_111264_e).func_111128_a(1.0D);
+        super.applyEntityAttributes();
+        this.getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).setBaseValue(8.0D);
+        this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).setBaseValue(0.25D);
+        this.getEntityAttribute(SharedMonsterAttributes.ATTACK_DAMAGE).setBaseValue(1.0D);
     }
 
-    protected boolean func_70041_e_()
+    protected boolean canTriggerWalking()
     {
         return false;
     }
 
-    protected SoundEvent func_184639_G()
+    protected SoundEvent getAmbientSound()
     {
-        return SoundEvents.field_187793_eY;
+        return SoundEvents.ENTITY_SILVERFISH_AMBIENT;
     }
 
-    protected SoundEvent func_184601_bQ()
+    protected SoundEvent getHurtSound()
     {
-        return SoundEvents.field_187850_fa;
+        return SoundEvents.ENTITY_SILVERFISH_HURT;
     }
 
-    protected SoundEvent func_184615_bR()
+    protected SoundEvent getDeathSound()
     {
-        return SoundEvents.field_187795_eZ;
+        return SoundEvents.ENTITY_SILVERFISH_DEATH;
     }
 
-    protected void func_180429_a(BlockPos p_180429_1_, Block p_180429_2_)
+    protected void playStepSound(BlockPos pos, Block blockIn)
     {
-        this.func_184185_a(SoundEvents.field_187852_fb, 0.15F, 1.0F);
+        this.playSound(SoundEvents.ENTITY_SILVERFISH_STEP, 0.15F, 1.0F);
     }
 
-    public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
+    public boolean attackEntityFrom(DamageSource source, float amount)
     {
-        if (this.func_180431_b(p_70097_1_))
+        if (this.isEntityInvulnerable(source))
         {
             return false;
         }
         else
         {
-            if ((p_70097_1_ instanceof EntityDamageSource || p_70097_1_ == DamageSource.field_76376_m) && this.field_175460_b != null)
+            if ((source instanceof EntityDamageSource || source == DamageSource.MAGIC) && this.summonSilverfish != null)
             {
-                this.field_175460_b.func_179462_f();
+                this.summonSilverfish.notifyHurt();
             }
 
-            return super.func_70097_a(p_70097_1_, p_70097_2_);
+            return super.attackEntityFrom(source, amount);
         }
     }
 
     @Nullable
-    protected ResourceLocation func_184647_J()
+    protected ResourceLocation getLootTable()
     {
-        return LootTableList.field_186438_t;
+        return LootTableList.ENTITIES_SILVERFISH;
     }
 
-    public void func_70071_h_()
+    public void onUpdate()
     {
-        this.field_70761_aq = this.field_70177_z;
-        super.func_70071_h_();
+        this.renderYawOffset = this.rotationYaw;
+        super.onUpdate();
     }
 
-    public void func_181013_g(float p_181013_1_)
+    public void setRenderYawOffset(float offset)
     {
-        this.field_70177_z = p_181013_1_;
-        super.func_181013_g(p_181013_1_);
+        this.rotationYaw = offset;
+        super.setRenderYawOffset(offset);
     }
 
-    public float func_180484_a(BlockPos p_180484_1_)
+    public float getBlockPathWeight(BlockPos pos)
     {
-        return this.field_70170_p.func_180495_p(p_180484_1_.func_177977_b()).func_177230_c() == Blocks.field_150348_b ? 10.0F : super.func_180484_a(p_180484_1_);
+        return this.world.getBlockState(pos.down()).getBlock() == Blocks.STONE ? 10.0F : super.getBlockPathWeight(pos);
     }
 
-    protected boolean func_70814_o()
+    protected boolean isValidLightLevel()
     {
         return true;
     }
 
-    public boolean func_70601_bi()
+    public boolean getCanSpawnHere()
     {
-        if (super.func_70601_bi())
+        if (super.getCanSpawnHere())
         {
-            EntityPlayer entityplayer = this.field_70170_p.func_184136_b(this, 5.0D);
+            EntityPlayer entityplayer = this.world.getNearestPlayerNotCreative(this, 5.0D);
             return entityplayer == null;
         }
         else
@@ -154,76 +155,76 @@
         }
     }
 
-    public EnumCreatureAttribute func_70668_bt()
+    public EnumCreatureAttribute getCreatureAttribute()
     {
         return EnumCreatureAttribute.ARTHROPOD;
     }
 
     static class AIHideInStone extends EntityAIWander
         {
-            private EnumFacing field_179483_b;
-            private boolean field_179484_c;
+            private EnumFacing facing;
+            private boolean doMerge;
 
-            public AIHideInStone(EntitySilverfish p_i45827_1_)
+            public AIHideInStone(EntitySilverfish silverfishIn)
             {
-                super(p_i45827_1_, 1.0D, 10);
-                this.func_75248_a(1);
+                super(silverfishIn, 1.0D, 10);
+                this.setMutexBits(1);
             }
 
-            public boolean func_75250_a()
+            public boolean shouldExecute()
             {
-                if (this.field_75457_a.func_70638_az() != null)
+                if (this.entity.getAttackTarget() != null)
                 {
                     return false;
                 }
-                else if (!this.field_75457_a.func_70661_as().func_75500_f())
+                else if (!this.entity.getNavigator().noPath())
                 {
                     return false;
                 }
                 else
                 {
-                    Random random = this.field_75457_a.func_70681_au();
+                    Random random = this.entity.getRNG();
 
-                    if (this.field_75457_a.field_70170_p.func_82736_K().func_82766_b("mobGriefing") && random.nextInt(10) == 0)
+                    if (this.entity.world.getGameRules().getBoolean("mobGriefing") && random.nextInt(10) == 0)
                     {
-                        this.field_179483_b = EnumFacing.func_176741_a(random);
-                        BlockPos blockpos = (new BlockPos(this.field_75457_a.field_70165_t, this.field_75457_a.field_70163_u + 0.5D, this.field_75457_a.field_70161_v)).func_177972_a(this.field_179483_b);
-                        IBlockState iblockstate = this.field_75457_a.field_70170_p.func_180495_p(blockpos);
+                        this.facing = EnumFacing.random(random);
+                        BlockPos blockpos = (new BlockPos(this.entity.posX, this.entity.posY + 0.5D, this.entity.posZ)).offset(this.facing);
+                        IBlockState iblockstate = this.entity.world.getBlockState(blockpos);
 
-                        if (BlockSilverfish.func_176377_d(iblockstate))
+                        if (BlockSilverfish.canContainSilverfish(iblockstate))
                         {
-                            this.field_179484_c = true;
+                            this.doMerge = true;
                             return true;
                         }
                     }
 
-                    this.field_179484_c = false;
-                    return super.func_75250_a();
+                    this.doMerge = false;
+                    return super.shouldExecute();
                 }
             }
 
-            public boolean func_75253_b()
+            public boolean continueExecuting()
             {
-                return this.field_179484_c ? false : super.func_75253_b();
+                return this.doMerge ? false : super.continueExecuting();
             }
 
-            public void func_75249_e()
+            public void startExecuting()
             {
-                if (!this.field_179484_c)
+                if (!this.doMerge)
                 {
-                    super.func_75249_e();
+                    super.startExecuting();
                 }
                 else
                 {
-                    World world = this.field_75457_a.field_70170_p;
-                    BlockPos blockpos = (new BlockPos(this.field_75457_a.field_70165_t, this.field_75457_a.field_70163_u + 0.5D, this.field_75457_a.field_70161_v)).func_177972_a(this.field_179483_b);
-                    IBlockState iblockstate = world.func_180495_p(blockpos);
+                    World world = this.entity.world;
+                    BlockPos blockpos = (new BlockPos(this.entity.posX, this.entity.posY + 0.5D, this.entity.posZ)).offset(this.facing);
+                    IBlockState iblockstate = world.getBlockState(blockpos);
 
-                    if (BlockSilverfish.func_176377_d(iblockstate))
+                    if (BlockSilverfish.canContainSilverfish(iblockstate))
                     {
-                        world.func_180501_a(blockpos, Blocks.field_150418_aU.func_176223_P().func_177226_a(BlockSilverfish.field_176378_a, BlockSilverfish.EnumType.func_176878_a(iblockstate)), 3);
-                        this.field_75457_a.func_70656_aK();
-                        this.field_75457_a.func_70106_y();
+                        world.setBlockState(blockpos, Blocks.MONSTER_EGG.getDefaultState().withProperty(BlockSilverfish.VARIANT, BlockSilverfish.EnumType.forModelBlock(iblockstate)), 3);
+                        this.entity.spawnExplosionParticle();
+                        this.entity.setDead();
                     }
                 }
             }
@@ -231,36 +232,36 @@
 
     static class AISummonSilverfish extends EntityAIBase
         {
-            private final EntitySilverfish field_179464_a;
-            private int field_179463_b;
+            private final EntitySilverfish silverfish;
+            private int lookForFriends;
 
-            public AISummonSilverfish(EntitySilverfish p_i45826_1_)
+            public AISummonSilverfish(EntitySilverfish silverfishIn)
             {
-                this.field_179464_a = p_i45826_1_;
+                this.silverfish = silverfishIn;
             }
 
-            public void func_179462_f()
+            public void notifyHurt()
             {
-                if (this.field_179463_b == 0)
+                if (this.lookForFriends == 0)
                 {
-                    this.field_179463_b = 20;
+                    this.lookForFriends = 20;
                 }
             }
 
-            public boolean func_75250_a()
+            public boolean shouldExecute()
             {
-                return this.field_179463_b > 0;
+                return this.lookForFriends > 0;
             }
 
-            public void func_75246_d()
+            public void updateTask()
             {
-                --this.field_179463_b;
+                --this.lookForFriends;
 
-                if (this.field_179463_b <= 0)
+                if (this.lookForFriends <= 0)
                 {
-                    World world = this.field_179464_a.field_70170_p;
-                    Random random = this.field_179464_a.func_70681_au();
-                    BlockPos blockpos = new BlockPos(this.field_179464_a);
+                    World world = this.silverfish.world;
+                    Random random = this.silverfish.getRNG();
+                    BlockPos blockpos = new BlockPos(this.silverfish);
 
                     for (int i = 0; i <= 5 && i >= -5; i = (i <= 0 ? 1 : 0) - i)
                     {
@@ -268,18 +269,23 @@
                         {
                             for (int k = 0; k <= 10 && k >= -10; k = (k <= 0 ? 1 : 0) - k)
                             {
-                                BlockPos blockpos1 = blockpos.func_177982_a(j, i, k);
-                                IBlockState iblockstate = world.func_180495_p(blockpos1);
+                                BlockPos blockpos1 = blockpos.add(j, i, k);
+                                IBlockState iblockstate = world.getBlockState(blockpos1);
 
-                                if (iblockstate.func_177230_c() == Blocks.field_150418_aU)
+                                if (iblockstate.getBlock() == Blocks.MONSTER_EGG)
                                 {
-                                    if (world.func_82736_K().func_82766_b("mobGriefing"))
+                                    // CraftBukkit start
+                                    if (CraftEventFactory.callEntityChangeBlockEvent(this.silverfish, blockpos1, Blocks.AIR, 0).isCancelled()) {
+                                        continue;
+                                    }
+                                    // CraftBukkit end
+                                    if (world.getGameRules().getBoolean("mobGriefing"))
                                     {
-                                        world.func_175655_b(blockpos1, true);
+                                        world.destroyBlock(blockpos1, true);
                                     }
                                     else
                                     {
-                                        world.func_180501_a(blockpos1, ((BlockSilverfish.EnumType)iblockstate.func_177229_b(BlockSilverfish.field_176378_a)).func_176883_d(), 3);
+                                        world.setBlockState(blockpos1, ((BlockSilverfish.EnumType)iblockstate.getValue(BlockSilverfish.VARIANT)).getModelBlock(), 3);
                                     }
 
                                     if (random.nextBoolean())
