--- ../src-base/minecraft/net/minecraft/entity/monster/EntitySlime.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntitySlime.java
@@ -35,293 +35,305 @@
 import net.minecraft.world.biome.Biome;
 import net.minecraft.world.chunk.Chunk;
 import net.minecraft.world.storage.loot.LootTableList;
+import org.bukkit.event.entity.SlimeSplitEvent;
 
 public class EntitySlime extends EntityLiving implements IMob
 {
-    private static final DataParameter<Integer> field_184711_bt = EntityDataManager.<Integer>func_187226_a(EntitySlime.class, DataSerializers.field_187192_b);
-    public float field_70813_a;
-    public float field_70811_b;
-    public float field_70812_c;
-    private boolean field_175452_bi;
+    private static final DataParameter<Integer> SLIME_SIZE = EntityDataManager.<Integer>createKey(EntitySlime.class, DataSerializers.VARINT);
+    public float squishAmount;
+    public float squishFactor;
+    public float prevSquishFactor;
+    private boolean wasOnGround;
 
-    public EntitySlime(World p_i1742_1_)
+    public EntitySlime(World worldIn)
     {
-        super(p_i1742_1_);
-        this.field_70765_h = new EntitySlime.SlimeMoveHelper(this);
+        super(worldIn);
+        this.moveHelper = new EntitySlime.SlimeMoveHelper(this);
     }
 
-    protected void func_184651_r()
+    protected void initEntityAI()
     {
-        this.field_70714_bg.func_75776_a(1, new EntitySlime.AISlimeFloat(this));
-        this.field_70714_bg.func_75776_a(2, new EntitySlime.AISlimeAttack(this));
-        this.field_70714_bg.func_75776_a(3, new EntitySlime.AISlimeFaceRandom(this));
-        this.field_70714_bg.func_75776_a(5, new EntitySlime.AISlimeHop(this));
-        this.field_70715_bh.func_75776_a(1, new EntityAIFindEntityNearestPlayer(this));
-        this.field_70715_bh.func_75776_a(3, new EntityAIFindEntityNearest(this, EntityIronGolem.class));
+        this.tasks.addTask(1, new EntitySlime.AISlimeFloat(this));
+        this.tasks.addTask(2, new EntitySlime.AISlimeAttack(this));
+        this.tasks.addTask(3, new EntitySlime.AISlimeFaceRandom(this));
+        this.tasks.addTask(5, new EntitySlime.AISlimeHop(this));
+        this.targetTasks.addTask(1, new EntityAIFindEntityNearestPlayer(this));
+        this.targetTasks.addTask(3, new EntityAIFindEntityNearest(this, EntityIronGolem.class));
     }
 
-    protected void func_70088_a()
+    protected void entityInit()
     {
-        super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184711_bt, Integer.valueOf(1));
+        super.entityInit();
+        this.dataManager.register(SLIME_SIZE, Integer.valueOf(1));
     }
 
-    public void func_70799_a(int p_70799_1_, boolean p_70799_2_)
+    public void setSlimeSize(int size, boolean p_70799_2_)
     {
-        this.field_70180_af.func_187227_b(field_184711_bt, Integer.valueOf(p_70799_1_));
-        this.func_70105_a(0.51000005F * (float)p_70799_1_, 0.51000005F * (float)p_70799_1_);
-        this.func_70107_b(this.field_70165_t, this.field_70163_u, this.field_70161_v);
-        this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a((double)(p_70799_1_ * p_70799_1_));
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a((double)(0.2F + 0.1F * (float)p_70799_1_));
+        this.dataManager.set(SLIME_SIZE, Integer.valueOf(size));
+        this.setSize(0.51000005F * (float)size, 0.51000005F * (float)size);
+        this.setPosition(this.posX, this.posY, this.posZ);
+        this.getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).setBaseValue((double)(size * size));
+        this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).setBaseValue((double)(0.2F + 0.1F * (float)size));
 
         if (p_70799_2_)
         {
-            this.func_70606_j(this.func_110138_aP());
+            this.setHealth(this.getMaxHealth());
         }
 
-        this.field_70728_aV = p_70799_1_;
+        this.experienceValue = size;
     }
 
-    public int func_70809_q()
+    public int getSlimeSize()
     {
-        return ((Integer)this.field_70180_af.func_187225_a(field_184711_bt)).intValue();
+        return ((Integer)this.dataManager.get(SLIME_SIZE)).intValue();
     }
 
-    public static void func_189758_c(DataFixer p_189758_0_)
+    public static void registerFixesSlime(DataFixer fixer)
     {
-        EntityLiving.func_189752_a(p_189758_0_, EntitySlime.class);
+        EntityLiving.registerFixesMob(fixer, EntitySlime.class);
     }
 
-    public void func_70014_b(NBTTagCompound p_70014_1_)
+    public void writeEntityToNBT(NBTTagCompound compound)
     {
-        super.func_70014_b(p_70014_1_);
-        p_70014_1_.func_74768_a("Size", this.func_70809_q() - 1);
-        p_70014_1_.func_74757_a("wasOnGround", this.field_175452_bi);
+        super.writeEntityToNBT(compound);
+        compound.setInteger("Size", this.getSlimeSize() - 1);
+        compound.setBoolean("wasOnGround", this.wasOnGround);
     }
 
-    public void func_70037_a(NBTTagCompound p_70037_1_)
+    public void readEntityFromNBT(NBTTagCompound compound)
     {
-        super.func_70037_a(p_70037_1_);
-        int i = p_70037_1_.func_74762_e("Size");
+        super.readEntityFromNBT(compound);
+        int i = compound.getInteger("Size");
 
         if (i < 0)
         {
             i = 0;
         }
 
-        this.func_70799_a(i + 1, false);
-        this.field_175452_bi = p_70037_1_.func_74767_n("wasOnGround");
+        this.setSlimeSize(i + 1, false);
+        this.wasOnGround = compound.getBoolean("wasOnGround");
     }
 
-    public boolean func_189101_db()
+    public boolean isSmallSlime()
     {
-        return this.func_70809_q() <= 1;
+        return this.getSlimeSize() <= 1;
     }
 
-    protected EnumParticleTypes func_180487_n()
+    protected EnumParticleTypes getParticleType()
     {
         return EnumParticleTypes.SLIME;
     }
 
-    public void func_70071_h_()
+    public void onUpdate()
     {
-        if (!this.field_70170_p.field_72995_K && this.field_70170_p.func_175659_aa() == EnumDifficulty.PEACEFUL && this.func_70809_q() > 0)
+        if (!this.world.isRemote && this.world.getDifficulty() == EnumDifficulty.PEACEFUL && this.getSlimeSize() > 0)
         {
-            this.field_70128_L = true;
+            this.isDead = true;
         }
 
-        this.field_70811_b += (this.field_70813_a - this.field_70811_b) * 0.5F;
-        this.field_70812_c = this.field_70811_b;
-        super.func_70071_h_();
+        this.squishFactor += (this.squishAmount - this.squishFactor) * 0.5F;
+        this.prevSquishFactor = this.squishFactor;
+        super.onUpdate();
 
-        if (this.field_70122_E && !this.field_175452_bi)
+        if (this.onGround && !this.wasOnGround)
         {
-            int i = this.func_70809_q();
-
+            int i = this.getSlimeSize();
+            if (spawnCustomParticles()) { i = 0; } // don't spawn particles if it's handled by the implementation itself
             for (int j = 0; j < i * 8; ++j)
             {
-                float f = this.field_70146_Z.nextFloat() * ((float)Math.PI * 2F);
-                float f1 = this.field_70146_Z.nextFloat() * 0.5F + 0.5F;
-                float f2 = MathHelper.func_76126_a(f) * (float)i * 0.5F * f1;
-                float f3 = MathHelper.func_76134_b(f) * (float)i * 0.5F * f1;
-                World world = this.field_70170_p;
-                EnumParticleTypes enumparticletypes = this.func_180487_n();
-                double d0 = this.field_70165_t + (double)f2;
-                double d1 = this.field_70161_v + (double)f3;
-                world.func_175688_a(enumparticletypes, d0, this.func_174813_aQ().field_72338_b, d1, 0.0D, 0.0D, 0.0D, new int[0]);
+                float f = this.rand.nextFloat() * ((float)Math.PI * 2F);
+                float f1 = this.rand.nextFloat() * 0.5F + 0.5F;
+                float f2 = MathHelper.sin(f) * (float)i * 0.5F * f1;
+                float f3 = MathHelper.cos(f) * (float)i * 0.5F * f1;
+                World world = this.world;
+                EnumParticleTypes enumparticletypes = this.getParticleType();
+                double d0 = this.posX + (double)f2;
+                double d1 = this.posZ + (double)f3;
+                world.spawnParticle(enumparticletypes, d0, this.getEntityBoundingBox().minY, d1, 0.0D, 0.0D, 0.0D, new int[0]);
             }
 
-            this.func_184185_a(this.func_184709_cY(), this.func_70599_aP(), ((this.field_70146_Z.nextFloat() - this.field_70146_Z.nextFloat()) * 0.2F + 1.0F) / 0.8F);
-            this.field_70813_a = -0.5F;
+            this.playSound(this.getSquishSound(), this.getSoundVolume(), ((this.rand.nextFloat() - this.rand.nextFloat()) * 0.2F + 1.0F) / 0.8F);
+            this.squishAmount = -0.5F;
         }
-        else if (!this.field_70122_E && this.field_175452_bi)
+        else if (!this.onGround && this.wasOnGround)
         {
-            this.field_70813_a = 1.0F;
+            this.squishAmount = 1.0F;
         }
 
-        this.field_175452_bi = this.field_70122_E;
-        this.func_70808_l();
+        this.wasOnGround = this.onGround;
+        this.alterSquishAmount();
     }
 
-    protected void func_70808_l()
+    protected void alterSquishAmount()
     {
-        this.field_70813_a *= 0.6F;
+        this.squishAmount *= 0.6F;
     }
 
-    protected int func_70806_k()
+    protected int getJumpDelay()
     {
-        return this.field_70146_Z.nextInt(20) + 10;
+        return this.rand.nextInt(20) + 10;
     }
 
-    protected EntitySlime func_70802_j()
+    protected EntitySlime createInstance()
     {
-        return new EntitySlime(this.field_70170_p);
+        return new EntitySlime(this.world);
     }
 
-    public void func_184206_a(DataParameter<?> p_184206_1_)
+    public void notifyDataManagerChange(DataParameter<?> key)
     {
-        if (field_184711_bt.equals(p_184206_1_))
+        if (SLIME_SIZE.equals(key))
         {
-            int i = this.func_70809_q();
-            this.func_70105_a(0.51000005F * (float)i, 0.51000005F * (float)i);
-            this.field_70177_z = this.field_70759_as;
-            this.field_70761_aq = this.field_70759_as;
+            int i = this.getSlimeSize();
+            this.setSize(0.51000005F * (float)i, 0.51000005F * (float)i);
+            this.rotationYaw = this.rotationYawHead;
+            this.renderYawOffset = this.rotationYawHead;
 
-            if (this.func_70090_H() && this.field_70146_Z.nextInt(20) == 0)
+            if (this.isInWater() && this.rand.nextInt(20) == 0)
             {
-                this.func_71061_d_();
+                this.resetHeight();
             }
         }
 
-        super.func_184206_a(p_184206_1_);
+        super.notifyDataManagerChange(key);
     }
 
-    public void func_70106_y()
+    public void setDead()
     {
-        int i = this.func_70809_q();
+        int i = this.getSlimeSize();
 
-        if (!this.field_70170_p.field_72995_K && i > 1 && this.func_110143_aJ() <= 0.0F)
+        if (!this.world.isRemote && i > 1 && this.getHealth() <= 0.0F)
         {
-            int j = 2 + this.field_70146_Z.nextInt(3);
 
+            int j = 2 + this.rand.nextInt(3);
+            // CraftBukkit start
+            SlimeSplitEvent event = new SlimeSplitEvent((org.bukkit.entity.Slime) this.getBukkitEntity(), j);
+            this.world.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled() && event.getCount() > 0) {
+                j = event.getCount();
+            } else {
+                super.setDead();
+                return;
+            }
+            // CraftBukkit end
             for (int k = 0; k < j; ++k)
             {
                 float f = ((float)(k % 2) - 0.5F) * (float)i / 4.0F;
                 float f1 = ((float)(k / 2) - 0.5F) * (float)i / 4.0F;
-                EntitySlime entityslime = this.func_70802_j();
+                EntitySlime entityslime = this.createInstance();
 
-                if (this.func_145818_k_())
+                if (this.hasCustomName())
                 {
-                    entityslime.func_96094_a(this.func_95999_t());
+                    entityslime.setCustomNameTag(this.getCustomNameTag());
                 }
 
-                if (this.func_104002_bU())
+                if (this.isNoDespawnRequired())
                 {
-                    entityslime.func_110163_bv();
+                    entityslime.enablePersistence();
                 }
 
-                entityslime.func_70799_a(i / 2, true);
-                entityslime.func_70012_b(this.field_70165_t + (double)f, this.field_70163_u + 0.5D, this.field_70161_v + (double)f1, this.field_70146_Z.nextFloat() * 360.0F, 0.0F);
-                this.field_70170_p.func_72838_d(entityslime);
+                entityslime.setSlimeSize(i / 2, true);
+                entityslime.setLocationAndAngles(this.posX + (double)f, this.posY + 0.5D, this.posZ + (double)f1, this.rand.nextFloat() * 360.0F, 0.0F);
+                this.world.spawnEntity(entityslime);
             }
         }
 
-        super.func_70106_y();
+        super.setDead();
     }
 
-    public void func_70108_f(Entity p_70108_1_)
+    public void applyEntityCollision(Entity entityIn)
     {
-        super.func_70108_f(p_70108_1_);
+        super.applyEntityCollision(entityIn);
 
-        if (p_70108_1_ instanceof EntityIronGolem && this.func_70800_m())
+        if (entityIn instanceof EntityIronGolem && this.canDamagePlayer())
         {
-            this.func_175451_e((EntityLivingBase)p_70108_1_);
+            this.dealDamage((EntityLivingBase)entityIn);
         }
     }
 
-    public void func_70100_b_(EntityPlayer p_70100_1_)
+    public void onCollideWithPlayer(EntityPlayer entityIn)
     {
-        if (this.func_70800_m())
+        if (this.canDamagePlayer())
         {
-            this.func_175451_e(p_70100_1_);
+            this.dealDamage(entityIn);
         }
     }
 
-    protected void func_175451_e(EntityLivingBase p_175451_1_)
+    protected void dealDamage(EntityLivingBase entityIn)
     {
-        int i = this.func_70809_q();
+        int i = this.getSlimeSize();
 
-        if (this.func_70685_l(p_175451_1_) && this.func_70068_e(p_175451_1_) < 0.6D * (double)i * 0.6D * (double)i && p_175451_1_.func_70097_a(DamageSource.func_76358_a(this), (float)this.func_70805_n()))
+        if (this.canEntityBeSeen(entityIn) && this.getDistanceSqToEntity(entityIn) < 0.6D * (double)i * 0.6D * (double)i && entityIn.attackEntityFrom(DamageSource.causeMobDamage(this), (float)this.getAttackStrength()))
         {
-            this.func_184185_a(SoundEvents.field_187870_fk, 1.0F, (this.field_70146_Z.nextFloat() - this.field_70146_Z.nextFloat()) * 0.2F + 1.0F);
-            this.func_174815_a(this, p_175451_1_);
+            this.playSound(SoundEvents.ENTITY_SLIME_ATTACK, 1.0F, (this.rand.nextFloat() - this.rand.nextFloat()) * 0.2F + 1.0F);
+            this.applyEnchantments(this, entityIn);
         }
     }
 
-    public float func_70047_e()
+    public float getEyeHeight()
     {
-        return 0.625F * this.field_70131_O;
+        return 0.625F * this.height;
     }
 
-    protected boolean func_70800_m()
+    protected boolean canDamagePlayer()
     {
-        return !this.func_189101_db();
+        return !this.isSmallSlime();
     }
 
-    protected int func_70805_n()
+    protected int getAttackStrength()
     {
-        return this.func_70809_q();
+        return this.getSlimeSize();
     }
 
-    protected SoundEvent func_184601_bQ()
+    protected SoundEvent getHurtSound()
     {
-        return this.func_189101_db() ? SoundEvents.field_187898_fy : SoundEvents.field_187880_fp;
+        return this.isSmallSlime() ? SoundEvents.ENTITY_SMALL_SLIME_HURT : SoundEvents.ENTITY_SLIME_HURT;
     }
 
-    protected SoundEvent func_184615_bR()
+    protected SoundEvent getDeathSound()
     {
-        return this.func_189101_db() ? SoundEvents.field_187896_fx : SoundEvents.field_187874_fm;
+        return this.isSmallSlime() ? SoundEvents.ENTITY_SMALL_SLIME_DEATH : SoundEvents.ENTITY_SLIME_DEATH;
     }
 
-    protected SoundEvent func_184709_cY()
+    protected SoundEvent getSquishSound()
     {
-        return this.func_189101_db() ? SoundEvents.field_187900_fz : SoundEvents.field_187886_fs;
+        return this.isSmallSlime() ? SoundEvents.ENTITY_SMALL_SLIME_SQUISH : SoundEvents.ENTITY_SLIME_SQUISH;
     }
 
-    protected Item func_146068_u()
+    protected Item getDropItem()
     {
-        return this.func_70809_q() == 1 ? Items.field_151123_aH : null;
+        return this.getSlimeSize() == 1 ? Items.SLIME_BALL : null;
     }
 
     @Nullable
-    protected ResourceLocation func_184647_J()
+    protected ResourceLocation getLootTable()
     {
-        return this.func_70809_q() == 1 ? LootTableList.field_186378_ac : LootTableList.field_186419_a;
+        return this.getSlimeSize() == 1 ? LootTableList.ENTITIES_SLIME : LootTableList.EMPTY;
     }
 
-    public boolean func_70601_bi()
+    public boolean getCanSpawnHere()
     {
-        BlockPos blockpos = new BlockPos(MathHelper.func_76128_c(this.field_70165_t), 0, MathHelper.func_76128_c(this.field_70161_v));
-        Chunk chunk = this.field_70170_p.func_175726_f(blockpos);
+        BlockPos blockpos = new BlockPos(MathHelper.floor(this.posX), 0, MathHelper.floor(this.posZ));
+        Chunk chunk = this.world.getChunkFromBlockCoords(blockpos);
 
-        if (this.field_70170_p.func_72912_H().func_76067_t() == WorldType.field_77138_c && this.field_70146_Z.nextInt(4) != 1)
+        if (this.world.getWorldInfo().getTerrainType().handleSlimeSpawnReduction(rand, world))
         {
             return false;
         }
         else
         {
-            if (this.field_70170_p.func_175659_aa() != EnumDifficulty.PEACEFUL)
+            if (this.world.getDifficulty() != EnumDifficulty.PEACEFUL)
             {
-                Biome biome = this.field_70170_p.func_180494_b(blockpos);
+                Biome biome = this.world.getBiome(blockpos);
 
-                if (biome == Biomes.field_76780_h && this.field_70163_u > 50.0D && this.field_70163_u < 70.0D && this.field_70146_Z.nextFloat() < 0.5F && this.field_70146_Z.nextFloat() < this.field_70170_p.func_130001_d() && this.field_70170_p.func_175671_l(new BlockPos(this)) <= this.field_70146_Z.nextInt(8))
+                if (biome == Biomes.SWAMPLAND && this.posY > 50.0D && this.posY < 70.0D && this.rand.nextFloat() < 0.5F && this.rand.nextFloat() < this.world.getCurrentMoonPhaseFactor() && this.world.getLightFromNeighbors(new BlockPos(this)) <= this.rand.nextInt(8))
                 {
-                    return super.func_70601_bi();
+                    return super.getCanSpawnHere();
                 }
 
-                if (this.field_70146_Z.nextInt(10) == 0 && chunk.func_76617_a(987234911L).nextInt(10) == 0 && this.field_70163_u < 40.0D)
+                if (this.rand.nextInt(10) == 0 && chunk.getRandomWithSeed(987234911L).nextInt(10) == 0 && this.posY < 40.0D)
                 {
-                    return super.func_70601_bi();
+                    return super.getCanSpawnHere();
                 }
             }
 
@@ -329,230 +341,238 @@
         }
     }
 
-    protected float func_70599_aP()
+    protected float getSoundVolume()
     {
-        return 0.4F * (float)this.func_70809_q();
+        return 0.4F * (float)this.getSlimeSize();
     }
 
-    public int func_70646_bf()
+    public int getVerticalFaceSpeed()
     {
         return 0;
     }
 
-    protected boolean func_70807_r()
+    protected boolean makesSoundOnJump()
     {
-        return this.func_70809_q() > 0;
+        return this.getSlimeSize() > 0;
     }
 
-    protected void func_70664_aZ()
+    protected void jump()
     {
-        this.field_70181_x = 0.41999998688697815D;
-        this.field_70160_al = true;
+        this.motionY = 0.41999998688697815D;
+        this.isAirBorne = true;
     }
 
     @Nullable
-    public IEntityLivingData func_180482_a(DifficultyInstance p_180482_1_, @Nullable IEntityLivingData p_180482_2_)
+    public IEntityLivingData onInitialSpawn(DifficultyInstance difficulty, @Nullable IEntityLivingData livingdata)
     {
-        int i = this.field_70146_Z.nextInt(3);
+        int i = this.rand.nextInt(3);
 
-        if (i < 2 && this.field_70146_Z.nextFloat() < 0.5F * p_180482_1_.func_180170_c())
+        if (i < 2 && this.rand.nextFloat() < 0.5F * difficulty.getClampedAdditionalDifficulty())
         {
             ++i;
         }
 
         int j = 1 << i;
-        this.func_70799_a(j, true);
-        return super.func_180482_a(p_180482_1_, p_180482_2_);
+        this.setSlimeSize(j, true);
+        return super.onInitialSpawn(difficulty, livingdata);
     }
 
-    protected SoundEvent func_184710_cZ()
+    protected SoundEvent getJumpSound()
     {
-        return this.func_189101_db() ? SoundEvents.field_189110_fE : SoundEvents.field_187882_fq;
+        return this.isSmallSlime() ? SoundEvents.ENTITY_SMALL_SLIME_JUMP : SoundEvents.ENTITY_SLIME_JUMP;
     }
 
+    /* ======================================== FORGE START =====================================*/
+    /**
+     * Called when the slime spawns particles on landing, see onUpdate.
+     * Return true to prevent the spawning of the default particles.
+     */
+    protected boolean spawnCustomParticles() { return false; }
+    /* ======================================== FORGE END   =====================================*/
+
     static class AISlimeAttack extends EntityAIBase
         {
-            private final EntitySlime field_179466_a;
-            private int field_179465_b;
+            private final EntitySlime slime;
+            private int growTieredTimer;
 
-            public AISlimeAttack(EntitySlime p_i45824_1_)
+            public AISlimeAttack(EntitySlime slimeIn)
             {
-                this.field_179466_a = p_i45824_1_;
-                this.func_75248_a(2);
+                this.slime = slimeIn;
+                this.setMutexBits(2);
             }
 
-            public boolean func_75250_a()
+            public boolean shouldExecute()
             {
-                EntityLivingBase entitylivingbase = this.field_179466_a.func_70638_az();
-                return entitylivingbase == null ? false : (!entitylivingbase.func_70089_S() ? false : !(entitylivingbase instanceof EntityPlayer) || !((EntityPlayer)entitylivingbase).field_71075_bZ.field_75102_a);
+                EntityLivingBase entitylivingbase = this.slime.getAttackTarget();
+                return entitylivingbase == null ? false : (!entitylivingbase.isEntityAlive() ? false : !(entitylivingbase instanceof EntityPlayer) || !((EntityPlayer)entitylivingbase).capabilities.disableDamage);
             }
 
-            public void func_75249_e()
+            public void startExecuting()
             {
-                this.field_179465_b = 300;
-                super.func_75249_e();
+                this.growTieredTimer = 300;
+                super.startExecuting();
             }
 
-            public boolean func_75253_b()
+            public boolean continueExecuting()
             {
-                EntityLivingBase entitylivingbase = this.field_179466_a.func_70638_az();
-                return entitylivingbase == null ? false : (!entitylivingbase.func_70089_S() ? false : (entitylivingbase instanceof EntityPlayer && ((EntityPlayer)entitylivingbase).field_71075_bZ.field_75102_a ? false : --this.field_179465_b > 0));
+                EntityLivingBase entitylivingbase = this.slime.getAttackTarget();
+                return entitylivingbase == null ? false : (!entitylivingbase.isEntityAlive() ? false : (entitylivingbase instanceof EntityPlayer && ((EntityPlayer)entitylivingbase).capabilities.disableDamage ? false : --this.growTieredTimer > 0));
             }
 
-            public void func_75246_d()
+            public void updateTask()
             {
-                this.field_179466_a.func_70625_a(this.field_179466_a.func_70638_az(), 10.0F, 10.0F);
-                ((EntitySlime.SlimeMoveHelper)this.field_179466_a.func_70605_aq()).func_179920_a(this.field_179466_a.field_70177_z, this.field_179466_a.func_70800_m());
+                this.slime.faceEntity(this.slime.getAttackTarget(), 10.0F, 10.0F);
+                ((EntitySlime.SlimeMoveHelper)this.slime.getMoveHelper()).setDirection(this.slime.rotationYaw, this.slime.canDamagePlayer());
             }
         }
 
     static class AISlimeFaceRandom extends EntityAIBase
         {
-            private final EntitySlime field_179461_a;
-            private float field_179459_b;
-            private int field_179460_c;
+            private final EntitySlime slime;
+            private float chosenDegrees;
+            private int nextRandomizeTime;
 
-            public AISlimeFaceRandom(EntitySlime p_i45820_1_)
+            public AISlimeFaceRandom(EntitySlime slimeIn)
             {
-                this.field_179461_a = p_i45820_1_;
-                this.func_75248_a(2);
+                this.slime = slimeIn;
+                this.setMutexBits(2);
             }
 
-            public boolean func_75250_a()
+            public boolean shouldExecute()
             {
-                return this.field_179461_a.func_70638_az() == null && (this.field_179461_a.field_70122_E || this.field_179461_a.func_70090_H() || this.field_179461_a.func_180799_ab() || this.field_179461_a.func_70644_a(MobEffects.field_188424_y));
+                return this.slime.getAttackTarget() == null && (this.slime.onGround || this.slime.isInWater() || this.slime.isInLava() || this.slime.isPotionActive(MobEffects.LEVITATION));
             }
 
-            public void func_75246_d()
+            public void updateTask()
             {
-                if (--this.field_179460_c <= 0)
+                if (--this.nextRandomizeTime <= 0)
                 {
-                    this.field_179460_c = 40 + this.field_179461_a.func_70681_au().nextInt(60);
-                    this.field_179459_b = (float)this.field_179461_a.func_70681_au().nextInt(360);
+                    this.nextRandomizeTime = 40 + this.slime.getRNG().nextInt(60);
+                    this.chosenDegrees = (float)this.slime.getRNG().nextInt(360);
                 }
 
-                ((EntitySlime.SlimeMoveHelper)this.field_179461_a.func_70605_aq()).func_179920_a(this.field_179459_b, false);
+                ((EntitySlime.SlimeMoveHelper)this.slime.getMoveHelper()).setDirection(this.chosenDegrees, false);
             }
         }
 
     static class AISlimeFloat extends EntityAIBase
         {
-            private final EntitySlime field_179457_a;
+            private final EntitySlime slime;
 
-            public AISlimeFloat(EntitySlime p_i45823_1_)
+            public AISlimeFloat(EntitySlime slimeIn)
             {
-                this.field_179457_a = p_i45823_1_;
-                this.func_75248_a(5);
-                ((PathNavigateGround)p_i45823_1_.func_70661_as()).func_179693_d(true);
+                this.slime = slimeIn;
+                this.setMutexBits(5);
+                ((PathNavigateGround)slimeIn.getNavigator()).setCanSwim(true);
             }
 
-            public boolean func_75250_a()
+            public boolean shouldExecute()
             {
-                return this.field_179457_a.func_70090_H() || this.field_179457_a.func_180799_ab();
+                return this.slime.isInWater() || this.slime.isInLava();
             }
 
-            public void func_75246_d()
+            public void updateTask()
             {
-                if (this.field_179457_a.func_70681_au().nextFloat() < 0.8F)
+                if (this.slime.getRNG().nextFloat() < 0.8F)
                 {
-                    this.field_179457_a.func_70683_ar().func_75660_a();
+                    this.slime.getJumpHelper().setJumping();
                 }
 
-                ((EntitySlime.SlimeMoveHelper)this.field_179457_a.func_70605_aq()).func_179921_a(1.2D);
+                ((EntitySlime.SlimeMoveHelper)this.slime.getMoveHelper()).setSpeed(1.2D);
             }
         }
 
     static class AISlimeHop extends EntityAIBase
         {
-            private final EntitySlime field_179458_a;
+            private final EntitySlime slime;
 
-            public AISlimeHop(EntitySlime p_i45822_1_)
+            public AISlimeHop(EntitySlime slimeIn)
             {
-                this.field_179458_a = p_i45822_1_;
-                this.func_75248_a(5);
+                this.slime = slimeIn;
+                this.setMutexBits(5);
             }
 
-            public boolean func_75250_a()
+            public boolean shouldExecute()
             {
                 return true;
             }
 
-            public void func_75246_d()
+            public void updateTask()
             {
-                ((EntitySlime.SlimeMoveHelper)this.field_179458_a.func_70605_aq()).func_179921_a(1.0D);
+                ((EntitySlime.SlimeMoveHelper)this.slime.getMoveHelper()).setSpeed(1.0D);
             }
         }
 
     static class SlimeMoveHelper extends EntityMoveHelper
         {
-            private float field_179922_g;
-            private int field_179924_h;
-            private final EntitySlime field_179925_i;
-            private boolean field_179923_j;
+            private float yRot;
+            private int jumpDelay;
+            private final EntitySlime slime;
+            private boolean isAggressive;
 
-            public SlimeMoveHelper(EntitySlime p_i45821_1_)
+            public SlimeMoveHelper(EntitySlime slimeIn)
             {
-                super(p_i45821_1_);
-                this.field_179925_i = p_i45821_1_;
-                this.field_179922_g = 180.0F * p_i45821_1_.field_70177_z / (float)Math.PI;
+                super(slimeIn);
+                this.slime = slimeIn;
+                this.yRot = 180.0F * slimeIn.rotationYaw / (float)Math.PI;
             }
 
-            public void func_179920_a(float p_179920_1_, boolean p_179920_2_)
+            public void setDirection(float p_179920_1_, boolean p_179920_2_)
             {
-                this.field_179922_g = p_179920_1_;
-                this.field_179923_j = p_179920_2_;
+                this.yRot = p_179920_1_;
+                this.isAggressive = p_179920_2_;
             }
 
-            public void func_179921_a(double p_179921_1_)
+            public void setSpeed(double speedIn)
             {
-                this.field_75645_e = p_179921_1_;
-                this.field_188491_h = EntityMoveHelper.Action.MOVE_TO;
+                this.speed = speedIn;
+                this.action = EntityMoveHelper.Action.MOVE_TO;
             }
 
-            public void func_75641_c()
+            public void onUpdateMoveHelper()
             {
-                this.field_75648_a.field_70177_z = this.func_75639_a(this.field_75648_a.field_70177_z, this.field_179922_g, 90.0F);
-                this.field_75648_a.field_70759_as = this.field_75648_a.field_70177_z;
-                this.field_75648_a.field_70761_aq = this.field_75648_a.field_70177_z;
+                this.entity.rotationYaw = this.limitAngle(this.entity.rotationYaw, this.yRot, 90.0F);
+                this.entity.rotationYawHead = this.entity.rotationYaw;
+                this.entity.renderYawOffset = this.entity.rotationYaw;
 
-                if (this.field_188491_h != EntityMoveHelper.Action.MOVE_TO)
+                if (this.action != EntityMoveHelper.Action.MOVE_TO)
                 {
-                    this.field_75648_a.func_70657_f(0.0F);
+                    this.entity.setMoveForward(0.0F);
                 }
                 else
                 {
-                    this.field_188491_h = EntityMoveHelper.Action.WAIT;
+                    this.action = EntityMoveHelper.Action.WAIT;
 
-                    if (this.field_75648_a.field_70122_E)
+                    if (this.entity.onGround)
                     {
-                        this.field_75648_a.func_70659_e((float)(this.field_75645_e * this.field_75648_a.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111126_e()));
+                        this.entity.setAIMoveSpeed((float)(this.speed * this.entity.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).getAttributeValue()));
 
-                        if (this.field_179924_h-- <= 0)
+                        if (this.jumpDelay-- <= 0)
                         {
-                            this.field_179924_h = this.field_179925_i.func_70806_k();
+                            this.jumpDelay = this.slime.getJumpDelay();
 
-                            if (this.field_179923_j)
+                            if (this.isAggressive)
                             {
-                                this.field_179924_h /= 3;
+                                this.jumpDelay /= 3;
                             }
 
-                            this.field_179925_i.func_70683_ar().func_75660_a();
+                            this.slime.getJumpHelper().setJumping();
 
-                            if (this.field_179925_i.func_70807_r())
+                            if (this.slime.makesSoundOnJump())
                             {
-                                this.field_179925_i.func_184185_a(this.field_179925_i.func_184710_cZ(), this.field_179925_i.func_70599_aP(), ((this.field_179925_i.func_70681_au().nextFloat() - this.field_179925_i.func_70681_au().nextFloat()) * 0.2F + 1.0F) * 0.8F);
+                                this.slime.playSound(this.slime.getJumpSound(), this.slime.getSoundVolume(), ((this.slime.getRNG().nextFloat() - this.slime.getRNG().nextFloat()) * 0.2F + 1.0F) * 0.8F);
                             }
                         }
                         else
                         {
-                            this.field_179925_i.field_70702_br = 0.0F;
-                            this.field_179925_i.field_70701_bs = 0.0F;
-                            this.field_75648_a.func_70659_e(0.0F);
+                            this.slime.moveStrafing = 0.0F;
+                            this.slime.moveForward = 0.0F;
+                            this.entity.setAIMoveSpeed(0.0F);
                         }
                     }
                     else
                     {
-                        this.field_75648_a.func_70659_e((float)(this.field_75645_e * this.field_75648_a.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111126_e()));
+                        this.entity.setAIMoveSpeed((float)(this.speed * this.entity.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).getAttributeValue()));
                     }
                 }
             }
