--- ../src-base/minecraft/net/minecraft/entity/monster/EntitySpider.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntitySpider.java
@@ -37,120 +37,120 @@
 
 public class EntitySpider extends EntityMob
 {
-    private static final DataParameter<Byte> field_184729_a = EntityDataManager.<Byte>func_187226_a(EntitySpider.class, DataSerializers.field_187191_a);
+    private static final DataParameter<Byte> CLIMBING = EntityDataManager.<Byte>createKey(EntitySpider.class, DataSerializers.BYTE);
 
-    public EntitySpider(World p_i1743_1_)
+    public EntitySpider(World worldIn)
     {
-        super(p_i1743_1_);
-        this.func_70105_a(1.4F, 0.9F);
+        super(worldIn);
+        this.setSize(1.4F, 0.9F);
     }
 
-    public static void func_189774_d(DataFixer p_189774_0_)
+    public static void registerFixesSpider(DataFixer fixer)
     {
-        EntityLiving.func_189752_a(p_189774_0_, EntitySpider.class);
+        EntityLiving.registerFixesMob(fixer, EntitySpider.class);
     }
 
-    protected void func_184651_r()
+    protected void initEntityAI()
     {
-        this.field_70714_bg.func_75776_a(1, new EntityAISwimming(this));
-        this.field_70714_bg.func_75776_a(3, new EntityAILeapAtTarget(this, 0.4F));
-        this.field_70714_bg.func_75776_a(4, new EntitySpider.AISpiderAttack(this));
-        this.field_70714_bg.func_75776_a(5, new EntityAIWanderAvoidWater(this, 0.8D));
-        this.field_70714_bg.func_75776_a(6, new EntityAIWatchClosest(this, EntityPlayer.class, 8.0F));
-        this.field_70714_bg.func_75776_a(6, new EntityAILookIdle(this));
-        this.field_70715_bh.func_75776_a(1, new EntityAIHurtByTarget(this, false, new Class[0]));
-        this.field_70715_bh.func_75776_a(2, new EntitySpider.AISpiderTarget(this, EntityPlayer.class));
-        this.field_70715_bh.func_75776_a(3, new EntitySpider.AISpiderTarget(this, EntityIronGolem.class));
+        this.tasks.addTask(1, new EntityAISwimming(this));
+        this.tasks.addTask(3, new EntityAILeapAtTarget(this, 0.4F));
+        this.tasks.addTask(4, new EntitySpider.AISpiderAttack(this));
+        this.tasks.addTask(5, new EntityAIWanderAvoidWater(this, 0.8D));
+        this.tasks.addTask(6, new EntityAIWatchClosest(this, EntityPlayer.class, 8.0F));
+        this.tasks.addTask(6, new EntityAILookIdle(this));
+        this.targetTasks.addTask(1, new EntityAIHurtByTarget(this, false, new Class[0]));
+        this.targetTasks.addTask(2, new EntitySpider.AISpiderTarget(this, EntityPlayer.class));
+        this.targetTasks.addTask(3, new EntitySpider.AISpiderTarget(this, EntityIronGolem.class));
     }
 
-    public double func_70042_X()
+    public double getMountedYOffset()
     {
-        return (double)(this.field_70131_O * 0.5F);
+        return (double)(this.height * 0.5F);
     }
 
-    protected PathNavigate func_175447_b(World p_175447_1_)
+    protected PathNavigate createNavigator(World worldIn)
     {
-        return new PathNavigateClimber(this, p_175447_1_);
+        return new PathNavigateClimber(this, worldIn);
     }
 
-    protected void func_70088_a()
+    protected void entityInit()
     {
-        super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184729_a, Byte.valueOf((byte)0));
+        super.entityInit();
+        this.dataManager.register(CLIMBING, Byte.valueOf((byte)0));
     }
 
-    public void func_70071_h_()
+    public void onUpdate()
     {
-        super.func_70071_h_();
+        super.onUpdate();
 
-        if (!this.field_70170_p.field_72995_K)
+        if (!this.world.isRemote)
         {
-            this.func_70839_e(this.field_70123_F);
+            this.setBesideClimbableBlock(this.isCollidedHorizontally);
         }
     }
 
-    protected void func_110147_ax()
+    protected void applyEntityAttributes()
     {
-        super.func_110147_ax();
-        this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(16.0D);
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.30000001192092896D);
+        super.applyEntityAttributes();
+        this.getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).setBaseValue(16.0D);
+        this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).setBaseValue(0.30000001192092896D);
     }
 
-    protected SoundEvent func_184639_G()
+    protected SoundEvent getAmbientSound()
     {
-        return SoundEvents.field_187817_fK;
+        return SoundEvents.ENTITY_SPIDER_AMBIENT;
     }
 
-    protected SoundEvent func_184601_bQ()
+    protected SoundEvent getHurtSound()
     {
-        return SoundEvents.field_187821_fM;
+        return SoundEvents.ENTITY_SPIDER_HURT;
     }
 
-    protected SoundEvent func_184615_bR()
+    protected SoundEvent getDeathSound()
     {
-        return SoundEvents.field_187819_fL;
+        return SoundEvents.ENTITY_SPIDER_DEATH;
     }
 
-    protected void func_180429_a(BlockPos p_180429_1_, Block p_180429_2_)
+    protected void playStepSound(BlockPos pos, Block blockIn)
     {
-        this.func_184185_a(SoundEvents.field_187823_fN, 0.15F, 1.0F);
+        this.playSound(SoundEvents.ENTITY_SPIDER_STEP, 0.15F, 1.0F);
     }
 
     @Nullable
-    protected ResourceLocation func_184647_J()
+    protected ResourceLocation getLootTable()
     {
-        return LootTableList.field_186435_q;
+        return LootTableList.ENTITIES_SPIDER;
     }
 
-    public boolean func_70617_f_()
+    public boolean isOnLadder()
     {
-        return this.func_70841_p();
+        return this.isBesideClimbableBlock();
     }
 
-    public void func_70110_aj()
+    public void setInWeb()
     {
     }
 
-    public EnumCreatureAttribute func_70668_bt()
+    public EnumCreatureAttribute getCreatureAttribute()
     {
         return EnumCreatureAttribute.ARTHROPOD;
     }
 
-    public boolean func_70687_e(PotionEffect p_70687_1_)
+    public boolean isPotionApplicable(PotionEffect potioneffectIn)
     {
-        return p_70687_1_.func_188419_a() == MobEffects.field_76436_u ? false : super.func_70687_e(p_70687_1_);
+        return potioneffectIn.getPotion() == MobEffects.POISON ? false : super.isPotionApplicable(potioneffectIn);
     }
 
-    public boolean func_70841_p()
+    public boolean isBesideClimbableBlock()
     {
-        return (((Byte)this.field_70180_af.func_187225_a(field_184729_a)).byteValue() & 1) != 0;
+        return (((Byte)this.dataManager.get(CLIMBING)).byteValue() & 1) != 0;
     }
 
-    public void func_70839_e(boolean p_70839_1_)
+    public void setBesideClimbableBlock(boolean climbing)
     {
-        byte b0 = ((Byte)this.field_70180_af.func_187225_a(field_184729_a)).byteValue();
+        byte b0 = ((Byte)this.dataManager.get(CLIMBING)).byteValue();
 
-        if (p_70839_1_)
+        if (climbing)
         {
             b0 = (byte)(b0 | 1);
         }
@@ -159,116 +159,116 @@
             b0 = (byte)(b0 & -2);
         }
 
-        this.field_70180_af.func_187227_b(field_184729_a, Byte.valueOf(b0));
+        this.dataManager.set(CLIMBING, Byte.valueOf(b0));
     }
 
     @Nullable
-    public IEntityLivingData func_180482_a(DifficultyInstance p_180482_1_, @Nullable IEntityLivingData p_180482_2_)
+    public IEntityLivingData onInitialSpawn(DifficultyInstance difficulty, @Nullable IEntityLivingData livingdata)
     {
-        p_180482_2_ = super.func_180482_a(p_180482_1_, p_180482_2_);
+        livingdata = super.onInitialSpawn(difficulty, livingdata);
 
-        if (this.field_70170_p.field_73012_v.nextInt(100) == 0)
+        if (this.world.rand.nextInt(100) == 0)
         {
-            EntitySkeleton entityskeleton = new EntitySkeleton(this.field_70170_p);
-            entityskeleton.func_70012_b(this.field_70165_t, this.field_70163_u, this.field_70161_v, this.field_70177_z, 0.0F);
-            entityskeleton.func_180482_a(p_180482_1_, (IEntityLivingData)null);
-            this.field_70170_p.func_72838_d(entityskeleton);
-            entityskeleton.func_184220_m(this);
+            EntitySkeleton entityskeleton = new EntitySkeleton(this.world);
+            entityskeleton.setLocationAndAngles(this.posX, this.posY, this.posZ, this.rotationYaw, 0.0F);
+            entityskeleton.onInitialSpawn(difficulty, (IEntityLivingData)null);
+            this.world.addEntity(entityskeleton, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.JOCKEY); // CraftBukkit - add SpawnReason
+            entityskeleton.startRiding(this);
         }
 
-        if (p_180482_2_ == null)
+        if (livingdata == null)
         {
-            p_180482_2_ = new EntitySpider.GroupData();
+            livingdata = new EntitySpider.GroupData();
 
-            if (this.field_70170_p.func_175659_aa() == EnumDifficulty.HARD && this.field_70170_p.field_73012_v.nextFloat() < 0.1F * p_180482_1_.func_180170_c())
+            if (this.world.getDifficulty() == EnumDifficulty.HARD && this.world.rand.nextFloat() < 0.1F * difficulty.getClampedAdditionalDifficulty())
             {
-                ((EntitySpider.GroupData)p_180482_2_).func_111104_a(this.field_70170_p.field_73012_v);
+                ((EntitySpider.GroupData)livingdata).setRandomEffect(this.world.rand);
             }
         }
 
-        if (p_180482_2_ instanceof EntitySpider.GroupData)
+        if (livingdata instanceof EntitySpider.GroupData)
         {
-            Potion potion = ((EntitySpider.GroupData)p_180482_2_).field_188478_a;
+            Potion potion = ((EntitySpider.GroupData)livingdata).effect;
 
             if (potion != null)
             {
-                this.func_70690_d(new PotionEffect(potion, Integer.MAX_VALUE));
+                this.addPotionEffect(new PotionEffect(potion, Integer.MAX_VALUE));
             }
         }
 
-        return p_180482_2_;
+        return livingdata;
     }
 
-    public float func_70047_e()
+    public float getEyeHeight()
     {
         return 0.65F;
     }
 
     static class AISpiderAttack extends EntityAIAttackMelee
         {
-            public AISpiderAttack(EntitySpider p_i46676_1_)
+            public AISpiderAttack(EntitySpider spider)
             {
-                super(p_i46676_1_, 1.0D, true);
+                super(spider, 1.0D, true);
             }
 
-            public boolean func_75253_b()
+            public boolean continueExecuting()
             {
-                float f = this.field_75441_b.func_70013_c(1.0F);
+                float f = this.attacker.getBrightness(1.0F);
 
-                if (f >= 0.5F && this.field_75441_b.func_70681_au().nextInt(100) == 0)
+                if (f >= 0.5F && this.attacker.getRNG().nextInt(100) == 0)
                 {
-                    this.field_75441_b.func_70624_b((EntityLivingBase)null);
+                    this.attacker.setAttackTarget((EntityLivingBase)null);
                     return false;
                 }
                 else
                 {
-                    return super.func_75253_b();
+                    return super.continueExecuting();
                 }
             }
 
-            protected double func_179512_a(EntityLivingBase p_179512_1_)
+            protected double getAttackReachSqr(EntityLivingBase attackTarget)
             {
-                return (double)(4.0F + p_179512_1_.field_70130_N);
+                return (double)(4.0F + attackTarget.width);
             }
         }
 
     static class AISpiderTarget<T extends EntityLivingBase> extends EntityAINearestAttackableTarget<T>
         {
-            public AISpiderTarget(EntitySpider p_i45818_1_, Class<T> p_i45818_2_)
+            public AISpiderTarget(EntitySpider spider, Class<T> classTarget)
             {
-                super(p_i45818_1_, p_i45818_2_, true);
+                super(spider, classTarget, true);
             }
 
-            public boolean func_75250_a()
+            public boolean shouldExecute()
             {
-                float f = this.field_75299_d.func_70013_c(1.0F);
-                return f >= 0.5F ? false : super.func_75250_a();
+                float f = this.taskOwner.getBrightness(1.0F);
+                return f >= 0.5F ? false : super.shouldExecute();
             }
         }
 
     public static class GroupData implements IEntityLivingData
         {
-            public Potion field_188478_a;
+            public Potion effect;
 
-            public void func_111104_a(Random p_111104_1_)
+            public void setRandomEffect(Random rand)
             {
-                int i = p_111104_1_.nextInt(5);
+                int i = rand.nextInt(5);
 
                 if (i <= 1)
                 {
-                    this.field_188478_a = MobEffects.field_76424_c;
+                    this.effect = MobEffects.SPEED;
                 }
                 else if (i <= 2)
                 {
-                    this.field_188478_a = MobEffects.field_76420_g;
+                    this.effect = MobEffects.STRENGTH;
                 }
                 else if (i <= 3)
                 {
-                    this.field_188478_a = MobEffects.field_76428_l;
+                    this.effect = MobEffects.REGENERATION;
                 }
                 else if (i <= 4)
                 {
-                    this.field_188478_a = MobEffects.field_76441_p;
+                    this.effect = MobEffects.INVISIBILITY;
                 }
             }
         }
