--- ../src-base/minecraft/net/minecraft/entity/monster/EntitySpider.java
+++ ../src-work/minecraft/net/minecraft/entity/monster/EntitySpider.java
@@ -172,7 +172,7 @@
             EntitySkeleton entityskeleton = new EntitySkeleton(this.field_70170_p);
             entityskeleton.func_70012_b(this.field_70165_t, this.field_70163_u, this.field_70161_v, this.field_70177_z, 0.0F);
             entityskeleton.func_180482_a(p_180482_1_, (IEntityLivingData)null);
-            this.field_70170_p.func_72838_d(entityskeleton);
+            this.field_70170_p.addEntity(entityskeleton, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.JOCKEY); // CraftBukkit - add SpawnReason
             entityskeleton.func_184220_m(this);
         }
 
