--- ../src-base/minecraft/net/minecraft/server/management/UserListEntryBan.java
+++ ../src-work/minecraft/net/minecraft/server/management/UserListEntryBan.java
@@ -75,4 +75,29 @@
         p_152641_1_.addProperty("expires", this.field_73692_f == null ? "forever" : field_73698_a.format(this.field_73692_f));
         p_152641_1_.addProperty("reason", this.field_73693_g);
     }
+    // CraftBukkit start
+    public String getSource() {
+        return this.field_73695_e;
+    }
+
+    public Date getCreated() {
+        return this.field_73694_d;
+    }
+
+    private static <T> T checkExpiry(T object, JsonObject jsonobject) {
+        Date expires = null;
+
+        try {
+            expires = jsonobject.has("expires") ? field_73698_a.parse(jsonobject.get("expires").getAsString()) : null;
+        } catch (ParseException ex) {
+            // Guess we don't have a date
+        }
+
+        if (expires == null || expires.after(new Date())) {
+            return object;
+        } else {
+            return null;
+        }
+    }
+    // CraftBukkit end
 }
