--- ../src-base/minecraft/net/minecraft/entity/EntityAreaEffectCloud.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityAreaEffectCloud.java
@@ -2,10 +2,8 @@
 
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
-import java.util.UUID;
+
+import java.util.*;
 import java.util.Map.Entry;
 import javax.annotation.Nullable;
 import net.minecraft.block.material.EnumPushReaction;
@@ -23,6 +21,8 @@
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.World;
 import net.minecraft.world.WorldServer;
+import org.bukkit.craftbukkit.entity.CraftLivingEntity;
+import org.bukkit.entity.LivingEntity;
 
 public class EntityAreaEffectCloud extends Entity
 {
@@ -59,6 +59,21 @@
         this.func_184483_a(3.0F);
     }
 
+    // CraftBukkit start accessor methods
+    public void refreshEffects() {
+        if (!this.field_184508_au) {
+            this.func_184212_Q().func_187227_b(net.minecraft.entity.EntityAreaEffectCloud.field_184499_b, Integer.valueOf(PotionUtils.func_185181_a((Collection) PotionUtils.func_185186_a(this.field_184502_e, (Collection) this.field_184503_f)))); // PAIL: rename
+        }
+    }
+
+    public String getType() {
+        return ((ResourceLocation) PotionType.field_185176_a.func_177774_c(this.field_184502_e)).toString(); // PAIL: rename
+    }
+
+    public void setType(String string) {
+        func_184484_a(PotionType.field_185176_a.func_82594_a(new ResourceLocation(string))); // PAIL: rename
+    }
+    // CraftBukkit end
     public EntityAreaEffectCloud(World p_i46810_1_, double p_i46810_2_, double p_i46810_4_, double p_i46810_6_)
     {
         this(p_i46810_1_);
@@ -322,12 +337,31 @@
                 }
                 else
                 {
+
                     List<EntityLivingBase> list = this.field_70170_p.<EntityLivingBase>func_72872_a(EntityLivingBase.class, this.func_174813_aQ());
 
                     if (!list.isEmpty())
                     {
-                        for (EntityLivingBase entitylivingbase : list)
+                        List<LivingEntity> entities = new ArrayList<LivingEntity>(); // CraftBukkit
+                        for(EntityLivingBase entityliving : list)
                         {
+                            if (!this.field_184504_g.containsKey(entityliving) && entityliving.func_184603_cC()) {
+                                double d0 = entityliving.field_70165_t - this.field_70165_t;
+                                double d1 = entityliving.field_70161_v - this.field_70161_v;
+                                double d2 = d0 * d0 + d1 * d1;
+
+                                if (d2 <= (double) (f * f)) {
+                                    // CraftBukkit start
+                                    entities.add((LivingEntity) entityliving.getBukkitEntity());
+                                }
+                            }
+                        }
+                        org.bukkit.event.entity.AreaEffectCloudApplyEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callAreaEffectCloudApplyEvent(this, entities);
+
+                        for (LivingEntity e : event.getAffectedEntities())
+                        {
+                            EntityLivingBase entitylivingbase = ((CraftLivingEntity)e).getHandle();
+                            //CraftBukkit end
                             if (!this.field_184504_g.containsKey(entitylivingbase) && entitylivingbase.func_184603_cC())
                             {
                                 double d0 = entitylivingbase.field_70165_t - this.field_70165_t;
