--- ../src-base/minecraft/net/minecraft/entity/EntityLivingBase.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityLivingBase.java
@@ -1,14 +1,10 @@
 package net.minecraft.entity;
 
 import com.google.common.base.Objects;
+import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
-import java.util.Collection;
-import java.util.ConcurrentModificationException;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
-import java.util.Random;
-import java.util.UUID;
+
+import java.util.*;
 import javax.annotation.Nullable;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockLadder;
@@ -42,8 +38,7 @@
 import net.minecraft.item.ItemArmor;
 import net.minecraft.item.ItemElytra;
 import net.minecraft.item.ItemStack;
-import net.minecraft.nbt.NBTTagCompound;
-import net.minecraft.nbt.NBTTagList;
+import net.minecraft.nbt.*;
 import net.minecraft.network.datasync.DataParameter;
 import net.minecraft.network.datasync.DataSerializers;
 import net.minecraft.network.datasync.EntityDataManager;
@@ -75,6 +70,9 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.craftbukkit.attribute.CraftAttributeMap;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityRegainHealthEvent;
 
 public abstract class EntityLivingBase extends Entity
 {
@@ -82,7 +80,7 @@
     private static final UUID field_110156_b = UUID.fromString("662A6B8D-DA3E-4C1C-8813-96EA6097278D");
     private static final AttributeModifier field_110157_c = (new AttributeModifier(field_110156_b, "Sprinting speed boost", 0.30000001192092896D, 2)).func_111168_a(false);
     protected static final DataParameter<Byte> field_184621_as = EntityDataManager.<Byte>func_187226_a(EntityLivingBase.class, DataSerializers.field_187191_a);
-    private static final DataParameter<Float> field_184632_c = EntityDataManager.<Float>func_187226_a(EntityLivingBase.class, DataSerializers.field_187193_c);
+    public static final DataParameter<Float> field_184632_c = EntityDataManager.<Float>func_187226_a(EntityLivingBase.class, DataSerializers.field_187193_c);
     private static final DataParameter<Integer> field_184633_f = EntityDataManager.<Integer>func_187226_a(EntityLivingBase.class, DataSerializers.field_187192_b);
     private static final DataParameter<Boolean> field_184634_g = EntityDataManager.<Boolean>func_187226_a(EntityLivingBase.class, DataSerializers.field_187198_h);
     private static final DataParameter<Integer> field_184635_h = EntityDataManager.<Integer>func_187226_a(EntityLivingBase.class, DataSerializers.field_187192_b);
@@ -150,7 +148,14 @@
     private BlockPos field_184620_bC;
     private DamageSource field_189750_bF;
     private long field_189751_bG;
-
+    // CraftBukkit start
+    public int expToDrop;
+    public int maxAirTicks = 300;
+    protected boolean forceDrops;
+    protected ArrayList<org.bukkit.inventory.ItemStack> drops = new ArrayList<org.bukkit.inventory.ItemStack>();
+    public CraftAttributeMap craftAttributes;
+    public boolean collides = true;
+    // CraftBukkit end
     public void func_174812_G()
     {
         this.func_70097_a(DamageSource.field_76380_i, Float.MAX_VALUE);
@@ -199,10 +204,11 @@
         {
             float f = (float)MathHelper.func_76123_f(this.field_70143_R - 3.0F);
 
-            if (p_184231_4_.func_185904_a() != Material.field_151579_a)
+            if (!p_184231_4_.func_177230_c().isAir(p_184231_4_, field_70170_p, p_184231_5_))
             {
                 double d0 = Math.min((double)(0.2F + f / 15.0F), 2.5D);
                 int i = (int)(150.0D * d0);
+                if (!p_184231_4_.func_177230_c().addLandingEffects(p_184231_4_, (WorldServer)this.field_70170_p, p_184231_5_, p_184231_4_, this, i))
                 ((WorldServer)this.field_70170_p).func_175739_a(EnumParticleTypes.BLOCK_DUST, this.field_70165_t, this.field_70163_u, this.field_70161_v, i, 0.0D, 0.0D, 0.0D, 0.15000000596046448D, new int[] {Block.func_176210_f(p_184231_4_)});
             }
         }
@@ -279,7 +285,7 @@
                     }
                 }
 
-                if (!this.field_70170_p.field_72995_K && this.func_184218_aH() && this.func_184187_bx() instanceof EntityLivingBase)
+                if (!this.field_70170_p.field_72995_K && this.func_184218_aH() && this.func_184187_bx() != null && this.func_184187_bx().shouldDismountInWater(this))
                 {
                     this.func_184210_p();
                 }
@@ -353,7 +359,19 @@
         this.field_70127_C = this.field_70125_A;
         this.field_70170_p.field_72984_F.func_76319_b();
     }
+    // CraftBukkit start
+    public int getExpReward() {
+        int exp = this.func_70693_a(this.field_70717_bb);
 
+        if (!this.field_70170_p.field_72995_K && (this.field_70718_bc > 0 || this.func_70684_aJ()) && this.func_146066_aG() && this.field_70170_p.func_82736_K().func_82766_b("doMobLoot")) {
+            this.expToDrop = exp;
+            return exp;
+        } else {
+            this.expToDrop = 0;
+            return 0;
+        }
+    }
+    // CraftBukkit end
     protected void func_184594_b(BlockPos p_184594_1_)
     {
         int i = EnchantmentHelper.func_185284_a(Enchantments.field_185301_j, this);
@@ -375,18 +393,17 @@
 
         if (this.field_70725_aQ == 20)
         {
-            if (!this.field_70170_p.field_72995_K && (this.func_70684_aJ() || this.field_70718_bc > 0 && this.func_146066_aG() && this.field_70170_p.func_82736_K().func_82766_b("doMobLoot")))
+            //MCPCRevive start
+            int i = this.getExpReward();
+            i = net.minecraftforge.event.ForgeEventFactory.getExperienceDrop(this, this.field_70717_bb, i);
+            while (i > 0)
             {
-                int i = this.func_70693_a(this.field_70717_bb);
-
-                while (i > 0)
-                {
-                    int j = EntityXPOrb.func_70527_a(i);
-                    i -= j;
-                    this.field_70170_p.func_72838_d(new EntityXPOrb(this.field_70170_p, this.field_70165_t, this.field_70163_u, this.field_70161_v, j));
-                }
+                int j = EntityXPOrb.func_70527_a(i);
+                i -= j;
+                this.field_70170_p.func_72838_d(new EntityXPOrb(this.field_70170_p, this.field_70165_t, this.field_70163_u, this.field_70161_v, j));
             }
-
+            this.expToDrop = 0;
+            //MCPCRevive end
             this.func_70106_y();
 
             for (int k = 0; k < 20; ++k)
@@ -440,6 +457,7 @@
     {
         this.field_70755_b = p_70604_1_;
         this.field_70756_c = this.field_70173_aa;
+        net.minecraftforge.common.ForgeHooks.onLivingSetAttackTarget(this, p_70604_1_);
     }
 
     public EntityLivingBase func_110144_aD()
@@ -544,6 +562,16 @@
         {
             SharedMonsterAttributes.func_151475_a(this.func_110140_aT(), p_70037_1_.func_150295_c("Attributes", 10));
         }
+        // CraftBukkit start
+        if (p_70037_1_.func_74764_b("Bukkit.MaxHealth")) {
+            NBTBase nbtbase = p_70037_1_.func_74781_a("Bukkit.MaxHealth");
+            if (nbtbase.func_74732_a() == 5) {
+                this.func_110148_a(net.minecraft.entity.SharedMonsterAttributes.field_111267_a).func_111128_a(((NBTTagFloat) nbtbase).func_150286_g());
+            } else if (nbtbase.func_74732_a() == 3) {
+                this.func_110148_a(net.minecraft.entity.SharedMonsterAttributes.field_111267_a).func_111128_a(((NBTTagInt) nbtbase).func_150286_g());
+            }
+        }
+        // CraftBukkit end
 
         if (p_70037_1_.func_150297_b("ActiveEffects", 9))
         {
@@ -586,13 +614,17 @@
             this.func_70052_a(7, true);
         }
     }
-
+    // CraftBukkit start
+    private boolean isTickingEffects = false;
+    private List<Object> effectsToProcess = Lists.newArrayList();
+    // CraftBukkit end
     protected void func_70679_bo()
     {
         Iterator<Potion> iterator = this.field_70713_bf.keySet().iterator();
 
         try
         {
+            isTickingEffects = true;
             while (iterator.hasNext())
             {
                 Potion potion = (Potion)iterator.next();
@@ -614,9 +646,19 @@
         }
         catch (ConcurrentModificationException var11)
         {
-            ;
+            field_190632_a.warn("ticking potion",var11);
         }
-
+        // CraftBukkit start
+        isTickingEffects = false;
+        for (Object e : effectsToProcess) {
+            if (e instanceof PotionEffect) {
+                func_70690_d((PotionEffect) e);
+            } else {
+                func_184589_d((Potion) e);
+            }
+        }
+        effectsToProcess.clear();
+        // CraftBukkit end
         if (this.field_70752_e)
         {
             if (!this.field_70170_p.field_72995_K)
@@ -668,8 +710,10 @@
         else
         {
             Collection<PotionEffect> collection = this.field_70713_bf.values();
-            this.field_70180_af.func_187227_b(field_184634_g, Boolean.valueOf(func_184593_a(collection)));
-            this.field_70180_af.func_187227_b(field_184633_f, Integer.valueOf(PotionUtils.func_185181_a(collection)));
+            net.minecraftforge.event.entity.living.PotionColorCalculationEvent event = new net.minecraftforge.event.entity.living.PotionColorCalculationEvent(this, PotionUtils.func_185181_a(collection), func_184593_a(collection), collection);
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+            this.field_70180_af.func_187227_b(field_184634_g, event.areParticlesHidden());
+            this.field_70180_af.func_187227_b(field_184633_f, event.getColor());
             this.func_82142_c(this.func_70644_a(MobEffects.field_76441_p));
         }
     }
@@ -765,6 +809,12 @@
     @Nullable
     public PotionEffect func_184596_c(@Nullable Potion p_184596_1_)
     {
+        // CraftBukkit start
+        if (isTickingEffects) {
+            effectsToProcess.add(p_184596_1_);
+            return null;
+        }
+        // CraftBukkit end
         return (PotionEffect)this.field_70713_bf.remove(p_184596_1_);
     }
 
@@ -810,13 +860,23 @@
         }
     }
 
-    public void func_70691_i(float p_70691_1_)
+    // CraftBukkit start - Delegate so we can handle providing a reason for health being regained
+    public void func_70691_i(float p_70691_1_) {
+        p_70691_1_ = net.minecraftforge.event.ForgeEventFactory.onLivingHeal(this, p_70691_1_);
+        heal(p_70691_1_, EntityRegainHealthEvent.RegainReason.REGEN);
+    }
+    public void heal(float healAmount, EntityRegainHealthEvent.RegainReason reason)
     {
-        float f = this.func_110143_aJ();
+        float f1 = this.func_110143_aJ();
 
-        if (f > 0.0F)
-        {
-            this.func_70606_j(f + p_70691_1_);
+        if (f1 > 0.0F) {
+            EntityRegainHealthEvent event = new EntityRegainHealthEvent(this.getBukkitEntity(), healAmount, reason);
+            this.field_70170_p.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled()) {
+                this.func_70606_j((float) (this.func_110143_aJ() + event.getAmount()));
+            }
+            // CraftBukkit end
         }
     }
 
@@ -832,6 +892,7 @@
 
     public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
     {
+        if (!net.minecraftforge.common.ForgeHooks.onLivingAttack(this, p_70097_1_, p_70097_2_)) return false;
         if (this.func_180431_b(p_70097_1_))
         {
             return false;
@@ -918,9 +979,9 @@
                         this.field_70718_bc = 100;
                         this.field_70717_bb = (EntityPlayer)entity1;
                     }
-                    else if (entity1 instanceof EntityWolf)
+                    else if (entity1 instanceof net.minecraft.entity.passive.EntityTameable)
                     {
-                        EntityWolf entitywolf = (EntityWolf)entity1;
+                        net.minecraft.entity.passive.EntityTameable entitywolf = (net.minecraft.entity.passive.EntityTameable)entity1;
 
                         if (entitywolf.func_70909_n())
                         {
@@ -1089,7 +1150,7 @@
 
     public void func_70669_a(ItemStack p_70669_1_)
     {
-        this.func_184185_a(SoundEvents.field_187635_cQ, 0.8F, 0.8F + this.field_70170_p.field_73012_v.nextFloat() * 0.4F);
+        this.field_70170_p.func_184148_a(null, this.field_70165_t, this.field_70163_u, this.field_70161_v, SoundEvents.field_187635_cQ, this.func_184176_by(), 0.8F, 0.8F + this.field_70170_p.field_73012_v.nextFloat() * 0.4F); //Forge: Fix MC-2518 Items are not damaged on the client so client needs packet as well.
 
         for (int i = 0; i < 5; ++i)
         {
@@ -1101,12 +1162,16 @@
             vec3d1 = vec3d1.func_178789_a(-this.field_70125_A * 0.017453292F);
             vec3d1 = vec3d1.func_178785_b(-this.field_70177_z * 0.017453292F);
             vec3d1 = vec3d1.func_72441_c(this.field_70165_t, this.field_70163_u + (double)this.func_70047_e(), this.field_70161_v);
-            this.field_70170_p.func_175688_a(EnumParticleTypes.ITEM_CRACK, vec3d1.field_72450_a, vec3d1.field_72448_b, vec3d1.field_72449_c, vec3d.field_72450_a, vec3d.field_72448_b + 0.05D, vec3d.field_72449_c, new int[] {Item.func_150891_b(p_70669_1_.func_77973_b())});
+            if (this.field_70170_p instanceof WorldServer) //Forge: Fix MC-2518 spawnParticle is nooped on server, need to use server specific variant
+                ((WorldServer)this.field_70170_p).func_175739_a(EnumParticleTypes.ITEM_CRACK, vec3d1.field_72450_a, vec3d1.field_72448_b, vec3d1.field_72449_c, 0,  vec3d.field_72450_a, vec3d.field_72448_b + 0.05D, vec3d.field_72449_c, 0.0D, Item.func_150891_b(p_70669_1_.func_77973_b()), p_70669_1_.func_77960_j());
+            else //Fix the fact that spawning ItemCrack uses TWO arguments.
+                this.field_70170_p.func_175688_a(EnumParticleTypes.ITEM_CRACK, vec3d1.field_72450_a, vec3d1.field_72448_b, vec3d1.field_72449_c, vec3d.field_72450_a, vec3d.field_72448_b + 0.05D, vec3d.field_72449_c, Item.func_150891_b(p_70669_1_.func_77973_b()), p_70669_1_.func_77960_j());
         }
     }
 
     public void func_70645_a(DamageSource p_70645_1_)
     {
+        if (net.minecraftforge.common.ForgeHooks.onLivingDeath(this, p_70645_1_)) return;
         if (!this.field_70729_aU)
         {
             Entity entity = p_70645_1_.func_76346_g();
@@ -1127,18 +1192,32 @@
 
             if (!this.field_70170_p.field_72995_K)
             {
-                int i = 0;
+                int i = net.minecraftforge.common.ForgeHooks.getLootingLevel(this, entity, p_70645_1_);
 
-                if (entity instanceof EntityPlayer)
-                {
-                    i = EnchantmentHelper.func_185283_h((EntityLivingBase)entity);
-                }
+                captureDrops = true;
+                capturedDrops.clear();
 
                 if (this.func_146066_aG() && this.field_70170_p.func_82736_K().func_82766_b("doMobLoot"))
                 {
                     boolean flag = this.field_70718_bc > 0;
                     this.func_184610_a(flag, i, p_70645_1_);
+                    // CraftBukkit start - Call death event
+                    CraftEventFactory.callEntityDeathEvent(this, this.drops);
+                    this.drops = new ArrayList<org.bukkit.inventory.ItemStack>();
+                } else {
+                    CraftEventFactory.callEntityDeathEvent(this);
+                    // CraftBukkit end
                 }
+
+                captureDrops = false;
+
+                if (!net.minecraftforge.common.ForgeHooks.onLivingDrops(this, p_70645_1_, capturedDrops, i, field_70718_bc > 0))
+                {
+                    for (EntityItem item : capturedDrops)
+                    {
+                        field_70170_p.func_72838_d(item);
+                    }
+                }
             }
 
             this.field_70170_p.func_72960_a(this, (byte)3);
@@ -1215,7 +1294,7 @@
             BlockPos blockpos = new BlockPos(i, j, k);
             IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos);
             Block block = iblockstate.func_177230_c();
-            return block != Blocks.field_150468_ap && block != Blocks.field_150395_bd ? block instanceof BlockTrapDoor && this.func_184604_a(blockpos, iblockstate) : true;
+            return net.minecraftforge.common.ForgeHooks.isLivingOnLadder(iblockstate, field_70170_p, new BlockPos(i, j, k), this);
         }
     }
 
@@ -1241,6 +1320,9 @@
 
     public void func_180430_e(float p_180430_1_, float p_180430_2_)
     {
+        float[] ret = net.minecraftforge.common.ForgeHooks.onLivingFall(this, p_180430_1_, p_180430_2_);
+        if (ret == null) return;
+        p_180430_1_ = ret[0]; p_180430_2_ = ret[1];
         super.func_180430_e(p_180430_1_, p_180430_2_);
         PotionEffect potioneffect = this.func_70660_b(MobEffects.field_76430_j);
         float f = potioneffect == null ? 0.0F : (float)(potioneffect.func_76458_c() + 1);
@@ -1257,7 +1339,7 @@
 
             if (iblockstate.func_185904_a() != Material.field_151579_a)
             {
-                SoundType soundtype = iblockstate.func_177230_c().func_185467_w();
+                SoundType soundtype = iblockstate.func_177230_c().getSoundType(iblockstate, field_70170_p, new BlockPos(j, k, l), this);
                 this.func_184185_a(soundtype.func_185842_g(), soundtype.func_185843_a() * 0.5F, soundtype.func_185847_b() * 0.75F);
             }
         }
@@ -1334,6 +1416,8 @@
     {
         if (!this.func_180431_b(p_70665_1_))
         {
+            p_70665_2_ = net.minecraftforge.common.ForgeHooks.onLivingHurt(this, p_70665_1_, p_70665_2_);
+            if (p_70665_2_ <= 0) return;
             p_70665_2_ = this.func_70655_b(p_70665_1_, p_70665_2_);
             p_70665_2_ = this.func_70672_c(p_70665_1_, p_70665_2_);
             float f = p_70665_2_;
@@ -1383,6 +1467,11 @@
 
     public void func_184609_a(EnumHand p_184609_1_)
     {
+        ItemStack stack = this.func_184586_b(p_184609_1_);
+        if (stack != null && stack.func_77973_b() != null)
+        {
+            if (stack.func_77973_b().onEntitySwing(this, stack)) return;
+        }
         if (!this.field_82175_bq || this.field_110158_av >= this.func_82166_i() / 2 || this.field_110158_av < 0)
         {
             this.field_110158_av = -1;
@@ -1613,7 +1702,7 @@
 
                     if (!this.field_70170_p.func_184143_b(axisalignedbb1))
                     {
-                        if (this.field_70170_p.func_180495_p(new BlockPos(d11, this.field_70163_u, d12)).func_185896_q())
+                        if (this.field_70170_p.func_180495_p(new BlockPos(d11, this.field_70163_u, d12)).isSideSolid(field_70170_p, new BlockPos(d11, this.field_70163_u, d12), EnumFacing.UP))
                         {
                             this.func_70634_a(d11, this.field_70163_u + 1.0D, d12);
                             return;
@@ -1621,14 +1710,14 @@
 
                         BlockPos blockpos = new BlockPos(d11, this.field_70163_u - 1.0D, d12);
 
-                        if (this.field_70170_p.func_180495_p(blockpos).func_185896_q() || this.field_70170_p.func_180495_p(blockpos).func_185904_a() == Material.field_151586_h)
+                        if (this.field_70170_p.func_180495_p(blockpos).isSideSolid(field_70170_p, blockpos, EnumFacing.UP) || this.field_70170_p.func_180495_p(blockpos).func_185904_a() == Material.field_151586_h)
                         {
                             d1 = d11;
                             d13 = this.field_70163_u + 1.0D;
                             d14 = d12;
                         }
                     }
-                    else if (!this.field_70170_p.func_184143_b(axisalignedbb1.func_72317_d(0.0D, 1.0D, 0.0D)) && this.field_70170_p.func_180495_p(new BlockPos(d11, this.field_70163_u + 1.0D, d12)).func_185896_q())
+                    else if (!this.field_70170_p.func_184143_b(axisalignedbb1.func_72317_d(0.0D, 1.0D, 0.0D)) && this.field_70170_p.func_180495_p(new BlockPos(d11, this.field_70163_u + 1.0D, d12)).isSideSolid(field_70170_p, new BlockPos(d11, this.field_70163_u + 1.0D, d12), EnumFacing.UP))
                     {
                         d1 = d11;
                         d13 = this.field_70163_u + 2.0D;
@@ -1700,6 +1789,7 @@
         }
 
         this.field_70160_al = true;
+        net.minecraftforge.common.ForgeHooks.onLivingJump(this);
     }
 
     protected void func_70629_bd()
@@ -1972,6 +2062,7 @@
 
     public void func_70071_h_()
     {
+        if (net.minecraftforge.common.ForgeHooks.onLivingUpdate(this)) return;
         super.func_70071_h_();
         this.func_184608_ct();
 
@@ -2015,6 +2106,7 @@
                 if (!ItemStack.func_77989_b(itemstack1, itemstack))
                 {
                     ((WorldServer)this.field_70170_p).func_73039_n().func_151247_a(this, new SPacketEntityEquipment(this.func_145782_y(), entityequipmentslot, itemstack1));
+                    net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.living.LivingEquipmentChangeEvent(this, entityequipmentslot, itemstack, itemstack1));
 
                     if (!itemstack.func_190926_b())
                     {
@@ -2488,6 +2580,40 @@
         this.field_70752_e = true;
     }
 
+    /***
+     * Removes all potion effects that have curativeItem as a curative item for its effect
+     * @param curativeItem The itemstack we are using to cure potion effects
+     */
+    public void curePotionEffects(ItemStack curativeItem)
+    {
+        if (field_70170_p.field_72995_K) return;
+        Iterator<PotionEffect> iterator = this.field_70713_bf.values().iterator();
+
+        while (iterator.hasNext())
+        {
+            PotionEffect effect = iterator.next();
+
+            if (effect.isCurativeItem(curativeItem))
+            {
+                func_70688_c(effect);
+                iterator.remove();
+                this.field_70752_e = true;
+            }
+        }
+    }
+
+    /**
+     * Returns true if the entity's rider (EntityPlayer) should face forward when mounted.
+     * currently only used in vanilla code by pigs.
+     *
+     * @param player The player who is riding the entity.
+     * @return If the player should orient the same direction as this entity.
+     */
+    public boolean shouldRiderFaceForward(EntityPlayer player)
+    {
+        return this instanceof net.minecraft.entity.passive.EntityPig;
+    }
+
     public abstract EnumHandSide func_184591_cq();
 
     public boolean func_184587_cr()
@@ -2508,12 +2634,19 @@
 
             if (itemstack == this.field_184627_bm)
             {
+                if (this.field_184627_bm != null)
+                {
+                    field_184628_bn = net.minecraftforge.event.ForgeEventFactory.onItemUseTick(this, field_184627_bm, field_184628_bn);
+                    if (field_184628_bn > 0)
+                        field_184627_bm.func_77973_b().onUsingTick(field_184627_bm, this, field_184628_bn);
+                }
+
                 if (this.func_184605_cv() <= 25 && this.func_184605_cv() % 4 == 0)
                 {
                     this.func_184584_a(this.field_184627_bm, 5);
                 }
 
-                if (--this.field_184628_bn == 0 && !this.field_70170_p.field_72995_K)
+                if (--this.field_184628_bn <= 0 && !this.field_70170_p.field_72995_K)
                 {
                     this.func_71036_o();
                 }
@@ -2531,8 +2664,10 @@
 
         if (!itemstack.func_190926_b() && !this.func_184587_cr())
         {
+            int duration = net.minecraftforge.event.ForgeEventFactory.onItemUseStart(this, itemstack, itemstack.func_77988_m());
+            if (duration <= 0) return;
             this.field_184627_bm = itemstack;
-            this.field_184628_bn = itemstack.func_77988_m();
+            this.field_184628_bn = duration;
 
             if (!this.field_70170_p.field_72995_K)
             {
@@ -2613,7 +2748,9 @@
         if (!this.field_184627_bm.func_190926_b() && this.func_184587_cr())
         {
             this.func_184584_a(this.field_184627_bm, 16);
-            this.func_184611_a(this.func_184600_cs(), this.field_184627_bm.func_77950_b(this.field_70170_p, this));
+            ItemStack itemstack = this.field_184627_bm.func_77950_b(this.field_70170_p, this);
+            itemstack = net.minecraftforge.event.ForgeEventFactory.onItemUseFinish(this, field_184627_bm, func_184605_cv(), itemstack);
+            this.func_184611_a(this.func_184600_cs(), itemstack);
             this.func_184602_cy();
         }
     }
@@ -2637,7 +2774,8 @@
     {
         if (!this.field_184627_bm.func_190926_b())
         {
-            this.field_184627_bm.func_77974_b(this.field_70170_p, this, this.func_184605_cv());
+            if (!net.minecraftforge.event.ForgeEventFactory.onUseItemStop(this, field_184627_bm, this.func_184605_cv()))
+                this.field_184627_bm.func_77974_b(this.field_70170_p, this, this.func_184605_cv());
         }
 
         this.func_184602_cy();
@@ -2761,4 +2899,29 @@
     {
         return true;
     }
+
+    // FORGE
+    private final net.minecraftforge.items.IItemHandlerModifiable handHandler = new net.minecraftforge.items.wrapper.EntityHandsInvWrapper(this);
+    private final net.minecraftforge.items.IItemHandlerModifiable armorHandler = new net.minecraftforge.items.wrapper.EntityArmorInvWrapper(this);
+    private final net.minecraftforge.items.IItemHandler joinedHandler = new net.minecraftforge.items.wrapper.CombinedInvWrapper(armorHandler, handHandler);
+
+    @SuppressWarnings("unchecked")
+    @Override
+    @Nullable
+    public <T> T getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable net.minecraft.util.EnumFacing facing)
+    {
+        if (capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY)
+        {
+            if (facing == null) return (T) joinedHandler;
+            else if (facing.func_176740_k().func_176720_b()) return (T) handHandler;
+            else if (facing.func_176740_k().func_176722_c()) return (T) armorHandler;
+        }
+        return super.getCapability(capability, facing);
+    }
+
+    @Override
+    public boolean hasCapability(net.minecraftforge.common.capabilities.Capability<?> capability, @Nullable net.minecraft.util.EnumFacing facing)
+    {
+        return capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY || super.hasCapability(capability, facing);
+    }
 }
