--- ../src-base/minecraft/net/minecraft/entity/EntityLeashKnot.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityLeashKnot.java
@@ -4,8 +4,10 @@
 import javax.annotation.Nullable;
 import net.minecraft.block.BlockFence;
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.init.SoundEvents;
 import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.network.play.server.SPacketEntityAttach;
 import net.minecraft.util.EnumFacing;
 import net.minecraft.util.EnumHand;
 import net.minecraft.util.math.AxisAlignedBB;
@@ -14,83 +16,84 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class EntityLeashKnot extends EntityHanging
 {
-    public EntityLeashKnot(World p_i1592_1_)
+    public EntityLeashKnot(World worldIn)
     {
-        super(p_i1592_1_);
+        super(worldIn);
     }
 
-    public EntityLeashKnot(World p_i45851_1_, BlockPos p_i45851_2_)
+    public EntityLeashKnot(World worldIn, BlockPos hangingPositionIn)
     {
-        super(p_i45851_1_, p_i45851_2_);
-        this.func_70107_b((double)p_i45851_2_.func_177958_n() + 0.5D, (double)p_i45851_2_.func_177956_o() + 0.5D, (double)p_i45851_2_.func_177952_p() + 0.5D);
+        super(worldIn, hangingPositionIn);
+        this.setPosition((double)hangingPositionIn.getX() + 0.5D, (double)hangingPositionIn.getY() + 0.5D, (double)hangingPositionIn.getZ() + 0.5D);
         float f = 0.125F;
         float f1 = 0.1875F;
         float f2 = 0.25F;
-        this.func_174826_a(new AxisAlignedBB(this.field_70165_t - 0.1875D, this.field_70163_u - 0.25D + 0.125D, this.field_70161_v - 0.1875D, this.field_70165_t + 0.1875D, this.field_70163_u + 0.25D + 0.125D, this.field_70161_v + 0.1875D));
-        this.field_98038_p = true;
+        this.setEntityBoundingBox(new AxisAlignedBB(this.posX - 0.1875D, this.posY - 0.25D + 0.125D, this.posZ - 0.1875D, this.posX + 0.1875D, this.posY + 0.25D + 0.125D, this.posZ + 0.1875D));
+        this.forceSpawn = true;
     }
 
-    public void func_70107_b(double p_70107_1_, double p_70107_3_, double p_70107_5_)
+    public void setPosition(double x, double y, double z)
     {
-        super.func_70107_b((double)MathHelper.func_76128_c(p_70107_1_) + 0.5D, (double)MathHelper.func_76128_c(p_70107_3_) + 0.5D, (double)MathHelper.func_76128_c(p_70107_5_) + 0.5D);
+        super.setPosition((double)MathHelper.floor(x) + 0.5D, (double)MathHelper.floor(y) + 0.5D, (double)MathHelper.floor(z) + 0.5D);
     }
 
-    protected void func_174856_o()
+    protected void updateBoundingBox()
     {
-        this.field_70165_t = (double)this.field_174861_a.func_177958_n() + 0.5D;
-        this.field_70163_u = (double)this.field_174861_a.func_177956_o() + 0.5D;
-        this.field_70161_v = (double)this.field_174861_a.func_177952_p() + 0.5D;
+        this.posX = (double)this.hangingPosition.getX() + 0.5D;
+        this.posY = (double)this.hangingPosition.getY() + 0.5D;
+        this.posZ = (double)this.hangingPosition.getZ() + 0.5D;
     }
 
-    public void func_174859_a(EnumFacing p_174859_1_)
+    public void updateFacingWithBoundingBox(EnumFacing facingDirectionIn)
     {
     }
 
-    public int func_82329_d()
+    public int getWidthPixels()
     {
         return 9;
     }
 
-    public int func_82330_g()
+    public int getHeightPixels()
     {
         return 9;
     }
 
-    public float func_70047_e()
+    public float getEyeHeight()
     {
         return -0.0625F;
     }
 
     @SideOnly(Side.CLIENT)
-    public boolean func_70112_a(double p_70112_1_)
+    public boolean isInRangeToRenderDist(double distance)
     {
-        return p_70112_1_ < 1024.0D;
+        return distance < 1024.0D;
     }
 
-    public void func_110128_b(@Nullable Entity p_110128_1_)
+    public void onBroken(@Nullable Entity brokenEntity)
     {
-        this.func_184185_a(SoundEvents.field_187746_da, 1.0F, 1.0F);
+        this.playSound(SoundEvents.ENTITY_LEASHKNOT_BREAK, 1.0F, 1.0F);
     }
 
-    public boolean func_70039_c(NBTTagCompound p_70039_1_)
+    public boolean writeToNBTOptional(NBTTagCompound compound)
     {
         return false;
     }
 
-    public void func_70014_b(NBTTagCompound p_70014_1_)
+    public void writeEntityToNBT(NBTTagCompound compound)
     {
     }
 
-    public void func_70037_a(NBTTagCompound p_70037_1_)
+    public void readEntityFromNBT(NBTTagCompound compound)
     {
     }
 
-    public boolean func_184230_a(EntityPlayer p_184230_1_, EnumHand p_184230_2_)
+    public boolean processInitialInteract(EntityPlayer player, EnumHand hand)
     {
-        if (this.field_70170_p.field_72995_K)
+        if (this.world.isRemote)
         {
             return true;
         }
@@ -98,28 +101,38 @@
         {
             boolean flag = false;
             double d0 = 7.0D;
-            List<EntityLiving> list = this.field_70170_p.<EntityLiving>func_72872_a(EntityLiving.class, new AxisAlignedBB(this.field_70165_t - 7.0D, this.field_70163_u - 7.0D, this.field_70161_v - 7.0D, this.field_70165_t + 7.0D, this.field_70163_u + 7.0D, this.field_70161_v + 7.0D));
+            List<EntityLiving> list = this.world.<EntityLiving>getEntitiesWithinAABB(EntityLiving.class, new AxisAlignedBB(this.posX - 7.0D, this.posY - 7.0D, this.posZ - 7.0D, this.posX + 7.0D, this.posY + 7.0D, this.posZ + 7.0D));
 
             for (EntityLiving entityliving : list)
             {
-                if (entityliving.func_110167_bD() && entityliving.func_110166_bE() == p_184230_1_)
+                if (entityliving.getLeashed() && entityliving.getLeashedToEntity() == player)
                 {
-                    entityliving.func_110162_b(this, true);
+                    // CraftBukkit start
+                    if (CraftEventFactory.callPlayerLeashEntityEvent(entityliving, this, player).isCancelled()) {
+                        ((EntityPlayerMP) player).connection.sendPacket(new SPacketEntityAttach(entityliving, entityliving.getLeashedToEntity()));
+                        continue;
+                    }
+                    entityliving.setLeashedToEntity(this, true);
                     flag = true;
                 }
             }
 
             if (!flag)
             {
-                this.func_70106_y();
+                this.setDead();
 
-                if (p_184230_1_.field_71075_bZ.field_75098_d)
+                if (player.capabilities.isCreativeMode)
                 {
                     for (EntityLiving entityliving1 : list)
                     {
-                        if (entityliving1.func_110167_bD() && entityliving1.func_110166_bE() == this)
+                        if (entityliving1.getLeashed() && entityliving1.getLeashedToEntity() == this)
                         {
-                            entityliving1.func_110160_i(true, false);
+                            // CraftBukkit start
+                            if (CraftEventFactory.callPlayerUnleashEntityEvent(entityliving1, player).isCancelled()) {
+                                continue;
+                            }
+                            // CraftBukkit end
+                            entityliving1.clearLeashed(true, false);
                         }
                     }
                 }
@@ -129,29 +142,29 @@
         }
     }
 
-    public boolean func_70518_d()
+    public boolean onValidSurface()
     {
-        return this.field_70170_p.func_180495_p(this.field_174861_a).func_177230_c() instanceof BlockFence;
+        return this.world.getBlockState(this.hangingPosition).getBlock() instanceof BlockFence;
     }
 
-    public static EntityLeashKnot func_174862_a(World p_174862_0_, BlockPos p_174862_1_)
+    public static EntityLeashKnot createKnot(World worldIn, BlockPos fence)
     {
-        EntityLeashKnot entityleashknot = new EntityLeashKnot(p_174862_0_, p_174862_1_);
-        p_174862_0_.func_72838_d(entityleashknot);
-        entityleashknot.func_184523_o();
+        EntityLeashKnot entityleashknot = new EntityLeashKnot(worldIn, fence);
+        worldIn.spawnEntity(entityleashknot);
+        entityleashknot.playPlaceSound();
         return entityleashknot;
     }
 
     @Nullable
-    public static EntityLeashKnot func_174863_b(World p_174863_0_, BlockPos p_174863_1_)
+    public static EntityLeashKnot getKnotForPosition(World worldIn, BlockPos pos)
     {
-        int i = p_174863_1_.func_177958_n();
-        int j = p_174863_1_.func_177956_o();
-        int k = p_174863_1_.func_177952_p();
+        int i = pos.getX();
+        int j = pos.getY();
+        int k = pos.getZ();
 
-        for (EntityLeashKnot entityleashknot : p_174863_0_.func_72872_a(EntityLeashKnot.class, new AxisAlignedBB((double)i - 1.0D, (double)j - 1.0D, (double)k - 1.0D, (double)i + 1.0D, (double)j + 1.0D, (double)k + 1.0D)))
+        for (EntityLeashKnot entityleashknot : worldIn.getEntitiesWithinAABB(EntityLeashKnot.class, new AxisAlignedBB((double)i - 1.0D, (double)j - 1.0D, (double)k - 1.0D, (double)i + 1.0D, (double)j + 1.0D, (double)k + 1.0D)))
         {
-            if (entityleashknot.func_174857_n().equals(p_174863_1_))
+            if (entityleashknot.getHangingPosition().equals(pos))
             {
                 return entityleashknot;
             }
@@ -160,8 +173,8 @@
         return null;
     }
 
-    public void func_184523_o()
+    public void playPlaceSound()
     {
-        this.func_184185_a(SoundEvents.field_187748_db, 1.0F, 1.0F);
+        this.playSound(SoundEvents.ENTITY_LEASHKNOT_PLACE, 1.0F, 1.0F);
     }
 }
