--- ../src-base/minecraft/net/minecraft/entity/EntityLeashKnot.java
+++ ../src-work/minecraft/net/minecraft/entity/EntityLeashKnot.java
@@ -4,8 +4,10 @@
 import javax.annotation.Nullable;
 import net.minecraft.block.BlockFence;
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.init.SoundEvents;
 import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.network.play.server.SPacketEntityAttach;
 import net.minecraft.util.EnumFacing;
 import net.minecraft.util.EnumHand;
 import net.minecraft.util.math.AxisAlignedBB;
@@ -14,6 +16,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class EntityLeashKnot extends EntityHanging
 {
@@ -104,6 +107,11 @@
             {
                 if (entityliving.func_110167_bD() && entityliving.func_110166_bE() == p_184230_1_)
                 {
+                    // CraftBukkit start
+                    if (CraftEventFactory.callPlayerLeashEntityEvent(entityliving, this, p_184230_1_).isCancelled()) {
+                        ((EntityPlayerMP) p_184230_1_).field_71135_a.func_147359_a(new SPacketEntityAttach(entityliving, entityliving.func_110166_bE()));
+                        continue;
+                    }
                     entityliving.func_110162_b(this, true);
                     flag = true;
                 }
@@ -119,6 +127,11 @@
                     {
                         if (entityliving1.func_110167_bD() && entityliving1.func_110166_bE() == this)
                         {
+                            // CraftBukkit start
+                            if (CraftEventFactory.callPlayerUnleashEntityEvent(entityliving1, p_184230_1_).isCancelled()) {
+                                continue;
+                            }
+                            // CraftBukkit end
                             entityliving1.func_110160_i(true, false);
                         }
                     }
