--- ../src-base/minecraft/net/minecraft/entity/passive/AbstractHorse.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/AbstractHorse.java
@@ -54,292 +54,296 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.event.entity.EntityRegainHealthEvent;
 
 public abstract class AbstractHorse extends EntityAnimal implements IInventoryChangedListener, IJumpingMount
 {
-    private static final Predicate<Entity> field_110276_bu = new Predicate<Entity>()
+    private static final Predicate<Entity> IS_HORSE_BREEDING = new Predicate<Entity>()
     {
         public boolean apply(@Nullable Entity p_apply_1_)
         {
-            return p_apply_1_ instanceof AbstractHorse && ((AbstractHorse)p_apply_1_).func_110205_ce();
+            return p_apply_1_ instanceof AbstractHorse && ((AbstractHorse)p_apply_1_).isBreeding();
         }
     };
-    public static final IAttribute field_110271_bv = (new RangedAttribute((IAttribute)null, "horse.jumpStrength", 0.7D, 0.0D, 2.0D)).func_111117_a("Jump Strength").func_111112_a(true);
-    private static final DataParameter<Byte> field_184787_bE = EntityDataManager.<Byte>func_187226_a(AbstractHorse.class, DataSerializers.field_187191_a);
-    private static final DataParameter<Optional<UUID>> field_184790_bH = EntityDataManager.<Optional<UUID>>func_187226_a(AbstractHorse.class, DataSerializers.field_187203_m);
-    private int field_190689_bJ;
-    private int field_110290_bE;
-    private int field_110295_bF;
-    public int field_110278_bp;
-    public int field_110279_bq;
-    protected boolean field_110275_br;
-    public ContainerHorseChest field_110296_bG;
-    protected int field_110274_bs;
-    protected float field_110277_bt;
-    private boolean field_110294_bI;
-    private float field_110283_bJ;
-    private float field_110284_bK;
-    private float field_110281_bL;
-    private float field_110282_bM;
-    private float field_110287_bN;
-    private float field_110288_bO;
-    protected boolean field_190688_bE = true;
-    protected int field_110285_bP;
+    public static final IAttribute JUMP_STRENGTH = (new RangedAttribute((IAttribute)null, "horse.jumpStrength", 0.7D, 0.0D, 2.0D)).setDescription("Jump Strength").setShouldWatch(true);
+    private static final DataParameter<Byte> STATUS = EntityDataManager.<Byte>createKey(AbstractHorse.class, DataSerializers.BYTE);
+    private static final DataParameter<Optional<UUID>> OWNER_UNIQUE_ID = EntityDataManager.<Optional<UUID>>createKey(AbstractHorse.class, DataSerializers.OPTIONAL_UNIQUE_ID);
+    private int eatingCounter;
+    private int openMouthCounter;
+    private int jumpRearingCounter;
+    public int tailCounter;
+    public int sprintCounter;
+    protected boolean horseJumping;
+    public ContainerHorseChest horseChest;
+    protected int temper;
+    protected float jumpPower;
+    private boolean allowStandSliding;
+    private float headLean;
+    private float prevHeadLean;
+    private float rearingAmount;
+    private float prevRearingAmount;
+    private float mouthOpenness;
+    private float prevMouthOpenness;
+    protected boolean canGallop = true;
+    protected int gallopTime;
+    public int maxDomestication = 100; // CraftBukkit - store max domestication value
 
-    public AbstractHorse(World p_i47299_1_)
+
+    public AbstractHorse(World worldIn)
     {
-        super(p_i47299_1_);
-        this.func_70105_a(1.3964844F, 1.6F);
-        this.field_70138_W = 1.0F;
-        this.func_110226_cD();
+        super(worldIn);
+        this.setSize(1.3964844F, 1.6F);
+        this.stepHeight = 1.0F;
+        this.initHorseChest();
     }
 
-    protected void func_184651_r()
+    protected void initEntityAI()
     {
-        this.field_70714_bg.func_75776_a(0, new EntityAISwimming(this));
-        this.field_70714_bg.func_75776_a(1, new EntityAIPanic(this, 1.2D));
-        this.field_70714_bg.func_75776_a(1, new EntityAIRunAroundLikeCrazy(this, 1.2D));
-        this.field_70714_bg.func_75776_a(2, new EntityAIMate(this, 1.0D, AbstractHorse.class));
-        this.field_70714_bg.func_75776_a(4, new EntityAIFollowParent(this, 1.0D));
-        this.field_70714_bg.func_75776_a(6, new EntityAIWanderAvoidWater(this, 0.7D));
-        this.field_70714_bg.func_75776_a(7, new EntityAIWatchClosest(this, EntityPlayer.class, 6.0F));
-        this.field_70714_bg.func_75776_a(8, new EntityAILookIdle(this));
+        this.tasks.addTask(0, new EntityAISwimming(this));
+        this.tasks.addTask(1, new EntityAIPanic(this, 1.2D));
+        this.tasks.addTask(1, new EntityAIRunAroundLikeCrazy(this, 1.2D));
+        this.tasks.addTask(2, new EntityAIMate(this, 1.0D, AbstractHorse.class));
+        this.tasks.addTask(4, new EntityAIFollowParent(this, 1.0D));
+        this.tasks.addTask(6, new EntityAIWanderAvoidWater(this, 0.7D));
+        this.tasks.addTask(7, new EntityAIWatchClosest(this, EntityPlayer.class, 6.0F));
+        this.tasks.addTask(8, new EntityAILookIdle(this));
     }
 
-    protected void func_70088_a()
+    protected void entityInit()
     {
-        super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184787_bE, Byte.valueOf((byte)0));
-        this.field_70180_af.func_187214_a(field_184790_bH, Optional.<UUID>absent());
+        super.entityInit();
+        this.dataManager.register(STATUS, Byte.valueOf((byte)0));
+        this.dataManager.register(OWNER_UNIQUE_ID, Optional.<UUID>absent());
     }
 
-    protected boolean func_110233_w(int p_110233_1_)
+    protected boolean getHorseWatchableBoolean(int p_110233_1_)
     {
-        return (((Byte)this.field_70180_af.func_187225_a(field_184787_bE)).byteValue() & p_110233_1_) != 0;
+        return (((Byte)this.dataManager.get(STATUS)).byteValue() & p_110233_1_) != 0;
     }
 
-    protected void func_110208_b(int p_110208_1_, boolean p_110208_2_)
+    protected void setHorseWatchableBoolean(int p_110208_1_, boolean p_110208_2_)
     {
-        byte b0 = ((Byte)this.field_70180_af.func_187225_a(field_184787_bE)).byteValue();
+        byte b0 = ((Byte)this.dataManager.get(STATUS)).byteValue();
 
         if (p_110208_2_)
         {
-            this.field_70180_af.func_187227_b(field_184787_bE, Byte.valueOf((byte)(b0 | p_110208_1_)));
+            this.dataManager.set(STATUS, Byte.valueOf((byte)(b0 | p_110208_1_)));
         }
         else
         {
-            this.field_70180_af.func_187227_b(field_184787_bE, Byte.valueOf((byte)(b0 & ~p_110208_1_)));
+            this.dataManager.set(STATUS, Byte.valueOf((byte)(b0 & ~p_110208_1_)));
         }
     }
 
-    public boolean func_110248_bS()
+    public boolean isTame()
     {
-        return this.func_110233_w(2);
+        return this.getHorseWatchableBoolean(2);
     }
 
     @Nullable
-    public UUID func_184780_dh()
+    public UUID getOwnerUniqueId()
     {
-        return (UUID)((Optional)this.field_70180_af.func_187225_a(field_184790_bH)).orNull();
+        return (UUID)((Optional)this.dataManager.get(OWNER_UNIQUE_ID)).orNull();
     }
 
-    public void func_184779_b(@Nullable UUID p_184779_1_)
+    public void setOwnerUniqueId(@Nullable UUID uniqueId)
     {
-        this.field_70180_af.func_187227_b(field_184790_bH, Optional.<UUID>fromNullable(p_184779_1_));
+        this.dataManager.set(OWNER_UNIQUE_ID, Optional.<UUID>fromNullable(uniqueId));
     }
 
-    public float func_110254_bY()
+    public float getHorseSize()
     {
         return 0.5F;
     }
 
-    public void func_98054_a(boolean p_98054_1_)
+    public void setScaleForAge(boolean child)
     {
-        this.func_98055_j(p_98054_1_ ? this.func_110254_bY() : 1.0F);
+        this.setScale(child ? this.getHorseSize() : 1.0F);
     }
 
-    public boolean func_110246_bZ()
+    public boolean isHorseJumping()
     {
-        return this.field_110275_br;
+        return this.horseJumping;
     }
 
-    public void func_110234_j(boolean p_110234_1_)
+    public void setHorseTamed(boolean tamed)
     {
-        this.func_110208_b(2, p_110234_1_);
+        this.setHorseWatchableBoolean(2, tamed);
     }
 
-    public void func_110255_k(boolean p_110255_1_)
+    public void setHorseJumping(boolean jumping)
     {
-        this.field_110275_br = p_110255_1_;
+        this.horseJumping = jumping;
     }
 
-    public boolean func_184652_a(EntityPlayer p_184652_1_)
+    public boolean canBeLeashedTo(EntityPlayer player)
     {
-        return super.func_184652_a(p_184652_1_) && this.func_70668_bt() != EnumCreatureAttribute.UNDEAD;
+        return super.canBeLeashedTo(player) && this.getCreatureAttribute() != EnumCreatureAttribute.UNDEAD;
     }
 
-    protected void func_142017_o(float p_142017_1_)
+    protected void onLeashDistance(float p_142017_1_)
     {
-        if (p_142017_1_ > 6.0F && this.func_110204_cc())
+        if (p_142017_1_ > 6.0F && this.isEatingHaystack())
         {
-            this.func_110227_p(false);
+            this.setEatingHaystack(false);
         }
     }
 
-    public boolean func_110204_cc()
+    public boolean isEatingHaystack()
     {
-        return this.func_110233_w(16);
+        return this.getHorseWatchableBoolean(16);
     }
 
-    public boolean func_110209_cd()
+    public boolean isRearing()
     {
-        return this.func_110233_w(32);
+        return this.getHorseWatchableBoolean(32);
     }
 
-    public boolean func_110205_ce()
+    public boolean isBreeding()
     {
-        return this.func_110233_w(8);
+        return this.getHorseWatchableBoolean(8);
     }
 
-    public void func_110242_l(boolean p_110242_1_)
+    public void setBreeding(boolean breeding)
     {
-        this.func_110208_b(8, p_110242_1_);
+        this.setHorseWatchableBoolean(8, breeding);
     }
 
-    public void func_110251_o(boolean p_110251_1_)
+    public void setHorseSaddled(boolean saddled)
     {
-        this.func_110208_b(4, p_110251_1_);
+        this.setHorseWatchableBoolean(4, saddled);
     }
 
-    public int func_110252_cg()
+    public int getTemper()
     {
-        return this.field_110274_bs;
+        return this.temper;
     }
 
-    public void func_110238_s(int p_110238_1_)
+    public void setTemper(int temperIn)
     {
-        this.field_110274_bs = p_110238_1_;
+        this.temper = temperIn;
     }
 
-    public int func_110198_t(int p_110198_1_)
+    public int increaseTemper(int p_110198_1_)
     {
-        int i = MathHelper.func_76125_a(this.func_110252_cg() + p_110198_1_, 0, this.func_190676_dC());
-        this.func_110238_s(i);
+        int i = MathHelper.clamp(this.getTemper() + p_110198_1_, 0, this.getMaxTemper());
+        this.setTemper(i);
         return i;
     }
 
-    public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
+    public boolean attackEntityFrom(DamageSource source, float amount)
     {
-        Entity entity = p_70097_1_.func_76346_g();
-        return this.func_184207_aI() && entity != null && this.func_184215_y(entity) ? false : super.func_70097_a(p_70097_1_, p_70097_2_);
+        Entity entity = source.getEntity();
+        return this.isBeingRidden() && entity != null && this.isRidingOrBeingRiddenBy(entity) ? false : super.attackEntityFrom(source, amount);
     }
 
-    public boolean func_70104_M()
+    public boolean canBePushed()
     {
-        return !this.func_184207_aI();
+        return !this.isBeingRidden();
     }
 
-    private void func_110266_cB()
+    private void eatingHorse()
     {
-        this.func_110249_cI();
+        this.openHorseMouth();
 
-        if (!this.func_174814_R())
+        if (!this.isSilent())
         {
-            this.field_70170_p.func_184148_a((EntityPlayer)null, this.field_70165_t, this.field_70163_u, this.field_70161_v, SoundEvents.field_187711_cp, this.func_184176_by(), 1.0F, 1.0F + (this.field_70146_Z.nextFloat() - this.field_70146_Z.nextFloat()) * 0.2F);
+            this.world.playSound((EntityPlayer)null, this.posX, this.posY, this.posZ, SoundEvents.ENTITY_HORSE_EAT, this.getSoundCategory(), 1.0F, 1.0F + (this.rand.nextFloat() - this.rand.nextFloat()) * 0.2F);
         }
     }
 
-    public void func_180430_e(float p_180430_1_, float p_180430_2_)
+    public void fall(float distance, float damageMultiplier)
     {
-        if (p_180430_1_ > 1.0F)
+        if (distance > 1.0F)
         {
-            this.func_184185_a(SoundEvents.field_187723_ct, 0.4F, 1.0F);
+            this.playSound(SoundEvents.ENTITY_HORSE_LAND, 0.4F, 1.0F);
         }
 
-        int i = MathHelper.func_76123_f((p_180430_1_ * 0.5F - 3.0F) * p_180430_2_);
+        int i = MathHelper.ceil((distance * 0.5F - 3.0F) * damageMultiplier);
 
         if (i > 0)
         {
-            this.func_70097_a(DamageSource.field_76379_h, (float)i);
+            this.attackEntityFrom(DamageSource.FALL, (float)i);
 
-            if (this.func_184207_aI())
+            if (this.isBeingRidden())
             {
-                for (Entity entity : this.func_184182_bu())
+                for (Entity entity : this.getRecursivePassengers())
                 {
-                    entity.func_70097_a(DamageSource.field_76379_h, (float)i);
+                    entity.attackEntityFrom(DamageSource.FALL, (float)i);
                 }
             }
 
-            IBlockState iblockstate = this.field_70170_p.func_180495_p(new BlockPos(this.field_70165_t, this.field_70163_u - 0.2D - (double)this.field_70126_B, this.field_70161_v));
-            Block block = iblockstate.func_177230_c();
+            IBlockState iblockstate = this.world.getBlockState(new BlockPos(this.posX, this.posY - 0.2D - (double)this.prevRotationYaw, this.posZ));
+            Block block = iblockstate.getBlock();
 
-            if (iblockstate.func_185904_a() != Material.field_151579_a && !this.func_174814_R())
+            if (iblockstate.getMaterial() != Material.AIR && !this.isSilent())
             {
-                SoundType soundtype = block.func_185467_w();
-                this.field_70170_p.func_184148_a((EntityPlayer)null, this.field_70165_t, this.field_70163_u, this.field_70161_v, soundtype.func_185844_d(), this.func_184176_by(), soundtype.func_185843_a() * 0.5F, soundtype.func_185847_b() * 0.75F);
+                SoundType soundtype = block.getSoundType();
+                this.world.playSound((EntityPlayer)null, this.posX, this.posY, this.posZ, soundtype.getStepSound(), this.getSoundCategory(), soundtype.getVolume() * 0.5F, soundtype.getPitch() * 0.75F);
             }
         }
     }
 
-    protected int func_190686_di()
+    protected int getInventorySize()
     {
         return 2;
     }
 
-    protected void func_110226_cD()
+    public void initHorseChest()
     {
-        ContainerHorseChest containerhorsechest = this.field_110296_bG;
-        this.field_110296_bG = new ContainerHorseChest("HorseChest", this.func_190686_di());
-        this.field_110296_bG.func_110133_a(this.func_70005_c_());
+        ContainerHorseChest containerhorsechest = this.horseChest;
+        this.horseChest = new ContainerHorseChest("HorseChest", this.getInventorySize());
+        this.horseChest.setCustomName(this.getName());
 
         if (containerhorsechest != null)
         {
-            containerhorsechest.func_110132_b(this);
-            int i = Math.min(containerhorsechest.func_70302_i_(), this.field_110296_bG.func_70302_i_());
+            containerhorsechest.removeInventoryChangeListener(this);
+            int i = Math.min(containerhorsechest.getSizeInventory(), this.horseChest.getSizeInventory());
 
             for (int j = 0; j < i; ++j)
             {
-                ItemStack itemstack = containerhorsechest.func_70301_a(j);
+                ItemStack itemstack = containerhorsechest.getStackInSlot(j);
 
-                if (!itemstack.func_190926_b())
+                if (!itemstack.isEmpty())
                 {
-                    this.field_110296_bG.func_70299_a(j, itemstack.func_77946_l());
+                    this.horseChest.setInventorySlotContents(j, itemstack.copy());
                 }
             }
         }
 
-        this.field_110296_bG.func_110134_a(this);
-        this.func_110232_cE();
+        this.horseChest.addInventoryChangeListener(this);
+        this.updateHorseSlots();
+        this.itemHandler = new net.minecraftforge.items.wrapper.InvWrapper(this.horseChest);
     }
 
-    protected void func_110232_cE()
+    protected void updateHorseSlots()
     {
-        if (!this.field_70170_p.field_72995_K)
+        if (!this.world.isRemote)
         {
-            this.func_110251_o(!this.field_110296_bG.func_70301_a(0).func_190926_b() && this.func_190685_dA());
+            this.setHorseSaddled(!this.horseChest.getStackInSlot(0).isEmpty() && this.canBeSaddled());
         }
     }
 
-    public void func_76316_a(IInventory p_76316_1_)
+    public void onInventoryChanged(IInventory invBasic)
     {
-        boolean flag = this.func_110257_ck();
-        this.func_110232_cE();
+        boolean flag = this.isHorseSaddled();
+        this.updateHorseSlots();
 
-        if (this.field_70173_aa > 20 && !flag && this.func_110257_ck())
+        if (this.ticksExisted > 20 && !flag && this.isHorseSaddled())
         {
-            this.func_184185_a(SoundEvents.field_187726_cu, 0.5F, 1.0F);
+            this.playSound(SoundEvents.ENTITY_HORSE_SADDLE, 0.5F, 1.0F);
         }
     }
 
     @Nullable
-    protected AbstractHorse func_110250_a(Entity p_110250_1_, double p_110250_2_)
+    protected AbstractHorse getClosestHorse(Entity entityIn, double distance)
     {
         double d0 = Double.MAX_VALUE;
         Entity entity = null;
 
-        for (Entity entity1 : this.field_70170_p.func_175674_a(p_110250_1_, p_110250_1_.func_174813_aQ().func_72321_a(p_110250_2_, p_110250_2_, p_110250_2_), field_110276_bu))
+        for (Entity entity1 : this.world.getEntitiesInAABBexcluding(entityIn, entityIn.getEntityBoundingBox().addCoord(distance, distance, distance), IS_HORSE_BREEDING))
         {
-            double d1 = entity1.func_70092_e(p_110250_1_.field_70165_t, p_110250_1_.field_70163_u, p_110250_1_.field_70161_v);
+            double d1 = entity1.getDistanceSq(entityIn.posX, entityIn.posY, entityIn.posZ);
 
             if (d1 < d0)
             {
@@ -351,833 +355,833 @@
         return (AbstractHorse)entity;
     }
 
-    public double func_110215_cj()
+    public double getHorseJumpStrength()
     {
-        return this.func_110148_a(field_110271_bv).func_111126_e();
+        return this.getEntityAttribute(JUMP_STRENGTH).getAttributeValue();
     }
 
     @Nullable
-    protected SoundEvent func_184615_bR()
+    protected SoundEvent getDeathSound()
     {
-        this.func_110249_cI();
+        this.openHorseMouth();
         return null;
     }
 
     @Nullable
-    protected SoundEvent func_184601_bQ()
+    protected SoundEvent getHurtSound()
     {
-        this.func_110249_cI();
+        this.openHorseMouth();
 
-        if (this.field_70146_Z.nextInt(3) == 0)
+        if (this.rand.nextInt(3) == 0)
         {
-            this.func_110220_cK();
+            this.makeHorseRear();
         }
 
         return null;
     }
 
     @Nullable
-    protected SoundEvent func_184639_G()
+    protected SoundEvent getAmbientSound()
     {
-        this.func_110249_cI();
+        this.openHorseMouth();
 
-        if (this.field_70146_Z.nextInt(10) == 0 && !this.func_70610_aX())
+        if (this.rand.nextInt(10) == 0 && !this.isMovementBlocked())
         {
-            this.func_110220_cK();
+            this.makeHorseRear();
         }
 
         return null;
     }
 
-    public boolean func_190685_dA()
+    public boolean canBeSaddled()
     {
         return true;
     }
 
-    public boolean func_110257_ck()
+    public boolean isHorseSaddled()
     {
-        return this.func_110233_w(4);
+        return this.getHorseWatchableBoolean(4);
     }
 
     @Nullable
-    protected SoundEvent func_184785_dv()
+    protected SoundEvent getAngrySound()
     {
-        this.func_110249_cI();
-        this.func_110220_cK();
+        this.openHorseMouth();
+        this.makeHorseRear();
         return null;
     }
 
-    protected void func_180429_a(BlockPos p_180429_1_, Block p_180429_2_)
+    protected void playStepSound(BlockPos pos, Block blockIn)
     {
-        if (!p_180429_2_.func_176223_P().func_185904_a().func_76224_d())
+        if (!blockIn.getDefaultState().getMaterial().isLiquid())
         {
-            SoundType soundtype = p_180429_2_.func_185467_w();
+            SoundType soundtype = blockIn.getSoundType();
 
-            if (this.field_70170_p.func_180495_p(p_180429_1_.func_177984_a()).func_177230_c() == Blocks.field_150431_aC)
+            if (this.world.getBlockState(pos.up()).getBlock() == Blocks.SNOW_LAYER)
             {
-                soundtype = Blocks.field_150431_aC.func_185467_w();
+                soundtype = Blocks.SNOW_LAYER.getSoundType();
             }
 
-            if (this.func_184207_aI() && this.field_190688_bE)
+            if (this.isBeingRidden() && this.canGallop)
             {
-                ++this.field_110285_bP;
+                ++this.gallopTime;
 
-                if (this.field_110285_bP > 5 && this.field_110285_bP % 3 == 0)
+                if (this.gallopTime > 5 && this.gallopTime % 3 == 0)
                 {
-                    this.func_190680_a(soundtype);
+                    this.playGallopSound(soundtype);
                 }
-                else if (this.field_110285_bP <= 5)
+                else if (this.gallopTime <= 5)
                 {
-                    this.func_184185_a(SoundEvents.field_187732_cw, soundtype.func_185843_a() * 0.15F, soundtype.func_185847_b());
+                    this.playSound(SoundEvents.ENTITY_HORSE_STEP_WOOD, soundtype.getVolume() * 0.15F, soundtype.getPitch());
                 }
             }
-            else if (soundtype == SoundType.field_185848_a)
+            else if (soundtype == SoundType.WOOD)
             {
-                this.func_184185_a(SoundEvents.field_187732_cw, soundtype.func_185843_a() * 0.15F, soundtype.func_185847_b());
+                this.playSound(SoundEvents.ENTITY_HORSE_STEP_WOOD, soundtype.getVolume() * 0.15F, soundtype.getPitch());
             }
             else
             {
-                this.func_184185_a(SoundEvents.field_187729_cv, soundtype.func_185843_a() * 0.15F, soundtype.func_185847_b());
+                this.playSound(SoundEvents.ENTITY_HORSE_STEP, soundtype.getVolume() * 0.15F, soundtype.getPitch());
             }
         }
     }
 
-    protected void func_190680_a(SoundType p_190680_1_)
+    protected void playGallopSound(SoundType p_190680_1_)
     {
-        this.func_184185_a(SoundEvents.field_187714_cq, p_190680_1_.func_185843_a() * 0.15F, p_190680_1_.func_185847_b());
+        this.playSound(SoundEvents.ENTITY_HORSE_GALLOP, p_190680_1_.getVolume() * 0.15F, p_190680_1_.getPitch());
     }
 
-    protected void func_110147_ax()
+    protected void applyEntityAttributes()
     {
-        super.func_110147_ax();
-        this.func_110140_aT().func_111150_b(field_110271_bv);
-        this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(53.0D);
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.22499999403953552D);
+        super.applyEntityAttributes();
+        this.getAttributeMap().registerAttribute(JUMP_STRENGTH);
+        this.getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).setBaseValue(53.0D);
+        this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).setBaseValue(0.22499999403953552D);
     }
 
-    public int func_70641_bl()
+    public int getMaxSpawnedInChunk()
     {
         return 6;
     }
 
-    public int func_190676_dC()
+    public int getMaxTemper()
     {
-        return 100;
+        return maxDomestication;
     }
 
-    protected float func_70599_aP()
+    protected float getSoundVolume()
     {
         return 0.8F;
     }
 
-    public int func_70627_aG()
+    public int getTalkInterval()
     {
         return 400;
     }
 
-    public void func_110199_f(EntityPlayer p_110199_1_)
+    public void openGUI(EntityPlayer playerEntity)
     {
-        if (!this.field_70170_p.field_72995_K && (!this.func_184207_aI() || this.func_184196_w(p_110199_1_)) && this.func_110248_bS())
+        if (!this.world.isRemote && (!this.isBeingRidden() || this.isPassenger(playerEntity)) && this.isTame())
         {
-            this.field_110296_bG.func_110133_a(this.func_70005_c_());
-            p_110199_1_.func_184826_a(this, this.field_110296_bG);
+            this.horseChest.setCustomName(this.getName());
+            playerEntity.openGuiHorseInventory(this, this.horseChest);
         }
     }
 
-    protected boolean func_190678_b(EntityPlayer p_190678_1_, ItemStack p_190678_2_)
+    protected boolean handleEating(EntityPlayer player, ItemStack stack)
     {
         boolean flag = false;
         float f = 0.0F;
         int i = 0;
         int j = 0;
-        Item item = p_190678_2_.func_77973_b();
+        Item item = stack.getItem();
 
-        if (item == Items.field_151015_O)
+        if (item == Items.WHEAT)
         {
             f = 2.0F;
             i = 20;
             j = 3;
         }
-        else if (item == Items.field_151102_aT)
+        else if (item == Items.SUGAR)
         {
             f = 1.0F;
             i = 30;
             j = 3;
         }
-        else if (item == Item.func_150898_a(Blocks.field_150407_cf))
+        else if (item == Item.getItemFromBlock(Blocks.HAY_BLOCK))
         {
             f = 20.0F;
             i = 180;
         }
-        else if (item == Items.field_151034_e)
+        else if (item == Items.APPLE)
         {
             f = 3.0F;
             i = 60;
             j = 3;
         }
-        else if (item == Items.field_151150_bK)
+        else if (item == Items.GOLDEN_CARROT)
         {
             f = 4.0F;
             i = 60;
             j = 5;
 
-            if (this.func_110248_bS() && this.func_70874_b() == 0 && !this.func_70880_s())
+            if (this.isTame() && this.getGrowingAge() == 0 && !this.isInLove())
             {
                 flag = true;
-                this.func_146082_f(p_190678_1_);
+                this.setInLove(player);
             }
         }
-        else if (item == Items.field_151153_ao)
+        else if (item == Items.GOLDEN_APPLE)
         {
             f = 10.0F;
             i = 240;
             j = 10;
 
-            if (this.func_110248_bS() && this.func_70874_b() == 0 && !this.func_70880_s())
+            if (this.isTame() && this.getGrowingAge() == 0 && !this.isInLove())
             {
                 flag = true;
-                this.func_146082_f(p_190678_1_);
+                this.setInLove(player);
             }
         }
 
-        if (this.func_110143_aJ() < this.func_110138_aP() && f > 0.0F)
+        if (this.getHealth() < this.getMaxHealth() && f > 0.0F)
         {
-            this.func_70691_i(f);
+            this.heal(1.0F, EntityRegainHealthEvent.RegainReason.EATING); // CraftBukkit
             flag = true;
         }
 
-        if (this.func_70631_g_() && i > 0)
+        if (this.isChild() && i > 0)
         {
-            this.field_70170_p.func_175688_a(EnumParticleTypes.VILLAGER_HAPPY, this.field_70165_t + (double)(this.field_70146_Z.nextFloat() * this.field_70130_N * 2.0F) - (double)this.field_70130_N, this.field_70163_u + 0.5D + (double)(this.field_70146_Z.nextFloat() * this.field_70131_O), this.field_70161_v + (double)(this.field_70146_Z.nextFloat() * this.field_70130_N * 2.0F) - (double)this.field_70130_N, 0.0D, 0.0D, 0.0D, new int[0]);
+            this.world.spawnParticle(EnumParticleTypes.VILLAGER_HAPPY, this.posX + (double)(this.rand.nextFloat() * this.width * 2.0F) - (double)this.width, this.posY + 0.5D + (double)(this.rand.nextFloat() * this.height), this.posZ + (double)(this.rand.nextFloat() * this.width * 2.0F) - (double)this.width, 0.0D, 0.0D, 0.0D, new int[0]);
 
-            if (!this.field_70170_p.field_72995_K)
+            if (!this.world.isRemote)
             {
-                this.func_110195_a(i);
+                this.addGrowth(i);
             }
 
             flag = true;
         }
 
-        if (j > 0 && (flag || !this.func_110248_bS()) && this.func_110252_cg() < this.func_190676_dC())
+        if (j > 0 && (flag || !this.isTame()) && this.getTemper() < this.getMaxTemper())
         {
             flag = true;
 
-            if (!this.field_70170_p.field_72995_K)
+            if (!this.world.isRemote)
             {
-                this.func_110198_t(j);
+                this.increaseTemper(j);
             }
         }
 
         if (flag)
         {
-            this.func_110266_cB();
+            this.eatingHorse();
         }
 
         return flag;
     }
 
-    protected void func_110237_h(EntityPlayer p_110237_1_)
+    protected void mountTo(EntityPlayer player)
     {
-        p_110237_1_.field_70177_z = this.field_70177_z;
-        p_110237_1_.field_70125_A = this.field_70125_A;
-        this.func_110227_p(false);
-        this.func_110219_q(false);
+        player.rotationYaw = this.rotationYaw;
+        player.rotationPitch = this.rotationPitch;
+        this.setEatingHaystack(false);
+        this.setRearing(false);
 
-        if (!this.field_70170_p.field_72995_K)
+        if (!this.world.isRemote)
         {
-            p_110237_1_.func_184220_m(this);
+            player.startRiding(this);
         }
     }
 
-    protected boolean func_70610_aX()
+    protected boolean isMovementBlocked()
     {
-        return super.func_70610_aX() && this.func_184207_aI() && this.func_110257_ck() || this.func_110204_cc() || this.func_110209_cd();
+        return super.isMovementBlocked() && this.isBeingRidden() && this.isHorseSaddled() || this.isEatingHaystack() || this.isRearing();
     }
 
-    public boolean func_70877_b(ItemStack p_70877_1_)
+    public boolean isBreedingItem(ItemStack stack)
     {
         return false;
     }
 
-    private void func_110210_cH()
+    private void moveTail()
     {
-        this.field_110278_bp = 1;
+        this.tailCounter = 1;
     }
 
-    public void func_70645_a(DamageSource p_70645_1_)
+    public void onDeath(DamageSource cause)
     {
-        super.func_70645_a(p_70645_1_);
+        super.onDeath(cause);
 
-        if (!this.field_70170_p.field_72995_K && this.field_110296_bG != null)
+        if (!this.world.isRemote && this.horseChest != null)
         {
-            for (int i = 0; i < this.field_110296_bG.func_70302_i_(); ++i)
+            for (int i = 0; i < this.horseChest.getSizeInventory(); ++i)
             {
-                ItemStack itemstack = this.field_110296_bG.func_70301_a(i);
+                ItemStack itemstack = this.horseChest.getStackInSlot(i);
 
-                if (!itemstack.func_190926_b())
+                if (!itemstack.isEmpty())
                 {
-                    this.func_70099_a(itemstack, 0.0F);
+                    this.entityDropItem(itemstack, 0.0F);
                 }
             }
         }
     }
 
-    public void func_70636_d()
+    public void onLivingUpdate()
     {
-        if (this.field_70146_Z.nextInt(200) == 0)
+        if (this.rand.nextInt(200) == 0)
         {
-            this.func_110210_cH();
+            this.moveTail();
         }
 
-        super.func_70636_d();
+        super.onLivingUpdate();
 
-        if (!this.field_70170_p.field_72995_K)
+        if (!this.world.isRemote)
         {
-            if (this.field_70146_Z.nextInt(900) == 0 && this.field_70725_aQ == 0)
+            if (this.rand.nextInt(900) == 0 && this.deathTime == 0)
             {
-                this.func_70691_i(1.0F);
+                this.heal(1.0F, EntityRegainHealthEvent.RegainReason.REGEN); // CraftBukkit
             }
 
-            if (this.func_190684_dE())
+            if (this.canEatGrass())
             {
-                if (!this.func_110204_cc() && !this.func_184207_aI() && this.field_70146_Z.nextInt(300) == 0 && this.field_70170_p.func_180495_p(new BlockPos(MathHelper.func_76128_c(this.field_70165_t), MathHelper.func_76128_c(this.field_70163_u) - 1, MathHelper.func_76128_c(this.field_70161_v))).func_177230_c() == Blocks.field_150349_c)
+                if (!this.isEatingHaystack() && !this.isBeingRidden() && this.rand.nextInt(300) == 0 && this.world.getBlockState(new BlockPos(MathHelper.floor(this.posX), MathHelper.floor(this.posY) - 1, MathHelper.floor(this.posZ))).getBlock() == Blocks.GRASS)
                 {
-                    this.func_110227_p(true);
+                    this.setEatingHaystack(true);
                 }
 
-                if (this.func_110204_cc() && ++this.field_190689_bJ > 50)
+                if (this.isEatingHaystack() && ++this.eatingCounter > 50)
                 {
-                    this.field_190689_bJ = 0;
-                    this.func_110227_p(false);
+                    this.eatingCounter = 0;
+                    this.setEatingHaystack(false);
                 }
             }
 
-            this.func_190679_dD();
+            this.followMother();
         }
     }
 
-    protected void func_190679_dD()
+    protected void followMother()
     {
-        if (this.func_110205_ce() && this.func_70631_g_() && !this.func_110204_cc())
+        if (this.isBreeding() && this.isChild() && !this.isEatingHaystack())
         {
-            AbstractHorse abstracthorse = this.func_110250_a(this, 16.0D);
+            AbstractHorse abstracthorse = this.getClosestHorse(this, 16.0D);
 
-            if (abstracthorse != null && this.func_70068_e(abstracthorse) > 4.0D)
+            if (abstracthorse != null && this.getDistanceSqToEntity(abstracthorse) > 4.0D)
             {
-                this.field_70699_by.func_75494_a(abstracthorse);
+                this.navigator.getPathToEntityLiving(abstracthorse);
             }
         }
     }
 
-    public boolean func_190684_dE()
+    public boolean canEatGrass()
     {
         return true;
     }
 
-    public void func_70071_h_()
+    public void onUpdate()
     {
-        super.func_70071_h_();
+        super.onUpdate();
 
-        if (this.field_110290_bE > 0 && ++this.field_110290_bE > 30)
+        if (this.openMouthCounter > 0 && ++this.openMouthCounter > 30)
         {
-            this.field_110290_bE = 0;
-            this.func_110208_b(64, false);
+            this.openMouthCounter = 0;
+            this.setHorseWatchableBoolean(64, false);
         }
 
-        if (this.func_184186_bw() && this.field_110295_bF > 0 && ++this.field_110295_bF > 20)
+        if (this.canPassengerSteer() && this.jumpRearingCounter > 0 && ++this.jumpRearingCounter > 20)
         {
-            this.field_110295_bF = 0;
-            this.func_110219_q(false);
+            this.jumpRearingCounter = 0;
+            this.setRearing(false);
         }
 
-        if (this.field_110278_bp > 0 && ++this.field_110278_bp > 8)
+        if (this.tailCounter > 0 && ++this.tailCounter > 8)
         {
-            this.field_110278_bp = 0;
+            this.tailCounter = 0;
         }
 
-        if (this.field_110279_bq > 0)
+        if (this.sprintCounter > 0)
         {
-            ++this.field_110279_bq;
+            ++this.sprintCounter;
 
-            if (this.field_110279_bq > 300)
+            if (this.sprintCounter > 300)
             {
-                this.field_110279_bq = 0;
+                this.sprintCounter = 0;
             }
         }
 
-        this.field_110284_bK = this.field_110283_bJ;
+        this.prevHeadLean = this.headLean;
 
-        if (this.func_110204_cc())
+        if (this.isEatingHaystack())
         {
-            this.field_110283_bJ += (1.0F - this.field_110283_bJ) * 0.4F + 0.05F;
+            this.headLean += (1.0F - this.headLean) * 0.4F + 0.05F;
 
-            if (this.field_110283_bJ > 1.0F)
+            if (this.headLean > 1.0F)
             {
-                this.field_110283_bJ = 1.0F;
+                this.headLean = 1.0F;
             }
         }
         else
         {
-            this.field_110283_bJ += (0.0F - this.field_110283_bJ) * 0.4F - 0.05F;
+            this.headLean += (0.0F - this.headLean) * 0.4F - 0.05F;
 
-            if (this.field_110283_bJ < 0.0F)
+            if (this.headLean < 0.0F)
             {
-                this.field_110283_bJ = 0.0F;
+                this.headLean = 0.0F;
             }
         }
 
-        this.field_110282_bM = this.field_110281_bL;
+        this.prevRearingAmount = this.rearingAmount;
 
-        if (this.func_110209_cd())
+        if (this.isRearing())
         {
-            this.field_110283_bJ = 0.0F;
-            this.field_110284_bK = this.field_110283_bJ;
-            this.field_110281_bL += (1.0F - this.field_110281_bL) * 0.4F + 0.05F;
+            this.headLean = 0.0F;
+            this.prevHeadLean = this.headLean;
+            this.rearingAmount += (1.0F - this.rearingAmount) * 0.4F + 0.05F;
 
-            if (this.field_110281_bL > 1.0F)
+            if (this.rearingAmount > 1.0F)
             {
-                this.field_110281_bL = 1.0F;
+                this.rearingAmount = 1.0F;
             }
         }
         else
         {
-            this.field_110294_bI = false;
-            this.field_110281_bL += (0.8F * this.field_110281_bL * this.field_110281_bL * this.field_110281_bL - this.field_110281_bL) * 0.6F - 0.05F;
+            this.allowStandSliding = false;
+            this.rearingAmount += (0.8F * this.rearingAmount * this.rearingAmount * this.rearingAmount - this.rearingAmount) * 0.6F - 0.05F;
 
-            if (this.field_110281_bL < 0.0F)
+            if (this.rearingAmount < 0.0F)
             {
-                this.field_110281_bL = 0.0F;
+                this.rearingAmount = 0.0F;
             }
         }
 
-        this.field_110288_bO = this.field_110287_bN;
+        this.prevMouthOpenness = this.mouthOpenness;
 
-        if (this.func_110233_w(64))
+        if (this.getHorseWatchableBoolean(64))
         {
-            this.field_110287_bN += (1.0F - this.field_110287_bN) * 0.7F + 0.05F;
+            this.mouthOpenness += (1.0F - this.mouthOpenness) * 0.7F + 0.05F;
 
-            if (this.field_110287_bN > 1.0F)
+            if (this.mouthOpenness > 1.0F)
             {
-                this.field_110287_bN = 1.0F;
+                this.mouthOpenness = 1.0F;
             }
         }
         else
         {
-            this.field_110287_bN += (0.0F - this.field_110287_bN) * 0.7F - 0.05F;
+            this.mouthOpenness += (0.0F - this.mouthOpenness) * 0.7F - 0.05F;
 
-            if (this.field_110287_bN < 0.0F)
+            if (this.mouthOpenness < 0.0F)
             {
-                this.field_110287_bN = 0.0F;
+                this.mouthOpenness = 0.0F;
             }
         }
     }
 
-    private void func_110249_cI()
+    private void openHorseMouth()
     {
-        if (!this.field_70170_p.field_72995_K)
+        if (!this.world.isRemote)
         {
-            this.field_110290_bE = 1;
-            this.func_110208_b(64, true);
+            this.openMouthCounter = 1;
+            this.setHorseWatchableBoolean(64, true);
         }
     }
 
-    public void func_110227_p(boolean p_110227_1_)
+    public void setEatingHaystack(boolean p_110227_1_)
     {
-        this.func_110208_b(16, p_110227_1_);
+        this.setHorseWatchableBoolean(16, p_110227_1_);
     }
 
-    public void func_110219_q(boolean p_110219_1_)
+    public void setRearing(boolean rearing)
     {
-        if (p_110219_1_)
+        if (rearing)
         {
-            this.func_110227_p(false);
+            this.setEatingHaystack(false);
         }
 
-        this.func_110208_b(32, p_110219_1_);
+        this.setHorseWatchableBoolean(32, rearing);
     }
 
-    private void func_110220_cK()
+    private void makeHorseRear()
     {
-        if (this.func_184186_bw())
+        if (this.canPassengerSteer())
         {
-            this.field_110295_bF = 1;
-            this.func_110219_q(true);
+            this.jumpRearingCounter = 1;
+            this.setRearing(true);
         }
     }
 
-    public void func_190687_dF()
+    public void makeMad()
     {
-        this.func_110220_cK();
-        SoundEvent soundevent = this.func_184785_dv();
+        this.makeHorseRear();
+        SoundEvent soundevent = this.getAngrySound();
 
         if (soundevent != null)
         {
-            this.func_184185_a(soundevent, this.func_70599_aP(), this.func_70647_i());
+            this.playSound(soundevent, this.getSoundVolume(), this.getSoundPitch());
         }
     }
 
-    public boolean func_110263_g(EntityPlayer p_110263_1_)
+    public boolean setTamedBy(EntityPlayer player)
     {
-        this.func_184779_b(p_110263_1_.func_110124_au());
-        this.func_110234_j(true);
-        this.field_70170_p.func_72960_a(this, (byte)7);
+        this.setOwnerUniqueId(player.getUniqueID());
+        this.setHorseTamed(true);
+        this.world.setEntityState(this, (byte)7);
         return true;
     }
 
-    public void func_70612_e(float p_70612_1_, float p_70612_2_)
+    public void moveEntityWithHeading(float strafe, float forward)
     {
-        if (this.func_184207_aI() && this.func_82171_bF() && this.func_110257_ck())
+        if (this.isBeingRidden() && this.canBeSteered() && this.isHorseSaddled())
         {
-            EntityLivingBase entitylivingbase = (EntityLivingBase)this.func_184179_bs();
-            this.field_70177_z = entitylivingbase.field_70177_z;
-            this.field_70126_B = this.field_70177_z;
-            this.field_70125_A = entitylivingbase.field_70125_A * 0.5F;
-            this.func_70101_b(this.field_70177_z, this.field_70125_A);
-            this.field_70761_aq = this.field_70177_z;
-            this.field_70759_as = this.field_70761_aq;
-            p_70612_1_ = entitylivingbase.field_70702_br * 0.5F;
-            p_70612_2_ = entitylivingbase.field_70701_bs;
+            EntityLivingBase entitylivingbase = (EntityLivingBase)this.getControllingPassenger();
+            this.rotationYaw = entitylivingbase.rotationYaw;
+            this.prevRotationYaw = this.rotationYaw;
+            this.rotationPitch = entitylivingbase.rotationPitch * 0.5F;
+            this.setRotation(this.rotationYaw, this.rotationPitch);
+            this.renderYawOffset = this.rotationYaw;
+            this.rotationYawHead = this.renderYawOffset;
+            strafe = entitylivingbase.moveStrafing * 0.5F;
+            forward = entitylivingbase.moveForward;
 
-            if (p_70612_2_ <= 0.0F)
+            if (forward <= 0.0F)
             {
-                p_70612_2_ *= 0.25F;
-                this.field_110285_bP = 0;
+                forward *= 0.25F;
+                this.gallopTime = 0;
             }
 
-            if (this.field_70122_E && this.field_110277_bt == 0.0F && this.func_110209_cd() && !this.field_110294_bI)
+            if (this.onGround && this.jumpPower == 0.0F && this.isRearing() && !this.allowStandSliding)
             {
-                p_70612_1_ = 0.0F;
-                p_70612_2_ = 0.0F;
+                strafe = 0.0F;
+                forward = 0.0F;
             }
 
-            if (this.field_110277_bt > 0.0F && !this.func_110246_bZ() && this.field_70122_E)
+            if (this.jumpPower > 0.0F && !this.isHorseJumping() && this.onGround)
             {
-                this.field_70181_x = this.func_110215_cj() * (double)this.field_110277_bt;
+                this.motionY = this.getHorseJumpStrength() * (double)this.jumpPower;
 
-                if (this.func_70644_a(MobEffects.field_76430_j))
+                if (this.isPotionActive(MobEffects.JUMP_BOOST))
                 {
-                    this.field_70181_x += (double)((float)(this.func_70660_b(MobEffects.field_76430_j).func_76458_c() + 1) * 0.1F);
+                    this.motionY += (double)((float)(this.getActivePotionEffect(MobEffects.JUMP_BOOST).getAmplifier() + 1) * 0.1F);
                 }
 
-                this.func_110255_k(true);
-                this.field_70160_al = true;
+                this.setHorseJumping(true);
+                this.isAirBorne = true;
 
-                if (p_70612_2_ > 0.0F)
+                if (forward > 0.0F)
                 {
-                    float f = MathHelper.func_76126_a(this.field_70177_z * 0.017453292F);
-                    float f1 = MathHelper.func_76134_b(this.field_70177_z * 0.017453292F);
-                    this.field_70159_w += (double)(-0.4F * f * this.field_110277_bt);
-                    this.field_70179_y += (double)(0.4F * f1 * this.field_110277_bt);
-                    this.func_184185_a(SoundEvents.field_187720_cs, 0.4F, 1.0F);
+                    float f = MathHelper.sin(this.rotationYaw * 0.017453292F);
+                    float f1 = MathHelper.cos(this.rotationYaw * 0.017453292F);
+                    this.motionX += (double)(-0.4F * f * this.jumpPower);
+                    this.motionZ += (double)(0.4F * f1 * this.jumpPower);
+                    this.playSound(SoundEvents.ENTITY_HORSE_JUMP, 0.4F, 1.0F);
                 }
 
-                this.field_110277_bt = 0.0F;
+                this.jumpPower = 0.0F;
             }
 
-            this.field_70747_aH = this.func_70689_ay() * 0.1F;
+            this.jumpMovementFactor = this.getAIMoveSpeed() * 0.1F;
 
-            if (this.func_184186_bw())
+            if (this.canPassengerSteer())
             {
-                this.func_70659_e((float)this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111126_e());
-                super.func_70612_e(p_70612_1_, p_70612_2_);
+                this.setAIMoveSpeed((float)this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).getAttributeValue());
+                super.moveEntityWithHeading(strafe, forward);
             }
             else if (entitylivingbase instanceof EntityPlayer)
             {
-                this.field_70159_w = 0.0D;
-                this.field_70181_x = 0.0D;
-                this.field_70179_y = 0.0D;
+                this.motionX = 0.0D;
+                this.motionY = 0.0D;
+                this.motionZ = 0.0D;
             }
 
-            if (this.field_70122_E)
+            if (this.onGround)
             {
-                this.field_110277_bt = 0.0F;
-                this.func_110255_k(false);
+                this.jumpPower = 0.0F;
+                this.setHorseJumping(false);
             }
 
-            this.field_184618_aE = this.field_70721_aZ;
-            double d1 = this.field_70165_t - this.field_70169_q;
-            double d0 = this.field_70161_v - this.field_70166_s;
-            float f2 = MathHelper.func_76133_a(d1 * d1 + d0 * d0) * 4.0F;
+            this.prevLimbSwingAmount = this.limbSwingAmount;
+            double d1 = this.posX - this.prevPosX;
+            double d0 = this.posZ - this.prevPosZ;
+            float f2 = MathHelper.sqrt(d1 * d1 + d0 * d0) * 4.0F;
 
             if (f2 > 1.0F)
             {
                 f2 = 1.0F;
             }
 
-            this.field_70721_aZ += (f2 - this.field_70721_aZ) * 0.4F;
-            this.field_184619_aG += this.field_70721_aZ;
+            this.limbSwingAmount += (f2 - this.limbSwingAmount) * 0.4F;
+            this.limbSwing += this.limbSwingAmount;
         }
         else
         {
-            this.field_70747_aH = 0.02F;
-            super.func_70612_e(p_70612_1_, p_70612_2_);
+            this.jumpMovementFactor = 0.02F;
+            super.moveEntityWithHeading(strafe, forward);
         }
     }
 
-    public static void func_190683_c(DataFixer p_190683_0_, Class<?> p_190683_1_)
+    public static void registerFixesAbstractHorse(DataFixer fixer, Class<?> entityClass)
     {
-        EntityLiving.func_189752_a(p_190683_0_, p_190683_1_);
-        p_190683_0_.func_188258_a(FixTypes.ENTITY, new ItemStackData(p_190683_1_, new String[] {"SaddleItem"}));
+        EntityLiving.registerFixesMob(fixer, entityClass);
+        fixer.registerWalker(FixTypes.ENTITY, new ItemStackData(entityClass, new String[] {"SaddleItem"}));
     }
 
-    public void func_70014_b(NBTTagCompound p_70014_1_)
+    public void writeEntityToNBT(NBTTagCompound compound)
     {
-        super.func_70014_b(p_70014_1_);
-        p_70014_1_.func_74757_a("EatingHaystack", this.func_110204_cc());
-        p_70014_1_.func_74757_a("Bred", this.func_110205_ce());
-        p_70014_1_.func_74768_a("Temper", this.func_110252_cg());
-        p_70014_1_.func_74757_a("Tame", this.func_110248_bS());
+        super.writeEntityToNBT(compound);
+        compound.setBoolean("EatingHaystack", this.isEatingHaystack());
+        compound.setBoolean("Bred", this.isBreeding());
+        compound.setInteger("Temper", this.getTemper());
+        compound.setBoolean("Tame", this.isTame());
 
-        if (this.func_184780_dh() != null)
+        if (this.getOwnerUniqueId() != null)
         {
-            p_70014_1_.func_74778_a("OwnerUUID", this.func_184780_dh().toString());
+            compound.setString("OwnerUUID", this.getOwnerUniqueId().toString());
         }
 
-        if (!this.field_110296_bG.func_70301_a(0).func_190926_b())
+        if (!this.horseChest.getStackInSlot(0).isEmpty())
         {
-            p_70014_1_.func_74782_a("SaddleItem", this.field_110296_bG.func_70301_a(0).func_77955_b(new NBTTagCompound()));
+            compound.setTag("SaddleItem", this.horseChest.getStackInSlot(0).writeToNBT(new NBTTagCompound()));
         }
     }
 
-    public void func_70037_a(NBTTagCompound p_70037_1_)
+    public void readEntityFromNBT(NBTTagCompound compound)
     {
-        super.func_70037_a(p_70037_1_);
-        this.func_110227_p(p_70037_1_.func_74767_n("EatingHaystack"));
-        this.func_110242_l(p_70037_1_.func_74767_n("Bred"));
-        this.func_110238_s(p_70037_1_.func_74762_e("Temper"));
-        this.func_110234_j(p_70037_1_.func_74767_n("Tame"));
+        super.readEntityFromNBT(compound);
+        this.setEatingHaystack(compound.getBoolean("EatingHaystack"));
+        this.setBreeding(compound.getBoolean("Bred"));
+        this.setTemper(compound.getInteger("Temper"));
+        this.setHorseTamed(compound.getBoolean("Tame"));
         String s;
 
-        if (p_70037_1_.func_150297_b("OwnerUUID", 8))
+        if (compound.hasKey("OwnerUUID", 8))
         {
-            s = p_70037_1_.func_74779_i("OwnerUUID");
+            s = compound.getString("OwnerUUID");
         }
         else
         {
-            String s1 = p_70037_1_.func_74779_i("Owner");
-            s = PreYggdrasilConverter.func_187473_a(this.func_184102_h(), s1);
+            String s1 = compound.getString("Owner");
+            s = PreYggdrasilConverter.convertMobOwnerIfNeeded(this.getServer(), s1);
         }
 
         if (!s.isEmpty())
         {
-            this.func_184779_b(UUID.fromString(s));
+            this.setOwnerUniqueId(UUID.fromString(s));
         }
 
-        IAttributeInstance iattributeinstance = this.func_110140_aT().func_111152_a("Speed");
+        IAttributeInstance iattributeinstance = this.getAttributeMap().getAttributeInstanceByName("Speed");
 
         if (iattributeinstance != null)
         {
-            this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(iattributeinstance.func_111125_b() * 0.25D);
+            this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).setBaseValue(iattributeinstance.getBaseValue() * 0.25D);
         }
 
-        if (p_70037_1_.func_150297_b("SaddleItem", 10))
+        if (compound.hasKey("SaddleItem", 10))
         {
-            ItemStack itemstack = new ItemStack(p_70037_1_.func_74775_l("SaddleItem"));
+            ItemStack itemstack = new ItemStack(compound.getCompoundTag("SaddleItem"));
 
-            if (itemstack.func_77973_b() == Items.field_151141_av)
+            if (itemstack.getItem() == Items.SADDLE)
             {
-                this.field_110296_bG.func_70299_a(0, itemstack);
+                this.horseChest.setInventorySlotContents(0, itemstack);
             }
         }
 
-        this.func_110232_cE();
+        this.updateHorseSlots();
     }
 
-    public boolean func_70878_b(EntityAnimal p_70878_1_)
+    public boolean canMateWith(EntityAnimal otherAnimal)
     {
         return false;
     }
 
-    protected boolean func_110200_cJ()
+    protected boolean canMate()
     {
-        return !this.func_184207_aI() && !this.func_184218_aH() && this.func_110248_bS() && !this.func_70631_g_() && this.func_110143_aJ() >= this.func_110138_aP() && this.func_70880_s();
+        return !this.isBeingRidden() && !this.isRiding() && this.isTame() && !this.isChild() && this.getHealth() >= this.getMaxHealth() && this.isInLove();
     }
 
     @Nullable
-    public EntityAgeable func_90011_a(EntityAgeable p_90011_1_)
+    public EntityAgeable createChild(EntityAgeable ageable)
     {
         return null;
     }
 
-    protected void func_190681_a(EntityAgeable p_190681_1_, AbstractHorse p_190681_2_)
+    protected void setOffspringAttributes(EntityAgeable p_190681_1_, AbstractHorse p_190681_2_)
     {
-        double d0 = this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111125_b() + p_190681_1_.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111125_b() + (double)this.func_110267_cL();
-        p_190681_2_.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(d0 / 3.0D);
-        double d1 = this.func_110148_a(field_110271_bv).func_111125_b() + p_190681_1_.func_110148_a(field_110271_bv).func_111125_b() + this.func_110245_cM();
-        p_190681_2_.func_110148_a(field_110271_bv).func_111128_a(d1 / 3.0D);
-        double d2 = this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111125_b() + p_190681_1_.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111125_b() + this.func_110203_cN();
-        p_190681_2_.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(d2 / 3.0D);
+        double d0 = this.getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).getBaseValue() + p_190681_1_.getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).getBaseValue() + (double)this.getModifiedMaxHealth();
+        p_190681_2_.getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).setBaseValue(d0 / 3.0D);
+        double d1 = this.getEntityAttribute(JUMP_STRENGTH).getBaseValue() + p_190681_1_.getEntityAttribute(JUMP_STRENGTH).getBaseValue() + this.getModifiedJumpStrength();
+        p_190681_2_.getEntityAttribute(JUMP_STRENGTH).setBaseValue(d1 / 3.0D);
+        double d2 = this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).getBaseValue() + p_190681_1_.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).getBaseValue() + this.getModifiedMovementSpeed();
+        p_190681_2_.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).setBaseValue(d2 / 3.0D);
     }
 
-    public boolean func_82171_bF()
+    public boolean canBeSteered()
     {
-        return this.func_184179_bs() instanceof EntityLivingBase;
+        return this.getControllingPassenger() instanceof EntityLivingBase;
     }
 
     @SideOnly(Side.CLIENT)
-    public float func_110258_o(float p_110258_1_)
+    public float getGrassEatingAmount(float p_110258_1_)
     {
-        return this.field_110284_bK + (this.field_110283_bJ - this.field_110284_bK) * p_110258_1_;
+        return this.prevHeadLean + (this.headLean - this.prevHeadLean) * p_110258_1_;
     }
 
     @SideOnly(Side.CLIENT)
-    public float func_110223_p(float p_110223_1_)
+    public float getRearingAmount(float p_110223_1_)
     {
-        return this.field_110282_bM + (this.field_110281_bL - this.field_110282_bM) * p_110223_1_;
+        return this.prevRearingAmount + (this.rearingAmount - this.prevRearingAmount) * p_110223_1_;
     }
 
     @SideOnly(Side.CLIENT)
-    public float func_110201_q(float p_110201_1_)
+    public float getMouthOpennessAngle(float p_110201_1_)
     {
-        return this.field_110288_bO + (this.field_110287_bN - this.field_110288_bO) * p_110201_1_;
+        return this.prevMouthOpenness + (this.mouthOpenness - this.prevMouthOpenness) * p_110201_1_;
     }
 
     @SideOnly(Side.CLIENT)
-    public void func_110206_u(int p_110206_1_)
+    public void setJumpPower(int jumpPowerIn)
     {
-        if (this.func_110257_ck())
+        if (this.isHorseSaddled())
         {
-            if (p_110206_1_ < 0)
+            if (jumpPowerIn < 0)
             {
-                p_110206_1_ = 0;
+                jumpPowerIn = 0;
             }
             else
             {
-                this.field_110294_bI = true;
-                this.func_110220_cK();
+                this.allowStandSliding = true;
+                this.makeHorseRear();
             }
 
-            if (p_110206_1_ >= 90)
+            if (jumpPowerIn >= 90)
             {
-                this.field_110277_bt = 1.0F;
+                this.jumpPower = 1.0F;
             }
             else
             {
-                this.field_110277_bt = 0.4F + 0.4F * (float)p_110206_1_ / 90.0F;
+                this.jumpPower = 0.4F + 0.4F * (float)jumpPowerIn / 90.0F;
             }
         }
     }
 
-    public boolean func_184776_b()
+    public boolean canJump()
     {
-        return this.func_110257_ck();
+        return this.isHorseSaddled();
     }
 
-    public void func_184775_b(int p_184775_1_)
+    public void handleStartJump(int p_184775_1_)
     {
-        this.field_110294_bI = true;
-        this.func_110220_cK();
+        this.allowStandSliding = true;
+        this.makeHorseRear();
     }
 
-    public void func_184777_r_()
+    public void handleStopJump()
     {
     }
 
     @SideOnly(Side.CLIENT)
-    protected void func_110216_r(boolean p_110216_1_)
+    protected void spawnHorseParticles(boolean p_110216_1_)
     {
         EnumParticleTypes enumparticletypes = p_110216_1_ ? EnumParticleTypes.HEART : EnumParticleTypes.SMOKE_NORMAL;
 
         for (int i = 0; i < 7; ++i)
         {
-            double d0 = this.field_70146_Z.nextGaussian() * 0.02D;
-            double d1 = this.field_70146_Z.nextGaussian() * 0.02D;
-            double d2 = this.field_70146_Z.nextGaussian() * 0.02D;
-            this.field_70170_p.func_175688_a(enumparticletypes, this.field_70165_t + (double)(this.field_70146_Z.nextFloat() * this.field_70130_N * 2.0F) - (double)this.field_70130_N, this.field_70163_u + 0.5D + (double)(this.field_70146_Z.nextFloat() * this.field_70131_O), this.field_70161_v + (double)(this.field_70146_Z.nextFloat() * this.field_70130_N * 2.0F) - (double)this.field_70130_N, d0, d1, d2, new int[0]);
+            double d0 = this.rand.nextGaussian() * 0.02D;
+            double d1 = this.rand.nextGaussian() * 0.02D;
+            double d2 = this.rand.nextGaussian() * 0.02D;
+            this.world.spawnParticle(enumparticletypes, this.posX + (double)(this.rand.nextFloat() * this.width * 2.0F) - (double)this.width, this.posY + 0.5D + (double)(this.rand.nextFloat() * this.height), this.posZ + (double)(this.rand.nextFloat() * this.width * 2.0F) - (double)this.width, d0, d1, d2, new int[0]);
         }
     }
 
     @SideOnly(Side.CLIENT)
-    public void func_70103_a(byte p_70103_1_)
+    public void handleStatusUpdate(byte id)
     {
-        if (p_70103_1_ == 7)
+        if (id == 7)
         {
-            this.func_110216_r(true);
+            this.spawnHorseParticles(true);
         }
-        else if (p_70103_1_ == 6)
+        else if (id == 6)
         {
-            this.func_110216_r(false);
+            this.spawnHorseParticles(false);
         }
         else
         {
-            super.func_70103_a(p_70103_1_);
+            super.handleStatusUpdate(id);
         }
     }
 
-    public void func_184232_k(Entity p_184232_1_)
+    public void updatePassenger(Entity passenger)
     {
-        super.func_184232_k(p_184232_1_);
+        super.updatePassenger(passenger);
 
-        if (p_184232_1_ instanceof EntityLiving)
+        if (passenger instanceof EntityLiving)
         {
-            EntityLiving entityliving = (EntityLiving)p_184232_1_;
-            this.field_70761_aq = entityliving.field_70761_aq;
+            EntityLiving entityliving = (EntityLiving)passenger;
+            this.renderYawOffset = entityliving.renderYawOffset;
         }
 
-        if (this.field_110282_bM > 0.0F)
+        if (this.prevRearingAmount > 0.0F)
         {
-            float f3 = MathHelper.func_76126_a(this.field_70761_aq * 0.017453292F);
-            float f = MathHelper.func_76134_b(this.field_70761_aq * 0.017453292F);
-            float f1 = 0.7F * this.field_110282_bM;
-            float f2 = 0.15F * this.field_110282_bM;
-            p_184232_1_.func_70107_b(this.field_70165_t + (double)(f1 * f3), this.field_70163_u + this.func_70042_X() + p_184232_1_.func_70033_W() + (double)f2, this.field_70161_v - (double)(f1 * f));
+            float f3 = MathHelper.sin(this.renderYawOffset * 0.017453292F);
+            float f = MathHelper.cos(this.renderYawOffset * 0.017453292F);
+            float f1 = 0.7F * this.prevRearingAmount;
+            float f2 = 0.15F * this.prevRearingAmount;
+            passenger.setPosition(this.posX + (double)(f1 * f3), this.posY + this.getMountedYOffset() + passenger.getYOffset() + (double)f2, this.posZ - (double)(f1 * f));
 
-            if (p_184232_1_ instanceof EntityLivingBase)
+            if (passenger instanceof EntityLivingBase)
             {
-                ((EntityLivingBase)p_184232_1_).field_70761_aq = this.field_70761_aq;
+                ((EntityLivingBase)passenger).renderYawOffset = this.renderYawOffset;
             }
         }
     }
 
-    protected float func_110267_cL()
+    protected float getModifiedMaxHealth()
     {
-        return 15.0F + (float)this.field_70146_Z.nextInt(8) + (float)this.field_70146_Z.nextInt(9);
+        return 15.0F + (float)this.rand.nextInt(8) + (float)this.rand.nextInt(9);
     }
 
-    protected double func_110245_cM()
+    protected double getModifiedJumpStrength()
     {
-        return 0.4000000059604645D + this.field_70146_Z.nextDouble() * 0.2D + this.field_70146_Z.nextDouble() * 0.2D + this.field_70146_Z.nextDouble() * 0.2D;
+        return 0.4000000059604645D + this.rand.nextDouble() * 0.2D + this.rand.nextDouble() * 0.2D + this.rand.nextDouble() * 0.2D;
     }
 
-    protected double func_110203_cN()
+    protected double getModifiedMovementSpeed()
     {
-        return (0.44999998807907104D + this.field_70146_Z.nextDouble() * 0.3D + this.field_70146_Z.nextDouble() * 0.3D + this.field_70146_Z.nextDouble() * 0.3D) * 0.25D;
+        return (0.44999998807907104D + this.rand.nextDouble() * 0.3D + this.rand.nextDouble() * 0.3D + this.rand.nextDouble() * 0.3D) * 0.25D;
     }
 
-    public boolean func_70617_f_()
+    public boolean isOnLadder()
     {
         return false;
     }
 
-    public float func_70047_e()
+    public float getEyeHeight()
     {
-        return this.field_70131_O;
+        return this.height;
     }
 
-    public boolean func_190677_dK()
+    public boolean wearsArmor()
     {
         return false;
     }
 
-    public boolean func_190682_f(ItemStack p_190682_1_)
+    public boolean isArmor(ItemStack stack)
     {
         return false;
     }
 
-    public boolean func_174820_d(int p_174820_1_, ItemStack p_174820_2_)
+    public boolean replaceItemInInventory(int inventorySlot, ItemStack itemStackIn)
     {
-        int i = p_174820_1_ - 400;
+        int i = inventorySlot - 400;
 
-        if (i >= 0 && i < 2 && i < this.field_110296_bG.func_70302_i_())
+        if (i >= 0 && i < 2 && i < this.horseChest.getSizeInventory())
         {
-            if (i == 0 && p_174820_2_.func_77973_b() != Items.field_151141_av)
+            if (i == 0 && itemStackIn.getItem() != Items.SADDLE)
             {
                 return false;
             }
-            else if (i != 1 || this.func_190677_dK() && this.func_190682_f(p_174820_2_))
+            else if (i != 1 || this.wearsArmor() && this.isArmor(itemStackIn))
             {
-                this.field_110296_bG.func_70299_a(i, p_174820_2_);
-                this.func_110232_cE();
+                this.horseChest.setInventorySlotContents(i, itemStackIn);
+                this.updateHorseSlots();
                 return true;
             }
             else
@@ -1187,11 +1191,11 @@
         }
         else
         {
-            int j = p_174820_1_ - 500 + 2;
+            int j = inventorySlot - 500 + 2;
 
-            if (j >= 2 && j < this.field_110296_bG.func_70302_i_())
+            if (j >= 2 && j < this.horseChest.getSizeInventory())
             {
-                this.field_110296_bG.func_70299_a(j, p_174820_2_);
+                this.horseChest.setInventorySlotContents(j, itemStackIn);
                 return true;
             }
             else
@@ -1202,21 +1206,39 @@
     }
 
     @Nullable
-    public Entity func_184179_bs()
+    public Entity getControllingPassenger()
     {
-        return this.func_184188_bt().isEmpty() ? null : (Entity)this.func_184188_bt().get(0);
+        return this.getPassengers().isEmpty() ? null : (Entity)this.getPassengers().get(0);
     }
 
     @Nullable
-    public IEntityLivingData func_180482_a(DifficultyInstance p_180482_1_, @Nullable IEntityLivingData p_180482_2_)
+    public IEntityLivingData onInitialSpawn(DifficultyInstance difficulty, @Nullable IEntityLivingData livingdata)
     {
-        p_180482_2_ = super.func_180482_a(p_180482_1_, p_180482_2_);
+        livingdata = super.onInitialSpawn(difficulty, livingdata);
 
-        if (this.field_70146_Z.nextInt(5) == 0)
+        if (this.rand.nextInt(5) == 0)
         {
-            this.func_70873_a(-24000);
+            this.setGrowingAge(-24000);
         }
 
-        return p_180482_2_;
+        return livingdata;
     }
+
+    // FORGE
+    private net.minecraftforge.items.IItemHandler itemHandler = null; // Initialized by initHorseChest above.
+
+    @SuppressWarnings("unchecked")
+    @Override
+    @Nullable
+    public <T> T getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable net.minecraft.util.EnumFacing facing)
+    {
+        if (capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY) return (T) itemHandler;
+        return super.getCapability(capability, facing);
+    }
+
+    @Override
+    public boolean hasCapability(net.minecraftforge.common.capabilities.Capability<?> capability, @Nullable net.minecraft.util.EnumFacing facing)
+    {
+        return capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY || super.hasCapability(capability, facing);
+    }
 }
