--- ../src-base/minecraft/net/minecraft/entity/passive/EntitySheep.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntitySheep.java
@@ -24,6 +24,7 @@
 import net.minecraft.init.Items;
 import net.minecraft.init.SoundEvents;
 import net.minecraft.inventory.Container;
+import net.minecraft.inventory.InventoryCraftResult;
 import net.minecraft.inventory.InventoryCrafting;
 import net.minecraft.item.EnumDyeColor;
 import net.minecraft.item.Item;
@@ -44,12 +45,21 @@
 import net.minecraft.world.storage.loot.LootTableList;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.entity.Player;
+import org.bukkit.event.entity.SheepRegrowWoolEvent;
+import org.bukkit.event.player.PlayerShearEntityEvent;
+import org.bukkit.inventory.InventoryView;
 
-public class EntitySheep extends EntityAnimal
+public class EntitySheep extends EntityAnimal implements net.minecraftforge.common.IShearable
 {
     private static final DataParameter<Byte> field_184774_bv = EntityDataManager.<Byte>func_187226_a(EntitySheep.class, DataSerializers.field_187191_a);
     private final InventoryCrafting field_90016_e = new InventoryCrafting(new Container()
     {
+        @Override
+        public InventoryView getBukkitView() {
+            return null;
+        }
+
         public boolean func_75145_c(EntityPlayer p_75145_1_)
         {
             return false;
@@ -70,6 +80,8 @@
         this.func_70105_a(0.9F, 1.3F);
         this.field_90016_e.func_70299_a(0, new ItemStack(Items.field_151100_aR));
         this.field_90016_e.func_70299_a(1, new ItemStack(Items.field_151100_aR));
+        this.field_90016_e.resultInventory = new InventoryCraftResult(); // CraftBukkit - add result slot for event
+
     }
 
     protected void func_184651_r()
@@ -180,10 +192,19 @@
     {
         ItemStack itemstack = p_184645_1_.func_184586_b(p_184645_2_);
 
-        if (itemstack.func_77973_b() == Items.field_151097_aZ && !this.func_70892_o() && !this.func_70631_g_())
+        if (false && itemstack.func_77973_b() == Items.field_151097_aZ && !this.func_70892_o() && !this.func_70631_g_())   //Forge: Moved to onSheared
         {
             if (!this.field_70170_p.field_72995_K)
             {
+                // CraftBukkit start
+                PlayerShearEntityEvent event = new PlayerShearEntityEvent((Player) p_184645_1_.getBukkitEntity(), this.getBukkitEntity());
+                this.field_70170_p.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return false;
+                }
+                // CraftBukkit end
+
                 this.func_70893_e(true);
                 int i = 1 + this.field_70146_Z.nextInt(3);
 
@@ -308,6 +329,12 @@
 
     public void func_70615_aA()
     {
+        // CraftBukkit start
+        SheepRegrowWoolEvent event = new SheepRegrowWoolEvent((org.bukkit.entity.Sheep) this.getBukkitEntity());
+        this.field_70170_p.getServer().getPluginManager().callEvent(event);
+
+        if (event.isCancelled()) return;
+        // CraftBukkit end
         this.func_70893_e(false);
 
         if (this.func_70631_g_())
@@ -369,4 +396,19 @@
         field_175514_bm.put(EnumDyeColor.RED, new float[] {0.6F, 0.2F, 0.2F});
         field_175514_bm.put(EnumDyeColor.BLACK, new float[] {0.1F, 0.1F, 0.1F});
     }
+
+    @Override public boolean isShearable(ItemStack item, net.minecraft.world.IBlockAccess world, BlockPos pos){ return !this.func_70892_o() && !this.func_70631_g_(); }
+    @Override
+    public java.util.List<ItemStack> onSheared(ItemStack item, net.minecraft.world.IBlockAccess world, BlockPos pos, int fortune)
+    {
+        this.func_70893_e(true);
+        int i = 1 + this.field_70146_Z.nextInt(3);
+
+        java.util.List<ItemStack> ret = new java.util.ArrayList<ItemStack>();
+        for (int j = 0; j < i; ++j)
+            ret.add(new ItemStack(Item.func_150898_a(Blocks.field_150325_L), 1, this.func_175509_cj().func_176765_a()));
+
+        this.func_184185_a(SoundEvents.field_187763_eJ, 1.0F, 1.0F);
+        return ret;
+    }
 }
