--- ../src-base/minecraft/net/minecraft/entity/passive/EntityCow.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityCow.java
@@ -24,6 +24,10 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 import net.minecraft.world.storage.loot.LootTableList;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.player.PlayerBucketFillEvent;
 
 public class EntityCow extends EntityAnimal
 {
@@ -94,18 +98,23 @@
 
         if (itemstack.func_77973_b() == Items.field_151133_ar && !p_184645_1_.field_71075_bZ.field_75098_d && !this.func_70631_g_())
         {
-            p_184645_1_.func_184185_a(SoundEvents.field_187564_an, 1.0F, 1.0F);
-            itemstack.func_190918_g(1);
+            // CraftBukkit start - Got milk?
+            Location loc = this.getBukkitEntity().getLocation();
+            PlayerBucketFillEvent event = CraftEventFactory.callPlayerBucketFillEvent(p_184645_1_, loc.getBlockX(), loc.getBlockY(), loc.getBlockZ(), null, itemstack, Items.field_151117_aB);
 
-            if (itemstack.func_190926_b())
-            {
-                p_184645_1_.func_184611_a(p_184645_2_, new ItemStack(Items.field_151117_aB));
+            if (event.isCancelled()) {
+                return false;
             }
-            else if (!p_184645_1_.field_71071_by.func_70441_a(new ItemStack(Items.field_151117_aB)))
-            {
-                p_184645_1_.func_71019_a(new ItemStack(Items.field_151117_aB), false);
-            }
 
+            ItemStack result = CraftItemStack.asNMSCopy(event.getItemStack());
+            p_184645_1_.func_184185_a(SoundEvents.field_187564_an, 1.0F, 1.0F);
+            itemstack.func_190918_g(1);
+            if (itemstack.func_190926_b()) {
+                p_184645_1_.func_184611_a(p_184645_2_, result);
+            } else if (!p_184645_1_.field_71071_by.func_70441_a(result)) {
+                p_184645_1_.func_71019_a(result, false);
+            }
+            // CraftBukkit end
             return true;
         }
         else
