--- ../src-base/minecraft/net/minecraft/entity/passive/EntityRabbit.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityRabbit.java
@@ -52,58 +52,63 @@
 import net.minecraft.world.storage.loot.LootTableList;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class EntityRabbit extends EntityAnimal
 {
-    private static final DataParameter<Integer> field_184773_bv = EntityDataManager.<Integer>func_187226_a(EntityRabbit.class, DataSerializers.field_187192_b);
-    private int field_175540_bm;
-    private int field_175535_bn;
-    private boolean field_175537_bp;
-    private int field_175538_bq;
-    private int field_175541_bs;
+    private static final DataParameter<Integer> RABBIT_TYPE = EntityDataManager.<Integer>createKey(EntityRabbit.class, DataSerializers.VARINT);
+    private int jumpTicks;
+    private int jumpDuration;
+    private boolean wasOnGround;
+    private int currentMoveTypeDuration;
+    private int carrotTicks;
 
-    public EntityRabbit(World p_i45869_1_)
+    public EntityRabbit(World worldIn)
     {
-        super(p_i45869_1_);
-        this.func_70105_a(0.4F, 0.5F);
-        this.field_70767_i = new EntityRabbit.RabbitJumpHelper(this);
-        this.field_70765_h = new EntityRabbit.RabbitMoveHelper(this);
-        this.func_175515_b(0.0D);
+        super(worldIn);
+        this.setSize(0.4F, 0.5F);
+        this.jumpHelper = new EntityRabbit.RabbitJumpHelper(this);
+        this.moveHelper = new EntityRabbit.RabbitMoveHelper(this);
+        initializePathFinderGoals();
     }
-
-    protected void func_184651_r()
+    // CraftBukkit start - code from constructor
+    public void initializePathFinderGoals(){
+        this.setMovementSpeed(0.0D);
+    }
+    // CraftBukkit end
+    protected void initEntityAI()
     {
-        this.field_70714_bg.func_75776_a(1, new EntityAISwimming(this));
-        this.field_70714_bg.func_75776_a(1, new EntityRabbit.AIPanic(this, 2.2D));
-        this.field_70714_bg.func_75776_a(2, new EntityAIMate(this, 0.8D));
-        this.field_70714_bg.func_75776_a(3, new EntityAITempt(this, 1.0D, Items.field_151172_bF, false));
-        this.field_70714_bg.func_75776_a(3, new EntityAITempt(this, 1.0D, Items.field_151150_bK, false));
-        this.field_70714_bg.func_75776_a(3, new EntityAITempt(this, 1.0D, Item.func_150898_a(Blocks.field_150327_N), false));
-        this.field_70714_bg.func_75776_a(4, new EntityRabbit.AIAvoidEntity(this, EntityPlayer.class, 8.0F, 2.2D, 2.2D));
-        this.field_70714_bg.func_75776_a(4, new EntityRabbit.AIAvoidEntity(this, EntityWolf.class, 10.0F, 2.2D, 2.2D));
-        this.field_70714_bg.func_75776_a(4, new EntityRabbit.AIAvoidEntity(this, EntityMob.class, 4.0F, 2.2D, 2.2D));
-        this.field_70714_bg.func_75776_a(5, new EntityRabbit.AIRaidFarm(this));
-        this.field_70714_bg.func_75776_a(6, new EntityAIWanderAvoidWater(this, 0.6D));
-        this.field_70714_bg.func_75776_a(11, new EntityAIWatchClosest(this, EntityPlayer.class, 10.0F));
+        this.tasks.addTask(1, new EntityAISwimming(this));
+        this.tasks.addTask(1, new EntityRabbit.AIPanic(this, 2.2D));
+        this.tasks.addTask(2, new EntityAIMate(this, 0.8D));
+        this.tasks.addTask(3, new EntityAITempt(this, 1.0D, Items.CARROT, false));
+        this.tasks.addTask(3, new EntityAITempt(this, 1.0D, Items.GOLDEN_CARROT, false));
+        this.tasks.addTask(3, new EntityAITempt(this, 1.0D, Item.getItemFromBlock(Blocks.YELLOW_FLOWER), false));
+        this.tasks.addTask(4, new EntityRabbit.AIAvoidEntity(this, EntityPlayer.class, 8.0F, 2.2D, 2.2D));
+        this.tasks.addTask(4, new EntityRabbit.AIAvoidEntity(this, EntityWolf.class, 10.0F, 2.2D, 2.2D));
+        this.tasks.addTask(4, new EntityRabbit.AIAvoidEntity(this, EntityMob.class, 4.0F, 2.2D, 2.2D));
+        this.tasks.addTask(5, new EntityRabbit.AIRaidFarm(this));
+        this.tasks.addTask(6, new EntityAIWanderAvoidWater(this, 0.6D));
+        this.tasks.addTask(11, new EntityAIWatchClosest(this, EntityPlayer.class, 10.0F));
     }
 
-    protected float func_175134_bD()
+    protected float getJumpUpwardsMotion()
     {
-        if (!this.field_70123_F && (!this.field_70765_h.func_75640_a() || this.field_70765_h.func_179919_e() <= this.field_70163_u + 0.5D))
+        if (!this.isCollidedHorizontally && (!this.moveHelper.isUpdating() || this.moveHelper.getY() <= this.posY + 0.5D))
         {
-            Path path = this.field_70699_by.func_75505_d();
+            Path path = this.navigator.getPath();
 
-            if (path != null && path.func_75873_e() < path.func_75874_d())
+            if (path != null && path.getCurrentPathIndex() < path.getCurrentPathLength())
             {
-                Vec3d vec3d = path.func_75878_a(this);
+                Vec3d vec3d = path.getPosition(this);
 
-                if (vec3d.field_72448_b > this.field_70163_u + 0.5D)
+                if (vec3d.yCoord > this.posY + 0.5D)
                 {
                     return 0.5F;
                 }
             }
 
-            return this.field_70765_h.func_75638_b() <= 0.6D ? 0.2F : 0.3F;
+            return this.moveHelper.getSpeed() <= 0.6D ? 0.2F : 0.3F;
         }
         else
         {
@@ -111,500 +116,514 @@
         }
     }
 
-    protected void func_70664_aZ()
+    protected void jump()
     {
-        super.func_70664_aZ();
-        double d0 = this.field_70765_h.func_75638_b();
+        super.jump();
+        double d0 = this.moveHelper.getSpeed();
 
         if (d0 > 0.0D)
         {
-            double d1 = this.field_70159_w * this.field_70159_w + this.field_70179_y * this.field_70179_y;
+            double d1 = this.motionX * this.motionX + this.motionZ * this.motionZ;
 
             if (d1 < 0.010000000000000002D)
             {
-                this.func_70060_a(0.0F, 1.0F, 0.1F);
+                this.moveRelative(0.0F, 1.0F, 0.1F);
             }
         }
 
-        if (!this.field_70170_p.field_72995_K)
+        if (!this.world.isRemote)
         {
-            this.field_70170_p.func_72960_a(this, (byte)1);
+            this.world.setEntityState(this, (byte)1);
         }
     }
 
     @SideOnly(Side.CLIENT)
-    public float func_175521_o(float p_175521_1_)
+    public float setJumpCompletion(float p_175521_1_)
     {
-        return this.field_175535_bn == 0 ? 0.0F : ((float)this.field_175540_bm + p_175521_1_) / (float)this.field_175535_bn;
+        return this.jumpDuration == 0 ? 0.0F : ((float)this.jumpTicks + p_175521_1_) / (float)this.jumpDuration;
     }
 
-    public void func_175515_b(double p_175515_1_)
+    public void setMovementSpeed(double newSpeed)
     {
-        this.func_70661_as().func_75489_a(p_175515_1_);
-        this.field_70765_h.func_75642_a(this.field_70765_h.func_179917_d(), this.field_70765_h.func_179919_e(), this.field_70765_h.func_179918_f(), p_175515_1_);
+        this.getNavigator().setSpeed(newSpeed);
+        this.moveHelper.setMoveTo(this.moveHelper.getX(), this.moveHelper.getY(), this.moveHelper.getZ(), newSpeed);
     }
 
-    public void func_70637_d(boolean p_70637_1_)
+    public void setJumping(boolean jumping)
     {
-        super.func_70637_d(p_70637_1_);
+        super.setJumping(jumping);
 
-        if (p_70637_1_)
+        if (jumping)
         {
-            this.func_184185_a(this.func_184771_da(), this.func_70599_aP(), ((this.field_70146_Z.nextFloat() - this.field_70146_Z.nextFloat()) * 0.2F + 1.0F) * 0.8F);
+            this.playSound(this.getJumpSound(), this.getSoundVolume(), ((this.rand.nextFloat() - this.rand.nextFloat()) * 0.2F + 1.0F) * 0.8F);
         }
     }
 
-    public void func_184770_cZ()
+    public void startJumping()
     {
-        this.func_70637_d(true);
-        this.field_175535_bn = 10;
-        this.field_175540_bm = 0;
+        this.setJumping(true);
+        this.jumpDuration = 10;
+        this.jumpTicks = 0;
     }
 
-    protected void func_70088_a()
+    protected void entityInit()
     {
-        super.func_70088_a();
-        this.field_70180_af.func_187214_a(field_184773_bv, Integer.valueOf(0));
+        super.entityInit();
+        this.dataManager.register(RABBIT_TYPE, Integer.valueOf(0));
     }
 
-    public void func_70619_bc()
+    public void updateAITasks()
     {
-        if (this.field_175538_bq > 0)
+        if (this.currentMoveTypeDuration > 0)
         {
-            --this.field_175538_bq;
+            --this.currentMoveTypeDuration;
         }
 
-        if (this.field_175541_bs > 0)
+        if (this.carrotTicks > 0)
         {
-            this.field_175541_bs -= this.field_70146_Z.nextInt(3);
+            this.carrotTicks -= this.rand.nextInt(3);
 
-            if (this.field_175541_bs < 0)
+            if (this.carrotTicks < 0)
             {
-                this.field_175541_bs = 0;
+                this.carrotTicks = 0;
             }
         }
 
-        if (this.field_70122_E)
+        if (this.onGround)
         {
-            if (!this.field_175537_bp)
+            if (!this.wasOnGround)
             {
-                this.func_70637_d(false);
-                this.func_175517_cu();
+                this.setJumping(false);
+                this.checkLandingDelay();
             }
 
-            if (this.func_175531_cl() == 99 && this.field_175538_bq == 0)
+            if (this.getRabbitType() == 99 && this.currentMoveTypeDuration == 0)
             {
-                EntityLivingBase entitylivingbase = this.func_70638_az();
+                EntityLivingBase entitylivingbase = this.getAttackTarget();
 
-                if (entitylivingbase != null && this.func_70068_e(entitylivingbase) < 16.0D)
+                if (entitylivingbase != null && this.getDistanceSqToEntity(entitylivingbase) < 16.0D)
                 {
-                    this.func_175533_a(entitylivingbase.field_70165_t, entitylivingbase.field_70161_v);
-                    this.field_70765_h.func_75642_a(entitylivingbase.field_70165_t, entitylivingbase.field_70163_u, entitylivingbase.field_70161_v, this.field_70765_h.func_75638_b());
-                    this.func_184770_cZ();
-                    this.field_175537_bp = true;
+                    this.calculateRotationYaw(entitylivingbase.posX, entitylivingbase.posZ);
+                    this.moveHelper.setMoveTo(entitylivingbase.posX, entitylivingbase.posY, entitylivingbase.posZ, this.moveHelper.getSpeed());
+                    this.startJumping();
+                    this.wasOnGround = true;
                 }
             }
 
-            EntityRabbit.RabbitJumpHelper entityrabbit$rabbitjumphelper = (EntityRabbit.RabbitJumpHelper)this.field_70767_i;
+            EntityRabbit.RabbitJumpHelper entityrabbit$rabbitjumphelper = (EntityRabbit.RabbitJumpHelper)this.jumpHelper;
 
-            if (!entityrabbit$rabbitjumphelper.func_180067_c())
+            if (!entityrabbit$rabbitjumphelper.getIsJumping())
             {
-                if (this.field_70765_h.func_75640_a() && this.field_175538_bq == 0)
+                if (this.moveHelper.isUpdating() && this.currentMoveTypeDuration == 0)
                 {
-                    Path path = this.field_70699_by.func_75505_d();
-                    Vec3d vec3d = new Vec3d(this.field_70765_h.func_179917_d(), this.field_70765_h.func_179919_e(), this.field_70765_h.func_179918_f());
+                    Path path = this.navigator.getPath();
+                    Vec3d vec3d = new Vec3d(this.moveHelper.getX(), this.moveHelper.getY(), this.moveHelper.getZ());
 
-                    if (path != null && path.func_75873_e() < path.func_75874_d())
+                    if (path != null && path.getCurrentPathIndex() < path.getCurrentPathLength())
                     {
-                        vec3d = path.func_75878_a(this);
+                        vec3d = path.getPosition(this);
                     }
 
-                    this.func_175533_a(vec3d.field_72450_a, vec3d.field_72449_c);
-                    this.func_184770_cZ();
+                    this.calculateRotationYaw(vec3d.xCoord, vec3d.zCoord);
+                    this.startJumping();
                 }
             }
-            else if (!entityrabbit$rabbitjumphelper.func_180065_d())
+            else if (!entityrabbit$rabbitjumphelper.canJump())
             {
-                this.func_175518_cr();
+                this.enableJumpControl();
             }
         }
 
-        this.field_175537_bp = this.field_70122_E;
+        this.wasOnGround = this.onGround;
     }
 
-    public void func_174830_Y()
+    public void spawnRunningParticles()
     {
     }
 
-    private void func_175533_a(double p_175533_1_, double p_175533_3_)
+    private void calculateRotationYaw(double x, double z)
     {
-        this.field_70177_z = (float)(MathHelper.func_181159_b(p_175533_3_ - this.field_70161_v, p_175533_1_ - this.field_70165_t) * (180D / Math.PI)) - 90.0F;
+        this.rotationYaw = (float)(MathHelper.atan2(z - this.posZ, x - this.posX) * (180D / Math.PI)) - 90.0F;
     }
 
-    private void func_175518_cr()
+    private void enableJumpControl()
     {
-        ((EntityRabbit.RabbitJumpHelper)this.field_70767_i).func_180066_a(true);
+        ((EntityRabbit.RabbitJumpHelper)this.jumpHelper).setCanJump(true);
     }
 
-    private void func_175520_cs()
+    private void disableJumpControl()
     {
-        ((EntityRabbit.RabbitJumpHelper)this.field_70767_i).func_180066_a(false);
+        ((EntityRabbit.RabbitJumpHelper)this.jumpHelper).setCanJump(false);
     }
 
-    private void func_175530_ct()
+    private void updateMoveTypeDuration()
     {
-        if (this.field_70765_h.func_75638_b() < 2.2D)
+        if (this.moveHelper.getSpeed() < 2.2D)
         {
-            this.field_175538_bq = 10;
+            this.currentMoveTypeDuration = 10;
         }
         else
         {
-            this.field_175538_bq = 1;
+            this.currentMoveTypeDuration = 1;
         }
     }
 
-    private void func_175517_cu()
+    private void checkLandingDelay()
     {
-        this.func_175530_ct();
-        this.func_175520_cs();
+        this.updateMoveTypeDuration();
+        this.disableJumpControl();
     }
 
-    public void func_70636_d()
+    public void onLivingUpdate()
     {
-        super.func_70636_d();
+        super.onLivingUpdate();
 
-        if (this.field_175540_bm != this.field_175535_bn)
+        if (this.jumpTicks != this.jumpDuration)
         {
-            ++this.field_175540_bm;
+            ++this.jumpTicks;
         }
-        else if (this.field_175535_bn != 0)
+        else if (this.jumpDuration != 0)
         {
-            this.field_175540_bm = 0;
-            this.field_175535_bn = 0;
-            this.func_70637_d(false);
+            this.jumpTicks = 0;
+            this.jumpDuration = 0;
+            this.setJumping(false);
         }
     }
 
-    protected void func_110147_ax()
+    protected void applyEntityAttributes()
     {
-        super.func_110147_ax();
-        this.func_110148_a(SharedMonsterAttributes.field_111267_a).func_111128_a(3.0D);
-        this.func_110148_a(SharedMonsterAttributes.field_111263_d).func_111128_a(0.30000001192092896D);
+        super.applyEntityAttributes();
+        this.getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).setBaseValue(3.0D);
+        this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).setBaseValue(0.30000001192092896D);
     }
 
-    public static void func_189801_b(DataFixer p_189801_0_)
+    public static void registerFixesRabbit(DataFixer fixer)
     {
-        EntityLiving.func_189752_a(p_189801_0_, EntityRabbit.class);
+        EntityLiving.registerFixesMob(fixer, EntityRabbit.class);
     }
 
-    public void func_70014_b(NBTTagCompound p_70014_1_)
+    public void writeEntityToNBT(NBTTagCompound compound)
     {
-        super.func_70014_b(p_70014_1_);
-        p_70014_1_.func_74768_a("RabbitType", this.func_175531_cl());
-        p_70014_1_.func_74768_a("MoreCarrotTicks", this.field_175541_bs);
+        super.writeEntityToNBT(compound);
+        compound.setInteger("RabbitType", this.getRabbitType());
+        compound.setInteger("MoreCarrotTicks", this.carrotTicks);
     }
 
-    public void func_70037_a(NBTTagCompound p_70037_1_)
+    public void readEntityFromNBT(NBTTagCompound compound)
     {
-        super.func_70037_a(p_70037_1_);
-        this.func_175529_r(p_70037_1_.func_74762_e("RabbitType"));
-        this.field_175541_bs = p_70037_1_.func_74762_e("MoreCarrotTicks");
+        super.readEntityFromNBT(compound);
+        this.setRabbitType(compound.getInteger("RabbitType"));
+        this.carrotTicks = compound.getInteger("MoreCarrotTicks");
     }
 
-    protected SoundEvent func_184771_da()
+    protected SoundEvent getJumpSound()
     {
-        return SoundEvents.field_187824_en;
+        return SoundEvents.ENTITY_RABBIT_JUMP;
     }
 
-    protected SoundEvent func_184639_G()
+    protected SoundEvent getAmbientSound()
     {
-        return SoundEvents.field_187816_ej;
+        return SoundEvents.ENTITY_RABBIT_AMBIENT;
     }
 
-    protected SoundEvent func_184601_bQ()
+    protected SoundEvent getHurtSound()
     {
-        return SoundEvents.field_187822_em;
+        return SoundEvents.ENTITY_RABBIT_HURT;
     }
 
-    protected SoundEvent func_184615_bR()
+    protected SoundEvent getDeathSound()
     {
-        return SoundEvents.field_187820_el;
+        return SoundEvents.ENTITY_RABBIT_DEATH;
     }
 
-    public boolean func_70652_k(Entity p_70652_1_)
+    public boolean attackEntityAsMob(Entity entityIn)
     {
-        if (this.func_175531_cl() == 99)
+        if (this.getRabbitType() == 99)
         {
-            this.func_184185_a(SoundEvents.field_187818_ek, 1.0F, (this.field_70146_Z.nextFloat() - this.field_70146_Z.nextFloat()) * 0.2F + 1.0F);
-            return p_70652_1_.func_70097_a(DamageSource.func_76358_a(this), 8.0F);
+            this.playSound(SoundEvents.ENTITY_RABBIT_ATTACK, 1.0F, (this.rand.nextFloat() - this.rand.nextFloat()) * 0.2F + 1.0F);
+            return entityIn.attackEntityFrom(DamageSource.causeMobDamage(this), 8.0F);
         }
         else
         {
-            return p_70652_1_.func_70097_a(DamageSource.func_76358_a(this), 3.0F);
+            return entityIn.attackEntityFrom(DamageSource.causeMobDamage(this), 3.0F);
         }
     }
 
-    public SoundCategory func_184176_by()
+    public SoundCategory getSoundCategory()
     {
-        return this.func_175531_cl() == 99 ? SoundCategory.HOSTILE : SoundCategory.NEUTRAL;
+        return this.getRabbitType() == 99 ? SoundCategory.HOSTILE : SoundCategory.NEUTRAL;
     }
 
-    public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_)
+    public boolean attackEntityFrom(DamageSource source, float amount)
     {
-        return this.func_180431_b(p_70097_1_) ? false : super.func_70097_a(p_70097_1_, p_70097_2_);
+        return this.isEntityInvulnerable(source) ? false : super.attackEntityFrom(source, amount);
     }
 
     @Nullable
-    protected ResourceLocation func_184647_J()
+    protected ResourceLocation getLootTable()
     {
-        return LootTableList.field_186393_A;
+        return LootTableList.ENTITIES_RABBIT;
     }
 
-    private boolean func_175525_a(Item p_175525_1_)
+    private boolean isRabbitBreedingItem(Item itemIn)
     {
-        return p_175525_1_ == Items.field_151172_bF || p_175525_1_ == Items.field_151150_bK || p_175525_1_ == Item.func_150898_a(Blocks.field_150327_N);
+        return itemIn == Items.CARROT || itemIn == Items.GOLDEN_CARROT || itemIn == Item.getItemFromBlock(Blocks.YELLOW_FLOWER);
     }
 
-    public EntityRabbit func_90011_a(EntityAgeable p_90011_1_)
+    public EntityRabbit createChild(EntityAgeable ageable)
     {
-        EntityRabbit entityrabbit = new EntityRabbit(this.field_70170_p);
-        int i = this.func_184772_dk();
+        EntityRabbit entityrabbit = new EntityRabbit(this.world);
+        int i = this.getRandomRabbitType();
 
-        if (this.field_70146_Z.nextInt(20) != 0)
+        if (this.rand.nextInt(20) != 0)
         {
-            if (p_90011_1_ instanceof EntityRabbit && this.field_70146_Z.nextBoolean())
+            if (ageable instanceof EntityRabbit && this.rand.nextBoolean())
             {
-                i = ((EntityRabbit)p_90011_1_).func_175531_cl();
+                i = ((EntityRabbit)ageable).getRabbitType();
             }
             else
             {
-                i = this.func_175531_cl();
+                i = this.getRabbitType();
             }
         }
 
-        entityrabbit.func_175529_r(i);
+        entityrabbit.setRabbitType(i);
         return entityrabbit;
     }
 
-    public boolean func_70877_b(ItemStack p_70877_1_)
+    public boolean isBreedingItem(ItemStack stack)
     {
-        return this.func_175525_a(p_70877_1_.func_77973_b());
+        return this.isRabbitBreedingItem(stack.getItem());
     }
 
-    public int func_175531_cl()
+    public int getRabbitType()
     {
-        return ((Integer)this.field_70180_af.func_187225_a(field_184773_bv)).intValue();
+        return ((Integer)this.dataManager.get(RABBIT_TYPE)).intValue();
     }
 
-    public void func_175529_r(int p_175529_1_)
+    public void setRabbitType(int rabbitTypeId)
     {
-        if (p_175529_1_ == 99)
+        if (rabbitTypeId == 99)
         {
-            this.func_110148_a(SharedMonsterAttributes.field_188791_g).func_111128_a(8.0D);
-            this.field_70714_bg.func_75776_a(4, new EntityRabbit.AIEvilAttack(this));
-            this.field_70715_bh.func_75776_a(1, new EntityAIHurtByTarget(this, false, new Class[0]));
-            this.field_70715_bh.func_75776_a(2, new EntityAINearestAttackableTarget(this, EntityPlayer.class, true));
-            this.field_70715_bh.func_75776_a(2, new EntityAINearestAttackableTarget(this, EntityWolf.class, true));
+            this.getEntityAttribute(SharedMonsterAttributes.ARMOR).setBaseValue(8.0D);
+            this.tasks.addTask(4, new EntityRabbit.AIEvilAttack(this));
+            this.targetTasks.addTask(1, new EntityAIHurtByTarget(this, false, new Class[0]));
+            this.targetTasks.addTask(2, new EntityAINearestAttackableTarget(this, EntityPlayer.class, true));
+            this.targetTasks.addTask(2, new EntityAINearestAttackableTarget(this, EntityWolf.class, true));
 
-            if (!this.func_145818_k_())
+            if (!this.hasCustomName())
             {
-                this.func_96094_a(I18n.func_74838_a("entity.KillerBunny.name"));
+                this.setCustomNameTag(I18n.translateToLocal("entity.KillerBunny.name"));
             }
         }
 
-        this.field_70180_af.func_187227_b(field_184773_bv, Integer.valueOf(p_175529_1_));
+        this.dataManager.set(RABBIT_TYPE, Integer.valueOf(rabbitTypeId));
     }
 
     @Nullable
-    public IEntityLivingData func_180482_a(DifficultyInstance p_180482_1_, @Nullable IEntityLivingData p_180482_2_)
+    public IEntityLivingData onInitialSpawn(DifficultyInstance difficulty, @Nullable IEntityLivingData livingdata)
     {
-        p_180482_2_ = super.func_180482_a(p_180482_1_, p_180482_2_);
-        int i = this.func_184772_dk();
+        livingdata = super.onInitialSpawn(difficulty, livingdata);
+        int i = this.getRandomRabbitType();
         boolean flag = false;
 
-        if (p_180482_2_ instanceof EntityRabbit.RabbitTypeData)
+        if (livingdata instanceof EntityRabbit.RabbitTypeData)
         {
-            i = ((EntityRabbit.RabbitTypeData)p_180482_2_).field_179427_a;
+            i = ((EntityRabbit.RabbitTypeData)livingdata).typeData;
             flag = true;
         }
         else
         {
-            p_180482_2_ = new EntityRabbit.RabbitTypeData(i);
+            livingdata = new EntityRabbit.RabbitTypeData(i);
         }
 
-        this.func_175529_r(i);
+        this.setRabbitType(i);
 
         if (flag)
         {
-            this.func_70873_a(-24000);
+            this.setGrowingAge(-24000);
         }
 
-        return p_180482_2_;
+        return livingdata;
     }
 
-    private int func_184772_dk()
+    private int getRandomRabbitType()
     {
-        Biome biome = this.field_70170_p.func_180494_b(new BlockPos(this));
-        int i = this.field_70146_Z.nextInt(100);
-        return biome.func_150559_j() ? (i < 80 ? 1 : 3) : (biome instanceof BiomeDesert ? 4 : (i < 50 ? 0 : (i < 90 ? 5 : 2)));
+        Biome biome = this.world.getBiome(new BlockPos(this));
+        int i = this.rand.nextInt(100);
+        return biome.isSnowyBiome() ? (i < 80 ? 1 : 3) : (biome instanceof BiomeDesert ? 4 : (i < 50 ? 0 : (i < 90 ? 5 : 2)));
     }
 
-    private boolean func_175534_cv()
+    private boolean isCarrotEaten()
     {
-        return this.field_175541_bs == 0;
+        return this.carrotTicks == 0;
     }
 
-    protected void func_175528_cn()
+    protected void createEatingParticles()
     {
-        BlockCarrot blockcarrot = (BlockCarrot)Blocks.field_150459_bM;
-        IBlockState iblockstate = blockcarrot.func_185528_e(blockcarrot.func_185526_g());
-        this.field_70170_p.func_175688_a(EnumParticleTypes.BLOCK_DUST, this.field_70165_t + (double)(this.field_70146_Z.nextFloat() * this.field_70130_N * 2.0F) - (double)this.field_70130_N, this.field_70163_u + 0.5D + (double)(this.field_70146_Z.nextFloat() * this.field_70131_O), this.field_70161_v + (double)(this.field_70146_Z.nextFloat() * this.field_70130_N * 2.0F) - (double)this.field_70130_N, 0.0D, 0.0D, 0.0D, new int[] {Block.func_176210_f(iblockstate)});
-        this.field_175541_bs = 40;
+        BlockCarrot blockcarrot = (BlockCarrot)Blocks.CARROTS;
+        IBlockState iblockstate = blockcarrot.withAge(blockcarrot.getMaxAge());
+        this.world.spawnParticle(EnumParticleTypes.BLOCK_DUST, this.posX + (double)(this.rand.nextFloat() * this.width * 2.0F) - (double)this.width, this.posY + 0.5D + (double)(this.rand.nextFloat() * this.height), this.posZ + (double)(this.rand.nextFloat() * this.width * 2.0F) - (double)this.width, 0.0D, 0.0D, 0.0D, new int[] {Block.getStateId(iblockstate)});
+        this.carrotTicks = 40;
     }
 
     @SideOnly(Side.CLIENT)
-    public void func_70103_a(byte p_70103_1_)
+    public void handleStatusUpdate(byte id)
     {
-        if (p_70103_1_ == 1)
+        if (id == 1)
         {
-            this.func_174808_Z();
-            this.field_175535_bn = 10;
-            this.field_175540_bm = 0;
+            this.createRunningParticles();
+            this.jumpDuration = 10;
+            this.jumpTicks = 0;
         }
         else
         {
-            super.func_70103_a(p_70103_1_);
+            super.handleStatusUpdate(id);
         }
     }
 
     static class AIAvoidEntity<T extends Entity> extends EntityAIAvoidEntity<T>
         {
-            private final EntityRabbit field_179511_d;
+            private final EntityRabbit entityInstance;
 
-            public AIAvoidEntity(EntityRabbit p_i46403_1_, Class<T> p_i46403_2_, float p_i46403_3_, double p_i46403_4_, double p_i46403_6_)
+            public AIAvoidEntity(EntityRabbit rabbit, Class<T> p_i46403_2_, float p_i46403_3_, double p_i46403_4_, double p_i46403_6_)
             {
-                super(p_i46403_1_, p_i46403_2_, p_i46403_3_, p_i46403_4_, p_i46403_6_);
-                this.field_179511_d = p_i46403_1_;
+                super(rabbit, p_i46403_2_, p_i46403_3_, p_i46403_4_, p_i46403_6_);
+                this.entityInstance = rabbit;
             }
 
-            public boolean func_75250_a()
+            public boolean shouldExecute()
             {
-                return this.field_179511_d.func_175531_cl() != 99 && super.func_75250_a();
+                return this.entityInstance.getRabbitType() != 99 && super.shouldExecute();
             }
         }
 
     static class AIEvilAttack extends EntityAIAttackMelee
         {
-            public AIEvilAttack(EntityRabbit p_i45867_1_)
+            public AIEvilAttack(EntityRabbit rabbit)
             {
-                super(p_i45867_1_, 1.4D, true);
+                super(rabbit, 1.4D, true);
             }
 
-            protected double func_179512_a(EntityLivingBase p_179512_1_)
+            protected double getAttackReachSqr(EntityLivingBase attackTarget)
             {
-                return (double)(4.0F + p_179512_1_.field_70130_N);
+                return (double)(4.0F + attackTarget.width);
             }
         }
 
     static class AIPanic extends EntityAIPanic
         {
-            private final EntityRabbit field_179486_b;
+            private final EntityRabbit theEntity;
 
-            public AIPanic(EntityRabbit p_i45861_1_, double p_i45861_2_)
+            public AIPanic(EntityRabbit rabbit, double speedIn)
             {
-                super(p_i45861_1_, p_i45861_2_);
-                this.field_179486_b = p_i45861_1_;
+                super(rabbit, speedIn);
+                this.theEntity = rabbit;
             }
 
-            public void func_75246_d()
+            public void updateTask()
             {
-                super.func_75246_d();
-                this.field_179486_b.func_175515_b(this.field_75265_b);
+                super.updateTask();
+                this.theEntity.setMovementSpeed(this.speed);
             }
         }
 
     static class AIRaidFarm extends EntityAIMoveToBlock
         {
-            private final EntityRabbit field_179500_c;
-            private boolean field_179498_d;
-            private boolean field_179499_e;
+            private final EntityRabbit rabbit;
+            private boolean wantsToRaid;
+            private boolean canRaid;
 
-            public AIRaidFarm(EntityRabbit p_i45860_1_)
+            public AIRaidFarm(EntityRabbit rabbitIn)
             {
-                super(p_i45860_1_, 0.699999988079071D, 16);
-                this.field_179500_c = p_i45860_1_;
+                super(rabbitIn, 0.699999988079071D, 16);
+                this.rabbit = rabbitIn;
             }
 
-            public boolean func_75250_a()
+            public boolean shouldExecute()
             {
-                if (this.field_179496_a <= 0)
+                if (this.runDelay <= 0)
                 {
-                    if (!this.field_179500_c.field_70170_p.func_82736_K().func_82766_b("mobGriefing"))
+                    if (!this.rabbit.world.getGameRules().getBoolean("mobGriefing"))
                     {
                         return false;
                     }
 
-                    this.field_179499_e = false;
-                    this.field_179498_d = this.field_179500_c.func_175534_cv();
-                    this.field_179498_d = true;
+                    this.canRaid = false;
+                    this.wantsToRaid = this.rabbit.isCarrotEaten();
+                    this.wantsToRaid = true;
                 }
 
-                return super.func_75250_a();
+                return super.shouldExecute();
             }
 
-            public boolean func_75253_b()
+            public boolean continueExecuting()
             {
-                return this.field_179499_e && super.func_75253_b();
+                return this.canRaid && super.continueExecuting();
             }
 
-            public void func_75246_d()
+            public void updateTask()
             {
-                super.func_75246_d();
-                this.field_179500_c.func_70671_ap().func_75650_a((double)this.field_179494_b.func_177958_n() + 0.5D, (double)(this.field_179494_b.func_177956_o() + 1), (double)this.field_179494_b.func_177952_p() + 0.5D, 10.0F, (float)this.field_179500_c.func_70646_bf());
+                super.updateTask();
+                this.rabbit.getLookHelper().setLookPosition((double)this.destinationBlock.getX() + 0.5D, (double)(this.destinationBlock.getY() + 1), (double)this.destinationBlock.getZ() + 0.5D, 10.0F, (float)this.rabbit.getVerticalFaceSpeed());
 
-                if (this.func_179487_f())
+                if (this.getIsAboveDestination())
                 {
-                    World world = this.field_179500_c.field_70170_p;
-                    BlockPos blockpos = this.field_179494_b.func_177984_a();
-                    IBlockState iblockstate = world.func_180495_p(blockpos);
-                    Block block = iblockstate.func_177230_c();
+                    World world = this.rabbit.world;
+                    BlockPos blockpos = this.destinationBlock.up();
+                    IBlockState iblockstate = world.getBlockState(blockpos);
+                    Block block = iblockstate.getBlock();
 
-                    if (this.field_179499_e && block instanceof BlockCarrot)
+                    if (this.canRaid && block instanceof BlockCarrot)
                     {
-                        Integer integer = (Integer)iblockstate.func_177229_b(BlockCarrot.field_176488_a);
+                        Integer integer = (Integer)iblockstate.getValue(BlockCarrot.AGE);
 
                         if (integer.intValue() == 0)
                         {
-                            world.func_180501_a(blockpos, Blocks.field_150350_a.func_176223_P(), 2);
-                            world.func_175655_b(blockpos, true);
+                            // CraftBukkit start
+                            if (CraftEventFactory.callEntityChangeBlockEvent(this.rabbit, blockpos, Blocks.AIR, 0).isCancelled()) {
+                                return;
+                            }
+                            // CraftBukkit end
+                            world.setBlockState(blockpos, Blocks.AIR.getDefaultState(), 2);
+                            world.destroyBlock(blockpos, true);
                         }
                         else
                         {
-                            world.func_180501_a(blockpos, iblockstate.func_177226_a(BlockCarrot.field_176488_a, Integer.valueOf(integer.intValue() - 1)), 2);
-                            world.func_175718_b(2001, blockpos, Block.func_176210_f(iblockstate));
+                            // CraftBukkit start
+                            if (CraftEventFactory.callEntityChangeBlockEvent(
+                                    this.rabbit,
+                                    blockpos,
+                                    block, block.getMetaFromState(iblockstate.withProperty(BlockCarrot.AGE, Integer.valueOf(integer.intValue() - 1)))
+                            ).isCancelled()) {
+                                return;
+                            }
+                            // CraftBukkit end
+                            world.setBlockState(blockpos, iblockstate.withProperty(BlockCarrot.AGE, Integer.valueOf(integer.intValue() - 1)), 2);
+                            world.playEvent(2001, blockpos, Block.getStateId(iblockstate));
                         }
 
-                        this.field_179500_c.func_175528_cn();
+                        this.rabbit.createEatingParticles();
                     }
 
-                    this.field_179499_e = false;
-                    this.field_179496_a = 10;
+                    this.canRaid = false;
+                    this.runDelay = 10;
                 }
             }
 
-            protected boolean func_179488_a(World p_179488_1_, BlockPos p_179488_2_)
+            protected boolean shouldMoveTo(World worldIn, BlockPos pos)
             {
-                Block block = p_179488_1_.func_180495_p(p_179488_2_).func_177230_c();
+                Block block = worldIn.getBlockState(pos).getBlock();
 
-                if (block == Blocks.field_150458_ak && this.field_179498_d && !this.field_179499_e)
+                if (block == Blocks.FARMLAND && this.wantsToRaid && !this.canRaid)
                 {
-                    p_179488_2_ = p_179488_2_.func_177984_a();
-                    IBlockState iblockstate = p_179488_1_.func_180495_p(p_179488_2_);
-                    block = iblockstate.func_177230_c();
+                    pos = pos.up();
+                    IBlockState iblockstate = worldIn.getBlockState(pos);
+                    block = iblockstate.getBlock();
 
-                    if (block instanceof BlockCarrot && ((BlockCarrot)block).func_185525_y(iblockstate))
+                    if (block instanceof BlockCarrot && ((BlockCarrot)block).isMaxAge(iblockstate))
                     {
-                        this.field_179499_e = true;
+                        this.canRaid = true;
                         return true;
                     }
                 }
@@ -615,88 +634,88 @@
 
     public class RabbitJumpHelper extends EntityJumpHelper
     {
-        private final EntityRabbit field_180070_c;
-        private boolean field_180068_d;
+        private final EntityRabbit theEntity;
+        private boolean canJump;
 
-        public RabbitJumpHelper(EntityRabbit p_i45863_2_)
+        public RabbitJumpHelper(EntityRabbit rabbit)
         {
-            super(p_i45863_2_);
-            this.field_180070_c = p_i45863_2_;
+            super(rabbit);
+            this.theEntity = rabbit;
         }
 
-        public boolean func_180067_c()
+        public boolean getIsJumping()
         {
-            return this.field_75662_b;
+            return this.isJumping;
         }
 
-        public boolean func_180065_d()
+        public boolean canJump()
         {
-            return this.field_180068_d;
+            return this.canJump;
         }
 
-        public void func_180066_a(boolean p_180066_1_)
+        public void setCanJump(boolean canJumpIn)
         {
-            this.field_180068_d = p_180066_1_;
+            this.canJump = canJumpIn;
         }
 
-        public void func_75661_b()
+        public void doJump()
         {
-            if (this.field_75662_b)
+            if (this.isJumping)
             {
-                this.field_180070_c.func_184770_cZ();
-                this.field_75662_b = false;
+                this.theEntity.startJumping();
+                this.isJumping = false;
             }
         }
     }
 
     static class RabbitMoveHelper extends EntityMoveHelper
         {
-            private final EntityRabbit field_179929_g;
-            private double field_188492_j;
+            private final EntityRabbit theEntity;
+            private double nextJumpSpeed;
 
-            public RabbitMoveHelper(EntityRabbit p_i45862_1_)
+            public RabbitMoveHelper(EntityRabbit rabbit)
             {
-                super(p_i45862_1_);
-                this.field_179929_g = p_i45862_1_;
+                super(rabbit);
+                this.theEntity = rabbit;
             }
 
-            public void func_75641_c()
+            public void onUpdateMoveHelper()
             {
-                if (this.field_179929_g.field_70122_E && !this.field_179929_g.field_70703_bu && !((EntityRabbit.RabbitJumpHelper)this.field_179929_g.field_70767_i).func_180067_c())
+                if (this.theEntity.onGround && !this.theEntity.isJumping && !((EntityRabbit.RabbitJumpHelper)this.theEntity.jumpHelper).getIsJumping())
                 {
-                    this.field_179929_g.func_175515_b(0.0D);
+                    this.theEntity.setMovementSpeed(0.0D);
                 }
-                else if (this.func_75640_a())
+                else if (this.isUpdating())
                 {
-                    this.field_179929_g.func_175515_b(this.field_188492_j);
+                    this.theEntity.setMovementSpeed(this.nextJumpSpeed);
                 }
 
-                super.func_75641_c();
+                super.onUpdateMoveHelper();
             }
 
-            public void func_75642_a(double p_75642_1_, double p_75642_3_, double p_75642_5_, double p_75642_7_)
+            public void setMoveTo(double x, double y, double z, double speedIn)
             {
-                if (this.field_179929_g.func_70090_H())
+                if (this.theEntity.isInWater())
                 {
-                    p_75642_7_ = 1.5D;
+                    speedIn = 1.5D;
                 }
 
-                super.func_75642_a(p_75642_1_, p_75642_3_, p_75642_5_, p_75642_7_);
+                super.setMoveTo(x, y, z, speedIn);
 
-                if (p_75642_7_ > 0.0D)
+                if (speedIn > 0.0D)
                 {
-                    this.field_188492_j = p_75642_7_;
+                    this.nextJumpSpeed = speedIn;
                 }
             }
         }
 
     public static class RabbitTypeData implements IEntityLivingData
         {
-            public int field_179427_a;
+            public int typeData;
 
-            public RabbitTypeData(int p_i45864_1_)
+            public RabbitTypeData(int type)
             {
-                this.field_179427_a = p_i45864_1_;
+                this.typeData = type;
             }
         }
 }
