--- ../src-base/minecraft/net/minecraft/entity/passive/EntityWolf.java
+++ ../src-work/minecraft/net/minecraft/entity/passive/EntityWolf.java
@@ -28,6 +28,7 @@
 import net.minecraft.entity.monster.AbstractSkeleton;
 import net.minecraft.entity.monster.EntityCreeper;
 import net.minecraft.entity.monster.EntityGhast;
+import net.minecraft.entity.monster.EntityPolarBear;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.projectile.EntityArrow;
 import net.minecraft.init.Items;
@@ -51,6 +52,7 @@
 import net.minecraft.world.storage.loot.LootTableList;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.event.entity.EntityTargetEvent;
 
 public class EntityWolf extends EntityTameable
 {
@@ -92,12 +94,27 @@
         {
             public boolean apply(@Nullable Entity p_apply_1_)
             {
-                return p_apply_1_ instanceof EntitySheep || p_apply_1_ instanceof EntityRabbit;
+                return p_apply_1_ instanceof EntityAnimal && !(p_apply_1_ instanceof EntityWaterMob && !(p_apply_1_ instanceof EntityWolf) && !(p_apply_1_ instanceof EntityPolarBear));
             }
         }));
         this.field_70715_bh.func_75776_a(5, new EntityAINearestAttackableTarget(this, AbstractSkeleton.class, false));
     }
 
+    // CraftBukkit - add overriden version
+    @Override
+    public boolean setGoalTarget(EntityLivingBase entityliving, EntityTargetEvent.TargetReason reason, boolean fire) {
+        if (!super.setGoalTarget(entityliving, reason, fire)) {
+            return false;
+        }
+        entityliving = func_70638_az();
+        if (entityliving == null) {
+            this.func_70916_h(false);
+        } else if (!this.func_70909_n()) {
+            this.func_70916_h(true);
+        }
+        return true;
+    }
+    // CraftBukkit end
     protected void func_110147_ax()
     {
         super.func_110147_ax();
@@ -411,7 +428,7 @@
                 this.field_70911_d.func_75270_a(!this.func_70906_o());
                 this.field_70703_bu = false;
                 this.field_70699_by.func_75499_g();
-                this.func_70624_b((EntityLivingBase)null);
+                this.setGoalTarget((EntityLivingBase) null, EntityTargetEvent.TargetReason.FORGOT_TARGET, true); // CraftBukkit - reason
             }
         }
         else if (itemstack.func_77973_b() == Items.field_151103_aS && !this.func_70919_bu())
@@ -423,7 +440,7 @@
 
             if (!this.field_70170_p.field_72995_K)
             {
-                if (this.field_70146_Z.nextInt(3) == 0)
+                if (this.field_70146_Z.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_184645_1_))
                 {
                     this.func_70903_f(true);
                     this.field_70699_by.func_75499_g();
