--- ../src-base/minecraft/net/minecraft/inventory/ContainerMerchant.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerMerchant.java
@@ -5,13 +5,25 @@
 import net.minecraft.entity.player.InventoryPlayer;
 import net.minecraft.item.ItemStack;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
 
 public class ContainerMerchant extends Container
 {
     private final IMerchant field_75178_e;
     private final InventoryMerchant field_75176_f;
     private final World field_75177_g;
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private InventoryPlayer player;
 
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity == null) {
+            bukkitEntity = new CraftInventoryView(this.player.field_70458_d.getBukkitEntity(), new org.bukkit.craftbukkit.inventory.CraftInventoryMerchant((net.minecraft.inventory.InventoryMerchant) field_75176_f), this);
+        }
+        return bukkitEntity;
+    }
+    // CraftBukkit end
     public ContainerMerchant(InventoryPlayer p_i1821_1_, IMerchant p_i1821_2_, World p_i1821_3_)
     {
         this.field_75178_e = p_i1821_2_;
