--- ../src-base/minecraft/net/minecraft/inventory/ContainerPlayer.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerPlayer.java
@@ -4,12 +4,16 @@
 import net.minecraft.enchantment.EnchantmentHelper;
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.entity.player.InventoryPlayer;
 import net.minecraft.item.ItemArmor;
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.crafting.CraftingManager;
+import net.minecraft.network.play.server.SPacketSetSlot;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.inventory.CraftInventoryCrafting;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
 
 public class ContainerPlayer extends Container
 {
@@ -18,13 +22,33 @@
     public IInventory field_75179_f = new InventoryCraftResult();
     public boolean field_75180_g;
     private final EntityPlayer field_82862_h;
+    // CraftBukkit start
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
 
+        CraftInventoryCrafting inventory = new CraftInventoryCrafting(this.field_75181_e, this.field_75179_f);
+        bukkitEntity = new CraftInventoryView(this.playerinv.field_70458_d.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private InventoryPlayer playerinv;
+    // CraftBukkit end
     public ContainerPlayer(final InventoryPlayer p_i1819_1_, boolean p_i1819_2_, EntityPlayer p_i1819_3_)
     {
         this.field_75180_g = p_i1819_2_;
         this.field_82862_h = p_i1819_3_;
+        //CraftBukkit start
+        this.field_75179_f = new net.minecraft.inventory.InventoryCraftResult(); // CraftBukkit - moved to before InventoryCrafting construction
+        this.field_75181_e = new net.minecraft.inventory.InventoryCrafting(this, 2, 2, p_i1819_1_.field_70458_d); // CraftBukkit - pass player
+        this.field_75181_e.resultInventory = this.field_75179_f; // CraftBukkit - let InventoryCrafting know about its result slot
+        this.playerinv = p_i1819_1_; // CraftBukkit - save player
+        // CraftBukkit end
         this.func_75146_a(new SlotCrafting(p_i1819_1_.field_70458_d, this.field_75181_e, this.field_75179_f, 0, 154, 28));
-
         for (int i = 0; i < 2; ++i)
         {
             for (int j = 0; j < 2; ++j)
@@ -44,7 +68,7 @@
                 }
                 public boolean func_75214_a(ItemStack p_75214_1_)
                 {
-                    return entityequipmentslot == EntityLiving.func_184640_d(p_75214_1_);
+                    return p_75214_1_.func_77973_b().isValidArmor(p_75214_1_, entityequipmentslot, field_82862_h);
                 }
                 public boolean func_82869_a(EntityPlayer p_82869_1_)
                 {
@@ -87,7 +111,17 @@
 
     public void func_75130_a(IInventory p_75130_1_)
     {
+        // this.resultInventory.setItem(0, CraftingManager.getInstance().craft(this.craftInventory, this.owner.world));
+        // CraftBukkit start (Note: the following line would cause an error if called during construction)
+        CraftingManager.func_77594_a().lastCraftView = getBukkitView();
         this.field_75179_f.func_70299_a(0, CraftingManager.func_77594_a().func_82787_a(this.field_75181_e, this.field_82862_h.field_70170_p));
+        //if (super.listeners.size() < 1) {
+        //    return;
+        //}
+
+        //EntityPlayerMP player = (EntityPlayerMP) super.listeners.get(0); // TODO: Is this _always_ correct? Seems like it.
+        //player.connection.sendPacket(new SPacketSetSlot(player.openContainer.windowId, 0, craftResult));
+        // CraftBukkit end
     }
 
     public void func_75134_a(EntityPlayer p_75134_1_)
