--- ../src-base/minecraft/net/minecraft/inventory/ContainerBrewingStand.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerBrewingStand.java
@@ -1,6 +1,7 @@
 package net.minecraft.inventory;
 
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.entity.player.InventoryPlayer;
 import net.minecraft.init.Items;
 import net.minecraft.init.PotionTypes;
@@ -12,6 +13,10 @@
 import net.minecraft.stats.AchievementList;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.craftbukkit.inventory.CraftInventoryBrewer;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+import org.bukkit.entity.HumanEntity;
 
 public class ContainerBrewingStand extends Container
 {
@@ -20,6 +25,22 @@
     private int field_184998_g;
     private int field_184999_h;
 
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private InventoryPlayer player;
+    // CraftBukkit end
+    // CraftBukkit start
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventoryBrewer inventory = new CraftInventoryBrewer(this.field_75188_e);
+        bukkitEntity = new CraftInventoryView((HumanEntity) new CraftPlayer(player.field_70458_d.func_184102_h().server, (EntityPlayerMP) player.field_70458_d), inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
     public ContainerBrewingStand(InventoryPlayer p_i45802_1_, IInventory p_i45802_2_)
     {
         this.field_75188_e = p_i45802_2_;
@@ -28,7 +49,7 @@
         this.func_75146_a(new ContainerBrewingStand.Potion(p_i45802_1_.field_70458_d, p_i45802_2_, 2, 102, 51));
         this.field_75186_f = this.func_75146_a(new ContainerBrewingStand.Ingredient(p_i45802_2_, 3, 79, 17));
         this.func_75146_a(new ContainerBrewingStand.Fuel(p_i45802_2_, 4, 17, 17));
-
+        this.player = p_i45802_1_;
         for (int i = 0; i < 3; ++i)
         {
             for (int j = 0; j < 9; ++j)
@@ -80,6 +101,8 @@
 
     public boolean func_75145_c(EntityPlayer p_75145_1_)
     {
+
+        if (!this.checkReachable) return true; // CraftBukkit
         return this.field_75188_e.func_70300_a(p_75145_1_);
     }
 
@@ -197,7 +220,7 @@
 
             public boolean func_75214_a(ItemStack p_75214_1_)
             {
-                return PotionHelper.func_185205_a(p_75214_1_);
+                return net.minecraftforge.common.brewing.BrewingRecipeRegistry.isValidIngredient(p_75214_1_);
             }
 
             public int func_75219_a()
@@ -232,6 +255,7 @@
 
                 if (potiontype != PotionTypes.field_185230_b && potiontype != PotionTypes.field_185229_a)
                 {
+                    net.minecraftforge.event.ForgeEventFactory.onPlayerBrewedPotion(p_190901_1_, p_190901_2_);
                     this.field_75244_a.func_71029_a(AchievementList.field_187970_B);
                 }
 
@@ -241,8 +265,7 @@
 
             public static boolean func_75243_a_(ItemStack p_75243_0_)
             {
-                Item item = p_75243_0_.func_77973_b();
-                return item == Items.field_151068_bn || item == Items.field_185155_bH || item == Items.field_185156_bI || item == Items.field_151069_bo;
+                return net.minecraftforge.common.brewing.BrewingRecipeRegistry.isValidInput(p_75243_0_);
             }
         }
 }
