--- ../src-base/minecraft/net/minecraft/inventory/ContainerPlayer.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerPlayer.java
@@ -4,58 +4,82 @@
 import net.minecraft.enchantment.EnchantmentHelper;
 import net.minecraft.entity.EntityLiving;
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.entity.player.InventoryPlayer;
 import net.minecraft.item.ItemArmor;
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.crafting.CraftingManager;
+import net.minecraft.network.play.server.SPacketSetSlot;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.inventory.CraftInventoryCrafting;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
 
 public class ContainerPlayer extends Container
 {
-    private static final EntityEquipmentSlot[] field_185003_h = new EntityEquipmentSlot[] {EntityEquipmentSlot.HEAD, EntityEquipmentSlot.CHEST, EntityEquipmentSlot.LEGS, EntityEquipmentSlot.FEET};
-    public InventoryCrafting field_75181_e = new InventoryCrafting(this, 2, 2);
-    public IInventory field_75179_f = new InventoryCraftResult();
-    public boolean field_75180_g;
-    private final EntityPlayer field_82862_h;
+    private static final EntityEquipmentSlot[] VALID_EQUIPMENT_SLOTS = new EntityEquipmentSlot[] {EntityEquipmentSlot.HEAD, EntityEquipmentSlot.CHEST, EntityEquipmentSlot.LEGS, EntityEquipmentSlot.FEET};
+    public InventoryCrafting craftMatrix = new InventoryCrafting(this, 2, 2);
+    public IInventory craftResult = new InventoryCraftResult();
+    public boolean isLocalWorld;
+    private final EntityPlayer player;
+    // CraftBukkit start
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
 
-    public ContainerPlayer(final InventoryPlayer p_i1819_1_, boolean p_i1819_2_, EntityPlayer p_i1819_3_)
+        CraftInventoryCrafting inventory = new CraftInventoryCrafting(this.craftMatrix, this.craftResult);
+        bukkitEntity = new CraftInventoryView(this.playerinv.player.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private InventoryPlayer playerinv;
+    // CraftBukkit end
+    public ContainerPlayer(final InventoryPlayer playerInventory, boolean localWorld, EntityPlayer playerIn)
     {
-        this.field_75180_g = p_i1819_2_;
-        this.field_82862_h = p_i1819_3_;
-        this.func_75146_a(new SlotCrafting(p_i1819_1_.field_70458_d, this.field_75181_e, this.field_75179_f, 0, 154, 28));
-
+        this.isLocalWorld = localWorld;
+        this.player = playerIn;
+        this.addSlotToContainer(new SlotCrafting(playerInventory.player, this.craftMatrix, this.craftResult, 0, 154, 28));
+        //CraftBukkit start
+        this.craftResult = new net.minecraft.inventory.InventoryCraftResult(); // CraftBukkit - moved to before InventoryCrafting construction
+        this.craftMatrix = new net.minecraft.inventory.InventoryCrafting(this, 2, 2, playerInventory.player); // CraftBukkit - pass player
+        this.craftMatrix.resultInventory = this.craftResult; // CraftBukkit - let InventoryCrafting know about its result slot
+        this.playerinv = playerInventory; // CraftBukkit - save player
+        // CraftBukkit end
         for (int i = 0; i < 2; ++i)
         {
             for (int j = 0; j < 2; ++j)
             {
-                this.func_75146_a(new Slot(this.field_75181_e, j + i * 2, 98 + j * 18, 18 + i * 18));
+                this.addSlotToContainer(new Slot(this.craftMatrix, j + i * 2, 98 + j * 18, 18 + i * 18));
             }
         }
 
         for (int k = 0; k < 4; ++k)
         {
-            final EntityEquipmentSlot entityequipmentslot = field_185003_h[k];
-            this.func_75146_a(new Slot(p_i1819_1_, 36 + (3 - k), 8, 8 + k * 18)
+            final EntityEquipmentSlot entityequipmentslot = VALID_EQUIPMENT_SLOTS[k];
+            this.addSlotToContainer(new Slot(playerInventory, 36 + (3 - k), 8, 8 + k * 18)
             {
-                public int func_75219_a()
+                public int getSlotStackLimit()
                 {
                     return 1;
                 }
-                public boolean func_75214_a(ItemStack p_75214_1_)
+                public boolean isItemValid(ItemStack stack)
                 {
-                    return entityequipmentslot == EntityLiving.func_184640_d(p_75214_1_);
+                    return stack.getItem().isValidArmor(stack, entityequipmentslot, player);
                 }
-                public boolean func_82869_a(EntityPlayer p_82869_1_)
+                public boolean canTakeStack(EntityPlayer playerIn)
                 {
-                    ItemStack itemstack = this.func_75211_c();
-                    return !itemstack.func_190926_b() && !p_82869_1_.func_184812_l_() && EnchantmentHelper.func_190938_b(itemstack) ? false : super.func_82869_a(p_82869_1_);
+                    ItemStack itemstack = this.getStack();
+                    return !itemstack.isEmpty() && !playerIn.isCreative() && EnchantmentHelper.hasBindingCurse(itemstack) ? false : super.canTakeStack(playerIn);
                 }
                 @Nullable
                 @SideOnly(Side.CLIENT)
-                public String func_178171_c()
+                public String getSlotTexture()
                 {
-                    return ItemArmor.field_94603_a[entityequipmentslot.func_188454_b()];
+                    return ItemArmor.EMPTY_SLOT_NAMES[entityequipmentslot.getIndex()];
                 }
             });
         }
@@ -64,150 +88,161 @@
         {
             for (int j1 = 0; j1 < 9; ++j1)
             {
-                this.func_75146_a(new Slot(p_i1819_1_, j1 + (l + 1) * 9, 8 + j1 * 18, 84 + l * 18));
+                this.addSlotToContainer(new Slot(playerInventory, j1 + (l + 1) * 9, 8 + j1 * 18, 84 + l * 18));
             }
         }
 
         for (int i1 = 0; i1 < 9; ++i1)
         {
-            this.func_75146_a(new Slot(p_i1819_1_, i1, 8 + i1 * 18, 142));
+            this.addSlotToContainer(new Slot(playerInventory, i1, 8 + i1 * 18, 142));
         }
 
-        this.func_75146_a(new Slot(p_i1819_1_, 40, 77, 62)
+        this.addSlotToContainer(new Slot(playerInventory, 40, 77, 62)
         {
             @Nullable
             @SideOnly(Side.CLIENT)
-            public String func_178171_c()
+            public String getSlotTexture()
             {
                 return "minecraft:items/empty_armor_slot_shield";
             }
         });
-        this.func_75130_a(this.field_75181_e);
+        this.onCraftMatrixChanged(this.craftMatrix);
     }
 
-    public void func_75130_a(IInventory p_75130_1_)
+    public void onCraftMatrixChanged(IInventory inventoryIn)
     {
-        this.field_75179_f.func_70299_a(0, CraftingManager.func_77594_a().func_82787_a(this.field_75181_e, this.field_82862_h.field_70170_p));
+        // this.resultInventory.setItem(0, CraftingManager.getInstance().craft(this.craftInventory, this.owner.world));
+        // CraftBukkit start (Note: the following line would cause an error if called during construction)
+        CraftingManager.getInstance().lastCraftView = getBukkitView();
+        ItemStack craftResult = CraftingManager.getInstance().findMatchingRecipe(this.craftMatrix, this.player.world);
+        this.craftResult.setInventorySlotContents(0, craftResult);
+        if (super.listeners.size() < 1) {
+            return;
+        }
+
+        EntityPlayerMP player = (EntityPlayerMP) super.listeners.get(0); // TODO: Is this _always_ correct? Seems like it.
+        player.connection.sendPacket(new SPacketSetSlot(player.openContainer.windowId, 0, craftResult));
+        // CraftBukkit end
     }
 
-    public void func_75134_a(EntityPlayer p_75134_1_)
+    public void onContainerClosed(EntityPlayer playerIn)
     {
-        super.func_75134_a(p_75134_1_);
+        super.onContainerClosed(playerIn);
 
         for (int i = 0; i < 4; ++i)
         {
-            ItemStack itemstack = this.field_75181_e.func_70304_b(i);
+            ItemStack itemstack = this.craftMatrix.removeStackFromSlot(i);
 
-            if (!itemstack.func_190926_b())
+            if (!itemstack.isEmpty())
             {
-                p_75134_1_.func_71019_a(itemstack, false);
+                playerIn.dropItem(itemstack, false);
             }
         }
 
-        this.field_75179_f.func_70299_a(0, ItemStack.field_190927_a);
+        this.craftResult.setInventorySlotContents(0, ItemStack.EMPTY);
     }
 
-    public boolean func_75145_c(EntityPlayer p_75145_1_)
+    public boolean canInteractWith(EntityPlayer playerIn)
     {
         return true;
     }
 
-    public ItemStack func_82846_b(EntityPlayer p_82846_1_, int p_82846_2_)
+    public ItemStack transferStackInSlot(EntityPlayer playerIn, int index)
     {
-        ItemStack itemstack = ItemStack.field_190927_a;
-        Slot slot = (Slot)this.field_75151_b.get(p_82846_2_);
+        ItemStack itemstack = ItemStack.EMPTY;
+        Slot slot = (Slot)this.inventorySlots.get(index);
 
-        if (slot != null && slot.func_75216_d())
+        if (slot != null && slot.getHasStack())
         {
-            ItemStack itemstack1 = slot.func_75211_c();
-            itemstack = itemstack1.func_77946_l();
-            EntityEquipmentSlot entityequipmentslot = EntityLiving.func_184640_d(itemstack);
+            ItemStack itemstack1 = slot.getStack();
+            itemstack = itemstack1.copy();
+            EntityEquipmentSlot entityequipmentslot = EntityLiving.getSlotForItemStack(itemstack);
 
-            if (p_82846_2_ == 0)
+            if (index == 0)
             {
-                if (!this.func_75135_a(itemstack1, 9, 45, true))
+                if (!this.mergeItemStack(itemstack1, 9, 45, true))
                 {
-                    return ItemStack.field_190927_a;
+                    return ItemStack.EMPTY;
                 }
 
-                slot.func_75220_a(itemstack1, itemstack);
+                slot.onSlotChange(itemstack1, itemstack);
             }
-            else if (p_82846_2_ >= 1 && p_82846_2_ < 5)
+            else if (index >= 1 && index < 5)
             {
-                if (!this.func_75135_a(itemstack1, 9, 45, false))
+                if (!this.mergeItemStack(itemstack1, 9, 45, false))
                 {
-                    return ItemStack.field_190927_a;
+                    return ItemStack.EMPTY;
                 }
             }
-            else if (p_82846_2_ >= 5 && p_82846_2_ < 9)
+            else if (index >= 5 && index < 9)
             {
-                if (!this.func_75135_a(itemstack1, 9, 45, false))
+                if (!this.mergeItemStack(itemstack1, 9, 45, false))
                 {
-                    return ItemStack.field_190927_a;
+                    return ItemStack.EMPTY;
                 }
             }
-            else if (entityequipmentslot.func_188453_a() == EntityEquipmentSlot.Type.ARMOR && !((Slot)this.field_75151_b.get(8 - entityequipmentslot.func_188454_b())).func_75216_d())
+            else if (entityequipmentslot.getSlotType() == EntityEquipmentSlot.Type.ARMOR && !((Slot)this.inventorySlots.get(8 - entityequipmentslot.getIndex())).getHasStack())
             {
-                int i = 8 - entityequipmentslot.func_188454_b();
+                int i = 8 - entityequipmentslot.getIndex();
 
-                if (!this.func_75135_a(itemstack1, i, i + 1, false))
+                if (!this.mergeItemStack(itemstack1, i, i + 1, false))
                 {
-                    return ItemStack.field_190927_a;
+                    return ItemStack.EMPTY;
                 }
             }
-            else if (entityequipmentslot == EntityEquipmentSlot.OFFHAND && !((Slot)this.field_75151_b.get(45)).func_75216_d())
+            else if (entityequipmentslot == EntityEquipmentSlot.OFFHAND && !((Slot)this.inventorySlots.get(45)).getHasStack())
             {
-                if (!this.func_75135_a(itemstack1, 45, 46, false))
+                if (!this.mergeItemStack(itemstack1, 45, 46, false))
                 {
-                    return ItemStack.field_190927_a;
+                    return ItemStack.EMPTY;
                 }
             }
-            else if (p_82846_2_ >= 9 && p_82846_2_ < 36)
+            else if (index >= 9 && index < 36)
             {
-                if (!this.func_75135_a(itemstack1, 36, 45, false))
+                if (!this.mergeItemStack(itemstack1, 36, 45, false))
                 {
-                    return ItemStack.field_190927_a;
+                    return ItemStack.EMPTY;
                 }
             }
-            else if (p_82846_2_ >= 36 && p_82846_2_ < 45)
+            else if (index >= 36 && index < 45)
             {
-                if (!this.func_75135_a(itemstack1, 9, 36, false))
+                if (!this.mergeItemStack(itemstack1, 9, 36, false))
                 {
-                    return ItemStack.field_190927_a;
+                    return ItemStack.EMPTY;
                 }
             }
-            else if (!this.func_75135_a(itemstack1, 9, 45, false))
+            else if (!this.mergeItemStack(itemstack1, 9, 45, false))
             {
-                return ItemStack.field_190927_a;
+                return ItemStack.EMPTY;
             }
 
-            if (itemstack1.func_190926_b())
+            if (itemstack1.isEmpty())
             {
-                slot.func_75215_d(ItemStack.field_190927_a);
+                slot.putStack(ItemStack.EMPTY);
             }
             else
             {
-                slot.func_75218_e();
+                slot.onSlotChanged();
             }
 
-            if (itemstack1.func_190916_E() == itemstack.func_190916_E())
+            if (itemstack1.getCount() == itemstack.getCount())
             {
-                return ItemStack.field_190927_a;
+                return ItemStack.EMPTY;
             }
 
-            ItemStack itemstack2 = slot.func_190901_a(p_82846_1_, itemstack1);
+            ItemStack itemstack2 = slot.onTake(playerIn, itemstack1);
 
-            if (p_82846_2_ == 0)
+            if (index == 0)
             {
-                p_82846_1_.func_71019_a(itemstack2, false);
+                playerIn.dropItem(itemstack2, false);
             }
         }
 
         return itemstack;
     }
 
-    public boolean func_94530_a(ItemStack p_94530_1_, Slot p_94530_2_)
+    public boolean canMergeSlot(ItemStack stack, Slot slotIn)
     {
-        return p_94530_2_.field_75224_c != this.field_75179_f && super.func_94530_a(p_94530_1_, p_94530_2_);
+        return slotIn.inventory != this.craftResult && super.canMergeSlot(stack, slotIn);
     }
 }
