--- ../src-base/minecraft/net/minecraft/inventory/ContainerHorseChest.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerHorseChest.java
@@ -1,19 +1,26 @@
 package net.minecraft.inventory;
 
+import net.minecraft.entity.passive.AbstractHorse;
 import net.minecraft.util.text.ITextComponent;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class ContainerHorseChest extends InventoryBasic
 {
-    public ContainerHorseChest(String p_i47268_1_, int p_i47268_2_)
+    public ContainerHorseChest(String inventoryTitle, int slotCount)
     {
-        super(p_i47268_1_, false, p_i47268_2_);
+        super(inventoryTitle, false, slotCount);
     }
 
     @SideOnly(Side.CLIENT)
-    public ContainerHorseChest(ITextComponent p_i47269_1_, int p_i47269_2_)
+    public ContainerHorseChest(ITextComponent inventoryTitle, int slotCount)
     {
-        super(p_i47269_1_, p_i47269_2_);
+        super(inventoryTitle, slotCount);
     }
+
+    // CraftBukkit start
+    public ContainerHorseChest(String s, int i, AbstractHorse owner) {
+        super(s, false, i, (org.bukkit.entity.AbstractHorse) owner.getBukkitEntity());
+        // CraftBukkit end
+    }
 }
