--- ../src-base/minecraft/net/minecraft/inventory/ContainerWorkbench.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerWorkbench.java
@@ -1,12 +1,17 @@
 package net.minecraft.inventory;
 
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.entity.player.InventoryPlayer;
 import net.minecraft.init.Blocks;
+import net.minecraft.init.Items;
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.crafting.CraftingManager;
+import net.minecraft.network.play.server.SPacketSetSlot;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.inventory.CraftInventoryCrafting;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
 
 public class ContainerWorkbench extends Container
 {
@@ -14,9 +19,30 @@
     public IInventory field_75160_f = new InventoryCraftResult();
     private final World field_75161_g;
     private final BlockPos field_178145_h;
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private InventoryPlayer player;
+    //CraftBukkit end
+    // CraftBukkit start
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
 
+        CraftInventoryCrafting inventory = new CraftInventoryCrafting(this.field_75162_e, this.field_75160_f);
+        bukkitEntity = new CraftInventoryView(this.player.field_70458_d.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
     public ContainerWorkbench(InventoryPlayer p_i45800_1_, World p_i45800_2_, BlockPos p_i45800_3_)
     {
+        // CraftBukkit start - Switched order of IInventory construction and stored player
+        this.field_75160_f = new net.minecraft.inventory.InventoryCraftResult();
+        this.field_75162_e = new net.minecraft.inventory.InventoryCrafting(this, 3, 3, p_i45800_1_.field_70458_d); // CraftBukkit - pass player
+        this.field_75162_e.resultInventory = this.field_75160_f;
+        this.player = p_i45800_1_;
+        // CraftBukkit end
         this.field_75161_g = p_i45800_2_;
         this.field_178145_h = p_i45800_3_;
         this.func_75146_a(new SlotCrafting(p_i45800_1_.field_70458_d, this.field_75162_e, this.field_75160_f, 0, 124, 35));
@@ -47,7 +73,21 @@
 
     public void func_75130_a(IInventory p_75130_1_)
     {
-        this.field_75160_f.func_70299_a(0, CraftingManager.func_77594_a().func_82787_a(this.field_75162_e, this.field_75161_g));
+// this.resultInventory.setItem(0, CraftingManager.getInstance().craft(this.craftInventory, this.g));
+        // CraftBukkit start
+        CraftingManager.func_77594_a().lastCraftView = getBukkitView();
+        ItemStack craftResult = CraftingManager.func_77594_a().func_82787_a(this.field_75162_e, this.field_75161_g);
+        this.field_75160_f.func_70299_a(0, craftResult);
+        if (super.field_75149_d.size() < 1) {
+            return;
+        }
+        // See CraftBukkit PR #39
+        if (craftResult != null && craftResult.func_77973_b() == Items.field_151098_aY) {
+            return;
+        }
+        EntityPlayerMP player = (EntityPlayerMP) super.field_75149_d.get(0); // TODO: Is this _always_ correct? Seems like it.
+        player.field_71135_a.func_147359_a(new SPacketSetSlot(player.field_71070_bA.field_75152_c, 0, craftResult));
+        // CraftBukkit end
     }
 
     public void func_75134_a(EntityPlayer p_75134_1_)
