--- ../src-base/minecraft/net/minecraft/inventory/ContainerHorseChest.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerHorseChest.java
@@ -1,5 +1,6 @@
 package net.minecraft.inventory;
 
+import net.minecraft.entity.passive.AbstractHorse;
 import net.minecraft.util.text.ITextComponent;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
@@ -16,4 +17,10 @@
     {
         super(p_i47269_1_, p_i47269_2_);
     }
+
+    // CraftBukkit start
+    public ContainerHorseChest(String s, int i, AbstractHorse owner) {
+        super(s, false, i, (org.bukkit.entity.AbstractHorse) owner.getBukkitEntity());
+        // CraftBukkit end
+    }
 }
