--- ../src-base/minecraft/net/minecraft/inventory/ContainerShulkerBox.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerShulkerBox.java
@@ -3,13 +3,29 @@
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.player.InventoryPlayer;
 import net.minecraft.item.ItemStack;
+import org.bukkit.craftbukkit.inventory.CraftInventory;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
 
 public class ContainerShulkerBox extends Container
 {
     private final IInventory field_190899_a;
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity;
+    private InventoryPlayer player;
 
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        bukkitEntity = new CraftInventoryView(this.player.field_70458_d.getBukkitEntity(), new CraftInventory(this.field_190899_a), this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
     public ContainerShulkerBox(InventoryPlayer p_i47266_1_, IInventory p_i47266_2_, EntityPlayer p_i47266_3_)
     {
+        this.player = p_i47266_1_; // CraftBukkit - save player
         this.field_190899_a = p_i47266_2_;
         p_i47266_2_.func_174889_b(p_i47266_3_);
         int i = 3;
