--- ../src-base/minecraft/net/minecraft/inventory/Container.java
+++ ../src-work/minecraft/net/minecraft/inventory/Container.java
@@ -2,19 +2,33 @@
 
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
+
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 import java.util.Set;
 import javax.annotation.Nullable;
 import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.entity.player.InventoryPlayer;
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
+import net.minecraft.network.play.server.SPacketSetSlot;
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.util.NonNullList;
 import net.minecraft.util.math.MathHelper;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.Material;
+import org.bukkit.craftbukkit.inventory.CraftInventory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.Event;
+import org.bukkit.event.inventory.InventoryDragEvent;
+import org.bukkit.event.inventory.InventoryType;
+import org.bukkit.inventory.InventoryView;
 
+import static net.minecraft.item.ItemStack.EMPTY;
+
 public abstract class Container
 {
     public NonNullList<ItemStack> field_75153_a = NonNullList.<ItemStack>func_191196_a();
@@ -27,12 +41,22 @@
     private final Set<Slot> field_94537_h = Sets.<Slot>newHashSet();
     protected List<IContainerListener> field_75149_d = Lists.<IContainerListener>newArrayList();
     private final Set<EntityPlayer> field_75148_f = Sets.<EntityPlayer>newHashSet();
-
+    // CraftBukkit start
+    public boolean checkReachable = true;
+    public abstract InventoryView getBukkitView();
+    public void transferTo(net.minecraft.inventory.Container other, org.bukkit.craftbukkit.entity.CraftHumanEntity player) {
+        InventoryView source = this.getBukkitView(), destination = other.getBukkitView();
+        ((CraftInventory) source.getTopInventory()).getInventory().onClose(player);
+        ((CraftInventory) source.getBottomInventory()).getInventory().onClose(player);
+        ((CraftInventory) destination.getTopInventory()).getInventory().onOpen(player);
+        ((CraftInventory) destination.getBottomInventory()).getInventory().onOpen(player);
+    }
+    // CraftBukkit end
     protected Slot func_75146_a(Slot p_75146_1_)
     {
         p_75146_1_.field_75222_d = this.field_75151_b.size();
         this.field_75151_b.add(p_75146_1_);
-        this.field_75153_a.add(ItemStack.field_190927_a);
+        this.field_75153_a.add(field_190927_a);
         return p_75146_1_;
     }
 
@@ -77,7 +101,7 @@
 
             if (!ItemStack.func_77989_b(itemstack1, itemstack))
             {
-                itemstack1 = itemstack.func_190926_b() ? ItemStack.field_190927_a : itemstack.func_77946_l();
+                itemstack1 = itemstack.func_190926_b() ? field_190927_a : itemstack.func_77946_l();
                 this.field_75153_a.set(i, itemstack1);
 
                 for (int j = 0; j < this.field_75149_d.size(); ++j)
@@ -117,229 +141,212 @@
     public ItemStack func_82846_b(EntityPlayer p_82846_1_, int p_82846_2_)
     {
         Slot slot = (Slot)this.field_75151_b.get(p_82846_2_);
-        return slot != null ? slot.func_75211_c() : ItemStack.field_190927_a;
+        return slot != null ? slot.func_75211_c() : field_190927_a;
     }
-
-    public ItemStack func_184996_a(int p_184996_1_, int p_184996_2_, ClickType p_184996_3_, EntityPlayer p_184996_4_)
-    {
-        ItemStack itemstack = ItemStack.field_190927_a;
+                                            //左键？右键？          //按键要干什么
+    public ItemStack func_184996_a(int p_184996_1_, int p_184996_2_, ClickType p_184996_3_, EntityPlayer p_184996_4_) {
+        ItemStack itemstack = field_190927_a;
         InventoryPlayer inventoryplayer = p_184996_4_.field_71071_by;
 
-        if (p_184996_3_ == ClickType.QUICK_CRAFT)
-        {
-            int i = this.field_94536_g;
+        if (p_184996_3_ == ClickType.QUICK_CRAFT) {
+            //快速合成
+            //如果拖动状态变化或者拖动不是按左键，则重置拖动
+            int oldDragEvent = this.field_94536_g;
             this.field_94536_g = func_94532_c(p_184996_2_);
-
-            if ((i != 1 || this.field_94536_g != 2) && i != this.field_94536_g)
-            {
+            if ((oldDragEvent != 1 || this.field_94536_g != 2) && oldDragEvent != this.field_94536_g) {
                 this.func_94533_d();
-            }
-            else if (inventoryplayer.func_70445_o().func_190926_b())
-            {
+            } else if (inventoryplayer.func_70445_o().func_190926_b()) {
                 this.func_94533_d();
-            }
-            else if (this.field_94536_g == 0)
-            {
+            } else if (this.field_94536_g == 0) {
+                //左键：进入"额外拖动模式"
                 this.field_94535_f = func_94529_b(p_184996_2_);
-
-                if (func_180610_a(this.field_94535_f, p_184996_4_))
-                {
+                //如果不可用则回到右键模式
+                if (func_180610_a(this.field_94535_f, p_184996_4_)) {
                     this.field_94536_g = 1;
                     this.field_94537_h.clear();
-                }
-                else
-                {
+                } else {
                     this.func_94533_d();
                 }
-            }
-            else if (this.field_94536_g == 1)
-            {
-                Slot slot = (Slot)this.field_75151_b.get(p_184996_1_);
-                ItemStack itemstack1 = inventoryplayer.func_70445_o();
+            } else if (this.field_94536_g == 1) {
+                //右键拖动
+                Slot slot = (Slot) this.field_75151_b.get(p_184996_1_);
+                ItemStack currentSelected = inventoryplayer.func_70445_o();
 
-                if (slot != null && func_94527_a(slot, itemstack1, true) && slot.func_75214_a(itemstack1) && (this.field_94535_f == 2 || itemstack1.func_190916_E() > this.field_94537_h.size()) && this.func_94531_b(slot))
-                {
+                if (slot != null && func_94527_a(slot, currentSelected, true) && slot.func_75214_a(currentSelected) && (this.field_94535_f == 2 || currentSelected.func_190916_E() > this.field_94537_h.size()) && this.func_94531_b(slot)) {
                     this.field_94537_h.add(slot);
                 }
-            }
-            else if (this.field_94536_g == 2)
-            {
-                if (!this.field_94537_h.isEmpty())
-                {
+            } else if (this.field_94536_g == 2) {
+                if (!this.field_94537_h.isEmpty()) {
+                    Map<Integer, ItemStack> draggedSlots = new HashMap<Integer, ItemStack>(); // CraftBukkit - Store slots from drag in map (raw slot id -> new stack)
+
                     ItemStack itemstack5 = inventoryplayer.func_70445_o().func_77946_l();
                     int l = inventoryplayer.func_70445_o().func_190916_E();
 
-                    for (Slot slot1 : this.field_94537_h)
-                    {
+                    for (Slot slot1 : this.field_94537_h) {
                         ItemStack itemstack2 = inventoryplayer.func_70445_o();
 
-                        if (slot1 != null && func_94527_a(slot1, itemstack2, true) && slot1.func_75214_a(itemstack2) && (this.field_94535_f == 2 || itemstack2.func_190916_E() >= this.field_94537_h.size()) && this.func_94531_b(slot1))
-                        {
+                        if (slot1 != null && func_94527_a(slot1, itemstack2, true) && slot1.func_75214_a(itemstack2) && (this.field_94535_f == 2 || itemstack2.func_190916_E() >= this.field_94537_h.size()) && this.func_94531_b(slot1)) {
                             ItemStack itemstack3 = itemstack5.func_77946_l();
                             int j = slot1.func_75216_d() ? slot1.func_75211_c().func_190916_E() : 0;
                             func_94525_a(this.field_94537_h, this.field_94535_f, itemstack3, j);
                             int k = Math.min(itemstack3.func_77976_d(), slot1.func_178170_b(itemstack3));
 
-                            if (itemstack3.func_190916_E() > k)
-                            {
+                            if (itemstack3.func_190916_E() > k) {
                                 itemstack3.func_190920_e(k);
                             }
 
                             l -= itemstack3.func_190916_E() - j;
-                            slot1.func_75215_d(itemstack3);
+                            // slot1.set(itemstack4);
+                            draggedSlots.put(slot1.field_75222_d, itemstack3); // CraftBukkit - Put in map instead of setting                        }
                         }
+
+                        //itemstack5.setCount(l);
+                        //inventoryplayer.setItemStack(itemstack5);
+                        // CraftBukkit start - InventoryDragEvent
+                        InventoryView view = getBukkitView();
+                        org.bukkit.inventory.ItemStack newcursor = CraftItemStack.asCraftMirror(itemstack2);
+                        newcursor.setAmount(l);
+                        Map<Integer, org.bukkit.inventory.ItemStack> eventmap = new HashMap<Integer, org.bukkit.inventory.ItemStack>();
+                        for (Map.Entry<Integer, ItemStack> ditem : draggedSlots.entrySet()) {
+                            eventmap.put(ditem.getKey(), CraftItemStack.asBukkitCopy(ditem.getValue()));
+                        }
+
+                        // It's essential that we set the cursor to the new value here to prevent item duplication if a plugin closes the inventory.
+                        ItemStack oldCursor = inventoryplayer.func_70445_o();
+                        inventoryplayer.func_70437_b(CraftItemStack.asNMSCopy(newcursor));
+
+                        InventoryDragEvent event = new InventoryDragEvent(view, (newcursor.getType() != Material.AIR ? newcursor : null), CraftItemStack.asBukkitCopy(oldCursor), this.field_94535_f == 1, eventmap);
+                        p_184996_4_.field_70170_p.getServer().getPluginManager().callEvent(event);
+
+                        // Whether or not a change was made to the inventory that requires an update.
+                        boolean needsUpdate = event.getResult() != Event.Result.DEFAULT;
+
+                        if (event.getResult() != Event.Result.DENY) {
+                            for (Map.Entry<Integer, ItemStack> dslot : draggedSlots.entrySet()) {
+                                view.setItem(dslot.getKey(), CraftItemStack.asBukkitCopy(dslot.getValue()));
+                            }
+                            // The only time the carried item will be set to null is if the inventory is closed by the server.
+                            // If the inventory is closed by the server, then the cursor items are dropped.  This is why we change the cursor early.
+                            if (inventoryplayer.func_70445_o() != null) {
+                                inventoryplayer.func_70437_b(CraftItemStack.asNMSCopy(event.getCursor()));
+                                needsUpdate = true;
+                            }
+                        } else {
+                            inventoryplayer.func_70437_b(oldCursor);
+                        }
+
+                        if (needsUpdate && p_184996_4_ instanceof EntityPlayerMP) {
+                            ((EntityPlayerMP) p_184996_4_).func_71120_a(this);
+                        }
+                        // CraftBukkit end
                     }
 
-                    itemstack5.func_190920_e(l);
-                    inventoryplayer.func_70437_b(itemstack5);
+                    this.func_94533_d();
+                } else {
+                    this.func_94533_d();
                 }
-
+            } else if (this.field_94536_g != 0) {
                 this.func_94533_d();
             }
-            else
-            {
-                this.func_94533_d();
-            }
         }
-        else if (this.field_94536_g != 0)
-        {
-            this.func_94533_d();
-        }
-        else if ((p_184996_3_ == ClickType.PICKUP || p_184996_3_ == ClickType.QUICK_MOVE) && (p_184996_2_ == 0 || p_184996_2_ == 1))
-        {
-            if (p_184996_1_ == -999)
-            {
-                if (!inventoryplayer.func_70445_o().func_190926_b())
-                {
-                    if (p_184996_2_ == 0)
-                    {
+        else if ((p_184996_3_ == ClickType.PICKUP || p_184996_3_ == ClickType.QUICK_MOVE) && (p_184996_2_ == 0 || p_184996_2_ == 1)) {
+            if (p_184996_1_ == -999) {
+                if (!inventoryplayer.func_70445_o().func_190926_b()) {
+                    if (p_184996_2_ == 0) {
                         p_184996_4_.func_71019_a(inventoryplayer.func_70445_o(), true);
-                        inventoryplayer.func_70437_b(ItemStack.field_190927_a);
+                        inventoryplayer.func_70437_b(field_190927_a);
                     }
 
-                    if (p_184996_2_ == 1)
-                    {
+                    if (p_184996_2_ == 1) {
                         p_184996_4_.func_71019_a(inventoryplayer.func_70445_o().func_77979_a(1), true);
                     }
                 }
-            }
-            else if (p_184996_3_ == ClickType.QUICK_MOVE)
-            {
-                if (p_184996_1_ < 0)
-                {
-                    return ItemStack.field_190927_a;
+            } else if (p_184996_3_ == ClickType.QUICK_MOVE) {
+                if (p_184996_1_ < 0) {
+                    return field_190927_a;
                 }
 
-                Slot slot6 = (Slot)this.field_75151_b.get(p_184996_1_);
+                Slot slot6 = (Slot) this.field_75151_b.get(p_184996_1_);
 
-                if (slot6 != null && slot6.func_82869_a(p_184996_4_))
-                {
+                if (slot6 != null && slot6.func_82869_a(p_184996_4_)) {
                     ItemStack itemstack10 = this.func_82846_b(p_184996_4_, p_184996_1_);
 
-                    if (!itemstack10.func_190926_b())
-                    {
+                    if (!itemstack10.func_190926_b()) {
                         Item item = itemstack10.func_77973_b();
                         itemstack = itemstack10.func_77946_l();
 
-                        if (slot6.func_75211_c().func_77973_b() == item)
-                        {
+                        if (slot6.func_75211_c().func_77973_b() == item) {
                             this.func_75133_b(p_184996_1_, p_184996_2_, true, p_184996_4_);
                         }
                     }
                 }
-            }
-            else
-            {
-                if (p_184996_1_ < 0)
-                {
-                    return ItemStack.field_190927_a;
+            } else {
+                if (p_184996_1_ < 0) {
+                    return field_190927_a;
                 }
 
-                Slot slot7 = (Slot)this.field_75151_b.get(p_184996_1_);
+                Slot slot7 = (Slot) this.field_75151_b.get(p_184996_1_);
 
-                if (slot7 != null)
-                {
+                if (slot7 != null) {
                     ItemStack itemstack11 = slot7.func_75211_c();
                     ItemStack itemstack13 = inventoryplayer.func_70445_o();
 
-                    if (!itemstack11.func_190926_b())
-                    {
+                    if (!itemstack11.func_190926_b()) {
                         itemstack = itemstack11.func_77946_l();
                     }
 
-                    if (itemstack11.func_190926_b())
-                    {
-                        if (!itemstack13.func_190926_b() && slot7.func_75214_a(itemstack13))
-                        {
+                    if (itemstack11.func_190926_b()) {
+                        if (!itemstack13.func_190926_b() && slot7.func_75214_a(itemstack13)) {
                             int l2 = p_184996_2_ == 0 ? itemstack13.func_190916_E() : 1;
 
-                            if (l2 > slot7.func_178170_b(itemstack13))
-                            {
+                            if (l2 > slot7.func_178170_b(itemstack13)) {
                                 l2 = slot7.func_178170_b(itemstack13);
                             }
 
                             slot7.func_75215_d(itemstack13.func_77979_a(l2));
                         }
-                    }
-                    else if (slot7.func_82869_a(p_184996_4_))
-                    {
-                        if (itemstack13.func_190926_b())
-                        {
-                            if (itemstack11.func_190926_b())
-                            {
-                                slot7.func_75215_d(ItemStack.field_190927_a);
-                                inventoryplayer.func_70437_b(ItemStack.field_190927_a);
-                            }
-                            else
-                            {
+                    } else if (slot7.func_82869_a(p_184996_4_)) {
+                        if (itemstack13.func_190926_b()) {
+                            if (itemstack11.func_190926_b()) {
+                                slot7.func_75215_d(field_190927_a);
+                                inventoryplayer.func_70437_b(field_190927_a);
+                            } else {
                                 int k2 = p_184996_2_ == 0 ? itemstack11.func_190916_E() : (itemstack11.func_190916_E() + 1) / 2;
                                 inventoryplayer.func_70437_b(slot7.func_75209_a(k2));
 
-                                if (itemstack11.func_190926_b())
-                                {
-                                    slot7.func_75215_d(ItemStack.field_190927_a);
+                                if (itemstack11.func_190926_b()) {
+                                    slot7.func_75215_d(field_190927_a);
                                 }
 
                                 slot7.func_190901_a(p_184996_4_, inventoryplayer.func_70445_o());
                             }
-                        }
-                        else if (slot7.func_75214_a(itemstack13))
-                        {
-                            if (itemstack11.func_77973_b() == itemstack13.func_77973_b() && itemstack11.func_77960_j() == itemstack13.func_77960_j() && ItemStack.func_77970_a(itemstack11, itemstack13))
-                            {
+                        } else if (slot7.func_75214_a(itemstack13)) {
+                            if (itemstack11.func_77973_b() == itemstack13.func_77973_b() && itemstack11.func_77960_j() == itemstack13.func_77960_j() && ItemStack.func_77970_a(itemstack11, itemstack13)) {
                                 int j2 = p_184996_2_ == 0 ? itemstack13.func_190916_E() : 1;
 
-                                if (j2 > slot7.func_178170_b(itemstack13) - itemstack11.func_190916_E())
-                                {
+                                if (j2 > slot7.func_178170_b(itemstack13) - itemstack11.func_190916_E()) {
                                     j2 = slot7.func_178170_b(itemstack13) - itemstack11.func_190916_E();
                                 }
 
-                                if (j2 > itemstack13.func_77976_d() - itemstack11.func_190916_E())
-                                {
+                                if (j2 > itemstack13.func_77976_d() - itemstack11.func_190916_E()) {
                                     j2 = itemstack13.func_77976_d() - itemstack11.func_190916_E();
                                 }
 
                                 itemstack13.func_190918_g(j2);
                                 itemstack11.func_190917_f(j2);
-                            }
-                            else if (itemstack13.func_190916_E() <= slot7.func_178170_b(itemstack13))
-                            {
+                            } else if (itemstack13.func_190916_E() <= slot7.func_178170_b(itemstack13)) {
                                 slot7.func_75215_d(itemstack13);
                                 inventoryplayer.func_70437_b(itemstack11);
                             }
-                        }
-                        else if (itemstack11.func_77973_b() == itemstack13.func_77973_b() && itemstack13.func_77976_d() > 1 && (!itemstack11.func_77981_g() || itemstack11.func_77960_j() == itemstack13.func_77960_j()) && ItemStack.func_77970_a(itemstack11, itemstack13) && !itemstack11.func_190926_b())
-                        {
+                        } else if (itemstack11.func_77973_b() == itemstack13.func_77973_b() && itemstack13.func_77976_d() > 1 && (!itemstack11.func_77981_g() || itemstack11.func_77960_j() == itemstack13.func_77960_j()) && ItemStack.func_77970_a(itemstack11, itemstack13) && !itemstack11.func_190926_b()) {
                             int i2 = itemstack11.func_190916_E();
 
-                            if (i2 + itemstack13.func_190916_E() <= itemstack13.func_77976_d())
-                            {
+                            if (i2 + itemstack13.func_190916_E() <= itemstack13.func_77976_d()) {
                                 itemstack13.func_190917_f(i2);
                                 itemstack11 = slot7.func_75209_a(i2);
 
-                                if (itemstack11.func_190926_b())
-                                {
-                                    slot7.func_75215_d(ItemStack.field_190927_a);
+                                if (itemstack11.func_190926_b()) {
+                                    slot7.func_75215_d(field_190927_a);
                                 }
 
                                 slot7.func_190901_a(p_184996_4_, inventoryplayer.func_70445_o());
@@ -348,118 +355,98 @@
                     }
 
                     slot7.func_75218_e();
+                    /*
+                    // CraftBukkit start - Make sure the client has the right slot contents
+                    if (player instanceof EntityPlayerMP && slot7.getSlotStackLimit() != 64) {
+                        ((EntityPlayerMP) player).connection.sendPacket(new SPacketSetSlot(this.windowId, slot7.slotNumber, slot7.getStack()));
+                        // Updating a crafting inventory makes the client reset the result slot, have to send it again
+                        if (this.getBukkitView().getType() == InventoryType.WORKBENCH || this.getBukkitView().getType() == InventoryType.CRAFTING) {
+                            ((EntityPlayerMP) player).connection.sendPacket(new SPacketSetSlot(this.windowId, 0, this.getSlot(0).getStack()));
+                        }
+                    }
+                    // CraftBukkit end
+                    */
                 }
             }
-        }
-        else if (p_184996_3_ == ClickType.SWAP && p_184996_2_ >= 0 && p_184996_2_ < 9)
-        {
-            Slot slot5 = (Slot)this.field_75151_b.get(p_184996_1_);
+        } else if (p_184996_3_ == ClickType.SWAP && p_184996_2_ >= 0 && p_184996_2_ < 9) {
+            Slot slot5 = (Slot) this.field_75151_b.get(p_184996_1_);
             ItemStack itemstack9 = inventoryplayer.func_70301_a(p_184996_2_);
             ItemStack itemstack12 = slot5.func_75211_c();
 
-            if (!itemstack9.func_190926_b() || !itemstack12.func_190926_b())
-            {
-                if (itemstack9.func_190926_b())
-                {
-                    if (slot5.func_82869_a(p_184996_4_))
-                    {
+            if (!itemstack9.func_190926_b() || !itemstack12.func_190926_b()) {
+                if (itemstack9.func_190926_b()) {
+                    if (slot5.func_82869_a(p_184996_4_)) {
                         inventoryplayer.func_70299_a(p_184996_2_, itemstack12);
                         slot5.func_190900_b(itemstack12.func_190916_E());
-                        slot5.func_75215_d(ItemStack.field_190927_a);
+                        slot5.func_75215_d(field_190927_a);
                         slot5.func_190901_a(p_184996_4_, itemstack12);
                     }
-                }
-                else if (itemstack12.func_190926_b())
-                {
-                    if (slot5.func_75214_a(itemstack9))
-                    {
+                } else if (itemstack12.func_190926_b()) {
+                    if (slot5.func_75214_a(itemstack9)) {
                         int k1 = slot5.func_178170_b(itemstack9);
 
-                        if (itemstack9.func_190916_E() > k1)
-                        {
+                        if (itemstack9.func_190916_E() > k1) {
                             slot5.func_75215_d(itemstack9.func_77979_a(k1));
-                        }
-                        else
-                        {
+                        } else {
                             slot5.func_75215_d(itemstack9);
-                            inventoryplayer.func_70299_a(p_184996_2_, ItemStack.field_190927_a);
+                            inventoryplayer.func_70299_a(p_184996_2_, field_190927_a);
                         }
                     }
-                }
-                else if (slot5.func_82869_a(p_184996_4_) && slot5.func_75214_a(itemstack9))
-                {
+                } else if (slot5.func_82869_a(p_184996_4_) && slot5.func_75214_a(itemstack9)) {
                     int l1 = slot5.func_178170_b(itemstack9);
 
-                    if (itemstack9.func_190916_E() > l1)
-                    {
+                    if (itemstack9.func_190916_E() > l1) {
                         slot5.func_75215_d(itemstack9.func_77979_a(l1));
                         slot5.func_190901_a(p_184996_4_, itemstack12);
 
-                        if (!inventoryplayer.func_70441_a(itemstack12))
-                        {
+                        if (!inventoryplayer.func_70441_a(itemstack12)) {
                             p_184996_4_.func_71019_a(itemstack12, true);
                         }
-                    }
-                    else
-                    {
+                    } else {
                         slot5.func_75215_d(itemstack9);
                         inventoryplayer.func_70299_a(p_184996_2_, itemstack12);
                         slot5.func_190901_a(p_184996_4_, itemstack12);
                     }
                 }
             }
-        }
-        else if (p_184996_3_ == ClickType.CLONE && p_184996_4_.field_71075_bZ.field_75098_d && inventoryplayer.func_70445_o().func_190926_b() && p_184996_1_ >= 0)
-        {
-            Slot slot4 = (Slot)this.field_75151_b.get(p_184996_1_);
+        } else if (p_184996_3_ == ClickType.CLONE && p_184996_4_.field_71075_bZ.field_75098_d && inventoryplayer.func_70445_o().func_190926_b() && p_184996_1_ >= 0) {
+            Slot slot4 = (Slot) this.field_75151_b.get(p_184996_1_);
 
-            if (slot4 != null && slot4.func_75216_d())
-            {
+            if (slot4 != null && slot4.func_75216_d()) {
                 ItemStack itemstack8 = slot4.func_75211_c().func_77946_l();
                 itemstack8.func_190920_e(itemstack8.func_77976_d());
                 inventoryplayer.func_70437_b(itemstack8);
             }
-        }
-        else if (p_184996_3_ == ClickType.THROW && inventoryplayer.func_70445_o().func_190926_b() && p_184996_1_ >= 0)
-        {
-            Slot slot3 = (Slot)this.field_75151_b.get(p_184996_1_);
+        } else if (p_184996_3_ == ClickType.THROW && inventoryplayer.func_70445_o().func_190926_b() && p_184996_1_ >= 0) {
+            Slot slot3 = (Slot) this.field_75151_b.get(p_184996_1_);
 
-            if (slot3 != null && slot3.func_75216_d() && slot3.func_82869_a(p_184996_4_))
-            {
+            if (slot3 != null && slot3.func_75216_d() && slot3.func_82869_a(p_184996_4_)) {
                 ItemStack itemstack7 = slot3.func_75209_a(p_184996_2_ == 0 ? 1 : slot3.func_75211_c().func_190916_E());
                 slot3.func_190901_a(p_184996_4_, itemstack7);
                 p_184996_4_.func_71019_a(itemstack7, true);
             }
-        }
-        else if (p_184996_3_ == ClickType.PICKUP_ALL && p_184996_1_ >= 0)
-        {
-            Slot slot2 = (Slot)this.field_75151_b.get(p_184996_1_);
+        } else if (p_184996_3_ == ClickType.PICKUP_ALL && p_184996_1_ >= 0) {
+            Slot slot2 = (Slot) this.field_75151_b.get(p_184996_1_);
             ItemStack itemstack6 = inventoryplayer.func_70445_o();
 
-            if (!itemstack6.func_190926_b() && (slot2 == null || !slot2.func_75216_d() || !slot2.func_82869_a(p_184996_4_)))
-            {
+            if (!itemstack6.func_190926_b() && (slot2 == null || !slot2.func_75216_d() || !slot2.func_82869_a(p_184996_4_))) {
                 int i1 = p_184996_2_ == 0 ? 0 : this.field_75151_b.size() - 1;
                 int j1 = p_184996_2_ == 0 ? 1 : -1;
 
-                for (int i3 = 0; i3 < 2; ++i3)
-                {
-                    for (int j3 = i1; j3 >= 0 && j3 < this.field_75151_b.size() && itemstack6.func_190916_E() < itemstack6.func_77976_d(); j3 += j1)
-                    {
-                        Slot slot8 = (Slot)this.field_75151_b.get(j3);
+                for (int i3 = 0; i3 < 2; ++i3) {
+                    for (int j3 = i1; j3 >= 0 && j3 < this.field_75151_b.size() && itemstack6.func_190916_E() < itemstack6.func_77976_d(); j3 += j1) {
+                        Slot slot8 = (Slot) this.field_75151_b.get(j3);
 
-                        if (slot8.func_75216_d() && func_94527_a(slot8, itemstack6, true) && slot8.func_82869_a(p_184996_4_) && this.func_94530_a(itemstack6, slot8))
-                        {
+                        if (slot8.func_75216_d() && func_94527_a(slot8, itemstack6, true) && slot8.func_82869_a(p_184996_4_) && this.func_94530_a(itemstack6, slot8)) {
                             ItemStack itemstack14 = slot8.func_75211_c();
 
-                            if (i3 != 0 || itemstack14.func_190916_E() != itemstack14.func_77976_d())
-                            {
+                            if (i3 != 0 || itemstack14.func_190916_E() != itemstack14.func_77976_d()) {
                                 int k3 = Math.min(itemstack6.func_77976_d() - itemstack6.func_190916_E(), itemstack14.func_190916_E());
                                 ItemStack itemstack4 = slot8.func_75209_a(k3);
                                 itemstack6.func_190917_f(k3);
 
-                                if (itemstack4.func_190926_b())
-                                {
-                                    slot8.func_75215_d(ItemStack.field_190927_a);
+                                if (itemstack4.func_190926_b()) {
+                                    slot8.func_75215_d(field_190927_a);
                                 }
 
                                 slot8.func_190901_a(p_184996_4_, itemstack4);
@@ -468,10 +455,8 @@
                     }
                 }
             }
-
             this.func_75142_b();
         }
-
         return itemstack;
     }
 
@@ -492,7 +477,7 @@
         if (!inventoryplayer.func_70445_o().func_190926_b())
         {
             p_75134_1_.func_71019_a(inventoryplayer.func_70445_o(), false);
-            inventoryplayer.func_70437_b(ItemStack.field_190927_a);
+            inventoryplayer.func_70437_b(field_190927_a);
         }
     }
 
@@ -578,18 +563,19 @@
                 if (!itemstack.func_190926_b() && itemstack.func_77973_b() == p_75135_1_.func_77973_b() && (!p_75135_1_.func_77981_g() || p_75135_1_.func_77960_j() == itemstack.func_77960_j()) && ItemStack.func_77970_a(p_75135_1_, itemstack))
                 {
                     int j = itemstack.func_190916_E() + p_75135_1_.func_190916_E();
+                    int maxSize = Math.min(slot.func_75219_a(), p_75135_1_.func_77976_d());
 
-                    if (j <= p_75135_1_.func_77976_d())
+                    if (j <= maxSize)
                     {
                         p_75135_1_.func_190920_e(0);
                         itemstack.func_190920_e(j);
                         slot.func_75218_e();
                         flag = true;
                     }
-                    else if (itemstack.func_190916_E() < p_75135_1_.func_77976_d())
+                    else if (itemstack.func_190916_E() < maxSize)
                     {
-                        p_75135_1_.func_190918_g(p_75135_1_.func_77976_d() - itemstack.func_190916_E());
-                        itemstack.func_190920_e(p_75135_1_.func_77976_d());
+                        p_75135_1_.func_190918_g(maxSize - itemstack.func_190916_E());
+                        itemstack.func_190920_e(maxSize);
                         slot.func_75218_e();
                         flag = true;
                     }
@@ -671,6 +657,11 @@
 
     public static int func_94532_c(int p_94532_0_)
     {
+        //3 -> 11
+        //clickedButton -> 00 01 10
+        //0 为按下左键拖动
+        //1 为按下右键拖动
+        //2 就是按着中拖动
         return p_94532_0_ & 3;
     }
 
@@ -708,7 +699,7 @@
                 p_94525_2_.func_190920_e(1);
                 break;
             case 2:
-                p_94525_2_.func_190920_e(p_94525_2_.func_77973_b().func_77639_j());
+                p_94525_2_.func_190920_e(p_94525_2_.func_77976_d());
         }
 
         p_94525_2_.func_190917_f(p_94525_3_);
