--- ../src-base/minecraft/net/minecraft/inventory/ContainerEnchantment.java
+++ ../src-work/minecraft/net/minecraft/inventory/ContainerEnchantment.java
@@ -1,6 +1,8 @@
 package net.minecraft.inventory;
 
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 import java.util.Random;
 import net.minecraft.enchantment.Enchantment;
 import net.minecraft.enchantment.EnchantmentData;
@@ -16,20 +18,34 @@
 import net.minecraft.util.SoundCategory;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import net.minecraftforge.common.ForgeHooks;
+import net.minecraftforge.event.ForgeEventFactory;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.inventory.CraftInventoryEnchanting;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.enchantments.EnchantmentOffer;
+import org.bukkit.enchantments.EnchantmentWrapper;
+import org.bukkit.entity.Player;
+import org.bukkit.event.enchantment.EnchantItemEvent;
+import org.bukkit.event.enchantment.PrepareItemEnchantEvent;
 
 public class ContainerEnchantment extends Container
 {
     public IInventory field_75168_e;
-    public final World field_75172_h;
+    public World field_75172_h;
     private final BlockPos field_178150_j;
     private final Random field_75169_l;
     public int field_178149_f;
     public int[] field_75167_g;
     public int[] field_185001_h;
     public int[] field_185002_i;
-
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private Player player;
+    // CraftBukkit end
     @SideOnly(Side.CLIENT)
     public ContainerEnchantment(InventoryPlayer p_i45797_1_, World p_i45797_2_)
     {
@@ -49,6 +65,12 @@
                 super.func_70296_d();
                 ContainerEnchantment.this.func_75130_a(this);
             }
+            // CraftBukkit start
+            @Override
+            public Location getLocation() {
+                return new Location(field_75172_h.getWorld(), field_178150_j.func_177958_n(), field_178150_j.func_177956_o(), field_178150_j.func_177952_p());
+            }
+            // CraftBukkit end
         };
         this.field_75169_l = new Random();
         this.field_75167_g = new int[3];
@@ -70,9 +92,12 @@
         });
         this.func_75146_a(new Slot(this.field_75168_e, 1, 35, 47)
         {
+            java.util.List<ItemStack> ores = net.minecraftforge.oredict.OreDictionary.getOres("gemLapis");
             public boolean func_75214_a(ItemStack p_75214_1_)
             {
-                return p_75214_1_.func_77973_b() == Items.field_151100_aR && EnumDyeColor.func_176766_a(p_75214_1_.func_77960_j()) == EnumDyeColor.BLUE;
+                for (ItemStack ore : ores)
+                    if (net.minecraftforge.oredict.OreDictionary.itemMatches(ore, p_75214_1_, false)) return true;
+                return false;
             }
         });
 
@@ -88,8 +113,22 @@
         {
             this.func_75146_a(new Slot(p_i45798_1_, k, 8 + k * 18, 142));
         }
+        // CraftBukkit start
+        player = (Player) p_i45798_1_.field_70458_d.getBukkitEntity();
+        // CraftBukkit end
     }
+    // CraftBukkit start
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
 
+        CraftInventoryEnchanting inventory = new CraftInventoryEnchanting(this.field_75168_e);
+        bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
     protected void func_185000_c(IContainerListener p_185000_1_)
     {
         p_185000_1_.func_71112_a(this, 0, this.field_75167_g[0]);
@@ -157,6 +196,7 @@
                 if (!this.field_75172_h.field_72995_K)
                 {
                     int l = 0;
+                    float power = 0;
 
                     for (int j = -1; j <= 1; ++j)
                     {
@@ -164,37 +204,14 @@
                         {
                             if ((j != 0 || k != 0) && this.field_75172_h.func_175623_d(this.field_178150_j.func_177982_a(k, 0, j)) && this.field_75172_h.func_175623_d(this.field_178150_j.func_177982_a(k, 1, j)))
                             {
-                                if (this.field_75172_h.func_180495_p(this.field_178150_j.func_177982_a(k * 2, 0, j * 2)).func_177230_c() == Blocks.field_150342_X)
-                                {
-                                    ++l;
-                                }
-
-                                if (this.field_75172_h.func_180495_p(this.field_178150_j.func_177982_a(k * 2, 1, j * 2)).func_177230_c() == Blocks.field_150342_X)
-                                {
-                                    ++l;
-                                }
-
+                                power += ForgeHooks.getEnchantPower(field_75172_h, field_178150_j.func_177982_a(k * 2, 0, j * 2));
+                                power += ForgeHooks.getEnchantPower(field_75172_h, field_178150_j.func_177982_a(k * 2, 1, j * 2));
                                 if (k != 0 && j != 0)
                                 {
-                                    if (this.field_75172_h.func_180495_p(this.field_178150_j.func_177982_a(k * 2, 0, j)).func_177230_c() == Blocks.field_150342_X)
-                                    {
-                                        ++l;
-                                    }
-
-                                    if (this.field_75172_h.func_180495_p(this.field_178150_j.func_177982_a(k * 2, 1, j)).func_177230_c() == Blocks.field_150342_X)
-                                    {
-                                        ++l;
-                                    }
-
-                                    if (this.field_75172_h.func_180495_p(this.field_178150_j.func_177982_a(k, 0, j * 2)).func_177230_c() == Blocks.field_150342_X)
-                                    {
-                                        ++l;
-                                    }
-
-                                    if (this.field_75172_h.func_180495_p(this.field_178150_j.func_177982_a(k, 1, j * 2)).func_177230_c() == Blocks.field_150342_X)
-                                    {
-                                        ++l;
-                                    }
+                                    power += ForgeHooks.getEnchantPower(field_75172_h, field_178150_j.func_177982_a(k * 2, 0, j));
+                                    power += ForgeHooks.getEnchantPower(field_75172_h, field_178150_j.func_177982_a(k * 2, 1, j));
+                                    power += ForgeHooks.getEnchantPower(field_75172_h, field_178150_j.func_177982_a(k, 0, j * 2));
+                                    power += ForgeHooks.getEnchantPower(field_75172_h, field_178150_j.func_177982_a(k, 1, j * 2));
                                 }
                             }
                         }
@@ -204,7 +221,7 @@
 
                     for (int i1 = 0; i1 < 3; ++i1)
                     {
-                        this.field_75167_g[i1] = EnchantmentHelper.func_77514_a(this.field_75169_l, i1, l, itemstack);
+                        this.field_75167_g[i1] = EnchantmentHelper.func_77514_a(this.field_75169_l, i1, (int)power, itemstack);
                         this.field_185001_h[i1] = -1;
                         this.field_185002_i[i1] = -1;
 
@@ -212,6 +229,7 @@
                         {
                             this.field_75167_g[i1] = 0;
                         }
+                        this.field_75167_g[i1] = ForgeEventFactory.onEnchantmentLevelSet(field_75172_h, field_178150_j, i1, (int)power, itemstack, field_75167_g[i1]);
                     }
 
                     for (int j1 = 0; j1 < 3; ++j1)
@@ -228,7 +246,44 @@
                             }
                         }
                     }
+                    // CraftBukkit start
+                    CraftItemStack item = CraftItemStack.asCraftMirror(itemstack);
+                    EnchantmentOffer[] offers = new EnchantmentOffer[3];
+                    for (int idx = 0; idx < 3; ++idx) {
+                        org.bukkit.enchantments.Enchantment enchantment = (this.field_185001_h[idx] >= 0) ? org.bukkit.enchantments.Enchantment.getById(this.field_185001_h[idx]) : null;
+                        offers[idx] = (enchantment != null) ? new EnchantmentOffer(enchantment, this.field_185002_i[idx], this.field_75167_g[idx]) : null;
+                    }
 
+                    PrepareItemEnchantEvent event = new PrepareItemEnchantEvent(player, this.getBukkitView(), this.field_75172_h.getWorld().getBlockAt(field_178150_j.func_177958_n(), field_178150_j.func_177956_o(), field_178150_j.func_177952_p()), item, offers, l);
+                    //event.setCancelled(!itemstack.isItemEnchantable()); //!!!!setCancelled? exm?
+                    this.field_75172_h.getServer().getPluginManager().callEvent(event);
+
+                    if (event.isCancelled()) {
+                        for (l = 0; l < 3; ++l) {
+                            this.field_75167_g[l] = 0;
+                            this.field_185001_h[l] = -1;
+                            this.field_185002_i[l] = -1;
+                        }
+                        return;
+                    }
+
+                    for (l = 0; l < 3; l++) {
+                        EnchantmentOffer offer = event.getOffers()[l];
+                        if (offer != null) {
+                            this.field_75167_g[l] = offer.getCost();
+                            this.field_185001_h[l] = offer.getEnchantment().getId();
+                            this.field_185002_i[l] = offer.getEnchantmentLevel();
+                        }
+                        /*
+                        else {
+                            this.enchantLevels[l] = 0;
+                            this.enchantClue[l] = -1;
+                            this.worldClue[l] = -1;
+                        }
+                        */
+                        //OK, we DO NOT change anything if event didn't give us any offer
+                    }
+                    // CraftBukkit end
                     this.func_75142_b();
                 }
             }
@@ -246,50 +301,68 @@
 
     public boolean func_75140_a(EntityPlayer p_75140_1_, int p_75140_2_)
     {
-        ItemStack itemstack = this.field_75168_e.func_70301_a(0);
-        ItemStack itemstack1 = this.field_75168_e.func_70301_a(1);
+        ItemStack itemToEnchant = this.field_75168_e.func_70301_a(0);
+        ItemStack Lapis = this.field_75168_e.func_70301_a(1);
         int i = p_75140_2_ + 1;
 
-        if ((itemstack1.func_190926_b() || itemstack1.func_190916_E() < i) && !p_75140_1_.field_71075_bZ.field_75098_d)
+        if ((Lapis.func_190926_b() || Lapis.func_190916_E() < i) && !p_75140_1_.field_71075_bZ.field_75098_d)
         {
             return false;
         }
-        else if (this.field_75167_g[p_75140_2_] > 0 && !itemstack.func_190926_b() && (p_75140_1_.field_71068_ca >= i && p_75140_1_.field_71068_ca >= this.field_75167_g[p_75140_2_] || p_75140_1_.field_71075_bZ.field_75098_d))
+        else if (this.field_75167_g[p_75140_2_] > 0 && !itemToEnchant.func_190926_b() && (p_75140_1_.field_71068_ca >= i && p_75140_1_.field_71068_ca >= this.field_75167_g[p_75140_2_] || p_75140_1_.field_71075_bZ.field_75098_d))
         {
             if (!this.field_75172_h.field_72995_K)
             {
-                List<EnchantmentData> list = this.func_178148_a(itemstack, p_75140_2_, this.field_75167_g[p_75140_2_]);
+                List<EnchantmentData> list = this.func_178148_a(itemToEnchant, p_75140_2_, this.field_75167_g[p_75140_2_]);
+                // CraftBukkit start
+                if (!list.isEmpty()) {
+                    boolean itemIsBook = itemToEnchant.func_77973_b() == Items.field_151122_aG;
+                    Map<org.bukkit.enchantments.Enchantment, Integer> enchants = new HashMap<org.bukkit.enchantments.Enchantment, Integer>();
+                    for (Object obj : list) {
+                        EnchantmentData instance = (EnchantmentData) obj;
+                        //bukkit getEnchantmentByID returns null??
+                        enchants.put(org.bukkit.enchantments.Enchantment.getById(Enchantment.func_185258_b(instance.field_76302_b)), instance.field_76303_c);
+                    }
+                    CraftItemStack item = CraftItemStack.asCraftMirror(itemToEnchant);
 
-                if (!list.isEmpty())
-                {
-                    p_75140_1_.func_71013_b(i);
-                    boolean flag = itemstack.func_77973_b() == Items.field_151122_aG;
+                    EnchantItemEvent event = new EnchantItemEvent((Player) p_75140_1_.getBukkitEntity(), this.getBukkitView(), this.field_75172_h.getWorld().getBlockAt(field_178150_j.func_177958_n(), field_178150_j.func_177956_o(), field_178150_j.func_177952_p()), item, this.field_75167_g[p_75140_2_], enchants, p_75140_2_);
+                    this.field_75172_h.getServer().getPluginManager().callEvent(event);
 
-                    if (flag)
-                    {
-                        itemstack = new ItemStack(Items.field_151134_bR);
-                        this.field_75168_e.func_70299_a(0, itemstack);
+                    int level = event.getExpLevelCost();
+                    if (event.isCancelled() || (level > p_75140_1_.field_71068_ca && !p_75140_1_.field_71075_bZ.field_75098_d) || event.getEnchantsToAdd().isEmpty()) {
+                        return false;
                     }
+                    if (itemIsBook) {
+                        itemToEnchant = new ItemStack(Items.field_151134_bR);
+                        this.field_75168_e.func_70299_a(0, itemToEnchant);
+                    }
 
-                    for (int j = 0; j < list.size(); ++j)
-                    {
-                        EnchantmentData enchantmentdata = (EnchantmentData)list.get(j);
+                    for (Map.Entry<org.bukkit.enchantments.Enchantment, Integer> entry : event.getEnchantsToAdd().entrySet()) {
+                        try {
+                            if (itemIsBook) {
+                                int enchantId = entry.getKey().getId();
+                                if (Enchantment.func_185262_c(enchantId) == null) {
+                                    continue;
+                                }
 
-                        if (flag)
-                        {
-                            Items.field_151134_bR.func_92115_a(itemstack, enchantmentdata);
+                                EnchantmentData weightedrandomenchant = new EnchantmentData(Enchantment.func_185262_c(enchantId), entry.getValue());
+                                Items.field_151134_bR.func_92115_a(itemToEnchant, weightedrandomenchant);
+                            } else {
+                                itemToEnchant.func_77966_a(Enchantment.func_185262_c(entry.getKey().getId()), entry.getValue());
+                            }
+                        } catch (IllegalArgumentException e) {
+                            /* Just swallow invalid enchantments */
                         }
-                        else
-                        {
-                            itemstack.func_77966_a(enchantmentdata.field_76302_b, enchantmentdata.field_76303_c);
-                        }
                     }
 
+                    p_75140_1_.func_71013_b(i);
+                    // CraftBukkit end
+
                     if (!p_75140_1_.field_71075_bZ.field_75098_d)
                     {
-                        itemstack1.func_190918_g(i);
+                        Lapis.func_190918_g(i);
 
-                        if (itemstack1.func_190926_b())
+                        if (Lapis.func_190926_b())
                         {
                             this.field_75168_e.func_70299_a(1, ItemStack.field_190927_a);
                         }
@@ -334,7 +407,11 @@
     public void func_75134_a(EntityPlayer p_75134_1_)
     {
         super.func_75134_a(p_75134_1_);
-
+        // CraftBukkit Start - If an enchantable was opened from a null location, set the world to the player's world, preventing a crash
+        if (this.field_75172_h == null) {
+            this.field_75172_h = p_75134_1_.func_130014_f_();
+        }
+        // CraftBukkit end
         if (!this.field_75172_h.field_72995_K)
         {
             for (int i = 0; i < this.field_75168_e.func_70302_i_(); ++i)
