--- ../src-base/minecraft/net/minecraft/inventory/InventoryEnderChest.java
+++ ../src-work/minecraft/net/minecraft/inventory/InventoryEnderChest.java
@@ -5,14 +5,21 @@
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.nbt.NBTTagList;
 import net.minecraft.tileentity.TileEntityEnderChest;
+import org.bukkit.inventory.InventoryHolder;
 
 public class InventoryEnderChest extends InventoryBasic
 {
     private TileEntityEnderChest field_70488_a;
+    // CraftBukkit start
+    private final EntityPlayer owner;
 
-    public InventoryEnderChest()
-    {
+    public InventoryHolder getBukkitOwner() {
+        return owner.getBukkitEntity();
+    }
+    public InventoryEnderChest(EntityPlayer owner) {
         super("container.enderchest", false, 27);
+        this.owner = owner;
+        // CraftBukkit end
     }
 
     public void func_146031_a(TileEntityEnderChest p_146031_1_)
