--- ../src-base/minecraft/net/minecraft/inventory/SlotFurnaceOutput.java
+++ ../src-work/minecraft/net/minecraft/inventory/SlotFurnaceOutput.java
@@ -6,7 +6,15 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.crafting.FurnaceRecipes;
 import net.minecraft.stats.AchievementList;
+import net.minecraft.tileentity.TileEntityFurnace;
+import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MathHelper;
+import org.bukkit.craftbukkit.CraftChunk;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.block.CraftFurnace;
+import org.bukkit.entity.Player;
+import org.bukkit.event.inventory.FurnaceExtractEvent;
+import org.bukkit.event.inventory.FurnaceSmeltEvent;
 
 public class SlotFurnaceOutput extends Slot
 {
@@ -71,7 +79,17 @@
 
                 i = j;
             }
+            // CraftBukkit start - fire FurnaceExtractEvent
+            Player bplayer = (Player) field_75229_a.getBukkitEntity();
+            TileEntityFurnace furnace = ((TileEntityFurnace) this.field_75224_c);
+            org.bukkit.block.Block block = field_75229_a.field_70170_p.getWorld().getBlockAt(furnace.func_174877_v().func_177958_n(), furnace.func_174877_v().func_177956_o(), furnace.func_174877_v().func_177952_p());
 
+            if (field_75228_b != 0) {
+                FurnaceExtractEvent event = new FurnaceExtractEvent(bplayer, block, org.bukkit.craftbukkit.util.CraftMagicNumbers.getMaterial(p_75208_1_.func_77973_b()), field_75228_b, i);
+                field_75229_a.field_70170_p.getServer().getPluginManager().callEvent(event);
+                i = event.getExpToDrop();
+            }
+            // CraftBukkit end
             while (i > 0)
             {
                 int k = EntityXPOrb.func_70527_a(i);
@@ -82,6 +100,8 @@
 
         this.field_75228_b = 0;
 
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerSmeltedEvent(field_75229_a, p_75208_1_);
+
         if (p_75208_1_.func_77973_b() == Items.field_151042_j)
         {
             this.field_75229_a.func_71029_a(AchievementList.field_187987_k);
