--- ../src-base/minecraft/net/minecraft/entity/ai/EntityAIMate.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/EntityAIMate.java
@@ -11,7 +11,11 @@
 import net.minecraft.stats.StatList;
 import net.minecraft.util.EnumParticleTypes;
 import net.minecraft.world.World;
-
+import net.minecraftforge.event.entity.living.BabyEntitySpawnEvent;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.entity.EntityBreedEvent;
+import net.minecraft.entity.EntityLivingBase;
 public class EntityAIMate extends EntityAIBase
 {
     private final EntityAnimal field_75390_d;
@@ -92,7 +96,44 @@
     private void func_75388_i()
     {
         EntityAgeable entityageable = this.field_75390_d.func_90011_a(this.field_75391_e);
+        int xptodrop = field_75390_d.func_70681_au().nextInt(7) + 1;
+        //fire bukkitevent first, MCPCRevive start
+        EntityLivingBase child = (EntityLivingBase) entityageable;
+        EntityLivingBase parentA = (EntityLivingBase) field_75390_d;
+        EntityLivingBase parentB = (EntityLivingBase) field_75391_e;
+        EntityPlayer playerMP = null;
 
+        if (parentA instanceof EntityAnimal) {
+            playerMP = ((EntityAnimal)parentA).func_146083_cb();
+        }
+
+        if (playerMP == null && parentB instanceof EntityAnimal)
+        {
+            playerMP = ((EntityAnimal)parentB).func_146083_cb();
+        }
+        EntityBreedEvent craftevent = CraftEventFactory.callEntityBreedEvent(child,parentA,parentB,(EntityLivingBase) playerMP, playerMP.func_184614_ca(),xptodrop);
+        CraftEventFactory.callEvent(craftevent);
+        if(craftevent.isCancelled())
+        {
+            this.field_75390_d.func_70873_a(6000);
+            this.field_75391_e.func_70873_a(6000);
+            this.field_75390_d.func_70875_t();
+            this.field_75391_e.func_70875_t();
+            return;
+        }
+        //MCPCRevive end
+        final net.minecraftforge.event.entity.living.BabyEntitySpawnEvent event = new net.minecraftforge.event.entity.living.BabyEntitySpawnEvent(field_75390_d, field_75391_e, entityageable);
+        final boolean cancelled = net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+        entityageable = event.getChild();
+        if (cancelled) {
+            //Reset the "inLove" state for the animals
+            this.field_75390_d.func_70873_a(6000);
+            this.field_75391_e.func_70873_a(6000);
+            this.field_75390_d.func_70875_t();
+            this.field_75391_e.func_70875_t();
+            return;
+        }
+
         if (entityageable != null)
         {
             EntityPlayer entityplayer = this.field_75390_d.func_146083_cb();
