--- ../src-base/minecraft/net/minecraft/entity/ai/attributes/RangedAttribute.java
+++ ../src-work/minecraft/net/minecraft/entity/ai/attributes/RangedAttribute.java
@@ -6,7 +6,7 @@
 public class RangedAttribute extends BaseAttribute
 {
     private final double field_111120_a;
-    private final double field_111118_b;
+    public double field_111118_b;
     private String field_111119_c;
 
     public RangedAttribute(@Nullable IAttribute p_i45891_1_, String p_i45891_2_, double p_i45891_3_, double p_i45891_5_, double p_i45891_7_)
